{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.1.7     2.4\n"
     ]
    }
   ],
   "source": [
    "import pyham\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "\n",
    "print(pyham.__version__, \"   \", nx.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def locus_to_id(file):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function extracts loci from a fasta file and store the information as keys in a dictionary \n",
    "    and the related protein ID as value.\n",
    "\n",
    "    Args:\n",
    "        file (:obj:`str`): path to the file of interest.\n",
    "\n",
    "    Returns:\n",
    "        a dictionary mapping the locus (:obj:`str`) with the protein ID (:obj:`str`).\n",
    "    \"\"\"\n",
    "    \n",
    "    locus_to_id = {}\n",
    "    for seq_record in SeqIO.parse(file, 'fasta'):\n",
    "        locus_to_id[seq_record.description.split('locus: ')[-1]] = seq_record.description\n",
    "        \n",
    "    return locus_to_id\n",
    "\n",
    "\n",
    "def get_locus(directory):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function goes through each file of a folder and build a dictionary containing loci \n",
    "    as keys and the related protein ID as values. For each entry in the dictionary, the function \n",
    "    updates the corresponding :obj:`pyham.abstractgene.Gene` by adding its locus.\n",
    "\n",
    "    Args:\n",
    "        directory (:obj:`str`): directory of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "\n",
    "    path = directory\n",
    "    for file in os.listdir(path):\n",
    "        ##print('File:', file)\n",
    "        path_to_file = path + '/' + file\n",
    "        locus_dictionary = locus_to_id(path_to_file)\n",
    " \n",
    "        for (loci, gene_id) in locus_dictionary.items():\n",
    "            gene = ham_analysis_no_name.get_genes_by_external_id(gene_id)\n",
    "            \n",
    "            # each gene should only map to 1 other gene, so should always be length 1...\n",
    "            assert len(gene) == 1, 'gene_id maps to multiple entries'\n",
    "            gene[0].locus = loci\n",
    "            #for element in gene:\n",
    "            #    element.locus = loci\n",
    "                \n",
    "        ham_analysis_no_name.get_extant_genome_by_name('.'.join(file.split('.')[:-1])).locus_assignment = True\n",
    "        \n",
    "\n",
    "def get_gene_by_locus(genome, locus):\n",
    "    \n",
    "    \"\"\"\n",
    "    Get a :obj:`pyham.abstractgene.Gene` that match the query locus. \n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant genome of interest.\n",
    "        locus (:obj:`int`): locus.\n",
    "\n",
    "    Returns:\n",
    "        :obj:`pyham.abstractgene.Gene`.\n",
    "    \"\"\"\n",
    "    \n",
    "    for gene in genome.genes:\n",
    "        if int(gene.locus) == locus:\n",
    "            return gene \n",
    "    \n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_neighbors(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function assigns neighbors to each :obj:`pyham.abstractgene.Gene` or :obj:`pyham.abstractgene.HOG`.\n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    if isinstance(genome, pyham.ExtantGenome):\n",
    "        \n",
    "        for extant_gene in genome.genes:\n",
    "            extant_gene.neighbors = (get_gene_by_locus(genome, int(extant_gene.locus)-1), \n",
    "                                     get_gene_by_locus(genome, int(extant_gene.locus)+1))            \n",
    "                        \n",
    "    if isinstance(genome, pyham.AncestralGenome):\n",
    "    \n",
    "        for ancestral_gene in genome.genes:  \n",
    "            \n",
    "            # test here, see if does the same thing, children should only be direct descendants?\n",
    "            #children = [child for child in ancestral_gene.children]\n",
    "            children = ancestral_gene.children\n",
    "            \n",
    "            ######################################################################################\n",
    "            # Dictionary containing genome names as keys (i.e. SE004) and a list of children as value #\n",
    "            ######################################################################################\n",
    "            \n",
    "            genome_dict = {}  \n",
    "            \n",
    "            for child in children:\n",
    "                if child.genome.name in genome_dict:\n",
    "                    genome_dict[child.genome.name].append(child) # if already in dictionary, add it (duplicate)\n",
    "                else:\n",
    "                    genome_dict[child.genome.name] = [child] # otherwise add the child\n",
    "            \n",
    "            ############################################\n",
    "            # Children are replaced by their neighbors #      \n",
    "            ############################################\n",
    "            \n",
    "            for genome, gene_collection in genome_dict.items(): #stoca \n",
    "                for i in range(len(gene_collection)):\n",
    "                    gene_collection[i] = list(gene_collection[i].neighbors)\n",
    "            \n",
    "            ################################################################################\n",
    "            # Dictionary containing genomes as keys (i.e. SE004) and a list of parent HOGs #\n",
    "            ################################################################################\n",
    "            \n",
    "            parent_dict = {} # gets a list of all potential neighboring hogs to the ancestral_gene\n",
    "            for genome, neighbors_collection in genome_dict.items(): #stoca \n",
    "                for neighbors_pair in neighbors_collection:\n",
    "                    for neighbor in neighbors_pair: # don't really need the for loop bc only ever 2 neighbors\n",
    "                        if genome in parent_dict:   ## KJG - I don't understand this if/else - if we already put the focal genome in the parent dictionary, we do the same thing as if we didn't?\n",
    "                            if neighbor == None:\n",
    "                                parent_dict[genome].add(None)\n",
    "                            else:\n",
    "                                if neighbor.parent != ancestral_gene: #if theres' a duplication this might be true\n",
    "                                    parent_dict[genome].add(neighbor.parent) #add the neighbors parents - the neighboring hogs\n",
    "                        else:\n",
    "                            if neighbor == None:\n",
    "                                parent_dict[genome] = {None}    ## ah, this makes focal genome 'none' while above does 'add(None)' in addition to current genome?\n",
    "                            else:\n",
    "                                if neighbor.parent != ancestral_gene: # fill the parent dict by looking to see \n",
    "                                    parent_dict[genome] = {neighbor.parent}  ## again, no 'add', just make\n",
    "\n",
    "            l = []  # List of sets\n",
    "            for genome, parent in parent_dict.items():  \n",
    "                l.append(parent)  # parent is a set\n",
    "                ancestral_gene.list_parents = l            # List of potential neighbors in case of conflict\n",
    "                \n",
    "            final_neighbors = list(l[0].intersection(*l))  # Intersection of the sets\n",
    "            \n",
    "            if len(final_neighbors) == 2:\n",
    "                ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "                print(\"fin\", ancestral_gene.neighbors)\n",
    "            \n",
    "            elif len(final_neighbors) == 1:\n",
    "                none_counter = 0\n",
    "                potential_neighbor = None\n",
    "                \n",
    "                for collection in l:\n",
    "                    if len(collection) == 2:\n",
    "                        for element in collection:\n",
    "                            if element == final_neighbors[0]:\n",
    "                                continue\n",
    "                            elif element == None:\n",
    "                                none_counter += 1\n",
    "                            else:\n",
    "                                potential_neighbor = element\n",
    "                \n",
    "                    else:\n",
    "                        break\n",
    "                \n",
    "                if none_counter == 1:  # To deal with missing orthology\n",
    "                    final_neighbors.append(potential_neighbor)\n",
    "                \n",
    "                if len(final_neighbors) == 1:\n",
    "                    final_neighbors.append(None)\n",
    "                    ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "                else:  \n",
    "                    ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "            \n",
    "            elif len(final_neighbors) == 0 or len(final_neighbors) > 2:\n",
    "                final_neighbors = [None, None]\n",
    "                ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "\n",
    "\n",
    "def get_synteny_blocks(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function aims to reconstruct synteny blocks using neighbors of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        list of lists containing :obj:`pyham.abstractgene.Gene` or :obj:`pyham.abstractgene.HOG` \n",
    "        in a specific order.\n",
    "    \"\"\"\n",
    "\n",
    "    syn_blocks = []\n",
    "    seen = []\n",
    "    gene_collection = genome.genes\n",
    "    tmp = []\n",
    "        \n",
    "    for abstract_gene in gene_collection: # go through all the genes (extant or ancestral) for the genome\n",
    "        \n",
    "        if abstract_gene in seen: # if the gene is already there, skip out of the loop to the next gene\n",
    "            continue\n",
    "                \n",
    "        block1 = []\n",
    "        block2 = []\n",
    "        \n",
    "        current1 = abstract_gene\n",
    "        current2 = abstract_gene  # Used for the block2\n",
    "        \n",
    "        block1.append(current1)  # [X] where X is the first abstract gene # add the current gene to block 1\n",
    "        seen.append(current1) # append the current gene to seen\n",
    "        \n",
    "        ######################\n",
    "        # First end (block1) #\n",
    "        ######################\n",
    "        \n",
    "        next_gene = current1.neighbors[0] # find the neighbors of the current gene\n",
    "        \n",
    "        while next_gene != None: # if there is a next gene\n",
    "            \n",
    "            if next_gene in seen:\n",
    "                next_gene = None  \n",
    "                continue\n",
    "            \n",
    "            previous = current1 # then put the current gene as previous\n",
    "            current1 = next_gene # and put the next gene as current\n",
    "            \n",
    "            if previous in current1.neighbors:   # if the previous gene is a neighbor of the new current gene\n",
    "                block1.append(current1) # append the new current gene to block 1\n",
    "                seen.append(current1)   # append the new current gene to seen\n",
    "                # this step should always be true and the else statement is a check that breaks if not true\n",
    "            \n",
    "            else:                                  \n",
    "                break\n",
    "                print(\"previous gene is not a neighbor of the current gene\")\n",
    "                \n",
    "            if previous == current1.neighbors[0]: # if the previous gene is the first neighbor of the current gene\n",
    "                next_gene = current1.neighbors[1] # then the next gene equals the 2nd neighbor of the current gene\n",
    "            \n",
    "            else:\n",
    "                next_gene = current1.neighbors[0] # otherwise the next gene is the first neighbor of the current gene\n",
    "            # then iterate the while loop\n",
    "\n",
    "        #######################\n",
    "        # Second end (block2) #\n",
    "        #######################\n",
    "        \n",
    "        next_gene = current2.neighbors[1] # current2 is the same original abstract gene from above\n",
    "                                          # take its 2nd neighbor and make it the next gene\n",
    "            \n",
    "        while next_gene != None:    # if there is a next gene\n",
    "            \n",
    "            if next_gene in seen:   # if it's already seen, next = none and iterate the while loop\n",
    "                next_gene = None\n",
    "                continue\n",
    "            \n",
    "            previous = current2    # make the current gene become the previous gene\n",
    "            current2 = next_gene   # make the next gene become the current gene\n",
    "            \n",
    "            if previous in current2.neighbors:   # if the previous gene is a neighbor of the new next current gene\n",
    "                block2.append(current2)  # append the new current gene to block 2\n",
    "                seen.append(current2)    # append the new current gene to seen\n",
    "            \n",
    "            else:                                  \n",
    "                break\n",
    "                print(\"previous gene is not a neighbor of the current2 gene\")\n",
    "                \n",
    "            if previous == current2.neighbors[0]: # if the previous gene is the first neighbor of the current gene\n",
    "                next_gene = current2.neighbors[1] # then the next gene equals the 2nd neighbor of the current gene\n",
    "            \n",
    "            else:\n",
    "                next_gene = current2.neighbors[0] # otherwise the next gene is the first neighbor of the current gene\n",
    "            # then iterate the while loop\n",
    "            \n",
    "        syn_blocks.append(block1[::-1] + block2)\n",
    "        \n",
    "    print(syn_blocks)\n",
    "    genome.synteny = syn_blocks\n",
    "\n",
    "\n",
    "def assign_block_to_genes(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    The function links each :obj:`pyham.abstractgene.AbstractGene` to the synteny block it belongs to. \n",
    "    It creates a dictionary containing the genome of :obj:`pyham.abstractgene` as key and the synteny \n",
    "    block as value.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    for block in genome.synteny:\n",
    "    \n",
    "        for abstract_gene in block:\n",
    "            genomes_to_block = {}  # Dictionary\n",
    "            genomes_to_block[genome.name] = block  \n",
    "            abstract_gene.blocks = genomes_to_block\n",
    "            \n",
    "\n",
    "def get_unique_children(genome, children):  \n",
    "    \n",
    "    \"\"\"\n",
    "    Due to tandem duplication, some :obj:`pyham.abstractgene.HOG` can have more than two children.\n",
    "    The function refines the list of children by removing duplicates.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        children (:obj:`list` of :obj:`pyham.abstractgene.AbstractGene`): a list of direct descendants AbstractGene.\n",
    "    \n",
    "    Returns:\n",
    "        list of :obj:`pyham.abstractgene.AbstractGene` without duplicates.\n",
    "    \"\"\"\n",
    "    \n",
    "    unique_genome = []\n",
    "    unique_children = []\n",
    "    \n",
    "    if len(children) == 2:  # Two children comming from two different genomes\n",
    "        \n",
    "        if children[0].genome != children[1].genome:\n",
    "            return children\n",
    "        \n",
    "        else:  # Two children comming from the same genome (i.e. duplication)\n",
    "            unique_children.append(None)\n",
    "            return unique_children\n",
    "    \n",
    "    else:  # In case of duplications, only one child per specie is taken into account\n",
    "        \n",
    "        for child in children:  \n",
    "            \n",
    "            if child.genome in unique_genome:\n",
    "                continue\n",
    "\n",
    "            else:\n",
    "                \n",
    "                for neighbor in child.neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        continue\n",
    "\n",
    "                    elif neighbor.parent:\n",
    "                        \n",
    "                        if neighbor.parent in child.parent.blocks[genome.name]:  # To make sure that, for both unique children, we go in the same direction. \n",
    "                            continue\n",
    "                        \n",
    "                        unique_genome.append(child.genome)\n",
    "                        unique_children.append(child)\n",
    "\n",
    "                    else:  # If the child has no parent HOG\n",
    "                        unique_genome.append(child.genome)\n",
    "                        unique_children.append(child)\n",
    "        \n",
    "        if len(unique_children) == 0:\n",
    "            unique_children.append(None)\n",
    "            return unique_children  \n",
    "        \n",
    "        else:\n",
    "            return unique_children \n",
    "        \n",
    "\n",
    "def check_unique_children_from_unit_block(genome, children, genes_seen):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function aims to be used after refining the :obj:`list` genes_seen. It uses the :obj:`list` \n",
    "    genes_seen to check whether the :obj:`list` current_children returned by the function get_unique_children \n",
    "    are correctly selected. Indeed, the function get_unique_children encounters a limitation with single HOG.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        children (:obj:`list` of :obj:`pyham.abstractgene.AbstractGene`): a list of direct descendants AbstractGene.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        \n",
    "    Returns:\n",
    "        list of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "    \"\"\"\n",
    "    \n",
    "    unique_genome = []\n",
    "    unique_children = []\n",
    "    \n",
    "    if len(children) == 2:  # Two children comming from two different genomes\n",
    "        \n",
    "        if children[0].genome != children[1].genome:\n",
    "            return children\n",
    "        \n",
    "        else:  # Two children comming from the same genome (i.e. duplication)\n",
    "            unique_children.append(None)\n",
    "            return unique_children\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        for child in children:  \n",
    "            \n",
    "            if child.genome in unique_genome:\n",
    "                continue\n",
    "            \n",
    "            for neighbor in child.neighbors:\n",
    "                \n",
    "                if neighbor == None:\n",
    "                    continue\n",
    "\n",
    "                if neighbor.parent in child.parent.blocks[genome.name] or neighbor in genes_seen:  # To make sure that for both unique children we go in the same direction \n",
    "                    continue\n",
    "                \n",
    "                unique_genome.append(child.genome)\n",
    "                unique_children.append(child)\n",
    "                \n",
    "        if len(unique_children) == 0:\n",
    "            unique_children.append(None)\n",
    "            return unique_children  # Return a list without duplicates\n",
    "        \n",
    "        else:\n",
    "            return unique_children  \n",
    "\n",
    "        \n",
    "class Link:\n",
    "    \n",
    "    \"\"\"\n",
    "    Link is a class representing linking regions. Linking regions are unresolved regions with a weak synteny \n",
    "    signal that links two synteny blocks together. A Link is defined by its collection of poorly resolved genes.\n",
    "    \n",
    "    Attributes:\n",
    "        | linking_region: dictionary of :obj:`pyham.abstractgene.AbstractGene`\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.linking_region = {}\n",
    "        \n",
    "    def get_linking_region(self, genome, block, common_block, starting_children):\n",
    "        \n",
    "        \"\"\"\n",
    "        Create a linking region filled with :obj:`pyham.abstractgene.AbstractGene`.\n",
    "        \n",
    "        Args:\n",
    "            genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "            block (:obj:`list`): the synteny block from which the assessement starts.\n",
    "            common_block (:obj:`list`): the synteny block to be linked with the initial block.\n",
    "            starting_children (:obj:`list`): list of children coming from the last :obj:`pyham.abstractgene.AbstractGene` of the initial block.\n",
    "        \n",
    "        Returns:\n",
    "            return linking_region: dictionary of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "        \"\"\"\n",
    "        \n",
    "        initial_children = starting_children\n",
    "        seen = []\n",
    "\n",
    "        for child in initial_children:  # Creation of the dictionary's entries ({species1: [], species2: []})\n",
    "\n",
    "            if child.genome.name in self.linking_region:\n",
    "                continue\n",
    "\n",
    "            else:\n",
    "                self.linking_region[child.genome.name] = [] \n",
    "\n",
    "        next_children = []\n",
    "\n",
    "        for child in initial_children:\n",
    "            current_child = child\n",
    "            growing_link = True\n",
    "\n",
    "            while growing_link:\n",
    "                \n",
    "                for neighbor in current_child.neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor != current_child.next_neighbor:\n",
    "                        continue\n",
    "\n",
    "                    if neighbor.parent:\n",
    "\n",
    "                        if neighbor.parent.blocks[genome.name] != common_block:\n",
    "                            \n",
    "                            if neighbor.parent not in self.linking_region[neighbor.genome.name]:\n",
    "                                self.linking_region[neighbor.genome.name].append(neighbor.parent)\n",
    "                                current_child = neighbor\n",
    "                                seen.append(neighbor)\n",
    "\n",
    "                            else:\n",
    "                                current_child = neighbor\n",
    "                                seen.append(neighbor)\n",
    "\n",
    "                        else:\n",
    "                            growing_link = False\n",
    "\n",
    "                    else:\n",
    "                        self.linking_region[neighbor.genome.name].append(neighbor)  # The child is added to the linking region anyway.\n",
    "                        current_child = neighbor\n",
    "                        seen.append(neighbor)\n",
    "\n",
    "    def __repr__(self):\n",
    "        return '<{}>'.format(self.__class__.__name__)\n",
    "    \n",
    "    \n",
    "def refine_genes_seen(genome, genes_seen, current_children):\n",
    "    \n",
    "    \"\"\"\n",
    "    Regarding the secondary reconstruction, when the growing process continue by starting from a single \n",
    "    :obj:`pyham.abstractgene.HOG`, both neighbors of one of its children can be within the \n",
    "    :obj:`list` genes_seen. The function will remove all children that come after the children of interest\n",
    "    within the :obj:`list` genes_seen.\n",
    "    \n",
    "    Ars:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        current_children (:obj:`list`): list of children to begin the investigation with at the next iteration.\n",
    "        \n",
    "    Returns:\n",
    "        return the refined :obj:`list` genes_seen.\n",
    "    \"\"\"\n",
    "    \n",
    "    if len(current_children) == 2:\n",
    "        tmp = genes_seen[:]  # Make an independent copy of the initial list\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == current_children[0]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == current_children[0].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:]\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == current_children[1]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == current_children[1].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:] \n",
    "\n",
    "    return genes_seen\n",
    "\n",
    "\n",
    "def adjust_genes_seen(genome, genes_seen, starting_children):\n",
    "    \n",
    "    \"\"\"\n",
    "    Same principle as for the function refine_genes_seen but used when the growing process continue by\n",
    "    sarting from an element at the extremity of a synteny block. Neighbors from one of its children can\n",
    "    be present within the :obj:`list` genes_seen. The function will remove all children that come after \n",
    "    the children of interest within the :obj:`list` genes_seen.\n",
    "    \n",
    "    Ars:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        starting_children (:obj:`list`): list of children to begin the investigation with at the next iteration.\n",
    "        \n",
    "    Returns:\n",
    "        return the refined :obj:`list` genes_seen.\n",
    "    \"\"\"\n",
    "    \n",
    "    if len(starting_children) == 2:\n",
    "        tmp = genes_seen[:]  # Make an independent copy of the initial list\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == starting_children[0]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == starting_children[0].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:]\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == starting_children[1]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == starting_children[1].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:] \n",
    "\n",
    "    return genes_seen\n",
    "\n",
    "\n",
    "def joining_blocks(window, genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    The function performs the secondary reconstruction. Synteny blocks from the primary reconstruction\n",
    "    are linked together using linking regions.\n",
    "    \n",
    "    Ars:\n",
    "        window (:obj:`int`): a window of assessment\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        \n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    for hog in genome.genes:  #### To be used with the function \"add_remaining_blocks\"\n",
    "        hog.processed = False\n",
    "    \n",
    "    new_blocks = []\n",
    "    block_record = []  # To keep track of the block already assessed. Useful to link several blocks in a row.\n",
    "    \n",
    "    for block in genome.synteny:\n",
    "            \n",
    "        if block in block_record or len(block) == 1:  \n",
    "            continue\n",
    "        \n",
    "        block2 = block  # Used to reconstruct the other side\n",
    "        \n",
    "        counter = 0\n",
    "        children = [child for child in block[-1].children]\n",
    "        current_children = get_unique_children(genome, children)  # Remove duplicates from children\n",
    "        starting_children = current_children  # To keep track of the initial children considered (used to build a linking region)\n",
    "        \n",
    "        if len(current_children) == 1 or len(current_children) > 2:  # If missing information (==1) or other special case (>2), the block is no longer taken into account\n",
    "            continue\n",
    "        \n",
    "        block_record.append(block)   \n",
    "        \n",
    "        blocks_seen = [set() for child in current_children]  # To save already assessed blocks and perform the intersection\n",
    "        genes_seen = []  # To keep track of already assessed genes\n",
    "        intersect = None\n",
    "        extended_block = []\n",
    "        next_children = [None, None]\n",
    "        genes_seen.append(current_children[0])  \n",
    "        genes_seen.append(current_children[1])  \n",
    "        \n",
    "        while counter < window:\n",
    "            none_type_as_neighbor = 0\n",
    "            block_found = False\n",
    "            reaching_already_assessed = False  \n",
    "            \n",
    "            if None in current_children:\n",
    "                break\n",
    "            \n",
    "            # Without the following block, when a common block is reached from one extremity, the next iteration \n",
    "            # starts from the opposite one. However, children originating from that extreme element are not\n",
    "            # added within the list of genes seen.\n",
    "            if current_children[0] and current_children[1] not in genes_seen:\n",
    "                genes_seen.append(current_children[0])\n",
    "                genes_seen.append(current_children[1])\n",
    "            \n",
    "            for i in range(len(current_children)):\n",
    "                \n",
    "                for neighbor in current_children[i].neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        none_type_as_neighbor += 1\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor.parent:  # Check whether the neighbor is a singleton (i.e. does it have a parent or not?)\n",
    "                        \n",
    "                        if neighbor.parent in block or neighbor in genes_seen:  # Avoid to take into account elements from the initial block or those already assessed\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            blocks_seen[i].add(tuple(neighbor.parent.blocks[genome.name]))\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region \n",
    "                            break\n",
    "                    \n",
    "                    else:\n",
    "                        \n",
    "                        if neighbor in genes_seen:\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region\n",
    "                \n",
    "            if none_type_as_neighbor > 0:  # If children have only one neighbor\n",
    "                break\n",
    "\n",
    "            counter += 1\n",
    "            current_children = next_children[:]\n",
    "            intersect = list(blocks_seen[0].intersection(*blocks_seen))  # Intersection\n",
    "            \n",
    "            if len(intersect) == 1:  # ex. intersect = [('a', 'c')]\n",
    "\n",
    "                    if list(intersect[0]) in block_record:\n",
    "                        reaching_already_assessed = True\n",
    "                        break\n",
    "                        \n",
    "                    block_found = True  \n",
    "                    counter = 0  # Reset the counter to pursue the assessement with further blocks\n",
    "                    common_block = list(intersect[0])  # Extraction of the last common block. \n",
    "                    block_record.append(common_block)\n",
    "\n",
    "                    link = Link()  # Class for linking regions\n",
    "                    link.get_linking_region(genome, block, common_block, starting_children)  # Creation of the linking region\n",
    "                    \n",
    "                    if not extended_block:  # If the list if empty (i.e. for a first block of a serie)\n",
    "                        extended_block.append(block)  # The initial block\n",
    "                        extended_block.append(link)  # The linking region\n",
    "                        extended_block.append(common_block)  # The common block\n",
    "\n",
    "                    else:  # For all the block but the first one, \"block\" corresponds to the previous common block\n",
    "                        extended_block.append(link)\n",
    "                        extended_block.append(common_block)\n",
    "\n",
    "                    block = common_block  # The new block becomes the last common block found within the while loop\n",
    "\n",
    "                    if len(block) > 1:\n",
    "                        \n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its first element (i.e. element 0), the next iteration should start from its last element (i.e. element -1) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block[0].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block[0].children)[0] and get_unique_children(genome, block[0].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block[-1].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)  \n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block[0].children)\n",
    "                            break\n",
    "\n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its last element (i.e. element -1), the next iteration should start from its first element (i.e. element 0) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block[-1].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block[-1].children)[0] and get_unique_children(genome, block[-1].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block[0].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block[-1].children)\n",
    "                            break\n",
    "                    \n",
    "                    ################################################################################################################\n",
    "                    # When we reach a common unit block, the next iteration should start from the same unique HOG (i.e. element 0) #\n",
    "                    ################################################################################################################\n",
    "\n",
    "                    elif len(block) == 1:  # If the block has only one HOG...\n",
    "                        current_children = get_unique_children(genome, block[0].children)                         \n",
    "                        \n",
    "                        if len(current_children) == 2:\n",
    "                            genes_seen = refine_genes_seen(genome, genes_seen, current_children)\n",
    "                            current_children = check_unique_children_from_unit_block(genome, block[0].children, genes_seen)\n",
    "\n",
    "                        starting_children = current_children\n",
    "                        \n",
    "                    blocks_seen = [set() for child in current_children]  # The variable is restarted for the next iteration\n",
    "\n",
    "            if reaching_already_assessed:  \n",
    "                break\n",
    "                \n",
    "            if len(current_children) == 2:\n",
    "                if current_children[0].genome != current_children[1].genome:\n",
    "                    continue\n",
    "                \n",
    "                break  # To exit the while loop\n",
    "                \n",
    "            else:  \n",
    "                break  # To exit the while loop\n",
    "        \n",
    "        tmp = [[], []]\n",
    "        if extended_block:\n",
    "            for item in extended_block:\n",
    "                if isinstance(item, Link):\n",
    "                    tmp[1].append(item)\n",
    "\n",
    "                else:\n",
    "                    for element in item:\n",
    "                        tmp[1].append(element)\n",
    "        \n",
    "        else:\n",
    "            for element in block:\n",
    "                tmp[1].append(element)  # If a new initial block reach directly an already assessed block (i.e. common block), the initial block is stored\n",
    "                \n",
    "###############################################################################################################\n",
    "\n",
    "        counter = 0\n",
    "        children = [child for child in block2[0].children]\n",
    "        current_children = get_unique_children(genome, children)  # Remove duplicates from children\n",
    "        starting_children = current_children\n",
    "        \n",
    "        if len(current_children) == 1 or len(current_children) > 2:  # If missing information (==1) or other special case (>2), the block is no longer taken into account\n",
    "            new_blocks.append(tmp[1])\n",
    "            continue\n",
    "        \n",
    "        blocks_seen = [set() for child in current_children]  # To save already assessed blocks and perform the intersection\n",
    "        genes_seen = []  # To keep track of already assessed genes (used for gene without parent to make to we go in one direction)\n",
    "        intersect = None\n",
    "        extended_block = []\n",
    "        next_children = [None, None]\n",
    "        genes_seen.append(current_children[0])  \n",
    "        genes_seen.append(current_children[1])  \n",
    "        \n",
    "        while counter < window:\n",
    "            none_type_as_neighbor = 0\n",
    "            block_found = False\n",
    "            reaching_already_assessed = False  \n",
    "            \n",
    "            if None in current_children:\n",
    "                break\n",
    "            \n",
    "            # Without the following block, when a common block is reached from one extremity, the next iteration \n",
    "            # starts from the opposite one. However, children originating from the that extreme element are not\n",
    "            # added within the list of genes seen.\n",
    "            if current_children[0] and current_children[1] not in genes_seen:\n",
    "                genes_seen.append(current_children[0])\n",
    "                genes_seen.append(current_children[1])\n",
    "            \n",
    "            for i in range(len(current_children)):\n",
    "                \n",
    "                for neighbor in current_children[i].neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        none_type_as_neighbor += 1\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor.parent:  # Check whether the neighbor is a singleton (i.e. does it have a parent or not?)\n",
    "                        \n",
    "                        if neighbor.parent in block2 or neighbor in genes_seen:  # Avoid to take into account elements from the initial block or those already assessed\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            blocks_seen[i].add(tuple(neighbor.parent.blocks[genome.name]))\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region \n",
    "                            break  \n",
    "\n",
    "                    else:\n",
    "                        \n",
    "                        if neighbor in genes_seen:\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region\n",
    "                \n",
    "            if none_type_as_neighbor > 0:  # If children have only one neighbor\n",
    "                break\n",
    "\n",
    "            counter += 1\n",
    "            current_children = next_children[:]\n",
    "            intersect = list(blocks_seen[0].intersection(*blocks_seen))  # Intersection\n",
    "            \n",
    "            if len(intersect) == 1:  # ex. intersect = [('a', 'c')]\n",
    "\n",
    "                    if list(intersect[0]) in block_record:\n",
    "                        reaching_already_assessed = True\n",
    "                        break\n",
    "                        \n",
    "                    block_found = True  \n",
    "                    counter = 0  # Reset the counter to pursue the assessement with further blocks\n",
    "                    common_block = list(intersect[0])  # Extraction of the last common block. \n",
    "                    block_record.append(common_block)\n",
    "\n",
    "                    link = Link()  # Class for linking regions\n",
    "                    link.get_linking_region(genome, block2, common_block, starting_children)  # Creation of the linking region\n",
    "                    \n",
    "                    extended_block.append(link)  # The linking region\n",
    "                    extended_block.append(common_block)  # The common block\n",
    "\n",
    "                    block2 = common_block  # The new block becomes the last common block found within the while loop\n",
    "\n",
    "                    if len(block2) > 1:\n",
    "                        \n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its first element (i.e. element 0), the next iteration should start from its last element (i.e. element -1) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block2[0].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block2[0].children)[0] and get_unique_children(genome, block2[0].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block2[-1].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block2[0].children)\n",
    "                            break\n",
    "\n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its last element (i.e. element -1), the next iteration should start from its first element (i.e. element 0) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block2[-1].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block2[-1].children)[0] and get_unique_children(genome, block2[-1].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block2[0].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  # 05.01.18\n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)  # 05.01.18\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block2[-1].children)\n",
    "                            break\n",
    "                    \n",
    "                    ################################################################################################################\n",
    "                    # When we reach a common unit block, the next iteration should start from the same unique HOG (i.e. element 0) #\n",
    "                    ################################################################################################################\n",
    "\n",
    "                    elif len(block2) == 1:  # If the block has only one HOG...\n",
    "                        current_children = get_unique_children(genome, block2[0].children) \n",
    "                        \n",
    "                        if len(current_children) == 2:\n",
    "                            genes_seen = refine_genes_seen(genome, genes_seen, current_children)\n",
    "                            current_children = check_unique_children_from_unit_block(genome, block2[0].children, genes_seen)\n",
    "\n",
    "                        starting_children = current_children\n",
    "                        \n",
    "                    blocks_seen = [set() for child in current_children]  # The variable is restarted for the next iteration\n",
    "\n",
    "            if reaching_already_assessed:  \n",
    "                break\n",
    "                \n",
    "            if len(current_children) == 2:\n",
    "                if current_children[0].genome != current_children[1].genome:\n",
    "                    continue\n",
    "                \n",
    "                break  # To exit the while loop\n",
    "                \n",
    "            else:  \n",
    "                break  # To exit the while loop\n",
    "\n",
    "        if extended_block:\n",
    "            for item in extended_block:\n",
    "                if isinstance(item, Link):\n",
    "                    tmp[0].append(item)\n",
    "\n",
    "                else:\n",
    "                    for element in item:\n",
    "                        tmp[0].append(element)\n",
    "        \n",
    "        # In tmp[1] there is always something, namely either block or extended block.\n",
    "        # The list tmp[0] can be empty\n",
    "        if tmp[0] and tmp[1]:\n",
    "            tmp[0] = tmp[0][::-1]\n",
    "            new_blocks.append(tmp[0] + tmp[1])\n",
    "            \n",
    "        elif tmp[1] and not tmp[0]:\n",
    "            new_blocks.append(tmp[1])\n",
    "            \n",
    "    genome.new_synteny = new_blocks\n",
    "\n",
    "    add_remaining_blocks(genome)\n",
    "    \n",
    "    \n",
    "def add_remaining_blocks(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    Once the secondary reconstruction is done. Some blocks or single HOGs from the primary reconstruction may\n",
    "    remained unassessed. The function add such items within the collection of extended synteny blocks.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        \n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    c = 0\n",
    "    \n",
    "    for block in genome.new_synteny:\n",
    "        \n",
    "        for element in block:\n",
    "            \n",
    "            if not isinstance(element, pyham.HOG):\n",
    "                \n",
    "                for key, value in element.linking_region.items():\n",
    "                    \n",
    "                    for gene in value:\n",
    "                        \n",
    "                        if isinstance(gene, pyham.HOG):\n",
    "                            gene.processed = True\n",
    "                            c += 1\n",
    "\n",
    "            else:\n",
    "                element.processed = True\n",
    "                c += 1\n",
    "    \n",
    "    not_processed = 0\n",
    "    \n",
    "    for gene in genome.genes:\n",
    "        \n",
    "        if gene.processed == False and gene.blocks[genome.name] not in genome.new_synteny:\n",
    "            genome.new_synteny.append(gene.blocks[genome.name])\n",
    "            not_processed += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#nwk_path = \"working_perfect_3spp/RealTree.nwk\"\n",
    "nwk_path = \"working_inversion_3spp/RealTree.nwk\"\n",
    "\n",
    "# The newick tree is extracted as a string\n",
    "tree_str = pyham.utils.get_newick_string(nwk_path, type=\"nwk\")\n",
    "\n",
    "# Path to OrthoXML file\n",
    "#orthoxml_path = \"working_perfect_3spp/Output/HierarchicalGroups.orthoxml\"\n",
    "orthoxml_path = \"working_inversion_3spp/Output/HierarchicalGroups.orthoxml\"\n",
    "\n",
    "ham_analysis_no_name = pyham.Ham(tree_str, orthoxml_path, use_internal_name=False)\n",
    "\n",
    "#get_locus('working_perfect_3spp/DB')\n",
    "get_locus('working_inversion_3spp/DB')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "neighbors assigned\n",
      "[[Gene(1), Gene(2), Gene(3), Gene(4), Gene(5), Gene(6), Gene(7), Gene(8), Gene(9), Gene(10), Gene(11), Gene(12), Gene(13), Gene(14), Gene(15), Gene(16), Gene(17), Gene(18), Gene(19), Gene(20), Gene(21), Gene(22), Gene(23), Gene(24), Gene(25)]]\n",
      "next node\n",
      "neighbors assigned\n",
      "[[Gene(26), Gene(27), Gene(28), Gene(29), Gene(30), Gene(31), Gene(32), Gene(33), Gene(34), Gene(35), Gene(36), Gene(37), Gene(38), Gene(39), Gene(40), Gene(41), Gene(42), Gene(43), Gene(44), Gene(45), Gene(46), Gene(47), Gene(48), Gene(49), Gene(50)]]\n",
      "next node\n",
      "neighbors assigned\n",
      "[[Gene(51), Gene(52), Gene(53), Gene(54), Gene(55), Gene(56), Gene(57), Gene(58), Gene(59), Gene(60), Gene(61), Gene(62), Gene(63), Gene(64), Gene(65), Gene(66), Gene(67), Gene(68), Gene(69), Gene(70), Gene(71), Gene(72), Gene(73), Gene(74), Gene(75)]]\n",
      "next node\n",
      "fin (<HOG(i2)>, None)\n",
      "fin (<HOG(i1)>, <HOG(i3)>)\n",
      "fin (<HOG(i2)>, <HOG(i4)>)\n",
      "fin (<HOG(i5)>, <HOG(i3)>)\n",
      "fin (<HOG(i6)>, <HOG(i4)>)\n",
      "fin (<HOG(i5)>, <HOG(i7)>)\n",
      "fin (<HOG(i6)>, <HOG(i8)>)\n",
      "fin (<HOG(i9)>, <HOG(i7)>)\n",
      "fin (<HOG(i10)>, <HOG(i8)>)\n",
      "fin (<HOG(i11)>, <HOG(i9)>)\n",
      "fin (<HOG(i10)>, <HOG(i13)>)\n",
      "fin (<HOG(i14)>, <HOG(i13)>)\n",
      "fin (<HOG(i11)>, <HOG(i12)>)\n",
      "fin (<HOG(i15)>, <HOG(i12)>)\n",
      "fin (<HOG(i16)>, <HOG(i14)>)\n",
      "fin (<HOG(i17)>, <HOG(i15)>)\n",
      "fin (<HOG(i16)>, <HOG(i18)>)\n",
      "fin (<HOG(i17)>, <HOG(i19)>)\n",
      "fin (<HOG(i18)>, <HOG(i20)>)\n",
      "fin (<HOG(i19)>, <HOG(i21)>)\n",
      "fin (<HOG(i20)>, <HOG(i22)>)\n",
      "fin (<HOG(i21)>, <HOG(i23)>)\n",
      "fin (<HOG(i22)>, <HOG(i24)>)\n",
      "fin (<HOG(i23)>, <HOG(i25)>)\n",
      "fin (<HOG(i24)>, None)\n",
      "neighbors assigned\n",
      "[[<HOG(i25)>, <HOG(i24)>, <HOG(i23)>, <HOG(i22)>, <HOG(i21)>, <HOG(i20)>, <HOG(i19)>, <HOG(i18)>, <HOG(i17)>, <HOG(i16)>, <HOG(i15)>, <HOG(i14)>, <HOG(i12)>, <HOG(i13)>, <HOG(i11)>, <HOG(i10)>, <HOG(i9)>, <HOG(i8)>, <HOG(i7)>, <HOG(i6)>, <HOG(i5)>, <HOG(i4)>, <HOG(i3)>, <HOG(i2)>, <HOG(i1)>]]\n",
      "next node\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEWCAYAAABrDZDcAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAeZ0lEQVR4nO3deZwdZZ3v8c83CyhkIyYikISghHEComALMiCLgrLIIqISVCZcrlEHlBlcBkdBBHcG9eKgkBFEEIPsN2IQlEvAhSWBQCCJ4QYMEgETlS1BtvCbP56nSXFy+vRJd9c53V3f9+tVr66qp5ZfVXfXr+qpqqcUEZiZWXUNaXcAZmbWXk4EZmYV50RgZlZxTgRmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgpZO0h6TfSXpC0t8k/VbSWyRNl7RW0uqabss831hJV0laI+lBSUfVLPeoPH6NpKsljc3jN5Z0Xi57StICSQe0aFv/Q9JXC/1/yNu0QtJPC9PNlfRMzXb/rFD+Dkm/l/S0pBslbV0o21jS+ZKelPSopBMLZVMlzZf0WO5+JWlqK7bdBi4nAiuVpFHANcB3gbHAVsCXgGfzJLdExIia7uFcdjbwHLA58EHg+5K2z8vdHjgX+HAufxr4Xp5vGPAQsBcwGjgZuFTS5BI3tdOBwBxJ/5xj2zciRgAdwA010x5fs90HA0gaB1yZ4x4LzAd+WpjvVGAKsDWwD/BZSfvnsoeBI/J844DZwCV9vpU2uESEO3eldaQD4ONdlE0HftNF2aakJLBdYdxFwNdz/1eBnxTKXpenH9nF8hYC7+0m1s1ISWsV8Fjun1AoPwZYAjwFPAB8tM78K4GhwH8B32mwrrnA/+6ibAbwu5p98Xfg9Xn4T8A7C+WnA5fUWc4w4Djg6SZ+TwcBC4AnSUn01Jryy4BHgSeAm4Ht2/235a7vOl8RWNnuA9ZK+pGkAyRt1uR82wFrI+K+wri7ge1z//Z5GICIuJ+cOGoXJGnzPH5RN+scAvyQdKY9iXTw/a9C+Urg3cAoUlL4tqSdC+XvAm6IiLXArcDRkj4jqUPS0G7WXVS7bWuA+4Ht8/7bsljOy/cLAJIeB54hXYl9tYl1rgGOBsaQksLHJR1WKL+WdBXyauBO4OIN2B7r55wIrFQR8SSwBxDAfwOrJM3OB2eAt0p6vNDdn8ePIJ19Fj0BjGyyHABJw0kHrR9FxO+7ifWvEXFFRDwdEU8BXyFVL3WW/zwi7o/kJuB64G2FRRwEzMnT/hj4BCk53ASslHRSzSrPqtn205vYthGF4S63OyLGkKrFjied6TcUEXMj4p6IeDEiFgKzarb9/Ih4KiKeJVVNvVHS6O6WawODE4GVLiKWRMT0iJgA7EA6o/1OLr41IsYUutfl8atJZ95Fo0jVMs2UI2kIqTrpOdIBsSFJm0g6N99kfpJUBTKm82w+X9Hcmm94P066HzCusK79gF8UtvviiNiXdJb9MeA0Se8qrPKTNdt+chPbtrowXHe7C+tfA5wDXCjp1d1s+675pvQqSU/keDu3baikr0u6P++X5Xm2cY2WaQOHE4G1VD4rv4CUEBq5DxgmaUph3BtZV72zKA8DIOm1wMZ5PiQJOI90I/m9EfF8E+F9CvgHYNeIGAXsuW7x2hi4AvhPYPN8xj0HUJ7mLcDyiFhVZ5ufj4jLSPcputvuetu2KekeyKKIeAx4pFjOy/dLrSHAJqSb9I38hHRjeWJEjCYlkM5tOwo4FNiXdJUxuTO0JrbFBgAnAiuVpNdL+pSkCXl4IjCNVIfepXw2eyXpLHpTSbuTDkYX5UkuBg6W9LZ8oDwNuDJX6QB8H/hH4OCI+HuT4Y4k3Rd4PD+K+sVC2UakRLMKeCE/jvrOQvlL1UJ5O6dLOkjSSElD8vTbA7c1EcdVwA6S3ivpFcApwMJC1daFwBckbSbp9cBHSMkVSftJ2imfxY8CvkW68b2kiW3/W0Q8I2kX0sG/WPYs8FdSUmnmnoMNJO2+W+1ucHekM9FLSU+6rMk/zyVVZ0wH1pKqO4rdW/K8Y4Gr83x/BI6qWfZRefwa4P8CY/P4rUn3JJ6pWe4Hu4l1S9LTPKtJVxYfzcsZlsuPA/4MPE5KSJcAX85l84GOwrIOB35LOgg/CdwDTC+Uz60T3x2F8n2B35MS01xgcqFsY+D8vNw/AycWyt6X51tNSlpzgB2b+D0dATxIqmK6hnST/Me5bETev0/laY7O+2Xbdv99ueubTvkXbWY9lG983wVsGf6HsgHIVUNmvTeadFbuJGADkhOBVUpu9qG2SYvVkq7t6TIj4r6ImNWXcZZB0qIutv2D7Y7N2stVQ2ZmFTes3QFsqHHjxsXkyZPbHYaZ2YByxx13/CUixtcrG3CJYPLkycyfP7/dYZiZDSiSHuyqzPcIzMwqzonAzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4kpLBPnj2isl3dtFuSSdJWmZpIU1X3oyM7MWKfOK4AJg/wblB5A+fTeF9I3W75cYi5mZdaG0RBARNwN/azDJocCFkdxK+hLUFmXFY2Zm9bXzzeKtgIcKwyvyuEdqJ5Q0g3TVwKRJk3q8wskn/bzH85qZtdvyrx9UynLbebO43mfu6raAFxEzI6IjIjrGj6/bVIaZmfVQOxPBCmBiYXgC8HCbYjEzq6x2JoLZwNH56aG3Ak9ExHrVQmZmVq7S7hFImgXsDYyTtIL0IfDhABFxDulbqgcCy4CngWPKisXMzLpWWiKIiGndlAfpY+BmZtZGfrPYzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4pwIzMwqzonAzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4pwIzMwqzonAzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4pwIzMwqzonAzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4pwIzMwqzonAzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4pwIzMwqzonAzKzinAjMzCqu1EQgaX9JSyUtk3RSnfJJkm6UtEDSQkkHlhmPmZmtr7REIGkocDZwADAVmCZpas1kXwAujYidgCOB75UVj5mZ1VfmFcEuwLKIeCAingMuAQ6tmSaAUbl/NPBwifGYmVkdZSaCrYCHCsMr8riiU4EPSVoBzAE+UW9BkmZImi9p/qpVq8qI1cyssspMBKozLmqGpwEXRMQE4EDgIknrxRQRMyOiIyI6xo8fX0KoZmbVVWYiWAFMLAxPYP2qn2OBSwEi4hbgFcC4EmMyM7MaZSaCecAUSdtI2oh0M3h2zTR/BN4BIOkfSYnAdT9mZi1UWiKIiBeA44HrgCWkp4MWSTpN0iF5sk8BH5F0NzALmB4RtdVHZmZWomFlLjwi5pBuAhfHnVLoXwzsXmYMZmbWmN8sNjOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4rrNhFI+qakUZKGS7pB0l8kfagVwZmZWfmauSJ4Z0Q8Cbyb1LT0dsBnSo3KzMxapplEMDz/PBCYFRF/KzEeMzNrsWZaH/2ZpN8Dfwf+RdJ44JlywzIzs1bp9oogIk4CdgM6IuJ54GngkMZzmZnZQNHMzeLzI+KxiFjbOQr4SblhmZlZqzRzj+BPkr4PIGkz4Hrgx6VGZWZmLdNM1dDJwJOSziElgTMj4oelR2ZmZi3R5c1iSYcXBm8HTs4/Q9LhEXFl2cGZmVn5Gj01dHDN8ALSo6QHAwE4EZiZDQJdJoKIOKaVgZiZWXs089TQjySNKQxvJun8csMyM7NWaeapoR0j4vHOgYh4DNipvJDMzKyVmkkEQ/JjowBIGktzbySbmdkA0MwB/Uzgd5Iuz8PvA75SXkhmZtZK3SaCiLhQ0nzg7XnU4RGxuNywzMysVZr9MM1wUtMSnf1mZjZINPPU0AnAxcA44NXAjyV9ouzAzMysNZq5R3AssGtErAGQ9A3gFuC7ZQZmZmat0UzVkIC1heG1rKsmMjOzAa6ZK4IfArdJuioPHwacV15IZmbWSs08NfQtSXOBPUhXAsdExIKyAzMzs9bosmpI0tjODlhO+gbBRcCDeVy3JO0vaamkZZJO6mKa90taLGmRJH/wxsysxRpdEdxBamW03v2AAF7baMGShgJnA/sBK4B5kmYX30GQNAX4HLB7RDwm6dUbGL+ZmfVSo9ZHt+nlsncBlkXEAwCSLgEOBYovo30EODu3X0RErOzlOs3MbAM11WZQ/kjNHqQrgV9HxNVNzLYV8FBheAWwa8002+Xl/xYYCpwaEb9oJiYzM+sb3SYCSd8DtgVm5VEfk7RfRBzX3ax1xkWd9U8B9gYmAL+WtEOxtdMcwwxgBsCkSZO6C9nMzDZAM1cEewE7RERA+j4BcE8T860AJhaGJwAP15nm1oh4HviDpKWkxDCvOFFEzARmAnR0dNQmEzMz64VmXihbChRPwycCC5uYbx4wRdI2kjYCjgRm10xzNbAPgKRxpKqiB5pYtpmZ9ZFGH6//GakqZzSwRNLteXhX4HfdLTgiXpB0PHAdqf7//IhYJOk0YH5EzM5l75S0mPTG8mci4q+93SgzM2teo6qh/+ztwiNiDjCnZtwphf4ATsydmZm1QaPHR29qZSBmZtYezX6PwMzMBiknAjOzimvU1tAN+ec3WheOmZm1WqObxVtI2gs4JDcP8bIXxCLizlIjMzOzlmiUCE4BTiK9CPatmrJg3cfszcxsAGv01NDlwOWSTo6I01sYk5mZtVAzH6Y5XdIhwJ551NyIuKbcsMzMrFW6fWpI0teAE0jNRy8GTsjjzMxsEGim0bmDgDdFxIvwUqNzC0gflDEzswGu2fcIxhT6R5cRiJmZtUczVwRfAxZIupH0COme+GrAzGzQaOZm8SxJc4G3kBLBv0fEo2UHZmZmrdHUpyoj4hHW/5aAmZkNAm5ryMys4pwIzMwqrmEikDRE0r2tCsbMzFqvYSLI7w7cLWlSo+nMzGzgauZm8RbAovzN4jWdIyPikNKiMjOzlmkmEXyp9CjMzKxtmnmP4CZJWwNTIuJXkjYBhpYfmpmZtUIzjc59BLgcODeP2gq4usygzMysdZp5fPQ4YHfgSYCI+P/Aq8sMyszMWqeZRPBsRDzXOSBpGOkLZWZmNgg0kwhukvQfwCsl7QdcBvys3LDMzKxVmkkEJwGrgHuAjwJzgC+UGZSZmbVOM08NvZg/RnMbqUpoaUS4asjMbJDoNhFIOgg4B7if1Az1NpI+GhHXlh2cmZmVr5kXys4E9omIZQCSXgf8HHAiMDMbBJq5R7CyMwlkDwArS4rHzMxarMsrAkmH595FkuYAl5LuEbwPmNeC2MzMrAUaVQ0dXOj/M7BX7l8FbFZaRGZm1lJdJoKIOKaVgZiZWXs009bQNpK+JelKSbM7u2YWLml/SUslLZN0UoPpjpAUkjo2JHgzM+u9Zp4auho4j/Q28YvNLljSUOBsYD9gBTBP0uyIWFwz3Ujgk6T3FMzMrMWaSQTPRMRZPVj2LsCyiHgAQNIlwKHA4prpTge+CXy6B+swM7Neaubx0f8j6YuSdpO0c2fXxHxbAQ8VhlfkcS+RtBMwMSKuabQgSTMkzZc0f9WqVU2s2szMmtXMFcEbgA8Db2dd1VDk4UZUZ9xLTVNIGgJ8G5jeXQARMROYCdDR0eHmLczM+lAzieA9wGuLTVE3aQUwsTA8AXi4MDwS2AGYKwngNcBsSYdExPwNXJeZmfVQM1VDdwNjerDsecCU/NTRRsCRwEtPG0XEExExLiImR8Rk4FbAScDMrMWauSLYHPi9pHnAs50jI+KQRjNFxAuSjgeuI33j+PyIWCTpNGB+RDT1CKqZmZWrmUTwxZ4uPCLmkL5fUBx3ShfT7t3T9ZiZWc818z2Cm1oRiJmZtUcz3yN4inVP+2wEDAfWRMSoMgMzM7PWaOaKYGRxWNJhpJfFzMxsEGjmqaGXiYir6f4dAjMzGyCaqRo6vDA4BOig8GKYmZkNbM08NVT8LsELwHJSm0FmZjYINHOPwN8lMDMbxBp9qrLu8/5ZRMTpJcRjZmYt1uiKYE2dcZsCxwKvIjUfbWZmA1yjT1We2dmfPx5zAnAMcAlwZlfzmZnZwNLwHoGkscCJwAeBHwE7R8RjrQjMzMxao9E9gjOAw0nfAXhDRKxuWVRmZtYyjV4o+xSwJfAF4GFJT+buKUlPtiY8MzMrW6N7BBv81rGZmQ08PtibmVWcE4GZWcU5EZiZVZwTgZlZxTkRmJlVnBOBmVnFORGYmVWcE4GZWcU5EZiZVZwTgZlZxTkRmJlVnBOBmVnFORGYmVWcE4GZWcU5EZiZVZwTgZlZxTkRmJlVnBOBmVnFlZoIJO0vaamkZZJOqlN+oqTFkhZKukHS1mXGY2Zm6ystEUgaCpwNHABMBaZJmloz2QKgIyJ2BC4HvllWPGZmVl+ZVwS7AMsi4oGIeA64BDi0OEFE3BgRT+fBW4EJJcZjZmZ1lJkItgIeKgyvyOO6cixwbb0CSTMkzZc0f9WqVX0YopmZlZkIVGdc1J1Q+hDQAZxRrzwiZkZER0R0jB8/vg9DNDOzYSUuewUwsTA8AXi4diJJ+wKfB/aKiGdLjMfMzOoo84pgHjBF0jaSNgKOBGYXJ5C0E3AucEhErCwxFjMz60JpiSAiXgCOB64DlgCXRsQiSadJOiRPdgYwArhM0l2SZnexODMzK0mZVUNExBxgTs24Uwr9+5a5fjMz657fLDYzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOruFITgaT9JS2VtEzSSXXKN5b001x+m6TJZcZjZmbrKy0RSBoKnA0cAEwFpkmaWjPZscBjEbEt8G3gG2XFY2Zm9ZV5RbALsCwiHoiI54BLgENrpjkU+FHuvxx4hySVGJOZmdUYVuKytwIeKgyvAHbtapqIeEHSE8CrgL8UJ5I0A5iRB1dLWtrDmMbVLrufcFwbxnFtuP4am+PaAPpGr+LauquCMhNBvTP76ME0RMRMYGavA5LmR0RHb5fT1xzXhnFcG66/xua4NkxZcZVZNbQCmFgYngA83NU0koYBo4G/lRiTmZnVKDMRzAOmSNpG0kbAkcDsmmlmA/+c+48A/l9ErHdFYGZm5SmtaijX+R8PXAcMBc6PiEWSTgPmR8Rs4DzgIknLSFcCR5YVT9br6qWSOK4N47g2XH+NzXFtmFLikk/AzcyqzW8Wm5lVnBOBmVnFDYpEIGmipBslLZG0SNIJNeWflhSSxnUx/yRJ1+f5F/dVUxe9iUvSPpLuKnTPSDqs3XHl8m/m+ZZIOquvXgLsg7i+Iene3H2gL2JqFJekUyX9qfA7OrCL+Rs2tdLGuM6XtFLSvX0VU2/j6u5voI1xvULS7ZLuzvN+qT/EVVjGUEkLJF3ToyAiYsB3wBbAzrl/JHAfMDUPTyTdsH4QGNfF/HOB/XL/CGCT/hBXYTljSTfT2x4X8E/Ab0kPAAwFbgH27gdxHQT8kvQAxKbAfGBUmXEBpwKf7mbeocD9wGuBjYC7O7epnXHlefYEdgbu7Yt4+mh/dfk30Oa4BIzI/cOB24C3tjuuwjJOBH4CXNOTGAbFFUFEPBIRd+b+p4AlpLeWIbVh9FnqvKgGoNT+0bCI+GWef3VEPN3uuGocAVzbT+IK4BWkg9rGpH+KP/eDuKYCN0XECxGxhnTA3b8FcXWnmaZW2hEXEXEzJby305u4ertNJcYVEbE6Dw7PXZ88adPbbZY0gXQi9IOexjAoEkFRrtbZCbhN0iHAnyLi7gazbAc8LunKfGl1hlKDee2Oq+hIYFZfx9STuCLiFuBG4JHcXRcRS9odF+nAf4CkTXLV0T68/IXGPo8rjzpe0sJczbJZnVnqNbXSJwe2XsbVEr2Jq868bY0rV7/cBawEfhkR/SIu4Dukk6QXe7zivri06S8dqVrnDuBwYJO8M0fnsuXUr1I4AniCdOk+DLgCOLbdcRXm3QJYBQzvJ/trW+Dned4RpKqhPdsdVy77PHAXqYroYuCEsuLKw5uTqn6GAF8hvStTO8/7gB8Uhj8MfLfdcRXmnUwfVw31UVwvm7e/xJWnH0M6Gdqh3XEB7wa+l/v3podVQ33+y29XR7pUuw44MQ+/gZS5l+fuBeCPwGtq5nsrMLcw/GHg7HbHVZj/BGBmP9pfnwFOLgyfAny23XHVWc5PgAPLiqtOed0DKrAb6aqpc/hzwOfaHVez5e2Iq7t527m/CtN9kSbr70v++/oa6SpzOfAo8DTw4w1ef1/v6HZ0pBs5FwLfaTDNcuqf4Q4lVSuMz8M/BI5rd1yF8luBffrR/voA8CvS1dNw4Abg4H4Q11DgVbl/R+Be0r2f0uICtij0/xtwSZ15hwEPANuw7mbx9u2Oq1De54mgl/ur27+BNsU1HhiT+18J/Bp4d7vjqpl+b6p8RQDsQbpxs5BUNXAXNWeDxQMI0MHLL9f3y/PeA1wAbNRP4poM/AkY0l/2F+mAey7phtZi4Fv9JK5X5HgWk5Lnm8qOC7go/80sJLWbtUWefktgTmH+A0lPgtwPfL4fxTWLdJ/nedJZZZ9UifYmrmb+BtoU147AgjzNvcAp/eX3WFjO3vQwEbiJCTOziht0Tw2ZmdmGcSIwM6s4JwIzs4pzIjAzqzgnAjOzinMisNJJek9uNfT1hXF7b0hLiZKmS9qyj+IZI+lf+mJZ/VWj/StpeVctuHazzAskHdH76Ky/cSKwVpgG/IbefYp0Oun56fX0oG2oMcCgSgSSSvvsrA1+TgRWKkkjgN2BY1k/EYySdJXSNyDOkTQkN+x1gdI3Be6R9G/5LLQDuDi3y/7KfFZ7iqTfAO+T9BFJ83J78VdI2iSvf/O8jrtz90/A14HX5WWdkdf7vdwW/DWS5nSe+Up6s6SbJN0h6TpJW+Txc5W+f3C7pPskvS2PH5qXOS83FvbRPH4LSTfndd7bOX3NvlpeWObtkrbN48fnbZqXu93z+FMlzZR0PenN1Frr7d866zxR677h8K+F8Ufn+O+WdFGd+U7PvycfQwaDvno7zp27eh3wIeC83P871rW7vjfwDKmxv6GkhuKOAN5Matmxc/7O1/rnAh2F8csptHFEbl4i938Z+ETu/ynwr7l/KDCammYV8nrnkE6MXgM8lscNzzF3Nj/yAXLDXzmeM3P/gcCvcv8M4Au5f2PSdxG2AT5Ffqs4xzGyzr5aXpjmaPJboqR2k/bI/ZOAJbn/VFIjZa+ss6y6+7ewnnF5X99D+n7DCGARqeXL7YGlrHuDe2z+eUHeL98kvV2udv99ueubzpeTVrZppGZyIbXFPw24Mw/fHhEPAEiaRXrV/gbgtZK+S2rl9PoGy/5poX8HSV8mVfuMIDXgBfB20kGViFgLPFGnOd89gMsi4kXgUUk35vH/AOwA/FLpI2xDSU0ydLoy/7yDlFwA3gnsWKhLHw1MAeYB50saDlwdEXd1sU2zCj+/nfv3BaZq3YfgRkkamftnR8Tfu1hWvf17ec12XxXp+w1IuhJ4G6m5g8sj4i8AEVH8ZsHJwG0RMaOLddoA5ERgpZH0KtKBeAdJQTqQhqTP5klq2zeJiHhM0huBdwHHAe8H/lcXq1hT6L8AOCwi7pY0nXRG3HSoDcYviojduih/Nv9cy7r/JZGuRq6rnVjSnqQPiFwk6YyIqFedE3X6hwC71R7wc2Io7oNGy6o33Gi7u2p7Zh7wZkljaxKEDWCu37MyHQFcGBFbR8TkiJgI/IF0Jgqwi6Rtcj3zB4Df5KdZhkTEFaSzz53ztE+RPuPXlZHAI/mM+4OF8TcAH4eX6u9H1VnWb4D35nsFm7MuiSwFxkvaLc8/XNL23WzzdcDHcxxI2k7SppK2BlZGxH8D5xW2q9YHCj9vyf3XA8d3TiDpTd3E0Gm9/VtTfjNwmNLHfDYF3kNqVfMG4P05kSNpbGGeX5Dusfy8cFViA5yvCKxM00gHjaIrgKNI1Tq35PI3kA5KV+X+HxZuQn4u/7wAOEfS30lt/Nc6mfQBmwdJ9d6dB6kTgJmSjiWduX88Im6R9Fulj7ZfC/w78A5Sq5L35eU8ERHP5SqesySNJv2/fIdUl96VH5Cqie5UOmVfBRxGSi6fkfQ8sJpcXVXHxpJuI52kTcvjPgmcLWlhjuFm4GMNYuhUb/++JCLulHQBcHtn7BGxAEDSV4CbJK0ltbo5vTDfZTkJzJZ0YIOqKRsg3PqoGenppohYnc+Cbwd2j4hHWxzDctIN8b+0cr1mviIwS66RNIb08ZjTW50EzNrJVwRmZhXnm8VmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV9z9bjXAkB1gpVAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fin (<HOG(2)>, None)\n",
      "fin (<HOG(1)>, <HOG(3)>)\n",
      "fin (<HOG(4)>, <HOG(2)>)\n",
      "fin (<HOG(5)>, <HOG(3)>)\n",
      "fin (<HOG(6)>, <HOG(4)>)\n",
      "fin (<HOG(5)>, <HOG(7)>)\n",
      "fin (<HOG(6)>, <HOG(8)>)\n",
      "fin (<HOG(9)>, <HOG(7)>)\n",
      "fin (<HOG(10)>, <HOG(8)>)\n",
      "fin (<HOG(9)>, <HOG(11)>)\n",
      "fin (<HOG(16)>, <HOG(14)>)\n",
      "fin (<HOG(17)>, <HOG(15)>)\n",
      "fin (<HOG(16)>, <HOG(18)>)\n",
      "fin (<HOG(17)>, <HOG(19)>)\n",
      "fin (<HOG(18)>, <HOG(20)>)\n",
      "fin (<HOG(19)>, <HOG(21)>)\n",
      "fin (<HOG(20)>, <HOG(22)>)\n",
      "fin (<HOG(21)>, <HOG(23)>)\n",
      "fin (<HOG(22)>, <HOG(24)>)\n",
      "fin (<HOG(23)>, <HOG(25)>)\n",
      "fin (<HOG(24)>, None)\n",
      "neighbors assigned\n",
      "[[<HOG(11)>, <HOG(10)>, <HOG(9)>, <HOG(8)>, <HOG(7)>, <HOG(6)>, <HOG(5)>, <HOG(4)>, <HOG(3)>, <HOG(2)>, <HOG(1)>], [<HOG(13)>, <HOG(12)>], [<HOG(25)>, <HOG(24)>, <HOG(23)>, <HOG(22)>, <HOG(21)>, <HOG(20)>, <HOG(19)>, <HOG(18)>, <HOG(17)>, <HOG(16)>, <HOG(15)>, <HOG(14)>]]\n",
      "next node\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEWCAYAAABrDZDcAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3dabgdVZn28f+dBFAgIwk0JEAAgwqIgAHkZVQQGTQgogLaGJo2iCLY0NJxAGnQboFGbBSE2ARkltmAQUBkUggkTIEwBgwSGRLGkDBJfN4Pax1S7Ox9TmWofTip+3dddZ2qWjU8VXuf/VStqlqliMDMzOqrV3cHYGZm3cuJwMys5pwIzMxqzonAzKzmnAjMzGrOicDMrOacCMzMas6JwMys5pwIakDSNpJuk/SKpBcl/VnS5pJGS5ovaW5Dt0aeb5CkKyTNk/SkpP0alrtfHj9P0pWSBhXKDpE0RdKbks5u47Z+T9J/Ffr/krdppqTfFKa7SdIbDdt9VaF8R0kPS3pN0o2S1i6UrSBpvKQ5kp6VdHih7OOSrs/7ebakSySt3qZtf1TS+pKGSbpM0vP5M79f0ug8zXBJ0eQz/1Iul6TjJb2QuxMkqbCOTSTdlffLXZI2KZR9W9ITeb88LelkSX3ase22ZJwIlnGS+gFXAz8HBgFDgf8E3syT3B4RKzd0T+eyU4G3gNWALwO/lLRhXu6GwBnAP+fy14DTCqt+GvgRML7K7WtiN2CipK/m2HaKiJWBkcANDdMe0rDdnwWQNBi4HDiKtM+mAL8pzHcMMAJYG/gEcKSkXXLZQGAcMDyXvwqctbQ3spGk9YBeEfEocC7wVF7/KsD+wHMNswxo2PaO7RsD7Al8FNgY+AxwUF7H8sBvgfNI2/lr4Ld5PMBVwGYR0Q/YKC/j0Cq215ayiHC3DHekH8CXW5SNBv7UomwlUhJYvzDuXOAnuf+/gAsKZevl6fs2LOdHwNklYx1ISlqzgZdy/7BC+QHAQ6Qf1yeAg5rMPwvoDfwC+Fkn67oJ+NcWZWOA2xr2xevAh/Lw34CdC+XHARe1WNZmwKsltn134B5gDulH/JiG8kuAZ4FXgFuADRvKDwVOyf1zgU1arGc4EECfFuW3AWMKwwcCk3L/znnbVSj/K7BLk+WsAvwBOK3Etv9v3uY5wF3AtoWyLYDbgZeBZ/Lnunx3/k8ti53PCJZ9jwLzJf1a0q6SBpacb31gfqQjzA73ARvm/g3zMAAR8Tg5cSxBrL1IR89rA2uRfnx/USifRTpC7UdKCidL2qxQ/mnghoiYD0wC9pf0HUkjJfVehDgat20e8DiwYd5/axTLefd+abQdMK3EOueRjtwHkJLCwZL2LJRfQzoLWRW4Gzi/Yf7dgN/l/knAqZL2kbRWiXUXvWvbWfgznxr5FzqbWijvqC6cAzxPOiM4o8Q6JwObkM6+LgAukfS+XDYf+DdgMLAVsCPwjUXcJuuCE8EyLiLmANuQjgJ/BcyWNEHSanmSj0t6udA9nsevTDr6LHoF6FuyfHFifSEiLouI1yLiVeDHwPaF8t9FxOOR3AxcB2xbWMTuwMQ87XnAt0jJ4WZglqSxDas8pWHbjyuxbSsXhhvL3kXSxsDRwHdKbPtNEXF/RPwjIqYCFzZs+/iIeDUi3iRVTX1UUv+8nhWBzfN2AnwBuJVUtfUXSfdK2rxhlc83bPuHW2z7K8DK+TpBl595RFwQqWpofeB0Fq6Sarbt5+XP/u2IOAlYAfhgLrsrIiblshmkxLJ9J4uzxeBEUAMR8VBEjI6IYaS62zWAn+XiSRExoNCtl8fPJR15F/UjVcuUKV9kklaUdEa+AD2HVAUyoONoPp/RTMoXYl8mHQUPzmW9gE8Bvy9s9/kRsRPpKPvrwLGSPl1Y5aEN235UiW2bWxhuud2SPkA6ij8sIm4tse1b5ovSsyW9kuPt2Lbekn4i6fG8X2bk2QbnvzuSqrLeyNv9UkSMjYgNSddv7gWuLF70BQY3bPtDLba9HzA3nwWU/swj4jHSmdBpjWVNtv0ISQ/lC9svA/0L276+pKvzRfk5pCrJwZ0tzxadE0HNRMTDwNmkhNCZR4E+kkYUxn2UBdUc0/IwAJLWJR3JFauSFtURpCPBLfNR5XYLFq8VgMuA/wFWi4gBpKP/jh+3zYEZETG7caER8feIuIRUjdHVdsPC27YS6RrItIh4iVRX/dHC9MX9gtIdRn8AjouIc0usD1KVyARgzYjoTzqa7ti2/YA9gJ1IP5LDO1aV/xarhd4lIp4n7bM1SFUvXXnXtrPwZ75xQ0LZmNZVX31I+60lSdsC/wF8ERiYP9dXWLBtvwQeBkbk78T3CmW2lDgRLOMkfSgfcQ3Lw2sC+5LqkVvK9eKXk46iV5K0NenHqOOH7Xzgs5K2zT+UxwKX5yodJPXJ9by9gd6S3lfiVsK+pOsCLyvdivrDQtnypEQzG3hb0q6ki5cd3qkWyusfLWl3SX0l9crTbwjc0UUMAFcAG0n6fN6Go0l14w/n8nOAH0gaKOlDwNdIyRVJQ4E/AqdGxOkl1lXc9hcj4g1JW5B+/ItlbwIvACuSjoqLdm3Y9uMlbZQ/g77AwcD0iHihRBznAIdLGqp0G/ERHdtGusA+HzhU6RbaQ/L4P+b1/qukVXP/BsB3WfhOrWbb/Tbpc+0j6WjefdbRl3QReW7e1weX2AZbVN19tdpdtR3pdtGLSXd7zMt/zyD9s40m/WPPbeg2z/MOAq7M8/0V2K9h2fvl8fNItxUOKpQdQ7ouUeyO6SLWNUg/NnNJZxYHUbjDBfgmqc75ZVJCugj4US6bAowsLGsv4M+ku4/mAPcDowvlNwFvNGz3XYXynUhHoq/naYcXylYg3RY7J8dzeKHshznmd+3TEp/T3sCTpGqWq0kXyc/LZSvn/ftqnmb/vI4PkM5wHmhY1s+Bx/K6Z+flfTiXDW8WX8c2kI62TwBezN0JvPsuoU1Jd/a8TrpovWmh7Ky8P+aRqq9OBN7XxXb3Bs7M+/IZ4Mg87065fLv8OcwlXfc4lhZ3urlb/E55Z5v1WPnC973AGlGzL7SkI0n1/Ud2dyzWc/mpP1sW9Ccd0dYqCWQzSA9ymS02nxFYW0n6HumCX6NbI2LXdsfTTpKmkZ6RaHRQRDQ+F7DMyBeEr2lWFumpb+tmTgRmZjXX46qGBg8eHMOHD+/uMMzMepS77rrr+YgY0qysxyWC4cOHM2XKlO4Ow8ysR5H0ZKsyP0dgZlZzTgRmZjXnRGBmVnNOBGZmNedEYGZWc04EZmY1V1kiUHq59yxJD7Qol6RTJE2XNLXhTVNmZtYmVZ4RnA3s0kn5rqRX740gvSP2lxXGYmZmLVSWCCLiFlIztq3sAZwTySTSm6hWryoeMzNrrjufLB4KPFUYnpnHPdM4oaQxpLMG1lprUd/FvcDwsU1f4mQGwIyf7N7dIfg7ap2q6jvanReLm71urmkLeBExLiJGRsTIIUOaNpVhZmaLqTsTwUxgzcLwMODpborFzKy2ujMRTAD2z3cPfRx4JSIWqhYyM7NqVXaNQNKFwA7AYEkzSe9yXQ4g0ku9JwK7AdOB14ADqorFzMxaqywRRMS+XZQH6WXkZmbWjfxksZlZzTkRmJnVnBOBmVnNORGYmdWcE4GZWc05EZiZ1ZwTgZlZzTkRmJnVnBOBmVnNORGYmdWcE4GZWc05EZiZ1ZwTgZlZzTkRmJnVnBOBmVnNORGYmdWcE4GZWc05EZiZ1ZwTgZlZzTkRmJnVnBOBmVnNORGYmdWcE4GZWc05EZiZ1ZwTgZlZzTkRmJnVnBOBmVnNORGYmdWcE4GZWc05EZiZ1ZwTgZlZzTkRmJnVXKWJQNIukh6RNF3S2Cbla0m6UdI9kqZK2q3KeMzMbGGVJQJJvYFTgV2BDYB9JW3QMNkPgIsjYlNgH+C0quIxM7Pmqjwj2AKYHhFPRMRbwEXAHg3TBNAv9/cHnq4wHjMza6LKRDAUeKowPDOPKzoG+IqkmcBE4FvNFiRpjKQpkqbMnj27iljNzGqrykSgJuOiYXhf4OyIGAbsBpwraaGYImJcRIyMiJFDhgypIFQzs/qqMhHMBNYsDA9j4aqfA4GLASLiduB9wOAKYzIzswZVJoLJwAhJ60hannQxeELDNH8FdgSQ9GFSInDdj5lZG1WWCCLibeAQ4FrgIdLdQdMkHStpVJ7sCOBrku4DLgRGR0Rj9ZGZmVWoT5ULj4iJpIvAxXFHF/ofBLauMgYzM+ucnyw2M6s5JwIzs5pzIjAzqzknAjOzmnMiMDOrOScCM7OacyIwM6s5JwIzs5pzIjAzqzknAjOzmusyEUg6QVI/SctJukHS85K+0o7gzMysemXOCHaOiDnAZ0hNS68PfKfSqMzMrG3KJILl8t/dgAsj4sUK4zEzszYr0/roVZIeBl4HviFpCPBGtWGZmVm7dHlGEBFjga2AkRHxd+A1YFTnc5mZWU9R5mLx+Ih4KSLmd4wCLqg2LDMza5cy1wj+JumXAJIGAtcB51UalZmZtU2ZqqGjgDmSTiclgZMi4qzKIzMzs7ZoebFY0l6FwTuBo/LfkLRXRFxedXBmZla9zu4a+mzD8D2kW0k/CwTgRGBmtgxomQgi4oB2BmJmZt2jzF1Dv5Y0oDA8UNL4asMyM7N2KXPX0MYR8XLHQES8BGxaXUhmZtZOZRJBr3zbKACSBlHuiWQzM+sByvygnwTcJunSPPwF4MfVhWRmZu3UZSKIiHMkTQE+mUftFREPVhuWmZm1S9kX0yxHalqio9/MzJYRZe4aOgw4HxgMrAqcJ+lbVQdmZmbtUeYawYHAlhExD0DS8cDtwM+rDMzMzNqjTNWQgPmF4fksqCYyM7MerswZwVnAHZKuyMN7AmdWF5KZmbVTmbuGfirpJmAb0pnAARFxT9WBmZlZe7SsGpI0qKMDZpDeQXAu8GQe1yVJu0h6RNJ0SWNbTPNFSQ9KmibJL7wxM2uzzs4I7iK1MtrsekAA63a2YEm9gVOBTwEzgcmSJhSfQZA0AvgusHVEvCRp1UWM38zMllBnrY+us4TL3gKYHhFPAEi6CNgDKD6M9jXg1Nx+ERExawnXaWZmi6hUm0H5JTXbkM4Ebo2IK0vMNhR4qjA8E9iyYZr18/L/DPQGjomI35eJyczMlo4uE4Gk04APABfmUV+X9KmI+GZXszYZF03WPwLYARgG3Cppo2JrpzmGMcAYgLXWWqurkM3MbBGUOSPYHtgoIgLS+wmA+0vMNxNYszA8DHi6yTSTIuLvwF8kPUJKDJOLE0XEOGAcwMiRIxuTiZmZLYEyD5Q9AhQPw9cEppaYbzIwQtI6kpYH9gEmNExzJfAJAEmDSVVFT5RYtpmZLSWdvbz+KlJVTn/gIUl35uEtgdu6WnBEvC3pEOBaUv3/+IiYJulYYEpETMhlO0t6kPTE8nci4oUl3SgzMyuvs6qh/1nShUfERGBiw7ijC/0BHJ47MzPrBp3dPnpzOwMxM7PuUfZ9BGZmtoxyIjAzq7nO2hq6If89vn3hmJlZu3V2sXh1SdsDo3LzEO96QCwi7q40MjMza4vOEsHRwFjSg2A/bSgLFrzM3szMerDO7hq6FLhU0lERcVwbYzIzszYq82Ka4ySNArbLo26KiKurDcvMzNqly7uGJP03cBip+egHgcPyODMzWwaUaXRud2CTiPgHvNPo3D2kF8qYmVkPV/Y5ggGF/v5VBGJmZt2jzBnBfwP3SLqRdAvpdvhswMxsmVHmYvGFkm4CNiclgv+IiGerDszMzNqj1KsqI+IZFn6XgJmZLQPc1pCZWc05EZiZ1VyniUBSL0kPtCsYMzNrv04TQX524D5Ja3U2nZmZ9VxlLhavDkzL7yye1zEyIkZVFpWZmbVNmUTwn5VHYWZm3abMcwQ3S1obGBERf5C0ItC7+tDMzKwdyjQ69zXgUuCMPGoocGWVQZmZWfuUuX30m8DWwByAiHgMWLXKoMzMrH3KJII3I+KtjgFJfUhvKDMzs2VAmURws6TvAe+X9CngEuCqasMyM7N2KZMIxgKzgfuBg4CJwA+qDMrMzNqnzF1D/8gvo7mDVCX0SES4asjMbBnRZSKQtDtwOvA4qRnqdSQdFBHXVB2cmZlVr8wDZScBn4iI6QCS1gN+BzgRmJktA8pcI5jVkQSyJ4BZFcVjZmZt1vKMQNJeuXeapInAxaRrBF8AJrchNjMza4POqoY+W+h/Dtg+988GBlYWkZmZtVXLRBARB7QzEDMz6x5l2hpaR9JPJV0uaUJHV2bhknaR9Iik6ZLGdjLd3pJC0shFCd7MzJZcmbuGrgTOJD1N/I+yC5bUGzgV+BQwE5gsaUJEPNgwXV/gUNJzCmZm1mZlEsEbEXHKYix7C2B6RDwBIOkiYA/gwYbpjgNOAP59MdZhZmZLqMzto/8r6YeStpK0WUdXYr6hwFOF4Zl53DskbQqsGRFXd7YgSWMkTZE0Zfbs2SVWbWZmZZU5I/gI8M/AJ1lQNRR5uDNqMu6dpikk9QJOBkZ3FUBEjAPGAYwcOdLNW5iZLUVlEsHngHWLTVGXNBNYszA8DHi6MNwX2Ai4SRLAPwETJI2KiCmLuC4zM1tMZaqG7gMGLMayJwMj8l1HywP7AO/cbRQRr0TE4IgYHhHDgUmAk4CZWZuVOSNYDXhY0mTgzY6RETGqs5ki4m1JhwDXkt5xPD4ipkk6FpgSEaVuQTUzs2qVSQQ/XNyFR8RE0vsLiuOObjHtDou7HjMzW3xl3kdwczsCMTOz7lHmfQSvsuBun+WB5YB5EdGvysDMzKw9ypwR9C0OS9qT9LCYmZktA8rcNfQuEXElXT9DYGZmPUSZqqG9CoO9gJEUHgwzM7OercxdQ8X3ErwNzCC1GWRmZsuAMtcI/F4CM7NlWGevqmx6v38WEXFcBfGYmVmbdXZGMK/JuJWAA4FVSM1Hm5lZD9fZqypP6ujPL485DDgAuAg4qdV8ZmbWs3R6jUDSIOBw4MvAr4HNIuKldgRmZmbt0dk1ghOBvUjvAfhIRMxtW1RmZtY2nT1QdgSwBvAD4GlJc3L3qqQ57QnPzMyq1tk1gkV+6tjMzHoe/9ibmdWcE4GZWc05EZiZ1ZwTgZlZzTkRmJnVnBOBmVnNORGYmdWcE4GZWc05EZiZ1ZwTgZlZzTkRmJnVnBOBmVnNORGYmdWcE4GZWc05EZiZ1ZwTgZlZzTkRmJnVnBOBmVnNVZoIJO0i6RFJ0yWNbVJ+uKQHJU2VdIOktauMx8zMFlZZIpDUGzgV2BXYANhX0gYNk90DjIyIjYFLgROqisfMzJqr8oxgC2B6RDwREW8BFwF7FCeIiBsj4rU8OAkYVmE8ZmbWRJWJYCjwVGF4Zh7XyoHANc0KJI2RNEXSlNmzZy/FEM3MrMpEoCbjoumE0leAkcCJzcojYlxEjIyIkUOGDFmKIZqZWZ8Klz0TWLMwPAx4unEiSTsB3we2j4g3K4zHzMyaqPKMYDIwQtI6kpYH9gEmFCeQtClwBjAqImZVGIuZmbVQWSKIiLeBQ4BrgYeAiyNimqRjJY3Kk50IrAxcIuleSRNaLM7MzCpSZdUQETERmNgw7uhC/05Vrt/MzLrmJ4vNzGrOicDMrOacCMzMas6JwMys5pwIzMxqzonAzKzmnAjMzGrOicDMrOacCMzMas6JwMys5pwIzMxqzonAzKzmnAjMzGrOicDMrOacCMzMas6JwMys5pwIzMxqzonAzKzmnAjMzGrOicDMrOacCMzMas6JwMys5pwIzMxqzonAzKzmnAjMzGrOicDMrOacCMzMas6JwMys5pwIzMxqzonAzKzmnAjMzGrOicDMrOacCMzMaq7SRCBpF0mPSJouaWyT8hUk/SaX3yFpeJXxmJnZwipLBJJ6A6cCuwIbAPtK2qBhsgOBlyLiA8DJwPFVxWNmZs1VeUawBTA9Ip6IiLeAi4A9GqbZA/h17r8U2FGSKozJzMwa9Klw2UOBpwrDM4EtW00TEW9LegVYBXi+OJGkMcCYPDhX0iOVRLz0DKZhG96jHGeBlvx8tKfsT+g5sTrOgiX8jq7dqqDKRNDsyD4WYxoiYhwwbmkE1Q6SpkTEyO6OoyuOc+nqKXFCz4nVcbZHlVVDM4E1C8PDgKdbTSOpD9AfeLHCmMzMrEGViWAyMELSOpKWB/YBJjRMMwH4au7fG/hjRCx0RmBmZtWprGoo1/kfAlwL9AbGR8Q0SccCUyJiAnAmcK6k6aQzgX2qiqfNeko1luNcunpKnNBzYnWcbSAfgJuZ1ZufLDYzqzknAjOzmnMi6ISk8ZJmSXqgMO4LkqZJ+oekpreLSfqgpHsL3RxJ385lx0j6W6FstwpjPVHSw5KmSrpC0oAW8zZtCiRf6L9D0mO5KZDluytOSWtKulHSQ3n/H1YoW+r7dAn35wxJ9+dYphTGD5J0fd6f10sa2F1xvoe+o8flOO+VdJ2kNVrM+9W83x6T9NXC+I/lfT1d0ilL44HUxY1T0iaSbs/fz6mSvlQoO1vSXwr7dJMljXOpigh3LTpgO2Az4IHCuA8DHwRuAkaWWEZv4Flg7Tx8DPDvbYp1Z6BP7j8eOL5FfI8D6wLLA/cBG+Syi4F9cv/pwMHdGOfqwGa5vy/waCHOpb5PFzfOXDYDGNxk/AnA2Nw/ttX87YrzPfId7VfoPxQ4vcl8g4An8t+BuX9gLrsT2Ir0TNI1wK7dGOf6wIjcvwbwDDAgD58N7L209+nS6nxG0ImIuIWG5xoi4qGIWJQnm3cEHo+IJ5dqcA1axHpdRLydByeRnuVo1LQpkHxk9UlS0x+QmgLZs7vijIhnIuLu3P8q8BDpyfRKLMH+7EyxSZVu3Z8NuvM7OqcwuBJNHigFPg1cHxEvRsRLwPXALpJWJ/1A3x7p1/YcqtunXcYZEY9GxGO5/2lgFjBkSeNpByeC6u0DXNgw7pB86jh+aVQPlPQvpCOmRs2aAhlKaurj5cIPSsf4qrWK8x1KrdRuCtxRGN3ufdpZnAFcJ+kupeZROqwWEc9ASmzAqhXHCCX2J938HZX0Y0lPAV8Gjm4ySavv6NDc3zi+u+IsTrsF6Qz78cLoH+d9erKkFaqKc3E4EVQo16mPAi4pjP4lsB6wCenU8aQ2xPF94G3g/GbFTcZFJ+Mr00WcHdOsDFwGfLtwlNbWfVoizq0jYjNSy7vflLRdlfG0UnJ/dvt3NCK+HxFr5jgPaRZms9k6GV+JEnECkM9UzgUOiIh/5NHfBT4EbE6q4vqPquJcHE4E1doVuDsinusYERHPRcT8/AX5FalqpjL5wtpngC/n0+dGrZoCeR4YoNT0R3F8d8WJpOVISeD8iLi8Y3w792mZOHO1ABExC7iiEM9z+Uei48diVnfGmXX7d7TgAuDzTca3+o7O5N3VXpV+RwtaxYmkfsDvgB9ExKSO8blqMyLiTeAs2rdPS3EiqNa+NJxyd/wQZJ8DHqAiknYhHXmMiojXWkzWtCmQ/ONxI6npD0hNgfy2u+LM1yzOBB6KiJ82lLVln5aMcyVJfTv6SRduO+IpNqnSrfuzoLu/oyMKg6OAh5tMdi2ws6SBuZpqZ+DaXL32qqSP5+/H/lS3T7uMM///XAGcExGXNJR1HACIdB2jsn26WLr7avV7uSP9gzwD/J109HEg6R9jJvAm8BzpCwnpLoGJhXlXBF4A+jcs81zgfmAq6Ydh9QpjnU6qW703d6e3iHU30l04jwPfL4xfl3RXxnRS1cEK3RUnsA3ptH9qYbrdqtqnSxDnuqQ7r+4DpjXsz1WAG4DH8t9B3fy5vxe+o5eRfhSnAlcBQ/O0I4H/K8z7L3m7ppOqXChM90D+7v6C3FpCd8QJfCXPc2+h2ySX/THv0weA84CVl8Y+XVqdm5gwM6s5Vw2ZmdWcE4GZWc05EZiZ1ZwTgZlZzTkRmJnVnBOBVU7S5ySFpA8Vxu0g6epFWMboVi1TLkY8AyR9Y2ks672qs/2r1Drq4MVY5tmS9u56SutpnAisHfYF/sSSvYp0NOk++IVI6r2IyxoALFOJoPAEuNkicyKwSuW2gbYmPZTTmAj6KbWX/6Ck0yX1ktQ7H3k+kNuZ/7d8FDoSOD+35f7+fFR7tKQ/AV+Q9DVJkyXdJ+kySSvm9a+W13Ff7v4f8BNgvbysE/N6T8vtyF8taWLHka9Se/c35wbkri08IXqTpOMl3SnpUUnb5vG98zIn5wbGDsrjV5d0S17nAx3TN+yrGYVl3inpA3n8kLxNk3O3dR5/jKRxkq4jtbzZaKH922Sdh+d4HlB+H0Eev3+O/z5J5zaZ77j8Ofk3ZFnQ3U+0uVu2O9LTlmfm/ttY8E6BHYA3SE/i9iY1Lbw38DFSk8Md83e0534Thfc/kNr8P7IwvEqh/0fAt3L/b0gN1JHX0x8Yzrvbmt8bmEg6MPon4KU8brkc85A83ZeA8YV4Tsr9uwF/yP1jSO3MAKwATAHWAY4gP2Wc4+jbZF/NKEyzP3B17r8A2Cb3r0VqZgPSewPuAt7fZFlN929hPYPzvr6f1KzyyqQnoTcFNgQeIb9TgfwENLlNfdJ7Fc5gKTzF6+690fl00qq2L/Cz3H9RHr47D98ZEU8ASLqQ1IzEDcC6kn5Oarzruk6W/ZtC/0aSfkSq9lmZ1D4NpHcq7A8QEfOBV7Rws8rbAJdEamTtWUk35vEfBDYCrk9NxNCb1PRAh46G7+4iJRdI7eBsXKhL7w+MILXpNF6p4bwrI+LeFtt0YeHvybl/J2ADLXj5Vr+O9oxI7UK93mJZzfbvpYXybYArImJenuZyYFtSUx6XRsTzABFRbJv/KOCOiCg2r209nBOBVUbSKqQf4o0kBemHNCQdmSdpbN8kIuIlSR8lvYzkm8AXSe3MNDOv0H82sGdE3CdpNOmIuHSonYyfFhFbtSh/M/+dz4L/JZHORq5tnFipOerdgXMlnRgRzapzokl/L2Crxh/8nBiK+6CzZTUb7my7W7U9Mxn4mKRBDQnCejDX71mV9ia1xLh2RAyP1Jb7X0hHogBbKLV62otU7fKnfDdLr4i4jHT0uVme9lXSKypb6Qs8k4+4v1wYfwNwMLxTf9+vybL+BHw+XytYjQVJ5BFgiBhuob8AAAFzSURBVKSt8vzLSdqwi22+Fjg4x4Gk9ZVaJF0bmBURvyK1orpZi/m/VPh7e+6/jkL79yr/vtuF9m9D+S3AnpJWVGop9XPAraR99sWcyJE0qDDP70nXWH5XOCuxHs5nBFalfUk/GkWXAfuRqnVuz+UfIf0oXZH7zypchPxu/ns2cLqk10nvqG10FOmNZU+S6r07fqQOA8ZJOpB05H5wRNwu6c9KLye/htRk846kliEfzct5JSLeylU8p0jqT/p/+RmpLr2V/yNVE92tdMg+m9Ts8A7AdyT9HZhLrq5qYgVJd5AO0vbN4w4FTpU0NcdwC/D1TmLo0Gz/viMi7pZ0NqmFWUitaN4D6W1cwM2S5gP3kO7a6pjvkpwEJkjarZOqKesh3PqoGenupoiYm4+C7yS9ZezZNscwg3RB/Pl2rtfMZwRmydWSBpDeM3tcu5OAWXfyGYGZWc35YrGZWc05EZiZ1ZwTgZlZzTkRmJnVnBOBmVnN/X+ZRKsrF1uWFQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "counter = 1\n",
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    assign_neighbors(node.genome)\n",
    "    print(\"neighbors assigned\")\n",
    "    get_synteny_blocks(node.genome)\n",
    "    print(\"next node\")\n",
    "    assign_block_to_genes(node.genome)\n",
    "    \n",
    "    # here we go through all nodes, and if it is an ancestral, note, we make some plots\n",
    "    # in my 3 species example, there are 2 ancestral nodes, so two plots\n",
    "    \n",
    "    if isinstance(node.genome, pyham.AncestralGenome):\n",
    "        # links together the primary reconstruction's synteny blocks, using windows of 20\n",
    "        joining_blocks(20, node.genome)\n",
    "                \n",
    "        ### Plot ###        \n",
    "        title = node.genome.name  \n",
    "        raw_values = []\n",
    "        for block in node.genome.new_synteny:\n",
    "            count = 0\n",
    "            \n",
    "            for element in block:\n",
    "                # goes through the elements of that block, the HOGs         \n",
    "                # if that element (HOG) is a pyham.HOG, increase count, it's the number of HOGs we've alignd the synteny of in this block\n",
    "                if isinstance(element, pyham.HOG):\n",
    "                    count += 1  \n",
    "                    \n",
    "            raw_values.append(count)\n",
    "            \n",
    "        label = list(set(raw_values))\n",
    "        values = [0 for i in range(len(label))]\n",
    "        \n",
    "        for i in range(len(label)):\n",
    "            for element in raw_values:                \n",
    "                if label[i] == element:\n",
    "                    values[i] += 1\n",
    "\n",
    "        plt.bar(label, values)\n",
    "        plt.title(title)\n",
    "        plt.xlabel('Abstractgenes per block')\n",
    "        plt.ylabel('Number of blocks')\n",
    "        plt.savefig('results/secondary_reconstruction_plot{}.png'.format(counter))\n",
    "        counter += 1\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Graph Theory Approach"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make an empty graph with possibility of multiple edges per node\n",
    "G = nx.MultiGraph()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 269,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "G.number_of_nodes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 270,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "G.number_of_edges()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_synteny_graph(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function makes network graphs at the given level (genome) in the species tree.\n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        either a single graph for extant genes or multiple graphs for ancestral genes. \n",
    "        a graph for the HOGs and for each set of children from those HOGs at an ancestral node\n",
    "    \"\"\"\n",
    "    \n",
    "    if isinstance(genome, pyham.ExtantGenome):\n",
    "        # make the graphs\n",
    "        all_graphs = []\n",
    "        extant_gene_graph = nx.Graph()\n",
    "        parent_gene_graph = nx.Graph()\n",
    "        ##extant_gene_graph = nx.MultiGraph() # only if I needed multiple edges for the same pair of nodes\n",
    "        \n",
    "        extant_gene_graph.add_node(\"start\")\n",
    "        parent_gene_graph.add_node(\"start\")\n",
    "\n",
    "        old_node = \"start\"\n",
    "        old_parent_node = \"start\"\n",
    "        for extant_gene in genome.genes:            \n",
    "            new_node = (get_gene_by_locus(genome, int(extant_gene.locus)))\n",
    "            new_parent_node = new_node.parent\n",
    "                ##print(new_node, new_parent_node)\n",
    "                ##print(\"extant\", new_node)        \n",
    "            \n",
    "            # add the nodes to a graph\n",
    "            extant_gene_graph.add_node(new_node)\n",
    "            parent_gene_graph.add_node(new_parent_node)\n",
    "                ##print(\"node.number\", extant_gene_graph.number_of_nodes())\n",
    "                ##print(\"old\", old_node)\n",
    "                ##print(\"nodes\", extant_gene_graph.nodes)\n",
    "\n",
    "            # create edges between the correct nodes\n",
    "            extant_gene_graph.add_edge(old_node, new_node)\n",
    "            parent_gene_graph.add_edge(old_parent_node, new_parent_node)\n",
    "                ##print(\"edge.number\", extant_gene_graph.number_of_edges())\n",
    "                ##print(\"edges\", extant_gene_graph.edges)\n",
    "\n",
    "            # save the old node to create the next edge\n",
    "            old_node = new_node\n",
    "            old_parent_node = new_parent_node\n",
    "        \n",
    "        extant_gene_graph.remove_node(\"start\") # remove it, was only there to make the edges as we went along\n",
    "        parent_gene_graph.remove_node(\"start\") # remove it, was only there to make the edges as we went along\n",
    "        \n",
    "        all_graphs.append(extant_gene_graph)\n",
    "        all_graphs.append(parent_gene_graph)\n",
    "        \n",
    "        return(all_graphs)\n",
    "\n",
    "    \n",
    "    if isinstance(genome, pyham.AncestralGenome):\n",
    "        ancestral_gene_graph = nx.Graph()\n",
    "        child1_gene_graph = nx.Graph()\n",
    "        child2_gene_graph = nx.Graph()\n",
    "        \n",
    "        # make a list of lists\n",
    "        #     - a parent graph\n",
    "        #     - a graph for each set of children (maybe more than 2 children if polytomies?)\n",
    "        all_graphs = []\n",
    "        hog_list = []\n",
    "        child1_list = []\n",
    "        child2_list = []        \n",
    "        \n",
    "        for ancestral_gene in genome.genes:\n",
    "            hog_list.append(ancestral_gene)\n",
    "            child1_list.append(ancestral_gene.children[0])\n",
    "            child2_list.append(ancestral_gene.children[1])\n",
    "            if len(ancestral_gene.children) > 2:\n",
    "                print(\"fix problem, there is a polytomy\")\n",
    "            \n",
    "        # make a graph from the list of nodes, complete graph makes edges between all nodes\n",
    "        ancestral_gene_graph = nx.complete_graph(hog_list)     \n",
    "        child1_gene_graph = nx.complete_graph(child1_list)     \n",
    "        child2_gene_graph = nx.complete_graph(child2_list)\n",
    "        \n",
    "        all_graphs.append(ancestral_gene_graph)\n",
    "        all_graphs.append(child1_gene_graph)\n",
    "        all_graphs.append(child2_gene_graph)\n",
    "        \n",
    "        return(all_graphs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_synteny_graphs(tree_node):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function compares network graphs at the given level (genome) in the species tree to its descendant graphs.\n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        either a single graph for extant genes or multiple graphs for ancestral genes. \n",
    "    \"\"\"\n",
    "    \n",
    "    if isinstance(tree_node.genome, pyham.ExtantGenome):\n",
    "        print(\"synteny already assigned for extant genome\")\n",
    "        \n",
    "    if isinstance(tree_node.genome, pyham.AncestralGenome):\n",
    "        print(\"ancestor\", tree_node.graph[2].nodes)\n",
    "        print(\"find synteny\")\n",
    "        \n",
    "        # take this node ( go node by node)\n",
    "        # get the descendant nodes\n",
    "        # compare the synteny of the children graphs from this node to each descendant\n",
    "        # use the children graphs to assign synteny to this hog graph\n",
    "\n",
    "    gene2 = ham_analysis_no_name.get_extant_genome_by_name(\"SE002_aa\")\n",
    "    hog = ham_analysis_no_name.get_ancestral_genome_by_name(\"SE002_aa/SE003_aa\")\n",
    "    print(hog.graph)\n",
    "        #R.remove_nodes_from(n for n in G if n not in H)\n",
    "        \n",
    "# Then you get the corresponding mrca ancestral genomes \n",
    "#genome_rodents_3 = ham_analysis.get_ancestral_genome_by_mrca_of_genome_set({genome_rat, genome_mouse})\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene(1) <HOG(1)>\n",
      "Gene(2) <HOG(2)>\n",
      "Gene(3) <HOG(3)>\n",
      "Gene(4) <HOG(4)>\n",
      "Gene(5) <HOG(5)>\n",
      "Gene(6) <HOG(6)>\n",
      "Gene(7) <HOG(7)>\n",
      "Gene(8) <HOG(8)>\n",
      "Gene(9) <HOG(9)>\n",
      "Gene(10) <HOG(10)>\n",
      "Gene(11) <HOG(11)>\n",
      "Gene(12) <HOG(12)>\n",
      "Gene(13) <HOG(13)>\n",
      "Gene(14) <HOG(14)>\n",
      "Gene(15) <HOG(15)>\n",
      "Gene(16) <HOG(16)>\n",
      "Gene(17) <HOG(17)>\n",
      "Gene(18) <HOG(18)>\n",
      "Gene(19) <HOG(19)>\n",
      "Gene(20) <HOG(20)>\n",
      "Gene(21) <HOG(21)>\n",
      "Gene(22) <HOG(22)>\n",
      "Gene(23) <HOG(23)>\n",
      "Gene(24) <HOG(24)>\n",
      "Gene(25) <HOG(25)>\n",
      "Gene(26) <HOG(i1)>\n",
      "Gene(27) <HOG(i2)>\n",
      "Gene(28) <HOG(i3)>\n",
      "Gene(29) <HOG(i4)>\n",
      "Gene(30) <HOG(i5)>\n",
      "Gene(31) <HOG(i6)>\n",
      "Gene(32) <HOG(i7)>\n",
      "Gene(33) <HOG(i8)>\n",
      "Gene(34) <HOG(i9)>\n",
      "Gene(35) <HOG(i10)>\n",
      "Gene(36) <HOG(i11)>\n",
      "Gene(37) <HOG(i13)>\n",
      "Gene(38) <HOG(i12)>\n",
      "Gene(39) <HOG(i14)>\n",
      "Gene(40) <HOG(i15)>\n",
      "Gene(41) <HOG(i16)>\n",
      "Gene(42) <HOG(i17)>\n",
      "Gene(43) <HOG(i18)>\n",
      "Gene(44) <HOG(i19)>\n",
      "Gene(45) <HOG(i20)>\n",
      "Gene(46) <HOG(i21)>\n",
      "Gene(47) <HOG(i22)>\n",
      "Gene(48) <HOG(i23)>\n",
      "Gene(49) <HOG(i24)>\n",
      "Gene(50) <HOG(i25)>\n",
      "Gene(51) <HOG(i1)>\n",
      "Gene(52) <HOG(i2)>\n",
      "Gene(53) <HOG(i3)>\n",
      "Gene(54) <HOG(i4)>\n",
      "Gene(55) <HOG(i5)>\n",
      "Gene(56) <HOG(i6)>\n",
      "Gene(57) <HOG(i7)>\n",
      "Gene(58) <HOG(i8)>\n",
      "Gene(59) <HOG(i9)>\n",
      "Gene(60) <HOG(i10)>\n",
      "Gene(61) <HOG(i11)>\n",
      "Gene(62) <HOG(i13)>\n",
      "Gene(63) <HOG(i12)>\n",
      "Gene(64) <HOG(i14)>\n",
      "Gene(65) <HOG(i15)>\n",
      "Gene(66) <HOG(i16)>\n",
      "Gene(67) <HOG(i17)>\n",
      "Gene(68) <HOG(i18)>\n",
      "Gene(69) <HOG(i19)>\n",
      "Gene(70) <HOG(i20)>\n",
      "Gene(71) <HOG(i21)>\n",
      "Gene(72) <HOG(i22)>\n",
      "Gene(73) <HOG(i23)>\n",
      "Gene(74) <HOG(i24)>\n",
      "Gene(75) <HOG(i25)>\n",
      "synteny already assigned for extant genome\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'AncestralGenome' object has no attribute 'graph'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-267-c52d134b9700>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mtree_node\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mham_analysis_no_name\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtaxonomy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtraverse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'postorder'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m     \u001b[0mcompare_synteny_graphs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtree_node\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;31m#    if isinstance(tree_node.genome, pyham.ExtantGenome): # then we already have the synteny and all is good\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-266-6e942f5aff57>\u001b[0m in \u001b[0;36mcompare_synteny_graphs\u001b[0;34m(tree_node)\u001b[0m\n\u001b[1;32m     25\u001b[0m     \u001b[0mgene2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mham_analysis_no_name\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_extant_genome_by_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"SE002_aa\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m     \u001b[0mhog\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mham_analysis_no_name\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_ancestral_genome_by_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"SE002_aa/SE003_aa\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 27\u001b[0;31m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhog\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgraph\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     28\u001b[0m         \u001b[0;31m#R.remove_nodes_from(n for n in G if n not in H)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'AncestralGenome' object has no attribute 'graph'"
     ]
    }
   ],
   "source": [
    "## Make initial graphs\n",
    "\n",
    "for tree_node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    tree_node.graph = make_synteny_graph(tree_node.genome)\n",
    "    \n",
    "for tree_node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    compare_synteny_graphs(tree_node)\n",
    "    \n",
    "#    if isinstance(tree_node.genome, pyham.ExtantGenome): # then we already have the synteny and all is good\n",
    "#        print(tree_node.graph.nodes)\n",
    "#        print(\"nodes\", tree_node.graph.number_of_nodes())\n",
    "#        print(\"edges\", tree_node.graph.number_of_edges())\n",
    "        \n",
    "#    if isinstance(tree_node.genome, pyham.AncestralGenome): # then we want to compare to assign the proper synteny\n",
    "#        print(\"ancestor\", tree_node.graph[0].nodes)\n",
    "#        print(\"ancestor\", tree_node.graph[1].nodes)\n",
    "#        print(\"ancestor\", tree_node.graph[2].nodes)\n",
    "#        print(\"nodes\", tree_node.graph[0].number_of_nodes())\n",
    "#        print(\"edges\", tree_node.graph[0].number_of_edges())\n",
    "\n",
    "##print(ham_analysis_no_name.taxonomy.internal_nodes)\n",
    "##print(ham_analysis_no_name.taxonomy.leaves)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "   /-SE002_aa\n",
      "--|\n",
      "   \\-SE003_aa\n",
      "\n",
      "   /-SE001_aa\n",
      "--|\n",
      "  |   /-SE002_aa\n",
      "   \\-|\n",
      "      \\-SE003_aa\n"
     ]
    }
   ],
   "source": [
    "# construct proper synteny at internal nodes\n",
    "\n",
    "for tree_node in ham_analysis_no_name.taxonomy.internal_nodes: # NEED TO BE SURE IT AWAYS GOES IN POSTORDER?\n",
    "    print(tree_node)\n",
    "#ham_analysis_no_name.taxonomy.internal_nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py3]",
   "language": "python",
   "name": "conda-env-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
