{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.1.7'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pyham\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Path to nwk file\n",
    "##nwk_path = \"test_data/EstimatedSpeciesTree.nwk\"\n",
    "#nwk_path = \"my_3spp_simulated_test_data/RealTree.nwk\"\n",
    "#nwk_path = \"lower_mut_3spp/RealTree.nwk\"\n",
    "nwk_path = \"working_perfect_3spp/RealTree.nwk\"\n",
    "\n",
    "# The newick tree is extracted as a string\n",
    "tree_str = pyham.utils.get_newick_string(nwk_path, type=\"nwk\")\n",
    "\n",
    "# Path to OrthoXML file\n",
    "##orthoxml_path = \"test_data/HierarchicalGroups.orthoxml\"\n",
    "#orthoxml_path = \"my_3spp_simulated_test_data/Output/HierarchicalGroups.orthoxml\"\n",
    "#orthoxml_path = \"lower_mut_3spp/Output/HierarchicalGroups.orthoxml\"\n",
    "orthoxml_path = \"working_perfect_3spp/Output/HierarchicalGroups.orthoxml\"\n",
    "\n",
    "#ham_analysis = pyham.Ham(tree_str, orthoxml_path, use_internal_name=True)\n",
    "ham_analysis_no_name = pyham.Ham(tree_str, orthoxml_path, use_internal_name=False)\n",
    "\n",
    "pyham.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def locus_to_id(file):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function extracts loci from a fasta file and store the information as keys in a dictionary \n",
    "    and the related protein ID as value.\n",
    "\n",
    "    Args:\n",
    "        file (:obj:`str`): path to the file of interest.\n",
    "\n",
    "    Returns:\n",
    "        a dictionary mapping the locus (:obj:`str`) with the protein ID (:obj:`str`).\n",
    "    \"\"\"\n",
    "    \n",
    "    locus_to_id = {}\n",
    "    for seq_record in SeqIO.parse(file, 'fasta'):\n",
    "        locus_to_id[seq_record.description.split('locus: ')[-1]] = seq_record.description\n",
    "        \n",
    "    return locus_to_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# locus_to_id(\"test_data/DB/SE001_aa.fa\")\n",
    "##locus_to_id(\"my_3spp_simulated_test_data/DB/SE001_aa.fa\")\n",
    "\n",
    "##locus_to_id(\"working_perfect_3spp/DB/SE001_aa.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_locus(directory):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function goes through each file of a folder and build a dictionary containing loci \n",
    "    as keys and the related protein ID as values. For each entry in the dictionary, the function \n",
    "    updates the corresponding :obj:`pyham.abstractgene.Gene` by adding its locus.\n",
    "\n",
    "    Args:\n",
    "        directory (:obj:`str`): directory of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "\n",
    "    path = directory\n",
    "    for file in os.listdir(path):\n",
    "        ##print('File:', file)\n",
    "        path_to_file = path + '/' + file\n",
    "        locus_dictionary = locus_to_id(path_to_file)\n",
    " \n",
    "        for (loci, gene_id) in locus_dictionary.items():\n",
    "            gene = ham_analysis_no_name.get_genes_by_external_id(gene_id)\n",
    "            \n",
    "            # each gene should only map to 1 other gene, so should always be length 1...\n",
    "            assert len(gene) == 1, 'gene_id maps to multiple entries'\n",
    "            gene[0].locus = loci\n",
    "            #for element in gene:\n",
    "            #    element.locus = loci\n",
    "#                print(element.locus)\n",
    "                \n",
    "        ham_analysis_no_name.get_extant_genome_by_name('.'.join(file.split('.')[:-1])).locus_assignment = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get_locus('test_data/DB')\n",
    "##get_locus('my_3spp_simulated_test_data/DB')\n",
    "#get_locus('lower_mut_3spp/DB')\n",
    "get_locus('working_perfect_3spp/DB')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_by_locus(genome, locus):\n",
    "    \n",
    "    \"\"\"\n",
    "    Get a :obj:`pyham.abstractgene.Gene` that match the query locus. \n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant genome of interest.\n",
    "        locus (:obj:`int`): locus.\n",
    "\n",
    "    Returns:\n",
    "        :obj:`pyham.abstractgene.Gene`.\n",
    "    \"\"\"\n",
    "    \n",
    "    for gene in genome.genes:\n",
    "        if int(gene.locus) == locus:\n",
    "            return gene \n",
    "    \n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_neighbors(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function assigns neighbors to each :obj:`pyham.abstractgene.Gene` or :obj:`pyham.abstractgene.HOG`.\n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    if isinstance(genome, pyham.ExtantGenome):\n",
    "        \n",
    "        for extant_gene in genome.genes:\n",
    "            extant_gene.neighbors = (get_gene_by_locus(genome, int(extant_gene.locus)-1), \n",
    "                                     get_gene_by_locus(genome, int(extant_gene.locus)+1))            \n",
    "                        \n",
    "    if isinstance(genome, pyham.AncestralGenome):\n",
    "    \n",
    "        for ancestral_gene in genome.genes:  \n",
    "            \n",
    "            # test here, see if does the same thing, children should only be direct descendants?\n",
    "            #children = [child for child in ancestral_gene.children]\n",
    "            children = ancestral_gene.children\n",
    "            \n",
    "            ######################################################################################\n",
    "            # Dictionary containing genome names as keys (i.e. SE004) and a list of children as value #\n",
    "            ######################################################################################\n",
    "            \n",
    "            genome_dict = {}  \n",
    "            \n",
    "            for child in children:\n",
    "                if child.genome.name in genome_dict:\n",
    "                    genome_dict[child.genome.name].append(child) # if already in dictionary, add it (duplicate)\n",
    "                else:\n",
    "                    genome_dict[child.genome.name] = [child] # otherwise add the child\n",
    "\n",
    "##            print(getattr(ancestral_gene,'id'))\n",
    "#            print(genome, children)\n",
    "#            print(genome_dict)\n",
    "            \n",
    "            ############################################\n",
    "            # Children are replaced by their neighbors #      \n",
    "            ############################################\n",
    "            \n",
    "            for genome, gene_collection in genome_dict.items(): #stoca \n",
    "                for i in range(len(gene_collection)):\n",
    "                    gene_collection[i] = list(gene_collection[i].neighbors)\n",
    "            \n",
    "#            print(genome_dict)\n",
    "\n",
    "            ################################################################################\n",
    "            # Dictionary containing genomes as keys (i.e. SE004) and a list of parent HOGs #\n",
    "            ################################################################################\n",
    "            \n",
    "            parent_dict = {} # gets a list of all potential neighboring hogs to the ancestral_gene\n",
    "            for genome, neighbors_collection in genome_dict.items(): #stoca \n",
    "###                print(parent_dict)\n",
    "                for neighbors_pair in neighbors_collection:\n",
    "                    for neighbor in neighbors_pair: # don't really need the for loop bc only ever 2 neighbors\n",
    "          ##              print(neighbor, neighbor.parent)#, genome, ancestral_gene, parent_dict)\n",
    "                        if genome in parent_dict:   ## KJG - I don't understand this if/else - if we already put the focal genome in the parent dictionary, we do the same thing as if we didn't?\n",
    "          ##                  print(\"genome in parent dict\")\n",
    "                            if neighbor == None:\n",
    "                                parent_dict[genome].add(None)\n",
    "                            else:\n",
    "                                if neighbor.parent != ancestral_gene: #if theres' a duplication this might be true\n",
    "                                    parent_dict[genome].add(neighbor.parent) #add the neighbors parents - the neighboring hogs\n",
    "###                                    print(neighbor, neighbor.parent, \"in parent dict\")\n",
    "                        else:\n",
    "                            if neighbor == None:\n",
    "                                parent_dict[genome] = {None}    ## ah, this makes focal genome 'none' while above does 'add(None)' in addition to current genome?\n",
    "                            else:\n",
    "                                if neighbor.parent != ancestral_gene: # fill the parent dict by looking to see \n",
    "                                    parent_dict[genome] = {neighbor.parent}  ## again, no 'add', just make\n",
    "###                                    print(neighbor, neighbor.parent, \"not in parent dict\")\n",
    "          ##              print(neighbor, neighbor.parent, genome, ancestral_gene, parent_dict)\n",
    "#parent_dict = defaultdict(list)\n",
    "#if neighbor == None:\n",
    "#     parent_dict[genome].add(None)\n",
    "#else:\n",
    "# if neighbor.parent != ancestral_gene:\n",
    "# parent_dict[genome].add(neighbor.parent)\n",
    "###            print(genome_dict)\n",
    "###            print(parent_dict)\n",
    "\n",
    "            l = []  # List of sets\n",
    "            for genome, parent in parent_dict.items():  \n",
    "###                print(genome, parent, l)\n",
    "                l.append(parent)  # parent is a set\n",
    "                ancestral_gene.list_parents = l            # List of potential neighbors in case of conflict\n",
    "                \n",
    "            final_neighbors = list(l[0].intersection(*l))  # Intersection of the sets\n",
    "            \n",
    "            if len(final_neighbors) == 2:\n",
    "                ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "                print(\"fin\", ancestral_gene.neighbors)\n",
    "            \n",
    "            elif len(final_neighbors) == 1:\n",
    "                none_counter = 0\n",
    "                potential_neighbor = None\n",
    "                \n",
    "                for collection in l:\n",
    "                    if len(collection) == 2:\n",
    "                        for element in collection:\n",
    "                            if element == final_neighbors[0]:\n",
    "                                continue\n",
    "                            elif element == None:\n",
    "                                none_counter += 1\n",
    "                            else:\n",
    "                                potential_neighbor = element\n",
    "                \n",
    "                    else:\n",
    "                        break\n",
    "                \n",
    "                if none_counter == 1:  # To deal with missing orthology\n",
    "                    final_neighbors.append(potential_neighbor)\n",
    "                \n",
    "                if len(final_neighbors) == 1:\n",
    "                    final_neighbors.append(None)\n",
    "                    ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "                    print(\"len1\", ancestral_gene.neighbors)\n",
    "                else:  \n",
    "                    ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "                    print(\"else\", ancestral_gene.neighbors)\n",
    "            \n",
    "            elif len(final_neighbors) == 0 or len(final_neighbors) > 2:\n",
    "                final_neighbors = [None, None]\n",
    "                ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "\n",
    "        ##        print(\"anc\", ancestral_gene.neighbors)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_synteny_blocks(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function aims to reconstruct synteny blocks using neighbors of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        list of lists containing :obj:`pyham.abstractgene.Gene` or :obj:`pyham.abstractgene.HOG` \n",
    "        in a specific order.\n",
    "    \"\"\"\n",
    "\n",
    "    syn_blocks = []\n",
    "    seen = []\n",
    "    gene_collection = genome.genes\n",
    "    tmp = []\n",
    "        \n",
    "    for abstract_gene in gene_collection: # go through all the genes (extant or ancestral) for the genome\n",
    "        \n",
    "        if abstract_gene in seen: # if the gene is already there, skip out of the loop to the next gene\n",
    "            continue\n",
    "                \n",
    "        block1 = []\n",
    "        block2 = []\n",
    "        \n",
    "        current1 = abstract_gene\n",
    "        current2 = abstract_gene  # Used for the block2\n",
    "        \n",
    "        block1.append(current1)  # [X] where X is the first abstract gene # add the current gene to block 1\n",
    "        seen.append(current1) # append the current gene to seen\n",
    "        \n",
    "        ######################\n",
    "        # First end (block1) #\n",
    "        ######################\n",
    "        \n",
    "        next_gene = current1.neighbors[0] # find the neighbors of the current gene\n",
    "        \n",
    "        while next_gene != None: # if there is a next gene\n",
    "            \n",
    "            if next_gene in seen:\n",
    "                next_gene = None  \n",
    "                continue\n",
    "            \n",
    "            previous = current1 # then put the current gene as previous\n",
    "            current1 = next_gene # and put the next gene as current\n",
    "            \n",
    "            if previous in current1.neighbors:   # if the previous gene is a neighbor of the new current gene\n",
    "                block1.append(current1) # append the new current gene to block 1\n",
    "                seen.append(current1)   # append the new current gene to seen\n",
    "                # this step should always be true and the else statement is a check that breaks if not true\n",
    "            \n",
    "            else:                                  \n",
    "                break\n",
    "                print(\"previous gene is not a neighbor of the current gene\")\n",
    "                \n",
    "            if previous == current1.neighbors[0]: # if the previous gene is the first neighbor of the current gene\n",
    "                next_gene = current1.neighbors[1] # then the next gene equals the 2nd neighbor of the current gene\n",
    "            \n",
    "            else:\n",
    "                next_gene = current1.neighbors[0] # otherwise the next gene is the first neighbor of the current gene\n",
    "            # then iterate the while loop\n",
    "\n",
    "        #######################\n",
    "        # Second end (block2) #\n",
    "        #######################\n",
    "        \n",
    "        next_gene = current2.neighbors[1] # current2 is the same original abstract gene from above\n",
    "                                          # take its 2nd neighbor and make it the next gene\n",
    "            \n",
    "        while next_gene != None:    # if there is a next gene\n",
    "            \n",
    "            if next_gene in seen:   # if it's already seen, next = none and iterate the while loop\n",
    "                next_gene = None\n",
    "                continue\n",
    "            \n",
    "            previous = current2    # make the current gene become the previous gene\n",
    "            current2 = next_gene   # make the next gene become the current gene\n",
    "            \n",
    "            if previous in current2.neighbors:   # if the previous gene is a neighbor of the new next current gene\n",
    "                block2.append(current2)  # append the new current gene to block 2\n",
    "                seen.append(current2)    # append the new current gene to seen\n",
    "            \n",
    "            else:                                  \n",
    "                break\n",
    "                print(\"previous gene is not a neighbor of the current2 gene\")\n",
    "                \n",
    "            if previous == current2.neighbors[0]: # if the previous gene is the first neighbor of the current gene\n",
    "                next_gene = current2.neighbors[1] # then the next gene equals the 2nd neighbor of the current gene\n",
    "            \n",
    "            else:\n",
    "                next_gene = current2.neighbors[0] # otherwise the next gene is the first neighbor of the current gene\n",
    "            # then iterate the while loop\n",
    "            \n",
    "#        print(abstract_gene)\n",
    "#        print(current1.neighbors)\n",
    "#        print(block1)\n",
    "#        print(block2)\n",
    "        syn_blocks.append(block1[::-1] + block2)\n",
    "        \n",
    "    print(syn_blocks)\n",
    "    genome.synteny = syn_blocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_block_to_genes(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    The function links each :obj:`pyham.abstractgene.AbstractGene` to the synteny block it belongs to. \n",
    "    It creates a dictionary containing the genome of :obj:`pyham.abstractgene` as key and the synteny \n",
    "    block as value.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    for block in genome.synteny:\n",
    "    \n",
    "        for abstract_gene in block:\n",
    "            genomes_to_block = {}  # Dictionary\n",
    "            genomes_to_block[genome.name] = block  \n",
    "            abstract_gene.blocks = genomes_to_block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_children(genome, children):  \n",
    "    \n",
    "    \"\"\"\n",
    "    Due to tandem duplication, some :obj:`pyham.abstractgene.HOG` can have more than two children.\n",
    "    The function refines the list of children by removing duplicates.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        children (:obj:`list` of :obj:`pyham.abstractgene.AbstractGene`): a list of direct descendants AbstractGene.\n",
    "    \n",
    "    Returns:\n",
    "        list of :obj:`pyham.abstractgene.AbstractGene` without duplicates.\n",
    "    \"\"\"\n",
    "    \n",
    "    unique_genome = []\n",
    "    unique_children = []\n",
    "    \n",
    "    if len(children) == 2:  # Two children comming from two different genomes\n",
    "        \n",
    "        if children[0].genome != children[1].genome:\n",
    "            return children\n",
    "        \n",
    "        else:  # Two children comming from the same genome (i.e. duplication)\n",
    "            unique_children.append(None)\n",
    "            return unique_children\n",
    "    \n",
    "    else:  # In case of duplications, only one child per specie is taken into account\n",
    "        \n",
    "        for child in children:  \n",
    "            \n",
    "            if child.genome in unique_genome:\n",
    "                continue\n",
    "\n",
    "            else:\n",
    "                \n",
    "                for neighbor in child.neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        continue\n",
    "\n",
    "                    elif neighbor.parent:\n",
    "                        \n",
    "                        if neighbor.parent in child.parent.blocks[genome.name]:  # To make sure that, for both unique children, we go in the same direction. \n",
    "                            continue\n",
    "                        \n",
    "                        unique_genome.append(child.genome)\n",
    "                        unique_children.append(child)\n",
    "\n",
    "                    else:  # If the child has no parent HOG\n",
    "                        unique_genome.append(child.genome)\n",
    "                        unique_children.append(child)\n",
    "        \n",
    "        if len(unique_children) == 0:\n",
    "            unique_children.append(None)\n",
    "            return unique_children  \n",
    "        \n",
    "        else:\n",
    "            return unique_children  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_unique_children_from_unit_block(genome, children, genes_seen):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function aims to be used after refining the :obj:`list` genes_seen. It uses the :obj:`list` \n",
    "    genes_seen to check whether the :obj:`list` current_children returned by the function get_unique_children \n",
    "    are correctly selected. Indeed, the function get_unique_children encounters a limitation with single HOG.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        children (:obj:`list` of :obj:`pyham.abstractgene.AbstractGene`): a list of direct descendants AbstractGene.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        \n",
    "    Returns:\n",
    "        list of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "    \"\"\"\n",
    "    \n",
    "    unique_genome = []\n",
    "    unique_children = []\n",
    "    \n",
    "    if len(children) == 2:  # Two children comming from two different genomes\n",
    "        \n",
    "        if children[0].genome != children[1].genome:\n",
    "            return children\n",
    "        \n",
    "        else:  # Two children comming from the same genome (i.e. duplication)\n",
    "            unique_children.append(None)\n",
    "            return unique_children\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        for child in children:  \n",
    "            \n",
    "            if child.genome in unique_genome:\n",
    "                continue\n",
    "            \n",
    "            for neighbor in child.neighbors:\n",
    "                \n",
    "                if neighbor == None:\n",
    "                    continue\n",
    "\n",
    "                if neighbor.parent in child.parent.blocks[genome.name] or neighbor in genes_seen:  # To make sure that for both unique children we go in the same direction \n",
    "                    continue\n",
    "                \n",
    "                unique_genome.append(child.genome)\n",
    "                unique_children.append(child)\n",
    "                \n",
    "        if len(unique_children) == 0:\n",
    "            unique_children.append(None)\n",
    "            return unique_children  # Return a list without duplicates\n",
    "        \n",
    "        else:\n",
    "            return unique_children  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Link:\n",
    "    \n",
    "    \"\"\"\n",
    "    Link is a class representing linking regions. Linking regions are unresolved regions with a weak synteny \n",
    "    signal that links two synteny blocks together. A Link is defined by its collection of poorly resolved genes.\n",
    "    \n",
    "    Attributes:\n",
    "        | linking_region: dictionary of :obj:`pyham.abstractgene.AbstractGene`\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.linking_region = {}\n",
    "        \n",
    "    def get_linking_region(self, genome, block, common_block, starting_children):\n",
    "        \n",
    "        \"\"\"\n",
    "        Create a linking region filled with :obj:`pyham.abstractgene.AbstractGene`.\n",
    "        \n",
    "        Args:\n",
    "            genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "            block (:obj:`list`): the synteny block from which the assessement starts.\n",
    "            common_block (:obj:`list`): the synteny block to be linked with the initial block.\n",
    "            starting_children (:obj:`list`): list of children coming from the last :obj:`pyham.abstractgene.AbstractGene` of the initial block.\n",
    "        \n",
    "        Returns:\n",
    "            return linking_region: dictionary of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "        \"\"\"\n",
    "        \n",
    "        initial_children = starting_children\n",
    "        seen = []\n",
    "\n",
    "        for child in initial_children:  # Creation of the dictionary's entries ({species1: [], species2: []})\n",
    "\n",
    "            if child.genome.name in self.linking_region:\n",
    "                continue\n",
    "\n",
    "            else:\n",
    "                self.linking_region[child.genome.name] = [] \n",
    "\n",
    "        next_children = []\n",
    "\n",
    "        for child in initial_children:\n",
    "            current_child = child\n",
    "            growing_link = True\n",
    "\n",
    "            while growing_link:\n",
    "                \n",
    "                for neighbor in current_child.neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor != current_child.next_neighbor:\n",
    "                        continue\n",
    "\n",
    "                    if neighbor.parent:\n",
    "\n",
    "                        if neighbor.parent.blocks[genome.name] != common_block:\n",
    "                            \n",
    "                            if neighbor.parent not in self.linking_region[neighbor.genome.name]:\n",
    "                                self.linking_region[neighbor.genome.name].append(neighbor.parent)\n",
    "                                current_child = neighbor\n",
    "                                seen.append(neighbor)\n",
    "\n",
    "                            else:\n",
    "                                current_child = neighbor\n",
    "                                seen.append(neighbor)\n",
    "\n",
    "                        else:\n",
    "                            growing_link = False\n",
    "\n",
    "                    else:\n",
    "                        self.linking_region[neighbor.genome.name].append(neighbor)  # The child is added to the linking region anyway.\n",
    "                        current_child = neighbor\n",
    "                        seen.append(neighbor)\n",
    "\n",
    "    def __repr__(self):\n",
    "        return '<{}>'.format(self.__class__.__name__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def refine_genes_seen(genome, genes_seen, current_children):\n",
    "    \n",
    "    \"\"\"\n",
    "    Regarding the secondary reconstruction, when the growing process continue by starting from a single \n",
    "    :obj:`pyham.abstractgene.HOG`, both neighbors of one of its children can be within the \n",
    "    :obj:`list` genes_seen. The function will remove all children that come after the children of interest\n",
    "    within the :obj:`list` genes_seen.\n",
    "    \n",
    "    Ars:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        current_children (:obj:`list`): list of children to begin the investigation with at the next iteration.\n",
    "        \n",
    "    Returns:\n",
    "        return the refined :obj:`list` genes_seen.\n",
    "    \"\"\"\n",
    "    \n",
    "    if len(current_children) == 2:\n",
    "        tmp = genes_seen[:]  # Make an independent copy of the initial list\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == current_children[0]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == current_children[0].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:]\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == current_children[1]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == current_children[1].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:] \n",
    "\n",
    "    return genes_seen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def adjust_genes_seen(genome, genes_seen, starting_children):\n",
    "    \n",
    "    \"\"\"\n",
    "    Same principle as for the function refine_genes_seen but used when the growing process continue by\n",
    "    sarting from an element at the extremity of a synteny block. Neighbors from one of its children can\n",
    "    be present within the :obj:`list` genes_seen. The function will remove all children that come after \n",
    "    the children of interest within the :obj:`list` genes_seen.\n",
    "    \n",
    "    Ars:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        starting_children (:obj:`list`): list of children to begin the investigation with at the next iteration.\n",
    "        \n",
    "    Returns:\n",
    "        return the refined :obj:`list` genes_seen.\n",
    "    \"\"\"\n",
    "    \n",
    "    if len(starting_children) == 2:\n",
    "        tmp = genes_seen[:]  # Make an independent copy of the initial list\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == starting_children[0]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == starting_children[0].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:]\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == starting_children[1]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == starting_children[1].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:] \n",
    "\n",
    "    return genes_seen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def joining_blocks(window, genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    The function performs the secondary reconstruction. Synteny blocks from the primary reconstruction\n",
    "    are linked together using linking regions.\n",
    "    \n",
    "    Ars:\n",
    "        window (:obj:`int`): a window of assessment\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        \n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    for hog in genome.genes:  #### To be used with the function \"add_remaining_blocks\"\n",
    "        hog.processed = False\n",
    "    \n",
    "    new_blocks = []\n",
    "    block_record = []  # To keep track of the block already assessed. Useful to link several blocks in a row.\n",
    "    \n",
    "    for block in genome.synteny:\n",
    "            \n",
    "        if block in block_record or len(block) == 1:  \n",
    "            continue\n",
    "        \n",
    "        block2 = block  # Used to reconstruct the other side\n",
    "        \n",
    "        counter = 0\n",
    "        children = [child for child in block[-1].children]\n",
    "        current_children = get_unique_children(genome, children)  # Remove duplicates from children\n",
    "        starting_children = current_children  # To keep track of the initial children considered (used to build a linking region)\n",
    "        \n",
    "        if len(current_children) == 1 or len(current_children) > 2:  # If missing information (==1) or other special case (>2), the block is no longer taken into account\n",
    "            continue\n",
    "        \n",
    "        block_record.append(block)   \n",
    "        \n",
    "        blocks_seen = [set() for child in current_children]  # To save already assessed blocks and perform the intersection\n",
    "        genes_seen = []  # To keep track of already assessed genes\n",
    "        intersect = None\n",
    "        extended_block = []\n",
    "        next_children = [None, None]\n",
    "        genes_seen.append(current_children[0])  \n",
    "        genes_seen.append(current_children[1])  \n",
    "        \n",
    "        while counter < window:\n",
    "            none_type_as_neighbor = 0\n",
    "            block_found = False\n",
    "            reaching_already_assessed = False  \n",
    "            \n",
    "            if None in current_children:\n",
    "                break\n",
    "            \n",
    "            # Without the following block, when a common block is reached from one extremity, the next iteration \n",
    "            # starts from the opposite one. However, children originating from that extreme element are not\n",
    "            # added within the list of genes seen.\n",
    "            if current_children[0] and current_children[1] not in genes_seen:\n",
    "                genes_seen.append(current_children[0])\n",
    "                genes_seen.append(current_children[1])\n",
    "            \n",
    "            for i in range(len(current_children)):\n",
    "                \n",
    "                for neighbor in current_children[i].neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        none_type_as_neighbor += 1\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor.parent:  # Check whether the neighbor is a singleton (i.e. does it have a parent or not?)\n",
    "                        \n",
    "                        if neighbor.parent in block or neighbor in genes_seen:  # Avoid to take into account elements from the initial block or those already assessed\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            blocks_seen[i].add(tuple(neighbor.parent.blocks[genome.name]))\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region \n",
    "                            break\n",
    "                    \n",
    "                    else:\n",
    "                        \n",
    "                        if neighbor in genes_seen:\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region\n",
    "                \n",
    "            if none_type_as_neighbor > 0:  # If children have only one neighbor\n",
    "                break\n",
    "\n",
    "            counter += 1\n",
    "            current_children = next_children[:]\n",
    "            intersect = list(blocks_seen[0].intersection(*blocks_seen))  # Intersection\n",
    "            \n",
    "            if len(intersect) == 1:  # ex. intersect = [('a', 'c')]\n",
    "\n",
    "                    if list(intersect[0]) in block_record:\n",
    "                        reaching_already_assessed = True\n",
    "                        break\n",
    "                        \n",
    "                    block_found = True  \n",
    "                    counter = 0  # Reset the counter to pursue the assessement with further blocks\n",
    "                    common_block = list(intersect[0])  # Extraction of the last common block. \n",
    "                    block_record.append(common_block)\n",
    "\n",
    "                    link = Link()  # Class for linking regions\n",
    "                    link.get_linking_region(genome, block, common_block, starting_children)  # Creation of the linking region\n",
    "                    \n",
    "                    if not extended_block:  # If the list if empty (i.e. for a first block of a serie)\n",
    "                        extended_block.append(block)  # The initial block\n",
    "                        extended_block.append(link)  # The linking region\n",
    "                        extended_block.append(common_block)  # The common block\n",
    "\n",
    "                    else:  # For all the block but the first one, \"block\" corresponds to the previous common block\n",
    "                        extended_block.append(link)\n",
    "                        extended_block.append(common_block)\n",
    "\n",
    "                    block = common_block  # The new block becomes the last common block found within the while loop\n",
    "\n",
    "                    if len(block) > 1:\n",
    "                        \n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its first element (i.e. element 0), the next iteration should start from its last element (i.e. element -1) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block[0].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block[0].children)[0] and get_unique_children(genome, block[0].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block[-1].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)  \n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block[0].children)\n",
    "                            break\n",
    "\n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its last element (i.e. element -1), the next iteration should start from its first element (i.e. element 0) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block[-1].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block[-1].children)[0] and get_unique_children(genome, block[-1].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block[0].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block[-1].children)\n",
    "                            break\n",
    "                    \n",
    "                    ################################################################################################################\n",
    "                    # When we reach a common unit block, the next iteration should start from the same unique HOG (i.e. element 0) #\n",
    "                    ################################################################################################################\n",
    "\n",
    "                    elif len(block) == 1:  # If the block has only one HOG...\n",
    "                        current_children = get_unique_children(genome, block[0].children)                         \n",
    "                        \n",
    "                        if len(current_children) == 2:\n",
    "                            genes_seen = refine_genes_seen(genome, genes_seen, current_children)\n",
    "                            current_children = check_unique_children_from_unit_block(genome, block[0].children, genes_seen)\n",
    "\n",
    "                        starting_children = current_children\n",
    "                        \n",
    "                    blocks_seen = [set() for child in current_children]  # The variable is restarted for the next iteration\n",
    "\n",
    "            if reaching_already_assessed:  \n",
    "                break\n",
    "                \n",
    "            if len(current_children) == 2:\n",
    "                if current_children[0].genome != current_children[1].genome:\n",
    "                    continue\n",
    "                \n",
    "                break  # To exit the while loop\n",
    "                \n",
    "            else:  \n",
    "                break  # To exit the while loop\n",
    "        \n",
    "        tmp = [[], []]\n",
    "        if extended_block:\n",
    "            for item in extended_block:\n",
    "                if isinstance(item, Link):\n",
    "                    tmp[1].append(item)\n",
    "\n",
    "                else:\n",
    "                    for element in item:\n",
    "                        tmp[1].append(element)\n",
    "        \n",
    "        else:\n",
    "            for element in block:\n",
    "                tmp[1].append(element)  # If a new initial block reach directly an already assessed block (i.e. common block), the initial block is stored\n",
    "                \n",
    "###############################################################################################################\n",
    "\n",
    "        counter = 0\n",
    "        children = [child for child in block2[0].children]\n",
    "        current_children = get_unique_children(genome, children)  # Remove duplicates from children\n",
    "        starting_children = current_children\n",
    "        \n",
    "        if len(current_children) == 1 or len(current_children) > 2:  # If missing information (==1) or other special case (>2), the block is no longer taken into account\n",
    "            new_blocks.append(tmp[1])\n",
    "            continue\n",
    "        \n",
    "        blocks_seen = [set() for child in current_children]  # To save already assessed blocks and perform the intersection\n",
    "        genes_seen = []  # To keep track of already assessed genes (used for gene without parent to make to we go in one direction)\n",
    "        intersect = None\n",
    "        extended_block = []\n",
    "        next_children = [None, None]\n",
    "        genes_seen.append(current_children[0])  \n",
    "        genes_seen.append(current_children[1])  \n",
    "        \n",
    "        while counter < window:\n",
    "            none_type_as_neighbor = 0\n",
    "            block_found = False\n",
    "            reaching_already_assessed = False  \n",
    "            \n",
    "            if None in current_children:\n",
    "                break\n",
    "            \n",
    "            # Without the following block, when a common block is reached from one extremity, the next iteration \n",
    "            # starts from the opposite one. However, children originating from the that extreme element are not\n",
    "            # added within the list of genes seen.\n",
    "            if current_children[0] and current_children[1] not in genes_seen:\n",
    "                genes_seen.append(current_children[0])\n",
    "                genes_seen.append(current_children[1])\n",
    "            \n",
    "            for i in range(len(current_children)):\n",
    "                \n",
    "                for neighbor in current_children[i].neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        none_type_as_neighbor += 1\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor.parent:  # Check whether the neighbor is a singleton (i.e. does it have a parent or not?)\n",
    "                        \n",
    "                        if neighbor.parent in block2 or neighbor in genes_seen:  # Avoid to take into account elements from the initial block or those already assessed\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            blocks_seen[i].add(tuple(neighbor.parent.blocks[genome.name]))\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region \n",
    "                            break  \n",
    "\n",
    "                    else:\n",
    "                        \n",
    "                        if neighbor in genes_seen:\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region\n",
    "                \n",
    "            if none_type_as_neighbor > 0:  # If children have only one neighbor\n",
    "                break\n",
    "\n",
    "            counter += 1\n",
    "            current_children = next_children[:]\n",
    "            intersect = list(blocks_seen[0].intersection(*blocks_seen))  # Intersection\n",
    "            \n",
    "            if len(intersect) == 1:  # ex. intersect = [('a', 'c')]\n",
    "\n",
    "                    if list(intersect[0]) in block_record:\n",
    "                        reaching_already_assessed = True\n",
    "                        break\n",
    "                        \n",
    "                    block_found = True  \n",
    "                    counter = 0  # Reset the counter to pursue the assessement with further blocks\n",
    "                    common_block = list(intersect[0])  # Extraction of the last common block. \n",
    "                    block_record.append(common_block)\n",
    "\n",
    "                    link = Link()  # Class for linking regions\n",
    "                    link.get_linking_region(genome, block2, common_block, starting_children)  # Creation of the linking region\n",
    "                    \n",
    "                    extended_block.append(link)  # The linking region\n",
    "                    extended_block.append(common_block)  # The common block\n",
    "\n",
    "                    block2 = common_block  # The new block becomes the last common block found within the while loop\n",
    "\n",
    "                    if len(block2) > 1:\n",
    "                        \n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its first element (i.e. element 0), the next iteration should start from its last element (i.e. element -1) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block2[0].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block2[0].children)[0] and get_unique_children(genome, block2[0].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block2[-1].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block2[0].children)\n",
    "                            break\n",
    "\n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its last element (i.e. element -1), the next iteration should start from its first element (i.e. element 0) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block2[-1].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block2[-1].children)[0] and get_unique_children(genome, block2[-1].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block2[0].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  # 05.01.18\n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)  # 05.01.18\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block2[-1].children)\n",
    "                            break\n",
    "                    \n",
    "                    ################################################################################################################\n",
    "                    # When we reach a common unit block, the next iteration should start from the same unique HOG (i.e. element 0) #\n",
    "                    ################################################################################################################\n",
    "\n",
    "                    elif len(block2) == 1:  # If the block has only one HOG...\n",
    "                        current_children = get_unique_children(genome, block2[0].children) \n",
    "                        \n",
    "                        if len(current_children) == 2:\n",
    "                            genes_seen = refine_genes_seen(genome, genes_seen, current_children)\n",
    "                            current_children = check_unique_children_from_unit_block(genome, block2[0].children, genes_seen)\n",
    "\n",
    "                        starting_children = current_children\n",
    "                        \n",
    "                    blocks_seen = [set() for child in current_children]  # The variable is restarted for the next iteration\n",
    "\n",
    "            if reaching_already_assessed:  \n",
    "                break\n",
    "                \n",
    "            if len(current_children) == 2:\n",
    "                if current_children[0].genome != current_children[1].genome:\n",
    "                    continue\n",
    "                \n",
    "                break  # To exit the while loop\n",
    "                \n",
    "            else:  \n",
    "                break  # To exit the while loop\n",
    "\n",
    "        if extended_block:\n",
    "            for item in extended_block:\n",
    "                if isinstance(item, Link):\n",
    "                    tmp[0].append(item)\n",
    "\n",
    "                else:\n",
    "                    for element in item:\n",
    "                        tmp[0].append(element)\n",
    "        \n",
    "        # In tmp[1] there is always something, namely either block or extended block.\n",
    "        # The list tmp[0] can be empty\n",
    "        if tmp[0] and tmp[1]:\n",
    "            tmp[0] = tmp[0][::-1]\n",
    "            new_blocks.append(tmp[0] + tmp[1])\n",
    "            \n",
    "        elif tmp[1] and not tmp[0]:\n",
    "            new_blocks.append(tmp[1])\n",
    "            \n",
    "    genome.new_synteny = new_blocks\n",
    "\n",
    "    add_remaining_blocks(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_remaining_blocks(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    Once the secondary reconstruction is done. Some blocks or single HOGs from the primary reconstruction may\n",
    "    remained unassessed. The function add such items within the collection of extended synteny blocks.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        \n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    c = 0\n",
    "    \n",
    "    for block in genome.new_synteny:\n",
    "        \n",
    "        for element in block:\n",
    "            \n",
    "            if not isinstance(element, pyham.HOG):\n",
    "                \n",
    "                for key, value in element.linking_region.items():\n",
    "                    \n",
    "                    for gene in value:\n",
    "                        \n",
    "                        if isinstance(gene, pyham.HOG):\n",
    "                            gene.processed = True\n",
    "                            c += 1\n",
    "\n",
    "            else:\n",
    "                element.processed = True\n",
    "                c += 1\n",
    "    \n",
    "    not_processed = 0\n",
    "    \n",
    "    for gene in genome.genes:\n",
    "        \n",
    "        if gene.processed == False and gene.blocks[genome.name] not in genome.new_synteny:\n",
    "            genome.new_synteny.append(gene.blocks[genome.name])\n",
    "            not_processed += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--SE001_aa\n",
      "SE001_aa\n",
      "[Gene(1), Gene(2), Gene(3), Gene(4), Gene(5), Gene(6), Gene(7), Gene(8), Gene(9), Gene(10), Gene(11), Gene(12), Gene(13), Gene(14), Gene(15), Gene(16), Gene(17), Gene(18), Gene(19), Gene(20), Gene(21), Gene(22), Gene(23), Gene(24), Gene(25)]\n",
      "\n",
      "--SE002_aa\n",
      "SE002_aa\n",
      "[Gene(26), Gene(27), Gene(28), Gene(29), Gene(30), Gene(31), Gene(32), Gene(33), Gene(34), Gene(35), Gene(36), Gene(37), Gene(38), Gene(39), Gene(40), Gene(41), Gene(42), Gene(43), Gene(44), Gene(45), Gene(46), Gene(47), Gene(48), Gene(49), Gene(50)]\n",
      "\n",
      "--SE003_aa\n",
      "SE003_aa\n",
      "[Gene(51), Gene(52), Gene(53), Gene(54), Gene(55), Gene(56), Gene(57), Gene(58), Gene(59), Gene(60), Gene(61), Gene(62), Gene(63), Gene(64), Gene(65), Gene(66), Gene(67), Gene(68), Gene(69), Gene(70), Gene(71), Gene(72), Gene(73), Gene(74), Gene(75)]\n",
      "\n",
      "   /-SE002_aa\n",
      "--|\n",
      "   \\-SE003_aa\n",
      "SE002_aa/SE003_aa\n",
      "[<HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>]\n",
      "\n",
      "   /-SE001_aa\n",
      "--|\n",
      "  |   /-SE002_aa\n",
      "   \\-|\n",
      "      \\-SE003_aa\n",
      "SE001_aa/SE002_aa/SE003_aa\n",
      "[<HOG(1)>, <HOG(2)>, <HOG(3)>, <HOG(4)>, <HOG(5)>, <HOG(6)>, <HOG(7)>, <HOG(8)>, <HOG(9)>, <HOG(10)>, <HOG(11)>, <HOG(12)>, <HOG(13)>, <HOG(14)>, <HOG(15)>, <HOG(16)>, <HOG(17)>, <HOG(18)>, <HOG(19)>, <HOG(20)>, <HOG(21)>, <HOG(22)>, <HOG(23)>, <HOG(24)>, <HOG(25)>]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    print(node)\n",
    "    print(node.genome)\n",
    "    print(node.genome.genes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    \n",
    "    genome=node.genome   \n",
    "    if isinstance(genome, pyham.ExtantGenome):\n",
    "        \n",
    "        for extant_gene in genome.genes:\n",
    "            extant_gene.neighbors = (get_gene_by_locus(genome, int(extant_gene.locus)-1), \n",
    "                                     get_gene_by_locus(genome, int(extant_gene.locus)+1))\n",
    "            #print(extant_gene.neighbors)\n",
    "                        \n",
    "    if isinstance(genome, pyham.AncestralGenome):\n",
    "    \n",
    "##        print(genome.genes)\n",
    "        for ancestral_gene in genome.genes:  \n",
    "            children = ancestral_gene.children\n",
    "            \n",
    "##            print(ancestral_gene)\n",
    "##            print(genome, children)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "neighbors assigned\n",
      "[[Gene(1), Gene(2), Gene(3), Gene(4), Gene(5), Gene(6), Gene(7), Gene(8), Gene(9), Gene(10), Gene(11), Gene(12), Gene(13), Gene(14), Gene(15), Gene(16), Gene(17), Gene(18), Gene(19), Gene(20), Gene(21), Gene(22), Gene(23), Gene(24), Gene(25)]]\n",
      "next node\n",
      "neighbors assigned\n",
      "[[Gene(26), Gene(27), Gene(28), Gene(29), Gene(30), Gene(31), Gene(32), Gene(33), Gene(34), Gene(35), Gene(36), Gene(37), Gene(38), Gene(39), Gene(40), Gene(41), Gene(42), Gene(43), Gene(44), Gene(45), Gene(46), Gene(47), Gene(48), Gene(49), Gene(50)]]\n",
      "next node\n",
      "neighbors assigned\n",
      "[[Gene(51), Gene(52), Gene(53), Gene(54), Gene(55), Gene(56), Gene(57), Gene(58), Gene(59), Gene(60), Gene(61), Gene(62), Gene(63), Gene(64), Gene(65), Gene(66), Gene(67), Gene(68), Gene(69), Gene(70), Gene(71), Gene(72), Gene(73), Gene(74), Gene(75)]]\n",
      "next node\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'HOG' object has no attribute 'id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-23-89e9186861c1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mcounter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mnode\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mham_analysis_no_name\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtaxonomy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtraverse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'postorder'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0massign_neighbors\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"neighbors assigned\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mget_synteny_blocks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-22-a14404fd50b0>\u001b[0m in \u001b[0;36massign_neighbors\u001b[0;34m(genome)\u001b[0m\n\u001b[1;32m     37\u001b[0m                     \u001b[0mgenome_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mchild\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mchild\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m# otherwise add the child\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mancestral_gene\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'id'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m \u001b[0;31m#            print(genome, children)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m \u001b[0;31m#            print(genome_dict)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'HOG' object has no attribute 'id'"
     ]
    }
   ],
   "source": [
    "counter = 1\n",
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    assign_neighbors(node.genome)\n",
    "    print(\"neighbors assigned\")\n",
    "    get_synteny_blocks(node.genome)\n",
    "    print(\"next node\")\n",
    "    assign_block_to_genes(node.genome)\n",
    "    \n",
    "    #print(node)\n",
    "    # here we go through all nodes, and if it is an ancestral, note, we make some plots\n",
    "    # in my 3 species example, there are 2 ancestral nodes, so two plots\n",
    "    \n",
    "    if isinstance(node.genome, pyham.AncestralGenome):\n",
    "        # links together the primary reconstruction's synteny blocks, using windows of 20\n",
    "        joining_blocks(20, node.genome)\n",
    "                \n",
    "        ### Plot ###\n",
    "        \n",
    "        title = node.genome.name  \n",
    "        raw_values = []\n",
    "        \n",
    "        for block in node.genome.new_synteny:\n",
    "            count = 0\n",
    "            \n",
    "            for element in block:\n",
    "                # goes through the elements of that block, the HOGs\n",
    " #               print(block, element)\n",
    "         \n",
    "                # if that element (HOG) is a pyham.HOG, increase count, it's the number of HOGs we've alignd the synteny of in this block\n",
    "                if isinstance(element, pyham.HOG):\n",
    "                    count += 1\n",
    "                    \n",
    " #               print(count)\n",
    "                    \n",
    "            raw_values.append(count)\n",
    "            \n",
    " #       print(raw_values)\n",
    "        \n",
    "        label = list(set(raw_values))\n",
    "        values = [0 for i in range(len(label))]\n",
    "        \n",
    "        for i in range(len(label)):\n",
    "            \n",
    "            for element in raw_values:\n",
    "                \n",
    "                if label[i] == element:\n",
    "                    values[i] += 1\n",
    "\n",
    "        plt.bar(label, values)\n",
    "        plt.title(title)\n",
    "        plt.xlabel('Abstractgenes per block')\n",
    "        plt.ylabel('Number of blocks')\n",
    "        plt.savefig('results/secondary_reconstruction_plot{}.png'.format(counter))\n",
    "        counter += 1\n",
    "        plt.show()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEWCAYAAABrDZDcAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAeZ0lEQVR4nO3deZwdZZ3v8c83CyhkIyYikISghHEComALMiCLgrLIIqISVCZcrlEHlBlcBkdBBHcG9eKgkBFEEIPsN2IQlEvAhSWBQCCJ4QYMEgETlS1BtvCbP56nSXFy+vRJd9c53V3f9+tVr66qp5ZfVXfXr+qpqqcUEZiZWXUNaXcAZmbWXk4EZmYV50RgZlZxTgRmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgpZO0h6TfSXpC0t8k/VbSWyRNl7RW0uqabss831hJV0laI+lBSUfVLPeoPH6NpKsljc3jN5Z0Xi57StICSQe0aFv/Q9JXC/1/yNu0QtJPC9PNlfRMzXb/rFD+Dkm/l/S0pBslbV0o21jS+ZKelPSopBMLZVMlzZf0WO5+JWlqK7bdBi4nAiuVpFHANcB3gbHAVsCXgGfzJLdExIia7uFcdjbwHLA58EHg+5K2z8vdHjgX+HAufxr4Xp5vGPAQsBcwGjgZuFTS5BI3tdOBwBxJ/5xj2zciRgAdwA010x5fs90HA0gaB1yZ4x4LzAd+WpjvVGAKsDWwD/BZSfvnsoeBI/J844DZwCV9vpU2uESEO3eldaQD4ONdlE0HftNF2aakJLBdYdxFwNdz/1eBnxTKXpenH9nF8hYC7+0m1s1ISWsV8Fjun1AoPwZYAjwFPAB8tM78K4GhwH8B32mwrrnA/+6ibAbwu5p98Xfg9Xn4T8A7C+WnA5fUWc4w4Djg6SZ+TwcBC4AnSUn01Jryy4BHgSeAm4Ht2/235a7vOl8RWNnuA9ZK+pGkAyRt1uR82wFrI+K+wri7ge1z//Z5GICIuJ+cOGoXJGnzPH5RN+scAvyQdKY9iXTw/a9C+Urg3cAoUlL4tqSdC+XvAm6IiLXArcDRkj4jqUPS0G7WXVS7bWuA+4Ht8/7bsljOy/cLAJIeB54hXYl9tYl1rgGOBsaQksLHJR1WKL+WdBXyauBO4OIN2B7r55wIrFQR8SSwBxDAfwOrJM3OB2eAt0p6vNDdn8ePIJ19Fj0BjGyyHABJw0kHrR9FxO+7ifWvEXFFRDwdEU8BXyFVL3WW/zwi7o/kJuB64G2FRRwEzMnT/hj4BCk53ASslHRSzSrPqtn205vYthGF4S63OyLGkKrFjied6TcUEXMj4p6IeDEiFgKzarb9/Ih4KiKeJVVNvVHS6O6WawODE4GVLiKWRMT0iJgA7EA6o/1OLr41IsYUutfl8atJZ95Fo0jVMs2UI2kIqTrpOdIBsSFJm0g6N99kfpJUBTKm82w+X9Hcmm94P066HzCusK79gF8UtvviiNiXdJb9MeA0Se8qrPKTNdt+chPbtrowXHe7C+tfA5wDXCjp1d1s+675pvQqSU/keDu3baikr0u6P++X5Xm2cY2WaQOHE4G1VD4rv4CUEBq5DxgmaUph3BtZV72zKA8DIOm1wMZ5PiQJOI90I/m9EfF8E+F9CvgHYNeIGAXsuW7x2hi4AvhPYPN8xj0HUJ7mLcDyiFhVZ5ufj4jLSPcputvuetu2KekeyKKIeAx4pFjOy/dLrSHAJqSb9I38hHRjeWJEjCYlkM5tOwo4FNiXdJUxuTO0JrbFBgAnAiuVpNdL+pSkCXl4IjCNVIfepXw2eyXpLHpTSbuTDkYX5UkuBg6W9LZ8oDwNuDJX6QB8H/hH4OCI+HuT4Y4k3Rd4PD+K+sVC2UakRLMKeCE/jvrOQvlL1UJ5O6dLOkjSSElD8vTbA7c1EcdVwA6S3ivpFcApwMJC1daFwBckbSbp9cBHSMkVSftJ2imfxY8CvkW68b2kiW3/W0Q8I2kX0sG/WPYs8FdSUmnmnoMNJO2+W+1ucHekM9FLSU+6rMk/zyVVZ0wH1pKqO4rdW/K8Y4Gr83x/BI6qWfZRefwa4P8CY/P4rUn3JJ6pWe4Hu4l1S9LTPKtJVxYfzcsZlsuPA/4MPE5KSJcAX85l84GOwrIOB35LOgg/CdwDTC+Uz60T3x2F8n2B35MS01xgcqFsY+D8vNw/AycWyt6X51tNSlpzgB2b+D0dATxIqmK6hnST/Me5bETev0/laY7O+2Xbdv99ueubTvkXbWY9lG983wVsGf6HsgHIVUNmvTeadFbuJGADkhOBVUpu9qG2SYvVkq7t6TIj4r6ImNWXcZZB0qIutv2D7Y7N2stVQ2ZmFTes3QFsqHHjxsXkyZPbHYaZ2YByxx13/CUixtcrG3CJYPLkycyfP7/dYZiZDSiSHuyqzPcIzMwqzonAzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4kpLBPnj2isl3dtFuSSdJWmZpIU1X3oyM7MWKfOK4AJg/wblB5A+fTeF9I3W75cYi5mZdaG0RBARNwN/azDJocCFkdxK+hLUFmXFY2Zm9bXzzeKtgIcKwyvyuEdqJ5Q0g3TVwKRJk3q8wskn/bzH85qZtdvyrx9UynLbebO43mfu6raAFxEzI6IjIjrGj6/bVIaZmfVQOxPBCmBiYXgC8HCbYjEzq6x2JoLZwNH56aG3Ak9ExHrVQmZmVq7S7hFImgXsDYyTtIL0IfDhABFxDulbqgcCy4CngWPKisXMzLpWWiKIiGndlAfpY+BmZtZGfrPYzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4pwIzMwqzonAzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4pwIzMwqzonAzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4pwIzMwqzonAzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4pwIzMwqzonAzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4pwIzMwqzonAzKzinAjMzCqu1EQgaX9JSyUtk3RSnfJJkm6UtEDSQkkHlhmPmZmtr7REIGkocDZwADAVmCZpas1kXwAujYidgCOB75UVj5mZ1VfmFcEuwLKIeCAingMuAQ6tmSaAUbl/NPBwifGYmVkdZSaCrYCHCsMr8riiU4EPSVoBzAE+UW9BkmZImi9p/qpVq8qI1cyssspMBKozLmqGpwEXRMQE4EDgIknrxRQRMyOiIyI6xo8fX0KoZmbVVWYiWAFMLAxPYP2qn2OBSwEi4hbgFcC4EmMyM7MaZSaCecAUSdtI2oh0M3h2zTR/BN4BIOkfSYnAdT9mZi1UWiKIiBeA44HrgCWkp4MWSTpN0iF5sk8BH5F0NzALmB4RtdVHZmZWomFlLjwi5pBuAhfHnVLoXwzsXmYMZmbWmN8sNjOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4rrNhFI+qakUZKGS7pB0l8kfagVwZmZWfmauSJ4Z0Q8Cbyb1LT0dsBnSo3KzMxapplEMDz/PBCYFRF/KzEeMzNrsWZaH/2ZpN8Dfwf+RdJ44JlywzIzs1bp9oogIk4CdgM6IuJ54GngkMZzmZnZQNHMzeLzI+KxiFjbOQr4SblhmZlZqzRzj+BPkr4PIGkz4Hrgx6VGZWZmLdNM1dDJwJOSziElgTMj4oelR2ZmZi3R5c1iSYcXBm8HTs4/Q9LhEXFl2cGZmVn5Gj01dHDN8ALSo6QHAwE4EZiZDQJdJoKIOKaVgZiZWXs089TQjySNKQxvJun8csMyM7NWaeapoR0j4vHOgYh4DNipvJDMzKyVmkkEQ/JjowBIGktzbySbmdkA0MwB/Uzgd5Iuz8PvA75SXkhmZtZK3SaCiLhQ0nzg7XnU4RGxuNywzMysVZr9MM1wUtMSnf1mZjZINPPU0AnAxcA44NXAjyV9ouzAzMysNZq5R3AssGtErAGQ9A3gFuC7ZQZmZmat0UzVkIC1heG1rKsmMjOzAa6ZK4IfArdJuioPHwacV15IZmbWSs08NfQtSXOBPUhXAsdExIKyAzMzs9bosmpI0tjODlhO+gbBRcCDeVy3JO0vaamkZZJO6mKa90taLGmRJH/wxsysxRpdEdxBamW03v2AAF7baMGShgJnA/sBK4B5kmYX30GQNAX4HLB7RDwm6dUbGL+ZmfVSo9ZHt+nlsncBlkXEAwCSLgEOBYovo30EODu3X0RErOzlOs3MbAM11WZQ/kjNHqQrgV9HxNVNzLYV8FBheAWwa8002+Xl/xYYCpwaEb9oJiYzM+sb3SYCSd8DtgVm5VEfk7RfRBzX3ax1xkWd9U8B9gYmAL+WtEOxtdMcwwxgBsCkSZO6C9nMzDZAM1cEewE7RERA+j4BcE8T860AJhaGJwAP15nm1oh4HviDpKWkxDCvOFFEzARmAnR0dNQmEzMz64VmXihbChRPwycCC5uYbx4wRdI2kjYCjgRm10xzNbAPgKRxpKqiB5pYtpmZ9ZFGH6//GakqZzSwRNLteXhX4HfdLTgiXpB0PHAdqf7//IhYJOk0YH5EzM5l75S0mPTG8mci4q+93SgzM2teo6qh/+ztwiNiDjCnZtwphf4ATsydmZm1QaPHR29qZSBmZtYezX6PwMzMBiknAjOzimvU1tAN+ec3WheOmZm1WqObxVtI2gs4JDcP8bIXxCLizlIjMzOzlmiUCE4BTiK9CPatmrJg3cfszcxsAGv01NDlwOWSTo6I01sYk5mZtVAzH6Y5XdIhwJ551NyIuKbcsMzMrFW6fWpI0teAE0jNRy8GTsjjzMxsEGim0bmDgDdFxIvwUqNzC0gflDEzswGu2fcIxhT6R5cRiJmZtUczVwRfAxZIupH0COme+GrAzGzQaOZm8SxJc4G3kBLBv0fEo2UHZmZmrdHUpyoj4hHW/5aAmZkNAm5ryMys4pwIzMwqrmEikDRE0r2tCsbMzFqvYSLI7w7cLWlSo+nMzGzgauZm8RbAovzN4jWdIyPikNKiMjOzlmkmEXyp9CjMzKxtmnmP4CZJWwNTIuJXkjYBhpYfmpmZtUIzjc59BLgcODeP2gq4usygzMysdZp5fPQ4YHfgSYCI+P/Aq8sMyszMWqeZRPBsRDzXOSBpGOkLZWZmNgg0kwhukvQfwCsl7QdcBvys3LDMzKxVmkkEJwGrgHuAjwJzgC+UGZSZmbVOM08NvZg/RnMbqUpoaUS4asjMbJDoNhFIOgg4B7if1Az1NpI+GhHXlh2cmZmVr5kXys4E9omIZQCSXgf8HHAiMDMbBJq5R7CyMwlkDwArS4rHzMxarMsrAkmH595FkuYAl5LuEbwPmNeC2MzMrAUaVQ0dXOj/M7BX7l8FbFZaRGZm1lJdJoKIOKaVgZiZWXs009bQNpK+JelKSbM7u2YWLml/SUslLZN0UoPpjpAUkjo2JHgzM+u9Zp4auho4j/Q28YvNLljSUOBsYD9gBTBP0uyIWFwz3Ujgk6T3FMzMrMWaSQTPRMRZPVj2LsCyiHgAQNIlwKHA4prpTge+CXy6B+swM7Neaubx0f8j6YuSdpO0c2fXxHxbAQ8VhlfkcS+RtBMwMSKuabQgSTMkzZc0f9WqVU2s2szMmtXMFcEbgA8Db2dd1VDk4UZUZ9xLTVNIGgJ8G5jeXQARMROYCdDR0eHmLczM+lAzieA9wGuLTVE3aQUwsTA8AXi4MDwS2AGYKwngNcBsSYdExPwNXJeZmfVQM1VDdwNjerDsecCU/NTRRsCRwEtPG0XEExExLiImR8Rk4FbAScDMrMWauSLYHPi9pHnAs50jI+KQRjNFxAuSjgeuI33j+PyIWCTpNGB+RDT1CKqZmZWrmUTwxZ4uPCLmkL5fUBx3ShfT7t3T9ZiZWc818z2Cm1oRiJmZtUcz3yN4inVP+2wEDAfWRMSoMgMzM7PWaOaKYGRxWNJhpJfFzMxsEGjmqaGXiYir6f4dAjMzGyCaqRo6vDA4BOig8GKYmZkNbM08NVT8LsELwHJSm0FmZjYINHOPwN8lMDMbxBp9qrLu8/5ZRMTpJcRjZmYt1uiKYE2dcZsCxwKvIjUfbWZmA1yjT1We2dmfPx5zAnAMcAlwZlfzmZnZwNLwHoGkscCJwAeBHwE7R8RjrQjMzMxao9E9gjOAw0nfAXhDRKxuWVRmZtYyjV4o+xSwJfAF4GFJT+buKUlPtiY8MzMrW6N7BBv81rGZmQ08PtibmVWcE4GZWcU5EZiZVZwTgZlZxTkRmJlVnBOBmVnFORGYmVWcE4GZWcU5EZiZVZwTgZlZxTkRmJlVnBOBmVnFORGYmVWcE4GZWcU5EZiZVZwTgZlZxTkRmJlVnBOBmVnFlZoIJO0vaamkZZJOqlN+oqTFkhZKukHS1mXGY2Zm6ystEUgaCpwNHABMBaZJmloz2QKgIyJ2BC4HvllWPGZmVl+ZVwS7AMsi4oGIeA64BDi0OEFE3BgRT+fBW4EJJcZjZmZ1lJkItgIeKgyvyOO6cixwbb0CSTMkzZc0f9WqVX0YopmZlZkIVGdc1J1Q+hDQAZxRrzwiZkZER0R0jB8/vg9DNDOzYSUuewUwsTA8AXi4diJJ+wKfB/aKiGdLjMfMzOoo84pgHjBF0jaSNgKOBGYXJ5C0E3AucEhErCwxFjMz60JpiSAiXgCOB64DlgCXRsQiSadJOiRPdgYwArhM0l2SZnexODMzK0mZVUNExBxgTs24Uwr9+5a5fjMz657fLDYzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOruFITgaT9JS2VtEzSSXXKN5b001x+m6TJZcZjZmbrKy0RSBoKnA0cAEwFpkmaWjPZscBjEbEt8G3gG2XFY2Zm9ZV5RbALsCwiHoiI54BLgENrpjkU+FHuvxx4hySVGJOZmdUYVuKytwIeKgyvAHbtapqIeEHSE8CrgL8UJ5I0A5iRB1dLWtrDmMbVLrufcFwbxnFtuP4am+PaAPpGr+LauquCMhNBvTP76ME0RMRMYGavA5LmR0RHb5fT1xzXhnFcG66/xua4NkxZcZVZNbQCmFgYngA83NU0koYBo4G/lRiTmZnVKDMRzAOmSNpG0kbAkcDsmmlmA/+c+48A/l9ErHdFYGZm5SmtaijX+R8PXAcMBc6PiEWSTgPmR8Rs4DzgIknLSFcCR5YVT9br6qWSOK4N47g2XH+NzXFtmFLikk/AzcyqzW8Wm5lVnBOBmVnFDYpEIGmipBslLZG0SNIJNeWflhSSxnUx/yRJ1+f5F/dVUxe9iUvSPpLuKnTPSDqs3XHl8m/m+ZZIOquvXgLsg7i+Iene3H2gL2JqFJekUyX9qfA7OrCL+Rs2tdLGuM6XtFLSvX0VU2/j6u5voI1xvULS7ZLuzvN+qT/EVVjGUEkLJF3ToyAiYsB3wBbAzrl/JHAfMDUPTyTdsH4QGNfF/HOB/XL/CGCT/hBXYTljSTfT2x4X8E/Ab0kPAAwFbgH27gdxHQT8kvQAxKbAfGBUmXEBpwKf7mbeocD9wGuBjYC7O7epnXHlefYEdgbu7Yt4+mh/dfk30Oa4BIzI/cOB24C3tjuuwjJOBH4CXNOTGAbFFUFEPBIRd+b+p4AlpLeWIbVh9FnqvKgGoNT+0bCI+GWef3VEPN3uuGocAVzbT+IK4BWkg9rGpH+KP/eDuKYCN0XECxGxhnTA3b8FcXWnmaZW2hEXEXEzJby305u4ertNJcYVEbE6Dw7PXZ88adPbbZY0gXQi9IOexjAoEkFRrtbZCbhN0iHAnyLi7gazbAc8LunKfGl1hlKDee2Oq+hIYFZfx9STuCLiFuBG4JHcXRcRS9odF+nAf4CkTXLV0T68/IXGPo8rjzpe0sJczbJZnVnqNbXSJwe2XsbVEr2Jq868bY0rV7/cBawEfhkR/SIu4Dukk6QXe7zivri06S8dqVrnDuBwYJO8M0fnsuXUr1I4AniCdOk+DLgCOLbdcRXm3QJYBQzvJ/trW+Dned4RpKqhPdsdVy77PHAXqYroYuCEsuLKw5uTqn6GAF8hvStTO8/7gB8Uhj8MfLfdcRXmnUwfVw31UVwvm7e/xJWnH0M6Gdqh3XEB7wa+l/v3podVQ33+y29XR7pUuw44MQ+/gZS5l+fuBeCPwGtq5nsrMLcw/GHg7HbHVZj/BGBmP9pfnwFOLgyfAny23XHVWc5PgAPLiqtOed0DKrAb6aqpc/hzwOfaHVez5e2Iq7t527m/CtN9kSbr70v++/oa6SpzOfAo8DTw4w1ef1/v6HZ0pBs5FwLfaTDNcuqf4Q4lVSuMz8M/BI5rd1yF8luBffrR/voA8CvS1dNw4Abg4H4Q11DgVbl/R+Be0r2f0uICtij0/xtwSZ15hwEPANuw7mbx9u2Oq1De54mgl/ur27+BNsU1HhiT+18J/Bp4d7vjqpl+b6p8RQDsQbpxs5BUNXAXNWeDxQMI0MHLL9f3y/PeA1wAbNRP4poM/AkY0l/2F+mAey7phtZi4Fv9JK5X5HgWk5Lnm8qOC7go/80sJLWbtUWefktgTmH+A0lPgtwPfL4fxTWLdJ/nedJZZZ9UifYmrmb+BtoU147AgjzNvcAp/eX3WFjO3vQwEbiJCTOziht0Tw2ZmdmGcSIwM6s4JwIzs4pzIjAzqzgnAjOzinMisNJJek9uNfT1hXF7b0hLiZKmS9qyj+IZI+lf+mJZ/VWj/StpeVctuHazzAskHdH76Ky/cSKwVpgG/IbefYp0Oun56fX0oG2oMcCgSgSSSvvsrA1+TgRWKkkjgN2BY1k/EYySdJXSNyDOkTQkN+x1gdI3Be6R9G/5LLQDuDi3y/7KfFZ7iqTfAO+T9BFJ83J78VdI2iSvf/O8jrtz90/A14HX5WWdkdf7vdwW/DWS5nSe+Up6s6SbJN0h6TpJW+Txc5W+f3C7pPskvS2PH5qXOS83FvbRPH4LSTfndd7bOX3NvlpeWObtkrbN48fnbZqXu93z+FMlzZR0PenN1Frr7d866zxR677h8K+F8Ufn+O+WdFGd+U7PvycfQwaDvno7zp27eh3wIeC83P871rW7vjfwDKmxv6GkhuKOAN5Matmxc/7O1/rnAh2F8csptHFEbl4i938Z+ETu/ynwr7l/KDCammYV8nrnkE6MXgM8lscNzzF3Nj/yAXLDXzmeM3P/gcCvcv8M4Au5f2PSdxG2AT5Ffqs4xzGyzr5aXpjmaPJboqR2k/bI/ZOAJbn/VFIjZa+ss6y6+7ewnnF5X99D+n7DCGARqeXL7YGlrHuDe2z+eUHeL98kvV2udv99ueubzpeTVrZppGZyIbXFPw24Mw/fHhEPAEiaRXrV/gbgtZK+S2rl9PoGy/5poX8HSV8mVfuMIDXgBfB20kGViFgLPFGnOd89gMsi4kXgUUk35vH/AOwA/FLpI2xDSU0ydLoy/7yDlFwA3gnsWKhLHw1MAeYB50saDlwdEXd1sU2zCj+/nfv3BaZq3YfgRkkamftnR8Tfu1hWvf17ec12XxXp+w1IuhJ4G6m5g8sj4i8AEVH8ZsHJwG0RMaOLddoA5ERgpZH0KtKBeAdJQTqQhqTP5klq2zeJiHhM0huBdwHHAe8H/lcXq1hT6L8AOCwi7pY0nXRG3HSoDcYviojduih/Nv9cy7r/JZGuRq6rnVjSnqQPiFwk6YyIqFedE3X6hwC71R7wc2Io7oNGy6o33Gi7u2p7Zh7wZkljaxKEDWCu37MyHQFcGBFbR8TkiJgI/IF0Jgqwi6Rtcj3zB4Df5KdZhkTEFaSzz53ztE+RPuPXlZHAI/mM+4OF8TcAH4eX6u9H1VnWb4D35nsFm7MuiSwFxkvaLc8/XNL23WzzdcDHcxxI2k7SppK2BlZGxH8D5xW2q9YHCj9vyf3XA8d3TiDpTd3E0Gm9/VtTfjNwmNLHfDYF3kNqVfMG4P05kSNpbGGeX5Dusfy8cFViA5yvCKxM00gHjaIrgKNI1Tq35PI3kA5KV+X+HxZuQn4u/7wAOEfS30lt/Nc6mfQBmwdJ9d6dB6kTgJmSjiWduX88Im6R9Fulj7ZfC/w78A5Sq5L35eU8ERHP5SqesySNJv2/fIdUl96VH5Cqie5UOmVfBRxGSi6fkfQ8sJpcXVXHxpJuI52kTcvjPgmcLWlhjuFm4GMNYuhUb/++JCLulHQBcHtn7BGxAEDSV4CbJK0ltbo5vTDfZTkJzJZ0YIOqKRsg3PqoGenppohYnc+Cbwd2j4hHWxzDctIN8b+0cr1mviIwS66RNIb08ZjTW50EzNrJVwRmZhXnm8VmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV9z9bjXAkB1gpVAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEWCAYAAABrDZDcAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3deZwdVZn/8c83CyBkIyQikEBQwiggAjYgwy6gLLLIoCwqE36MQRRBQRAXEEFHgUEdHBAyApFdWSdiEBQJoGxJ2EMAAwaJgASFhIRN4vP745wmlZvb3bfTXfcmXd/361WvrqpTy1N1k3qqTlWdUkRgZmbV1a/VAZiZWWs5EZiZVZwTgZlZxTkRmJlVnBOBmVnFORGYmVWcE4GZWcU5EZiZVZwTQQVI2k7SnZLmSfq7pD9I2lLSOEmLJC2o6dbO8w2XdJ2khZKelnRIzXIPyeMXSrpe0vBC2VGSpkl6Q9LEJm7r1yX9Z6H/T3mb5kj6eWG6KZJer9nuXxbKd5H0mKRXJd0qab1C2cqSLpQ0X9Lzko4tlH1I0m/yfp4r6SpJazVp25+QtKGkUZKukfRi/s0fljQuTzNGUtT5zQ/M5ZJ0uqS/5e4MSSqsYzNJ0/N+mS5ps0LZlyQ9lffLs5J+KGlAM7bdesaJoI+TNAS4AfgxMBxYB/g28Eae5K6IGFTTPZvLzgHeBNYEPgX8RNLGebkbA+cDn8nlrwLnFlb9LPAd4MIyt6+OPYHJkv49x7ZrRAwC2oBbaqY9qma79waQNAK4FjiJtM+mAT8vzHcKMBZYD9gZOEHS7rlsdWACMCaXvwJc1NsbWUvSe4B+EfEEcAnwTF7/GsChwF9rZhlWs+3t2zce2A/4ALAp8DHgiLyOlYD/Ay4lbefPgP/L4wF+CWwREUOATfIyji5je62XRYS7PtyRDoAvd1A2Dvh9B2WrkZLAhoVxlwDfz/3/CVxeKHtPnn5wzXK+A0xsMNbVSUlrLvBS7h9VKD8MmEk6uD4FHFFn/heA/sD/AD/qZF1TgP/ooGw8cGfNvngNeG8e/gvwkUL5acCVHSxrC+CVBrZ9L+B+YD7pIH5KTflVwPPAPOB2YOOa8qOBs3P/AmCzDtYzBghgQAfldwLjC8OHA3fn/o/kbVeh/M/A7nWWswbwW+DcBrb9v/M2zwemA9sXyrYC7gJeBp7Lv+tKrfw/1Rc7XxH0fU8AiyT9TNIeklZvcL4NgUWRzjDbPQhsnPs3zsMARMST5MTRg1j7kc6e1wPWJR18/6dQ/gLpDHUIKSn8UNIWhfKPArdExCLgbuBQScdLapPUvxtx1G7bQuBJYOO8/9YulrPkfqm1AzCjgXUuJJ25DyMlhSMl7Vcov5F0FfJO4D7gspr59wR+lfvvBs6RdJCkdRtYd9ES287Sv/lDkY/Q2UOF8vbqwvnAi6QrgvMbWOdUYDPS1dflwFWSVslli4AvAyOAbYBdgM93c5usC04EfVxEzAe2I50F/i8wV9IkSWvmST4k6eVC92QeP4h09lk0DxjcYPmyxPq3iLgmIl6NiFeA7wI7Fsp/FRFPRnIbcDOwfWERewGT87SXAl8kJYfbgBcknVizyrNrtv20BrZtUGG4tmwJkjYFTgaOb2Dbp0TEwxHxz4h4CLiiZtsvjIhXIuINUtXUByQNzetZFdgybyfAJ4A7SFVbf5L0gKQta1b5Ys22v6+DbZ8HDMr3Cbr8zSPi8khVQxsC57F0lVS9bb80//ZvRcRZwMrAv+Sy6RFxdy6bTUosO3ayOFsGTgQVEBEzI2JcRIwi1d2uDfwoF98dEcMK3Xvy+AWkM++iIaRqmUbKu03SqpLOzzeg55OqQIa1n83nK5q7843Yl0lnwSNyWT9gN+DXhe2+LCJ2JZ1lfw44VdJHC6s8umbbT2pg2xYUhjvcbkkbkM7ij4mIOxrY9q3zTem5kubleNu3rb+k70t6Mu+X2Xm2EfnvLqSqrNfzdr8UESdGxMak+zcPANcXb/oCI2q2fWYH2z4EWJCvAhr+zSPij6QroXNry+ps+3GSZuYb2y8DQwvbvqGkG/JN+fmkKskRnS3Pus+JoGIi4jFgIikhdOYJYICksYVxH2BxNceMPAyApHeTzuSKVUnddRzpTHDrfFa5w+LFa2XgGuC/gDUjYhjp7L/94LYlMDsi5tYuNCL+ERFXkaoxutpuWHrbViPdA5kRES+R6qo/UJi+uF9QesLot8BpEXFJA+uDVCUyCRgdEUNJZ9Pt23YIsC+wK+kgOaZ9VflvsVpoCRHxImmfrU2qeunKEtvO0r/5pjUJZVM6rvoaQNpvHZK0PfBV4JPA6vl3ncfibfsJ8BgwNv+b+HqhzHqJE0EfJ+m9+YxrVB4eDRxMqkfuUK4Xv5Z0Fr2apG1JB6P2A9tlwN6Sts8HylOBa3OVDpIG5Hre/kB/Sas08CjhYNJ9gZeVHkX9VqFsJVKimQu8JWkP0s3Ldm9XC+X1j5O0l6TBkvrl6TcG7ukiBoDrgE0k/VvehpNJdeOP5fKLgW9KWl3Se4HPkpIrktYBfgecExHnNbCu4rb/PSJel7QV6eBfLHsD+BuwKumsuGiPmm0/XdIm+TcYDBwJzIqIvzUQx8XAsZLWUXqM+Lj2bSPdYF8EHK30CO1Refzv8nr/Q9I7c/9GwNdY+kmtetv9Ful3HSDpZJa86hhMuom8IO/rIxvYBuuuVt+tdlduR3pc9Bekpz0W5r/nk/6zjSP9x15Q022Z5x0OXJ/n+zNwSM2yD8njF5IeKxxeKDuFdF+i2J3SRaxrkw42C0hXFkdQeMIF+AKpzvllUkK6EvhOLpsGtBWWtT/wB9LTR/OBh4FxhfIpwOs12z29UL4r6Uz0tTztmELZyqTHYufneI4tlH0rx7zEPm3gdzoAeJpUzXID6Sb5pblsUN6/r+RpDs3r2IB0hfNIzbJ+DPwxr3tuXt77ctmYevG1bwPpbPsM4O+5O4MlnxLanPRkz2ukm9abF8ouyvtjIan66kxglS62uz9wQd6XzwEn5Hl3zeU75N9hAem+x6l08KSbu2XvlHe22Qor3/h+AFg7KvYPWtIJpPr+E1odi624/Naf9QVDSWe0lUoC2WzSi1xmy8xXBNZUkr5OuuFX646I2KPZ8TSTpBmkdyRqHRERte8F9Bn5hvCN9coivfVtLeZEYGZWcStc1dCIESNizJgxrQ7DzGyFMn369BcjYmS9shUuEYwZM4Zp06a1OgwzsxWKpKc7KvN7BGZmFedEYGZWcU4EZmYV50RgZlZxTgRmZhXnRGBmVnGlJQKlj3u/IOmRDsol6WxJsyQ9VPOlKTMza5IyrwgmArt3Ur4H6dN7Y0nfiP1JibGYmVkHSksEEXE7qRnbjuwLXBzJ3aQvUa1VVjxmZlZfK98sXgd4pjA8J497rnZCSeNJVw2su253v8W92JgT637EycxshTD7+3uVstxW3iyu97m5ui3gRcSEiGiLiLaRI+s2lWFmZsuolYlgDjC6MDwKeLZFsZiZVVYrE8Ek4ND89NCHgHkRsVS1kJmZlau0ewSSrgB2AkZImkP6lutAgEgf9Z4M7AnMAl4FDisrFjMz61hpiSAiDu6iPEgfIzczsxbym8VmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgZlZxTgRmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgZlZxTgRmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgZlZxTgRmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgZlZxTgRmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgZlZxTgRmZhXnRGBmVnGlJgJJu0t6XNIsSSfWKV9X0q2S7pf0kKQ9y4zHzMyWVloikNQfOAfYA9gIOFjSRjWTfRP4RURsDhwEnFtWPGZmVl+ZVwRbAbMi4qmIeBO4Eti3ZpoAhuT+ocCzJcZjZmZ1lJkI1gGeKQzPyeOKTgE+LWkOMBn4Yr0FSRovaZqkaXPnzi0jVjOzyiozEajOuKgZPhiYGBGjgD2BSyQtFVNETIiItohoGzlyZAmhmplVV5mJYA4wujA8iqWrfg4HfgEQEXcBqwAjSozJzMxqlJkIpgJjJa0vaSXSzeBJNdP8GdgFQNL7SInAdT9mZk1UWiKIiLeAo4CbgJmkp4NmSDpV0j55suOAz0p6ELgCGBcRtdVHZmZWogFlLjwiJpNuAhfHnVzofxTYtswYzMysc36z2Mys4pwIzMwqzonAzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4pwIzMwqzonAzKzinAjMzCquy0Qg6QxJQyQNlHSLpBclfboZwZmZWfkauSL4SETMBz5Galp6Q+D4UqMyM7OmaSQRDMx/9wSuiIi/lxiPmZk1WSOtj/5S0mPAa8DnJY0EXi83LDMza5Yurwgi4kRgG6AtIv4BvArs0/lcZma2omjkZvGFEfFSRCxqHwVcXm5YZmbWLI3cI/iLpJ8ASFoduBm4tNSozMysaRqpGjoJmC/pPFISOCsiLio9MjMza4oObxZL2r8weC9wUv4bkvaPiGvLDs7MzMrX2VNDe9cM3096lHRvIAAnAjOzPqDDRBARhzUzEDMza41Gnhr6maRhheHVJV1YblhmZtYsjTw1tGlEvNw+EBEvAZuXF5KZmTVTI4mgX35sFABJw2nsjWQzM1sBNHJAPwu4U9LVefgTwHfLC8nMzJqpy0QQERdLmgZ8OI/aPyIeLTcsMzNrlkY/TDOQ1LREe7+ZmfURjTw1dAxwGTACeCdwqaQvlh2YmZk1RyP3CA4Hto6IhQCSTgfuAn5cZmBmZtYcjVQNCVhUGF7E4moiMzNbwTVyRXARcI+k6/LwfsAF5YVkZmbN1MhTQz+QNAXYjnQlcFhE3F92YGZm1hwdVg1JGt7eAbNJ3yC4BHg6j+uSpN0lPS5plqQTO5jmk5IelTRDkj94Y2bWZJ1dEUwntTJa735AAO/ubMGS+gPnALsBc4CpkiYV30GQNBb4GrBtRLwk6Z3djN/MzHqos9ZH1+/hsrcCZkXEUwCSrgT2BYovo30WOCe3X0REvNDDdZqZWTc11GZQ/kjNdqQrgTsi4voGZlsHeKYwPAfYumaaDfPy/wD0B06JiF83EpOZmfWOLhOBpHOBDYAr8qjPSdotIr7Q1ax1xkWd9Y8FdgJGAXdI2qTY2mmOYTwwHmDdddftKmQzM+uGRq4IdgQ2iYiA9H0C4OEG5psDjC4MjwKerTPN3RHxD+BPkh4nJYapxYkiYgIwAaCtra02mZiZWQ808kLZ40DxNHw08FAD800FxkpaX9JKwEHApJpprgd2BpA0glRV9FQDyzYzs17S2cfrf0mqyhkKzJR0bx7eGrizqwVHxFuSjgJuItX/XxgRMySdCkyLiEm57COSHiW9sXx8RPytpxtlZmaN66xq6L96uvCImAxMrhl3cqE/gGNzZ2ZmLdDZ46O3NTMQMzNrjUa/R2BmZn2UE4GZWcV11tbQLfnv6c0Lx8zMmq2zm8VrSdoR2Cc3D7HEC2IRcV+pkZmZWVN0lghOBk4kvQj2g5qyYPHH7M3MbAXW2VNDVwNXSzopIk5rYkxmZtZEjXyY5jRJ+wA75FFTIuKGcsMyM7Nm6fKpIUnfA44hNR/9KHBMHmdmZn1AI43O7QVsFhH/hLcbnbuf9EEZMzNbwTX6HsGwQv/QMgIxM7PWaOSK4HvA/ZJuJT1CugO+GjAz6zMauVl8haQpwJakRPDViHi+7MDMzKw5GvpUZUQ8x9LfEjAzsz7AbQ2ZmVWcE4GZWcV1mggk9ZP0SLOCMTOz5us0EeR3Bx6UtG5n05mZ2YqrkZvFawEz8jeLF7aPjIh9SovKzMyappFE8O3SozAzs5Zp5D2C2yStB4yNiN9KWhXoX35oZmbWDI00OvdZ4Grg/DxqHeD6MoMyM7PmaeTx0S8A2wLzASLij8A7ywzKzMyap5FE8EZEvNk+IGkA6QtlZmbWBzSSCG6T9HXgHZJ2A64CflluWGZm1iyNJIITgbnAw8ARwGTgm2UGZWZmzdPIU0P/zB+juYdUJfR4RLhqyMysj+gyEUjaCzgPeJLUDPX6ko6IiBvLDs7MzMrXyAtlZwE7R8QsAEnvAX4FOBGYmfUBjdwjeKE9CWRPAS+UFI+ZmTVZh1cEkvbPvTMkTQZ+QbpH8AlgahNiMzOzJuisamjvQv9fgR1z/1xg9dIiMjOzpuowEUTEYc0MxMzMWqORtobWl/QDSddKmtTeNbJwSbtLelzSLEkndjLdAZJCUlt3gjczs55r5Kmh64ELSG8T/7PRBUvqD5wD7AbMAaZKmhQRj9ZMNxg4mvSegpmZNVkjieD1iDh7GZa9FTArIp4CkHQlsC/waM10pwFnAF9ZhnWYmVkPNfL46H9L+pakbSRt0d41MN86wDOF4Tl53NskbQ6MjogbOluQpPGSpkmaNnfu3AZWbWZmjWrkiuD9wGeAD7O4aijycGdUZ9zbTVNI6gf8EBjXVQARMQGYANDW1ubmLczMelEjieDjwLuLTVE3aA4wujA8Cni2MDwY2ASYIgngXcAkSftExLRursvMzJZRI1VDDwLDlmHZU4Gx+amjlYCDgLefNoqIeRExIiLGRMQY4G7AScDMrMkauSJYE3hM0lTgjfaREbFPZzNFxFuSjgJuIn3j+MKImCHpVGBaRDT0CKqZmZWrkUTwrWVdeERMJn2/oDju5A6m3WlZ12NmZsuuke8R3NaMQMzMrDUa+R7BKyx+2mclYCCwMCKGlBmYmZk1RyNXBIOLw5L2I70sZmZmfUAjTw0tISKup+t3CMzMbAXRSNXQ/oXBfkAbhRfDzMxsxdbIU0PF7xK8BcwmtRlkZmZ9QCP3CPxdAjOzPqyzT1XWfd4/i4g4rYR4zMysyTq7IlhYZ9xqwOHAGqTmo83MbAXX2acqz2rvzx+POQY4DLgSOKuj+czMbMXS6T0CScOBY4FPAT8DtoiIl5oRmJmZNUdn9wjOBPYnfQfg/RGxoGlRmZlZ03T2QtlxwNrAN4FnJc3P3SuS5jcnPDMzK1tn9wi6/daxmZmteHywNzOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOzinMiMDOrOCcCM7OKcyIwM6s4JwIzs4pzIjAzqzgnAjOziis1EUjaXdLjkmZJOrFO+bGSHpX0kKRbJK1XZjxmZra00hKBpP7AOcAewEbAwZI2qpnsfqAtIjYFrgbOKCseMzOrr8wrgq2AWRHxVES8CVwJ7FucICJujYhX8+DdwKgS4zEzszrKTATrAM8UhufkcR05HLixXoGk8ZKmSZo2d+7cXgzRzMzKTASqMy7qTih9GmgDzqxXHhETIqItItpGjhzZiyGamdmAEpc9BxhdGB4FPFs7kaRdgW8AO0bEGyXGY2ZmdZR5RTAVGCtpfUkrAQcBk4oTSNocOB/YJyJeKDEWMzPrQGmJICLeAo4CbgJmAr+IiBmSTpW0T57sTGAQcJWkByRN6mBxZmZWkjKrhoiIycDkmnEnF/p3LXP9ZmbWNb9ZbGZWcU4EZmYV50RgZlZxTgRmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgZlZxTgRmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgZlZxTgRmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgZlZxTgRmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgZlZxTgRmZhXnRGBmVnFOBGZmFedEYGZWcU4EZmYV50RgZlZxpSYCSbtLelzSLEkn1ilfWdLPc/k9ksaUGY+ZmS2ttEQgqT9wDrAHsBFwsKSNaiY7HHgpIjYAfgicXlY8ZmZWX5lXBFsBsyLiqYh4E7gS2Ldmmn2Bn+X+q4FdJKnEmMzMrMaAEpe9DvBMYXgOsHVH00TEW5LmAWsALxYnkjQeGJ8HF0h6fBljGlG77OWE4+oex9V9y2tsjqsbdHqP4lqvo4IyE0G9M/tYhmmIiAnAhB4HJE2LiLaeLqe3Oa7ucVzdt7zG5ri6p6y4yqwamgOMLgyPAp7taBpJA4ChwN9LjMnMzGqUmQimAmMlrS9pJeAgYFLNNJOAf8/9BwC/i4ilrgjMzKw8pVUN5Tr/o4CbgP7AhRExQ9KpwLSImARcAFwiaRbpSuCgsuLJely9VBLH1T2Oq/uW19gcV/eUEpd8Am5mVm1+s9jMrOKcCMzMKq5PJAJJoyXdKmmmpBmSjqkp/4qkkDSig/nXlXRznv/R3mrqoidxSdpZ0gOF7nVJ+7U6rlx+Rp5vpqSze+slwF6I63RJj+TuwN6IqbO4JJ0i6S+F32jPDubvtKmVFsZ1oaQXJD3SWzH1NK6u/g20MK5VJN0r6cE877eXh7gKy+gv6X5JNyxTEBGxwnfAWsAWuX8w8ASwUR4eTbph/TQwooP5pwC75f5BwKrLQ1yF5Qwn3UxveVzAvwJ/ID0A0B+4C9hpOYhrL+A3pAcgVgOmAUPKjAs4BfhKF/P2B54E3g2sBDzYvk2tjCvPswOwBfBIb8TTS/urw38DLY5LwKDcPxC4B/hQq+MqLONY4HLghmWJoU9cEUTEcxFxX+5/BZhJemsZUhtGJ1DnRTUApfaPBkTEb/L8CyLi1VbHVeMA4MblJK4AViEd1FYm/af463IQ10bAbRHxVkQsJB1wd29CXF1ppKmVVsRFRNxOCe/t9CSunm5TiXFFRCzIgwNz1ytP2vR0myWNIp0I/XRZY+gTiaAoV+tsDtwjaR/gLxHxYCezbAi8LOnafGl1plKDea2Oq+gg4IrejmlZ4oqIu4Bbgedyd1NEzGx1XKQD/x6SVs1VRzuz5AuNvR5XHnWUpIdyNcvqdWap19RKrxzYehhXU/QkrjrztjSuXP3yAPAC8JuIWC7iAn5EOkn65zKvuDcubZaXjlStMx3YH1g178yhuWw29asUDgDmkS7dBwDXAIe3Oq7CvGsBc4GBy8n+2gD4VZ53EKlqaIdWx5XLvgE8QKoiugw4pqy48vCapKqffsB3Se/K1M7zCeCnheHPAD9udVyFecfQy1VDvRTXEvMuL3Hl6YeRToY2aXVcwMeAc3P/Tixj1VCv//it6kiXajcBx+bh95My9+zcvQX8GXhXzXwfAqYUhj8DnNPquArzHwNMWI721/HASYXhk4ETWh1XneVcDuxZVlx1yuseUIFtSFdN7cNfA77W6rgaLW9FXF3N28r9VZjuWzRYf1/yv6/vka4yZwPPA68Cl3Z7/b29o1vRkW7kXAz8qJNpZlP/DLc/qVphZB6+CPhCq+MqlN8N7Lwc7a8Dgd+Srp4GArcAey8HcfUH1sj9mwKPkO79lBYXsFah/8vAlXXmHQA8BazP4pvFG7c6rkJ5ryeCHu6vLv8NtCiukcCw3P8O4A7gY62Oq2b6najyFQGwHenGzUOkqoEHqDkbLB5AgDaWvFzfLc/7MDARWGk5iWsM8Beg3/Kyv0gH3PNJN7QeBX6wnMS1So7nUVLy3KzsuIBL8r+Zh0jtZq2Vp18bmFyYf0/SkyBPAt9YjuK6gnSf5x+ks8peqRLtSVyN/BtoUVybAvfnaR4BTl5efsfCcnZiGROBm5gwM6u4PvfUkJmZdY8TgZlZxTkRmJlVnBOBmVnFORGYmVWcE4GVTtLHc6uh7y2M26k7LSVKGidp7V6KZ5ikz/fGspZXne1fSbM7asG1i2VOlHRAz6Oz5Y0TgTXDwcDv6dmnSMeRnp9eyjK0DTUM6FOJQFJpn521vs+JwEolaRCwLXA4SyeCIZKuU/oGxHmS+uWGvSYqfVPgYUlfzmehbcBluV32d+Sz2pMl/R74hKTPSpqa24u/RtKqef1r5nU8mLt/Bb4PvCcv68y83nNzW/A3SJrcfuYr6YOSbpM0XdJNktbK46coff/gXklPSNo+j++flzk1NxZ2RB6/lqTb8zofaZ++Zl/NLizzXkkb5PEj8zZNzd22efwpkiZIupn0ZmqtpfZvnXUeq8XfcPhSYfyhOf4HJV1SZ77T8u/kY0hf0Ftvx7lzV68DPg1ckPvvZHG76zsBr5Ma++tPaijuAOCDpJYd2+dvf61/CtBWGD+bQhtH5OYlcv93gC/m/p8DX8r9/YGh1DSrkNc7mXRi9C7gpTxuYI65vfmRA8kNf+V4zsr9ewK/zf3jgW/m/pVJ30VYHziO/FZxjmNwnX01uzDNoeS3REntJm2X+9cFZub+U0iNlL2jzrLq7t/Cekbkff0w6fsNg4AZpJYvNwYeZ/Eb3MPz34l5v5xBertcrf735a53Ol9OWtkOJjWTC6kt/oOB+/LwvRHxFICkK0iv2t8CvFvSj0mtnN7cybJ/XujfRNJ3SNU+g0gNeAF8mHRQJSIWAfPqNOe7HXBVRPwTeF7SrXn8vwCbAL9R+ghbf1KTDO2uzX+nk5ILwEeATQt16UOBscBU4EJJA4HrI+KBDrbpisLfH+b+XYGNtPhDcEMkDc79kyLitQ6WVW//Xl2z3ddF+n4Dkq4Ftic1d3B1RLwIEBHFbxacBNwTEeM7WKetgJwIrDSS1iAdiDeRFKQDaUg6IU9S275JRMRLkj4AfBT4AvBJ4P91sIqFhf6JwH4R8aCkcaQz4oZD7WT8jIjYpoPyN/LfRSz+vyTS1chNtRNL2oH0AZFLJJ0ZEfWqc6JOfz9gm9oDfk4MxX3Q2bLqDXe23R21PTMV+KCk4TUJwlZgrt+zMh0AXBwR60XEmIgYDfyJdCYKsJWk9XM984HA7/PTLP0i4hrS2ecWedpXSJ/x68hg4Ll8xv2pwvhbgCPh7fr7IXWW9Xvg3/K9gjVZnEQeB0ZK2ibPP1DSxl1s803AkTkOJG0oaTVJ6wEvRMT/AhcUtqvWgYW/d+X+m4Gj2ieQtFkXMbRbav/WlN8O7Kf0MZ/VgI+TWtW8BfhkTuRIGl6Y59ekeyy/KlyV2ArOVwRWpoNJB42ia4BDSNU6d+Xy95MOStfl/osKNyG/lv9OBM6T9Bqpjf9aJ5E+YPM0qd67/SB1DDBB0uGkM0aSlIAAAADbSURBVPcjI+IuSX9Q+mj7jcBXgV1IrUo+kZczLyLezFU8Z0saSvr/8iNSXXpHfkqqJrpP6ZR9LrAfKbkcL+kfwAJydVUdK0u6h3SSdnAedzRwjqSHcgy3A5/rJIZ29fbv2yLiPkkTgXvbY4+I+wEkfRe4TdIiUqub4wrzXZWTwCRJe3ZSNWUrCLc+akZ6uikiFuSz4HuBbSPi+SbHMJt0Q/zFZq7XzFcEZskNkoaRPh5zWrOTgFkr+YrAzKzifLPYzKzinAjMzCrOicDMrOKcCMzMKs6JwMys4v4/2y+GWwEmjpIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plots for the primary reconstruction\n",
    "\n",
    "counter = 1\n",
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    \n",
    "    if isinstance(node.genome, pyham.AncestralGenome):\n",
    "        title = node.genome.name  \n",
    "        raw_values = [len(block) for block in node.genome.synteny]  \n",
    "        label = list(set(raw_values))\n",
    "        values = [0 for i in range(len(label))]\n",
    "        \n",
    "        for i in range(len(label)):\n",
    "            \n",
    "            for element in raw_values:\n",
    "                \n",
    "                if label[i] == element:\n",
    "                    values[i] += 1\n",
    "\n",
    "        plt.bar(list(label), values)\n",
    "        plt.title(title)\n",
    "        plt.xlabel('Abstractgenes per block')\n",
    "        plt.ylabel('Number of blocks')\n",
    "        plt.savefig('results/primary_reconstruction_plot{}.png'.format(counter))\n",
    "        counter += 1\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.1.7'"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## slightly more complicated 3 species example\n",
    "\n",
    "# Path to nwk file\n",
    "nwk_path = \"8may_medhighMut/RealTree.nwk\"\n",
    "\n",
    "# The newick tree is extracted as a string\n",
    "tree_str = pyham.utils.get_newick_string(nwk_path, type=\"nwk\")\n",
    "\n",
    "# Path to OrthoXML file\n",
    "orthoxml_path = \"8may_medhighMut/Output/HierarchicalGroups.orthoxml\"\n",
    "\n",
    "ham_analysis_no_name = pyham.Ham(tree_str, orthoxml_path, use_internal_name=False)\n",
    "\n",
    "pyham.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "neighbors assigned\n",
      "[[Gene(1), Gene(2), Gene(3), Gene(4), Gene(5), Gene(6), Gene(7), Gene(8), Gene(9), Gene(10), Gene(11), Gene(12), Gene(13), Gene(14), Gene(15), Gene(16), Gene(17), Gene(18), Gene(19), Gene(20), Gene(21), Gene(22), Gene(23), Gene(24), Gene(25)]]\n",
      "next node\n",
      "neighbors assigned\n",
      "[[Gene(26), Gene(27), Gene(28), Gene(29), Gene(30), Gene(31), Gene(32), Gene(33), Gene(34), Gene(35), Gene(36), Gene(37), Gene(38), Gene(39), Gene(40), Gene(41), Gene(42), Gene(43), Gene(44), Gene(45), Gene(46), Gene(47), Gene(48), Gene(49), Gene(50)]]\n",
      "next node\n",
      "neighbors assigned\n",
      "[[Gene(51), Gene(52), Gene(53), Gene(54), Gene(55), Gene(56), Gene(57), Gene(58), Gene(59), Gene(60), Gene(61), Gene(62), Gene(63), Gene(64), Gene(65), Gene(66), Gene(67), Gene(68), Gene(69), Gene(70), Gene(71), Gene(72), Gene(73), Gene(74), Gene(75)]]\n",
      "next node\n",
      "{}\n",
      "Gene(30) None not in parent dict\n",
      "Gene(32) <HOG()> in parent dict\n",
      "{'SE002_aa': {<HOG()>, None}}\n",
      "Gene(59) None not in parent dict\n",
      "Gene(61) <HOG(4)> in parent dict\n",
      "{'SE002_aa': [[Gene(30), Gene(32)]], 'SE003_aa': [[Gene(59), Gene(61)]]}\n",
      "{'SE002_aa': {<HOG()>, None}, 'SE003_aa': {<HOG(4)>, None}}\n",
      "SE002_aa {<HOG()>, None} []\n",
      "SE003_aa {<HOG(4)>, None} [{<HOG()>, None}]\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(33) None not in parent dict\n",
      "Gene(35) <HOG(3)> in parent dict\n",
      "{'SE002_aa': {<HOG(3)>, None}}\n",
      "Gene(67) None not in parent dict\n",
      "Gene(69) <HOG()> in parent dict\n",
      "{'SE002_aa': [[Gene(33), Gene(35)]], 'SE003_aa': [[Gene(67), Gene(69)]]}\n",
      "{'SE002_aa': {<HOG(3)>, None}, 'SE003_aa': {<HOG()>, None}}\n",
      "SE002_aa {<HOG(3)>, None} []\n",
      "SE003_aa {<HOG()>, None} [{<HOG(3)>, None}]\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(34) <HOG(2)> not in parent dict\n",
      "Gene(36) None in parent dict\n",
      "{'SE002_aa': {<HOG(2)>, None}}\n",
      "Gene(52) <HOG()> in parent dict\n",
      "{'SE002_aa': [[Gene(34), Gene(36)]], 'SE003_aa': [[None, Gene(52)]]}\n",
      "{'SE002_aa': {<HOG(2)>, None}, 'SE003_aa': {<HOG()>, None}}\n",
      "SE002_aa {<HOG(2)>, None} []\n",
      "SE003_aa {<HOG()>, None} [{<HOG(2)>, None}]\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(42) None not in parent dict\n",
      "Gene(44) None in parent dict\n",
      "{'SE002_aa': {None}}\n",
      "Gene(60) <HOG(1)> not in parent dict\n",
      "Gene(62) None in parent dict\n",
      "{'SE002_aa': [[Gene(42), Gene(44)]], 'SE003_aa': [[Gene(60), Gene(62)]]}\n",
      "{'SE002_aa': {None}, 'SE003_aa': {<HOG(1)>, None}}\n",
      "SE002_aa {None} []\n",
      "SE003_aa {<HOG(1)>, None} [{None}]\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(48) None not in parent dict\n",
      "Gene(50) <HOG()> in parent dict\n",
      "{'SE002_aa': {<HOG()>, None}}\n",
      "Gene(63) None not in parent dict\n",
      "Gene(65) None in parent dict\n",
      "{'SE002_aa': [[Gene(48), Gene(50)]], 'SE003_aa': [[Gene(63), Gene(65)]]}\n",
      "{'SE002_aa': {<HOG()>, None}, 'SE003_aa': {None}}\n",
      "SE002_aa {<HOG()>, None} []\n",
      "SE003_aa {None} [{<HOG()>, None}]\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(37) None not in parent dict\n",
      "Gene(39) None in parent dict\n",
      "{'SE002_aa': {None}}\n",
      "Gene(68) <HOG(2)> not in parent dict\n",
      "Gene(70) None in parent dict\n",
      "{'SE002_aa': [[Gene(37), Gene(39)]], 'SE003_aa': [[Gene(68), Gene(70)]]}\n",
      "{'SE002_aa': {None}, 'SE003_aa': {<HOG(2)>, None}}\n",
      "SE002_aa {None} []\n",
      "SE003_aa {<HOG(2)>, None} [{None}]\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(31) <HOG(1)> not in parent dict\n",
      "Gene(33) None in parent dict\n",
      "{'SE002_aa': [[Gene(31), Gene(33)]]}\n",
      "{'SE002_aa': {<HOG(1)>, None}}\n",
      "SE002_aa {<HOG(1)>, None} []\n",
      "fin (<HOG(1)>, None)\n",
      "{}\n",
      "Gene(49) <HOG(5)> not in parent dict\n",
      "{'SE002_aa': [[Gene(49), None]]}\n",
      "{'SE002_aa': {<HOG(5)>, None}}\n",
      "SE002_aa {<HOG(5)>, None} []\n",
      "fin (<HOG(5)>, None)\n",
      "{}\n",
      "Gene(57) None not in parent dict\n",
      "Gene(59) None in parent dict\n",
      "{'SE003_aa': [[Gene(57), Gene(59)]]}\n",
      "{'SE003_aa': {None}}\n",
      "SE003_aa {None} []\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(51) <HOG(3)> not in parent dict\n",
      "Gene(53) None in parent dict\n",
      "{'SE003_aa': [[Gene(51), Gene(53)]]}\n",
      "{'SE003_aa': {<HOG(3)>, None}}\n",
      "SE003_aa {<HOG(3)>, None} []\n",
      "fin (<HOG(3)>, None)\n",
      "{}\n",
      "Gene(72) None not in parent dict\n",
      "Gene(74) None in parent dict\n",
      "{'SE003_aa': [[Gene(72), Gene(74)]]}\n",
      "{'SE003_aa': {None}}\n",
      "SE003_aa {None} []\n",
      "len1 (None, None)\n",
      "neighbors assigned\n",
      "[[<HOG(1)>], [<HOG(2)>], [<HOG(3)>], [<HOG(4)>], [<HOG(5)>], [<HOG()>], [<HOG()>], [<HOG()>], [<HOG()>], [<HOG()>], [<HOG()>]]\n",
      "next node\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAb+0lEQVR4nO3dd7xdVZn/8c83CUVIQhITGHoAKQOIA15EBQFp0kHAElAMw88go4iKBecnZUQUR1FGRSFK6EWpgxQFMwQGqQmhJARQMCrNXKWEhB6e+WOtC5vDLfuWfc692d/367VfZ7ez1rNPcp+zztp7r62IwMzM6mNYqwMwM7PmcuI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34bcJK2kXSLpGclPSXp95K2lDRZ0hJJixqm1fL7xkm6XNJiSX+WdGBDuQfm9YslXSFpXF6/nKQz8rbnJM2WtFuTjvXfJX27MP+nfEyPSvplYb8Zkl5sOO5fF7bvKOkBSc9LukHS2oVty0maJmmhpCclfamwbWNJMyU9naffSdq4GcduQ5cTvw0oSaOBq4AfA+OA1YH/AF7Ku9waESMbpsfztlOBl4FVgIOAn0naJJe7CXA68Mm8/Xngp/l9I4C/AtsBKwHHAL+SNLHCQ+2wO3CNpE/l2HaKiJFAGzC9Yd/PNRz3XgCSxgOX5bjHATOBXxbedzywPrA28EHgq5J2zdseBw7I7xsPXAlcNOBHaUuXiPDkacAmUsJ7pottk4Gbu9i2Iinpb1BYdy5wUp7/NnBBYdt6ef9RXZR3L7B/D7GOJX1JtQNP5/k1CtsPAeYBzwGPAId18v4FwHDgJ8Ap3dQ1A/h/XWybAtzS8Fm8AGyUlx8DdilsPwG4qJNyRgCfBZ4v8e+0BzAbWEj60jy+YfvFwJPAs8BNwCat/r/laeAmt/htoD0ELJF0tqTdJI0t+b4NgCUR8VBh3T3AJnl+k7wMQEQ8TP6iaCxI0ip5/dwe6hwGnElqSa9FSrY/KWxfAOwJjCZ9CfxQ0haF7R8CpkfEEuA24GBJX5HUJml4D3UXNR7bYuBhYJP8+a1W3M6bPxcAJD0DvEj6pfXtEnUuBg4GxpC+BA6XtG9h+7WkXxkrA3cB5/fieGyQc+K3ARURC4FtgAB+DrRLujInY4D3SnqmMD2c148ktS6LngVGldwOgKRlSEnq7Ih4oIdY/xERl0bE8xHxHHAiqbuoY/vVEfFwJDcC1wEfKBSxB3BN3vc84AjSl8GNwAJJRzdU+aOGYz+hxLGNLCx3edwRMYbUzfU5Uku+WxExIyLui4jXIuJe4MKGY58WEc9FxEukrqZ3SVqpp3JtaHDitwEXEfMiYnJErAFsSmqxnpI33xYRYwrTenn9IlLLumg0qZulzHYkDSN1D71MSoDdkrSCpNPzSeGFpC6NMR2t9fyL5bZ8gvoZUn/++EJdOwO/KRz3+RGxE6kV/Rngm5I+VKjy8w3HfkyJY1tUWO70uAv1LwZOA86RtHIPx75VPoncLunZHG/HsQ2XdJKkh/PnMj+/bXx3ZdrQ4cRvlcqt7rNIXwDdeQgYIWn9wrp38UZ3zdy8DICkdYHl8vuQJOAM0onf/SPilRLhHQVsCGwVEaOBbd8oXssBlwLfB1bJLeprAOV9tgTmR0R7J8f8SkRcTDrP0NNxd3ZsK5LOYcyNiKeBJ4rbefPn0mgYsALppHp3LiCdCF4zIlYifWF0HNuBwD7ATqRfERM7QitxLDYEOPHbgJK0kaSjJK2Rl9cEJpH6wLuUW6uXkVrJK0rampR8zs27nA/sJekDOTF+E7gsd9EA/Az4Z2CviHihZLijSP36z+RLQ48rbFuW9MXSDryaLw/dpbD99W6efJyTJe0haZSkYXn/TYDbS8RxObCppP0lLQ8cC9xb6Ko6B/iGpLGSNgI+TfoyRdLOkjbPrfTRwA9IJ6rnlTj2pyLiRUnvISX74raXgH+QvkTKnDOwoaTVZ5c9LV0TqaX5K9KVKIvz6+mk7onJwBJS90Vx2jK/dxxwRX7fX4ADG8o+MK9fDPw3MC6vX5t0TuHFhnIP6iHW1UhX2ywi/XI4LJczIm//LPA34BnSF9BFwLfytplAW6Gs/YDfk5LuQuA+YHJh+4xO4ptV2L4T8ADpi2gGMLGwbTlgWi73b8CXCts+kt+3iPQldQ2wWYl/pwOAP5O6jK4indQ+L28bmT/f5/I+B+fP5R2t/v/laWAm5X9oMyspn6i+G1gt/AdkQ5C7esx6byVSq9tJ34YkJ35bquVhFBqHiFgk6dq+lhkRD0XEhQMZZxUkze3i2A9qdWzWWu7qMTOrmRGtDqCM8ePHx8SJE1sdhpnZkDJr1qy/R8SExvVDIvFPnDiRmTNntjoMM7MhRdKfO1vvPn4zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmhkSd+72x8Sjr251CGZmfTL/pD0qKdctfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrmcoSv6RpkhZImlNYN07S9ZL+kF/HVlW/mZl1rsoW/1nArg3rjgamR8T6wPS8bGZmTVRZ4o+Im4CnGlbvA5yd588G9q2qfjMz61yz+/hXiYgnAPLryk2u38ys9gbtyV1JUyTNlDSzvb291eGYmS01mp34/yZpVYD8uqCrHSNiakS0RUTbhAkTmhagmdnSrtmJ/0rgU3n+U8B/N7l+M7Paq/JyzguBW4ENJT0q6VDgJGBnSX8Ads7LZmbWRCOqKjgiJnWxaceq6jQzs54N2pO7ZmZWDSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczq5mWJH5JX5Q0V9IcSRdKWr4VcZiZ1VHTE7+k1YHPA20RsSkwHPh4s+MwM6urVnX1jADeJmkEsALweIviMDOrnaYn/oh4DPg+8BfgCeDZiLiucT9JUyTNlDSzvb292WGamS21ekz8kv5T0mhJy0iaLunvkj7R1woljQX2AdYBVgNW7Ky8iJgaEW0R0TZhwoS+VmdmZg3KtPh3iYiFwJ7Ao8AGwFf6UedOwJ8ioj0iXgEuA97fj/LMzKwXyiT+ZfLr7sCFEfFUP+v8C/BeSStIErAjMK+fZZqZWUllEv+vJT0AtAHTJU0AXuxrhRFxO3AJcBdwX45hal/LMzOz3hnR0w4RcbSk7wILI2KJpOeBvftTaUQcBxzXnzLMzKxvypzcnRYRT0fEko5VwAXVhmVmZlUp09XzmKSfwetX5FwHnFdpVGZmVpkeE39EHAMslHQaKemfHBFnVh6ZmZlVoss+fkn7FRbvAI7JryFpv4i4rOrgzMxs4HV3cnevhuXZpEs79wKCdP29mZkNMV0m/og4pJmBmJlZc5S5qudsSWMKy2MlTas2LDMzq0qZq3o2i4hnOhYi4mlg8+pCMjOzKpVJ/MPyZZwASBpHiRu/zMxscCqTwE8GbpF0SV7+CHBidSGZmVmVygzZcI6kmcAOedV+EXF/tWGZmVlVyj6IZRnSUA0d82ZmNkSVuarnSOB8YDywMnCepCOqDszMzKpRpo//UGCriFgMkEfqvBX4cZWBmZlZNcp09QhYUlhewhvdPmZmNsSUafGfCdwu6fK8vC9wRnUhmZlZlcpc1fMDSTOAbUgt/UMiYnbVgZmZWTW6G51zXGFxfp5e3zYAz941M7MW6K7FP4s0Cmdn/fkBrFtJRGZmVqnuRudcp5mBmJlZc5Qacyc/lGUbUkv/fyPiikqjMjOzypS5geunwGeA+4A5wGcknVp1YGZmVo0yLf7tgE0jIiCNz0/6EjAzsyGozA1cDwJrFZbXBO6tJhwzM6tad5dz/prUp78SME/SHXl5K+CW5oRnZmYDrbuunu83LQozM2ua7i7nvLGZgZiZWXOUHY/fzMyWEk78ZmY102XilzQ9v353oCuVNEbSJZIekDRP0vsGug4zM+tcdyd3V5W0HbC3pItoGLMnIu7qR73/BfwmIg6QtCywQj/KMjOzXugu8R8LHA2sAfygYVvwxsPXe0XSaGBbYDJARLwMvNyXsszMrPe6u6rnEuASScdExAkDWOe6QDtwpqR3kUYBPbLj0Y4dJE0BpgCstdZabynEzMz6pseTuxFxgqS9JX0/T3v2s84RwBbAzyJic2Ax6ZdFY71TI6ItItomTJjQzyrNzKxDmUHavgMcCdyfpyPzur56FHg0Im7Py5eQvgjMzKwJygzStgfwLxHxGrw+SNts4Ot9qTAinpT0V0kbRsSDwI6kLxQzM2uCUuPxA2OAjkctrjQA9R4BnJ+v6HkEOGQAyjQzsxLKJP7vALMl3UC6pHNb+tja7xARdwNt/SnDzMz6psfEHxEXSpoBbElK/F+LiCerDszMzKpRqqsnIp4Arqw4FjMzawKP1WNmVjNO/GZmNdNt4pc0TNKcZgVjZmbV6zbx52v375HkMRPMzJYSZU7urgrMzc/cfX08nYjYu7KozMysMmUS/39UHoWZmTVNmev4b5S0NrB+RPxO0grA8OpDMzOzKpQZpO3TpIHUTs+rVgeuqDIoMzOrTpnLOT8LbA0sBIiIPwArVxmUmZlVp0zifyk/JQsASSNIT+AyM7MhqEziv1HSvwNvk7QzcDHw62rDMjOzqpRJ/EeTHpV4H3AYcA3wjSqDMjOz6pS5que1/PCV20ldPA9GhLt6zMyGqB4Tv6Q9gNOAh0nDMq8j6bCIuLbq4MzMbOCVuYHrZOCDEfFHAEnrAVcDTvxmZkNQmT7+BR1JP3sEWFBRPGZmVrEuW/yS9suzcyVdA/yK1Mf/EeDOJsRmZmYV6K6rZ6/C/N+A7fJ8OzC2sojMzKxSXSb+iDikmYGYmVlzlLmqZx3gCGBicX8Py2xmNjSVuarnCuAM0t26r1UbjpmZVa1M4n8xIn5UeSRmZtYUZRL/f0k6DrgOeKljZUTcVVlUZmZWmTKJ/53AJ4EdeKOrJ/KymZkNMWUS/4eBdYtDM5uZ2dBV5s7de4AxVQdiZmbNUabFvwrwgKQ7eXMff78u55Q0HJgJPBYRe/anLDMzK69M4j+uorqPBOYBoysq38zMOlFmPP4bB7pSSWsAewAnAl8a6PLNzKxrPfbxS3pO0sI8vShpiaSF/az3FOCrdHNDmKQpkmZKmtne3t7P6szMrEOPiT8iRkXE6DwtD+wP/KSvFUrakzTU86we6p0aEW0R0TZhwoS+VmdmZg3KXNXzJhFxBf27hn9rYG9J84GLgB0kndeP8szMrBfKDNK2X2FxGNBGuoGrTyLi68DXc9nbA1+OiE/0tTwzM+udMlf1FMflfxWYD+xTSTRmZla5Mlf1VDYuf0TMAGZUVb6Zmb1Vd49ePLab90VEnFBBPGZmVrHuWvyLO1m3InAo8HbAid/MbAjq7tGLJ3fMSxpFutP2ENKVOCd39T4zMxvcuu3jlzSOdGftQcDZwBYR8XQzAjMzs2p018f/PWA/YCrwzohY1LSozMysMt3dwHUUsBrwDeDxwrANzw3AkA1mZtYi3fXx9/quXjMzG/yc3M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmmp74Ja0p6QZJ8yTNlXRks2MwM6uzES2o81XgqIi4S9IoYJak6yPi/hbEYmZWO01v8UfEExFxV55/DpgHrN7sOMzM6qqlffySJgKbA7d3sm2KpJmSZra3tzc7NDOzpVbLEr+kkcClwBciYmHj9oiYGhFtEdE2YcKE5gdoZraUaknil7QMKemfHxGXtSIGM7O6asVVPQLOAOZFxA+aXb+ZWd21osW/NfBJYAdJd+dp9xbEYWZWS02/nDMibgbU7HrNzCzxnbtmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM078ZmY105LEL2lXSQ9K+qOko1sRg5lZXTU98UsaDpwK7AZsDEyStHGz4zAzq6tWtPjfA/wxIh6JiJeBi4B9WhCHmVktjWhBnasDfy0sPwps1biTpCnAlLy4SNKDfaxvPPD3Pr63So6rdxxX7ziu3hmUcem7/Y5r7c5WtiLxq5N18ZYVEVOBqf2uTJoZEW39LWegOa7ecVy947h6p25xtaKr51FgzcLyGsDjLYjDzKyWWpH47wTWl7SOpGWBjwNXtiAOM7NaanpXT0S8KulzwG+B4cC0iJhbYZX97i6qiOPqHcfVO46rd2oVlyLe0r1uZmZLMd+5a2ZWM078ZmY1s9Qk/jLDQEj6qKT7Jc2VdMFgiEvSDyXdnaeHJD0zSOJaS9INkmZLulfS7oMkrrUlTc8xzZC0RhNimiZpgaQ5XWyXpB/lmO+VtEXVMZWMayNJt0p6SdKXmxFTybgOyp/TvZJukfSuQRTbPjmuuyXNlLTNYIirsN+WkpZIOqBfFUbEkJ9IJ4kfBtYFlgXuATZu2Gd9YDYwNi+vPBjiatj/CNLJ7pbHRTqpdHie3xiYP0jiuhj4VJ7fATi3CXFtC2wBzOli++7AtaR7VN4L3F51TCXjWhnYEjgR+HIzYioZ1/sLf4e7NevzKhnbSN4497kZ8MBgiCvvMxz4H+Aa4ID+1Le0tPjLDAPxaeDUiHgaICIWDJK4iiYBFw6SuAIYnedXojn3WpSJa2Ngep6/oZPtAy4ibgKe6maXfYBzIrkNGCNp1VbHFRELIuJO4JWqY2mot6e4bun4OwRuI93L0xQlYlsUOcsCK9LJzaWtiCs7ArgU6HfuWloSf2fDQKzesM8GwAaSfi/pNkm7DpK4gNSFAaxD+kYfDHEdD3xC0qOkFsYRgySue4D98/yHgVGS3t6E2LpT+t/Z3uJQ0q+lQUPShyU9AFwN/Gur4wGQtDrp//tpA1He0pL4ywwDMYLU3bM9qWX9C0ljBkFcHT4OXBIRSyqMp0OZuCYBZ0XEGqSujHMlVf3/pUxcXwa2kzQb2A54DHi14rh60pt/Z8skfZCU+L/W6liKIuLyiNgI2Bc4odXxZKcAXxuo/NCKsXqqUGYYiEeB2yLiFeBPedC39Ul3Ercyrg4fBz5bYSxFZeI6FNgVICJulbQ8aSCrKrvIeowrIh4H9gOQNBLYPyKerTCmMjwMSS9J2gz4BbBbRPyj1fF0JiJukrSepPER0eoB3NqAiyRB+jvcXdKrEXFFXwpbWlr8ZYaBuAL4IICk8aSun0cGQVxI2hAYC9xacTy9iesvwI45vn8GlgfaWx2XpPGFXx5fB6ZVHFMZVwIH56t73gs8GxFPtDqowUrSWsBlwCcj4qFWx1Mk6R3K2TVfnbUs0PIvpohYJyImRsRE4BLg3/qa9GEpafFHF8NASPomMDMirszbdpF0P7AE+ErVLY2ScUHqVrmocFKpUiXjOgr4uaQvkrotJlcdX8m4tge+IymAm2jCryRJF+Z6x+dzHscBy+SYTyOdA9kd+CPwPHBI1TGViUvSPwEzSSfpX5P0BdJVUgtbGRdwLPB24Kc5x74aTRoZs0Rs+5O+xF8BXgA+1oy/yxJxDWx9Tco1ZmY2SCwtXT1mZlaSE7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/VSLf9h6SNiqs217SVb0oY7Kk1QYonjGS/m0gyhqsuvt8Jc3P96/0tsyz+j0SpA06TvxWlUnAzaSbsPpqMtBp4pc0vJdljQGWqsQvaam4D8eaz4nfBlweSmFr0rAPjYl/tKTLlZ6LcJqkYZKG55blHEn3SfpibmW2AefnsdHfllutx0q6GfiIpE9LulPSPZIulbRCrn+VXMc9eXo/cBKwXi7re7nenyo9m+EqSdd0tGwlvVvSjZJmSfqt8kibSuP/f1fSHUrPTvhAXj88l3mn0ljuh+X1q0q6Kdc5p2P/hs9qfqHMOyS9I6+fkI/pzjxtndcfL2mqpOuAczr5+N/y+XZS55dyPHPyTV0d6w/O8d8j6dxO3ndC/ndy3hjqmjHWtKd6TcAngDPy/C3AFnl+e+BF0nj7w4HrgQOAdwPXF94/Jr/OANoK6+cDXy0sv70w/y3giDz/S+ALeX44aVjpiRTGOs/1XkNq/PwT8HRet0yOeULe72PkZyTkeE7O87sDv8vzU4Bv5PnlSHfLrkO6+/n/F+IY1clnNb+wz8HAVXn+AmCbPL8WMC/PHw/MAt7WSVmdfr6Fesbnz/o+0pDDI4G5wObAJsCDwPi8/7j8elb+XP4TOJ1806enoT35p6JVYRJpNEFIY+pPAu7Ky3dExCPw+m3q25DG119X0o9JQ+Fe103ZvyzMbyrpW6RunJGkoR4gPaDlYIBIoxk+K2lsQznbABdHxGvAk5JuyOs3BDYFrs/DCQwHiuPuXJZfZ5G+TAB2ATYr9IWvxBsDAE6TtAxwRUTc3cUxXVh4/WGe3wnYOMcAqSU/Ks9fGREvdFFWZ5/vJQ3HfXlELM77XAZ8gDQsxyWRByOLiOLY8MeQHpYypYs6bYhx4rcBpTQ2/g6kpBykxBmSvpp3aRwjJCLiaaXH732INPbOR+l6HPTFhfmzgH0j4h5Jk0kt3tKhdrN+bkS8r4vtL+XXJbzx9yPSr43fNu4saVtgD9Kw1t+LiM66Z6KT+WHA+xoTfP4iKH4G3ZXV2XJ3x93V+C13Au+WNK7hC8GGKPfV2UA7gPQ0qrUjjSa4JvAnUksT4D1Ko28OI3Wj3JyvNhkWEZeSWpcdz6x9DhhF10YBT+QW9UGF9dOBw+H1/vfRnZR1M7B/7utfhTe+NB4EJkh6X37/MpI26eGYfwscnuNA0gaSVlR6uM6CiPg5cEbhuBp9rPDaMULrdcDnOnaQ9C89xNDhLZ9vw/abgH0lrSBpRdLDPf6X9Jl9NH9xI2lc4T2/IZ0jubrwq8OGMLf4baBNIiWJokuBA0ndNLfm7e8kJaHL8/yZevNwy5Ba9KdJegHorAV+DHA78GdSv3VHUjoSmCrpUFLL/PBIzxT4vdLDrK8lPfxjR2AO8FAu59mIeDl32fxI0kqkv5FTSH3hXfkFqdvnLqUmeTvpIR7bA19RGulxEbn7qRPLSbqd1BCblNd9HjhV0r05hpuAz3QTQ4fOPt/XRcRdks4C7uiIPSJmA0g6EbhR0hLS86knF953cU76V0ravZuuJhsCPDqn1ZakkRGxKLdy7wC2jognmxzDfNIJ7FY/6MNqxC1+q7OrlB6/uSxwQrOTvlmruMVvZlYzPrlrZlYzTvxmZjXjxG9mVjNO/GZmNePEb2ZWM/8H6nKcTnfgFKwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{}\n",
      "Gene(8) None not in parent dict\n",
      "Gene(10) None in parent dict\n",
      "{'SE001_aa': {None}}\n",
      "{'SE001_aa': [[Gene(8), Gene(10)]], 'SE002_aa/SE003_aa': [[None, None]]}\n",
      "{'SE001_aa': {None}, 'SE002_aa/SE003_aa': {None}}\n",
      "SE001_aa {None} []\n",
      "SE002_aa/SE003_aa {None} [{None}]\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(13) None not in parent dict\n",
      "Gene(15) None in parent dict\n",
      "{'SE001_aa': {None}}\n",
      "<HOG(1)> None not in parent dict\n",
      "{'SE001_aa': [[Gene(13), Gene(15)]], 'SE002_aa/SE003_aa': [[<HOG(1)>, None]]}\n",
      "{'SE001_aa': {None}, 'SE002_aa/SE003_aa': {None}}\n",
      "SE001_aa {None} []\n",
      "SE002_aa/SE003_aa {None} [{None}]\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(18) None not in parent dict\n",
      "Gene(20) None in parent dict\n",
      "{'SE001_aa': {None}}\n",
      "<HOG(5)> None not in parent dict\n",
      "{'SE001_aa': [[Gene(18), Gene(20)]], 'SE002_aa/SE003_aa': [[<HOG(5)>, None]]}\n",
      "{'SE001_aa': {None}, 'SE002_aa/SE003_aa': {None}}\n",
      "SE001_aa {None} []\n",
      "SE002_aa/SE003_aa {None} [{None}]\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(20) None not in parent dict\n",
      "Gene(22) None in parent dict\n",
      "{'SE001_aa': {None}}\n",
      "{'SE001_aa': [[Gene(20), Gene(22)]], 'SE002_aa/SE003_aa': [[None, None]]}\n",
      "{'SE001_aa': {None}, 'SE002_aa/SE003_aa': {None}}\n",
      "SE001_aa {None} []\n",
      "SE002_aa/SE003_aa {None} [{None}]\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(22) None not in parent dict\n",
      "Gene(24) None in parent dict\n",
      "{'SE001_aa': {None}}\n",
      "<HOG(3)> None not in parent dict\n",
      "{'SE001_aa': [[Gene(22), Gene(24)]], 'SE002_aa/SE003_aa': [[<HOG(3)>, None]]}\n",
      "{'SE001_aa': {None}, 'SE002_aa/SE003_aa': {None}}\n",
      "SE001_aa {None} []\n",
      "SE002_aa/SE003_aa {None} [{None}]\n",
      "len1 (None, None)\n",
      "{}\n",
      "Gene(24) None not in parent dict\n",
      "{'SE001_aa': {None}}\n",
      "{'SE001_aa': [[Gene(24), None]], 'SE002_aa/SE003_aa': [[None, None]]}\n",
      "{'SE001_aa': {None}, 'SE002_aa/SE003_aa': {None}}\n",
      "SE001_aa {None} []\n",
      "SE002_aa/SE003_aa {None} [{None}]\n",
      "len1 (None, None)\n",
      "neighbors assigned\n",
      "[[<HOG(6)>], [<HOG(7)>], [<HOG(8)>], [<HOG(9)>], [<HOG(10)>], [<HOG(11)>]]\n",
      "next node\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXgAAAEWCAYAAABsY4yMAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAefElEQVR4nO3deZgdVZnH8e8vCyAkIYkJyB5AEAHZbFCGTYVBCCoIuIDKhGEMoiIqiujIMqKj6OA6IERZZFfCMqwKogERAnRYgywKRkXANEIIiYAQ3vnjnCaVm+7b1Ut13y5+n+epp6vq1K3zVt3k7dOnqk4pIjAzs/oZMdQBmJlZNZzgzcxqygnezKymnODNzGrKCd7MrKac4M3MasoJ3sysppzgzcxqygl+mJO0o6SbJT0j6SlJv5W0raRpkpZIWtQwrZk/N1HSpZIWS/qTpAMb9ntgXr9Y0mWSJhbKPimpXdILks4axGP9kqT/Lsz/MR/To5J+WthulqTnG477ikL5rpIekPQPSb+WtF6hbEVJZ0haKOkJSZ8tlL1V0nX5PHdIukjSGoN07A9J2ljS2pIulvRk/s7vlTQtbzNFUnTxnX8gl0vSiZL+nqdvSlKhjq0kzcnnZY6krQpln5b0SD4vj0n6jqRRg3Hs1ndO8MOYpHHAlcAPgInAWsB/AS/kTW6JiDEN02O57GTgn8DqwIeAH0raLO93M+A04CO5/B/AKYWqHwO+CpxR5fF1YSpwtaR/y7HtFhFjgDbg+oZtP9lw3O8GkDQJuAQ4hnTO2oGfFj53PLARsB7wduAoSXvksgnADGBKLn8WOHOgD7KRpA2BERHxEHAO8Jdc/2uBg4C/NXxkfMOxdx7fdGAfYEtgC+BdwKG5jhWA/wPOJR3nT4D/y+sBrgC2iYhxwOZ5H5+q4nhtAEWEp2E6kRLbgm7KpgE3dVO2Cim5b1xYdw7wjTz/38D5hbIN8/ZjG/bzVeCskrFOIP0y6gCezvNrF8oPBu4nJc1HgEO7+Px8YCTwv8B3m9Q1C/iPbsqmAzc3nIvngE3y8l+B3QvlJwAXdrOvbYBnSxz7XsCdwEJScj6+ofwi4AngGeBGYLOG8k8B38/zi4CtuqlnChDAqG7KbwamF5YPAWbn+d3zsatQ/mdgjy7281rgl8ApJY79e/mYFwJzgJ0KZdsBtwALgMfz97rCUP6fqtvkFvzw9hCwRNJPJO0paULJz20MLInUIux0N7BZnt8sLwMQEQ+TfyH0I9YRpNbuesC6pKT6v4Xy+aQW5ThSsv+OpG0K5e8Ero+IJcBs4CBJn5fUJmlkL+JoPLbFwMPAZvn8rVksZ9nz0mhn4L4SdS4mtbTHk5L9YZL2KZRfQ/qrYTXgDuC8hs9PBa7K87OBkyV9UNK6JeouWubYWf47vydy5s3uKZR3dtstBJ4kteBPK1Hn7cBWpL+WzgcukrRSLlsCfAaYBGwP7Ap8vJfHZE04wQ9jEbEQ2JHUavsR0CHpckmr503eKmlBYXo4rx9Dai0WPQOMLVnel1j/HhEXR8Q/IuJZ4GvALoXyqyLi4UhuAK4FdirsYi/g6rztucDhpKR/AzBf0tENVX6/4dhPKHFsYwrLjWXLkLQFcCzw+RLHPisi7o2IlyPiHuCChmM/IyKejYgXSF1EW0paNdezMrBtPk6A9wG/IXUx/VHSXZK2bajyyYZjf2M3x/4MMCb3w/f4nUfE+ZG6aDYGTmX5rqGujv3c/N2/FBEnASsCb8hlcyJidi6bR/qFsUuT3VkvOcEPcxFxf0RMi4i1SX2jawLfzcWzI2J8Ydowr19EaikXjSN1j5Qp7zVJK0s6LV+4XUjqihjf2frOf4HMzhcwF5BarZNy2QjgX4GfF477vIjYjdQq/hjwFUnvLFT5qYZjP6bEsS0qLHd73JJeT2p1HxERvylx7G/JF3M7JD2T4+08tpGSviHp4Xxe5uWPTco/dyV1KT2fj/vpiDg6IjYjXR+5C7iseLEUmNRw7Pd3c+zjgEW51V76O4+I35P+cjmlsayLYz9S0v35gvACYNXCsW8s6cp8MXshqWtwUrP9We84wddIRDwAnEVK9M08BIyStFFh3ZYs7W64Ly8DIGkDUsur2KXTW0eSWm5vya3AnZfuXisCFwP/A6weEeNJrfXOpLUtMC8iOhp3GhEvRsRFpO6Eno4blj+2VUjXGO6LiKdJfcFbFrYvnheU7rj5JXBCRJxToj5IXROXA+tExKqk1m/nsR0I7A3sRkp+Uzqryj+L3TPLiIgnSedsTVIXSE+WOXaW/863aPhFsQXdd0GNIp23bknaCfgC8H5gQv5en2Hpsf0QeADYKP+b+FKhzAaAE/wwJmmT3EJaOy+vAxxA6qftVu53voTU6l1F0g6kJNOZsM4D3i1pp5wAvwJckrtWkDQq96OOBEZKWqnELXNjSf3uC5RuuTyuULYC6RdIB/CSpD1JF/06vdI9k+ufJmkvSWMljcjbbwbc2kMMAJcCm0vaLx/DsaS+5wdy+dnAlyVNkLQJ8FHSL00krQX8Cjg5Ik4tUVfx2J+KiOclbUdK6sWyF4C/AyuTWrFFezYc+4mSNs/fwVjgMOAPEfH3EnGcDXxW0lpKt8se2XlspAvTS4BPKd0q+sm8/le53v+QtFqe3xT4IsvfudTVcb9E+l5HSTqWZf9KGEu6+Loon+vDShyD9cZQX+X11PeJdFvkz0h3PyzOP08j/SeaRvoPu6hh2jZ/diJwWf7cn4EDG/Z9YF6/mHT73MRC2fGkfv/idHwPsa5JSiKLSH8JHErhjg/gE6Q+3QWkXzQXAl/NZe1AW2Ff+wK/Jd2NsxC4F5hWKJ8FPN9w3HMK5buRWo7P5W2nFMpWJN3+uTDH89lC2XE55mXOaYnvaX/gT6TujitJF5fPzWVj8vl9Nm9zUK7j9aS/SOY27OsHwO9z3R15f2/MZVO6iq/zGEit428CT+Xpmyx718zWpDtdniNd7N26UHZmPh+LSd1I3wJW6uG4RwKn53P5OHBU/uxuuXzn/D0sIl1X+Ard3PnlqW+T8ok2a0n5gvFdwJrxKvvHKukoUn/6UUMdiw1PfhLNWt2qpBboqyq5Z/NIDxiZ9Ylb8DZgJH2JdKGs0W8iYs/BjmcwSbqPdI9/o0MjovG+9trIF1Kv6aos0lPGNoSc4M3MaqqlumgmTZoUU6ZMGeowzMyGjTlz5jwZEZO7KmupBD9lyhTa29uHOgwzs2FD0p+6K/N98GZmNeUEb2ZWU07wZmY15QRvZlZTTvBmZjXlBG9mVlOVJnhJ4yXNVHrB8f2Stq+yPjMzW6rq++C/B/w8IvbPL+9dueL6zMwsqyzBS+p8qcM0gIj4J+m9nmZmNgiqbMFvQBqv+kxJW5LGmT4i0ssmXiFpOulN96y7bm/fIbzUlKO7fOmNmVnLm/eNvSrZb5V98KOAbYAfRsTWpBcFNL4YmYiYERFtEdE2eXKXwymYmVkfVJngHwUejYjO16jNJCV8MzMbBJUl+Ih4AviLpDfkVbsCv6uqPjMzW1bVd9EcDpyX76B5BDi44vrMzCyrNMFHxF1AW5V1mJlZ1/wkq5lZTTnBm5nVlBO8mVlNOcGbmdWUE7yZWU05wZuZ1ZQTvJlZTTnBm5nVlBO8mVlNOcGbmdWUE7yZWU05wZuZ1ZQTvJlZTTnBm5nVlBO8mVlNOcGbmdWUE7yZWU05wZuZ1ZQTvJlZTTnBm5nVlBO8mVlNOcGbmdWUE7yZWU05wZuZ1ZQTvJlZTY2qcueS5gHPAkuAlyKircr6zMxsqUoTfPb2iHhyEOoxM7MCd9GYmdVU1Qk+gGslzZE0vasNJE2X1C6pvaOjo+JwzMxePapO8DtExDbAnsAnJO3cuEFEzIiItohomzx5csXhmJm9elSa4CPisfxzPnApsF2V9ZmZ2VKVJXhJq0ga2zkP7A7Mrao+MzNbVpV30awOXCqps57zI+LnFdZnZmYFlSX4iHgE2LKq/ZuZWXO+TdLMrKac4M3MasoJ3sysppzgzcxqygnezKymnODNzGrKCd7MrKac4M3MasoJ3sysppzgzcxqygnezKymnODNzGrKCd7MrKac4M3MasoJ3sysppzgzcxqygnezKymekzwkr4paZyk0ZKul/SkpA8PRnBmZtZ3ZVrwu0fEQuBdwKPAxsDnK43KzMz6rUyCH51/TgUuiIinKozHzMwGSJmXbl8h6QHgOeDjkiYDz1cblpmZ9VePLfiIOBrYHmiLiBeBfwDvqTowMzPrnzIXWc+IiKcjYknnKuD8asMyM7P+KtMH/1dJPwSQNAG4Fji30qjMzKzfynTRHAMslHQqKbmfFBFnVh6ZmZn1S7cXWSXtW1i8DTgm/wxJ+0bEJVUHZ2ZmfdfsLpp3NyzfSbpl8t1AAKUSvKSRQDvw14h4V1+CNDOz3us2wUfEwQNUxxHA/cC4AdqfmZmVUOYump9IGl9YniDpjDI7l7Q2sBfw476HaGZmfVHmLpotImJB50JEPA1sXXL/3wWOAl7ubgNJ0yW1S2rv6OgouVszM+tJmQQ/It8eCYCkiZR4AlbSu4D5ETGn2XYRMSMi2iKibfLkySXCMTOzMsoMVXAScLOkmXn5fcDXSnxuB+A9kqYCKwHjJJ0bER6J0sxsEJS5D/5sYD/gb3naNyLOKfG5L0bE2hExBfgg8CsndzOzwVP2hR+jSUMUdM6bmVmLK3MXzRHAecAkYDXgXEmH96aSiJjle+DNzAZXmT74Q4C3RMRiAEknArcAP6gyMDMz658yXTQClhSWl7C0u8bMzFpUmRb8mcCtki7Ny/sAp1cXkpmZDYQeE3xEfFvSLGBHUsv94Ii4s+rAzMysf5qNJjmxsDgvT6+U+d2sZmatrVkLfg5p1Miu+tsD2KCSiMzMbEA0G01y/cEMxMzMBlaZi6ydL//YkdRy/01EXFZpVGZm1m9lHnQ6BfgYcC8wF/iYpJOrDszMzPqnTAt+F2DziAhI48OTkr2ZmbWwMg86PQisW1heB7inmnDMzGygNLtN8gpSn/uqwP2SbsvLbwFuHpzwzMysr5p10fzPoEVhZmYDrtltkjcMZiBmZjawyo4Hb2Zmw4wTvJlZTXWb4CVdn3+eOHjhmJnZQGl2kXUNSbuQXpx9IQ1j0kTEHZVGZmZm/dIswR8LHA2sDXy7oSyAd1QVlJmZ9V+zu2hmAjMlHRMRJwxiTGZmNgDKvPDjBEnvAXbOq2ZFxJXVhmVmZv1VZrCxrwNHAL/L0xF5nZmZtbAyg43tBWwVES/DK4ON3Ql8scrAzMysf8reBz++ML9qFYGYmdnAKtOC/zpwp6Rfk26V3Bm33s3MWl6Zi6wXSJoFbEtK8F+IiCd6+pyklYAbgRVzPTMj4rj+hWtmZmWVemVfRDwOXN7Lfb8AvCMiFkkaDdwk6ZqImN3bIM3MrPdKJfi+yG+AWpQXR+cpqqrPzMyWVelgY5JGSroLmA9cFxG3drHNdEntkto7OjqqDMfM7FWlaYKXNELS3L7uPCKWRMRWpOEOtpO0eRfbzIiItohomzx5cl+rMjOzBk0TfL73/W5J6zbbricRsQCYBezRn/2YmVl5Zfrg1wDuy+9kXdy5MiLe0+xDkiYDL0bEAkmvAXYDPPSwmdkgKZPg/6uP+14D+ImkkaS/FH7mMWzMzAZPmfvgb5C0HrBRRPxS0srAyBKfuwfYegBiNDOzPigz2NhHgZnAaXnVWsBlVQZlZmb9V+Y2yU8AOwALASLi98BqVQZlZmb9VybBvxAR/+xckDQKP7BkZtbyyiT4GyR9CXiNpH8FLgKuqDYsMzPrrzIJ/migA7gXOBS4GvhylUGZmVn/lbmL5uX8ko9bSV0zD+ZxZszMrIX1mOAl7QWcCjxMGi54fUmHRsQ1VQdnZmZ9V+ZBp5OAt0fEHwAkbQhcBTjBm5m1sDJ98PM7k3v2CGl0SDMza2HdtuAl7Ztn75N0NfAzUh/8+4DbByE2MzPrh2ZdNO8uzP8N2CXPdwATKovIzMwGRLcJPiIOHsxAzMxsYJW5i2Z94HBgSnH7noYLNjOzoVXmLprLgNNJT6++XG04ZmY2UMok+Ocj4vuVR2JmZgOqTIL/nqTjgGuBFzpXRsQdlUVlZmb9VibBvwn4CPAOlnbRRF42M7MWVSbBvxfYoDhksJmZtb4yT7LeDYyvOhAzMxtYZVrwqwMPSLqdZfvgfZukmVkLK5Pgj6s8CjMzG3BlxoO/YTACMTOzgVXmSdZnWfoO1hWA0cDiiBhXZWBmZtY/ZVrwY4vLkvYBtqssIjMzGxBl7qJZRkRchu+BNzNreWW6aPYtLI4A2ljaZWNmZi2qzF00xXHhXwLmAXv39CFJ6wBnA68jPQE7IyK+14cYzcysD8r0wfd1XPiXgCMj4g5JY4E5kq6LiN/1cX9mZtYLzV7Zd2yTz0VEnNBsxxHxOPB4nn9W0v3AWoATvJnZIGh2kXVxFxPAIcAXelOJpCnA1sCtXZRNl9Quqb2jo6M3uzUzsyaavbLvpM753MVyBHAwcCFwUnefayRpDHAx8OmIWNhFPTOAGQBtbW2+eGtmNkCa9sFLmgh8FvgQ8BNgm4h4uuzOJY0mJffzIuKS/gRqZma906wP/lvAvqTW9ZsiYlFvdixJpFf93R8R3+5XlGZm1mvN+uCPBNYEvgw8Jmlhnp6VtFxXSxd2IL8oRNJdeZo6ADGbmVkJzfrge/2Ua8PnbwLUn32YmVnf9SuJm5lZ63KCNzOrKSd4M7OacoI3M6spJ3gzs5pygjczqykneDOzmnKCNzOrKSd4M7OacoI3M6spJ3gzs5pygjczqykneDOzmnKCNzOrKSd4M7OacoI3M6spJ3gzs5pygjczqykneDOzmnKCNzOrKSd4M7OacoI3M6spJ3gzs5pygjczqykneDOzmqoswUs6Q9J8SXOrqsPMzLpXZQv+LGCPCvdvZmZNVJbgI+JG4Kmq9m9mZs0NeR+8pOmS2iW1d3R0DHU4Zma1MeQJPiJmRERbRLRNnjx5qMMxM6uNIU/wZmZWDSd4M7OaqvI2yQuAW4A3SHpU0iFV1WVmZssbVdWOI+KAqvZtZmY9cxeNmVlNOcGbmdWUE7yZWU05wZuZ1ZQTvJlZTTnBm5nVlBO8mVlNOcGbmdWUE7yZWU05wZuZ1ZQTvJlZTTnBm5nVlBO8mVlNOcGbmdWUE7yZWU05wZuZ1ZQTvJlZTTnBm5nVlBO8mVlNOcGbmdWUE7yZWU05wZuZ1ZQTvJlZTTnBm5nVlBO8mVlNOcGbmdVUpQle0h6SHpT0B0lHV1mXmZktq7IEL2kkcDKwJ7ApcICkTauqz8zMllVlC3474A8R8UhE/BO4ENi7wvrMzKxgVIX7Xgv4S2H5UeAtjRtJmg5Mz4uLJD3Yx/omAU/28bNVcly947h6x3H1TkvGpRP7Fdd63RVUmeDVxbpYbkXEDGBGvyuT2iOirb/7GWiOq3ccV+84rt55tcVVZRfNo8A6heW1gccqrM/MzAqqTPC3AxtJWl/SCsAHgcsrrM/MzAoq66KJiJckfRL4BTASOCMi7quqPgagm6cijqt3HFfvOK7eeVXFpYjlusXNzKwG/CSrmVlNOcGbmdXUsErwZYY+kPR+Sb+TdJ+k81slNknfkXRXnh6StKBF4lpX0q8l3SnpHklTWySu9SRdn2OaJWntQYjpDEnzJc3tplySvp9jvkfSNlXHVDKuTSTdIukFSZ8bjJhKxvWhfJ7ukXSzpC1bJK69c0x3SWqXtGMrxFXYbltJSyTt3+9KI2JYTKQLtQ8DGwArAHcDmzZssxFwJzAhL6/WKrE1bH846aLzkMdFurhzWJ7fFJjXInFdBPxbnn8HcM4gxLUzsA0wt5vyqcA1pGc83grcOkj/vnqKazVgW+BrwOcGI6aScf1L4f/ini10vsaw9PrjFsADrRBX3mYk8CvgamD//tY5nFrwZYY++ChwckQ8DRAR81sotqIDgAtaJK4AxuX5VRmcZxXKxLUpcH2e/3UX5QMuIm4Enmqyyd7A2ZHMBsZLWmOo44qI+RFxO/Bi1bE01NtTXDd3/l8EZpOehWmFuBZFzqbAKnTxAOZQxJUdDlwMDEjuGk4JvquhD9Zq2GZjYGNJv5U0W9IeLRQbkLoegPVJv6VbIa7jgQ9LepTUaji8ReK6G9gvz78XGCvptYMQWzOlv2dbziGkv35agqT3SnoAuAr496GOB0DSWqR/66cO1D6HU4IvM/TBKFI3zdtIreQfSxpfcVxQcliG7IPAzIhYUmE8ncrEdQBwVkSsTeqCOEdS1f8uysT1OWAXSXcCuwB/BV6qOK6e9OZ7tkzS20kJ/gtDHUuniLg0IjYB9gFOGOp4su8CXxjI3FDlWDQDrczQB48CsyPiReCPeeCyjUhP1Q51bJ0+CHyi4ng6lYnrEGAPgIi4RdJKpAGZquze6jGuiHgM2BdA0hhgv4h4psKYyvDwG70kaQvgx8CeEfH3oY6nUUTcKGlDSZMiYqgHIWsDLpQE6f/gVEkvRcRlfd3hcGrBlxn64DLg7QCSJpG6bB5pkdiQ9AZgAnDLIMRUNq4/A7vm+N4IrAR0DHVckiYV/pL4InBGxTGVcTlwUL6b5q3AMxHx+FAH1aokrQtcAnwkIh4a6ng6SXq9chbNd0KtAAz5L5+IWD8ipkTEFGAm8PH+JHcYRi346GboA0lfAdoj4vJctruk3wFLgM8PRquhZGyQukMuLFzgaYW4jgR+JOkzpO6GaVXHVzKutwFflxTAjQzCXz2SLsj1TsrXJI4DRueYTyVdo5gK/AH4B3Bw1TGViUvS64B20sXylyV9mnRX0sKhjAs4FngtcErOpy/FIIzkWCKu/Ui/qF8EngM+MBj/J0vENfB1DlKuMTOzQTacumjMzKwXnODNzGrKCd7MrKac4M3MasoJ3sysppzgrc/y494haZPCurdJurIX+5gmac0Bime8pI8PxL5aVbPzK2lefv6jt/s8a0BGLrSW4wRv/XEAcBPpQaW+mgZ0meAljezlvsYDtUrwkobNsyrWepzgrU/y8AE7kIY6aEzw4yRdqjQu/6mSRkgamVuKcyXdK+kzudXYBpyXx+Z+TW6FHivpJuB9kj4q6XZJd0u6WNLKuf7Vcx135+lfgG8AG+Z9fSvXe4rSuwGulHR1Z0tV0psl3SBpjqRfKI8KqTT2/ImSblMat3+nvH5k3uftSmOJH5rXryHpxlzn3M7tG87VvMI+b5P0+rx+cj6m2/O0Q15/vKQZkq4Fzu7i9C93fruo87M5nrn5wafO9Qfl+O+WdE4Xnzshf0/ODXUwGOMge6rfBHwYOD3P3wxsk+ffBjxPGut9JHAdsD/wZuC6wufH55+zgLbC+nnAUYXl1xbmvwocnud/Cnw6z48kDXU8hcJY27neq0kNmdcBT+d1o3PMk/N2HyCPz5/jOSnPTwV+meenA1/O8yuSnhxdn/Qk8H8W4hjbxbmaV9jmIODKPH8+sGOeXxe4P88fD8wBXtPFvro8v4V6JuVzfS9pKNwxwH3A1sBmwIPApLz9xPzzrHxevgmcRn4A0tPwn/znn/XVAaTR7yCN534AcEdevi0iHoFXHs/ekTS2+waSfkAaovXaJvv+aWF+c0lfJXW/jCENbwDpJSAHAUQafe8ZSRMa9rMjcFFEvAw8IenXef0bgM2B6/Ij9COB4pgyl+Sfc0i/NAB2B7Yo9FWvytKB7M6QNBq4LCLu6uaYLij8/E6e3w3YNMcAqWU+Ns9fHhHPdbOvrs7vzIbjvjQiFudtLgF2Ig1FMTPyoFoRURyb/BjSCzmmd1OnDUNO8NZrSuOyv4OUfIOUIEPSUXmTxvEvIiKeVnpl2ztJ48q8n+7H4V5cmD8L2Cci7pY0jdSCLR1qk/X3RcT23ZS/kH8uYen/EZH+evhF48aSdgb2Ig21/K2I6KpbJbqYHwFs35jIc8IvnoNm++pqudlxdzc2ye3AmyVNbEj8Noy5n836Yn/Sm43WizT63TrAH0ktR4DtlEaKHEHq/rgp390xIiIuJrUWO99n+iwwlu6NBR7PLeQPFdZfDxwGr/SPj+tiXzcB++W++NVZ+svhQWCypO3z50dL2qyHY/4FcFiOA0kbS1pF6QUu8yPiR8DpheNq9IHCz87RRK8FPtm5gaSteoih03Lnt6H8RmAfSStLWoX0EonfkM7Z+/MvaCRNLHzm56RrGFcV/oqwYc4teOuLA0jJoOhi4EBS98otufxNpGRzaZ4/U8sOAQyphX6qpOeArlrUxwC3An8i9St3Jp8jgBmSDiG1tA+LNJ79b5VeanwN6QUTuwJzgYfyfp6JiH/mrpbvS1qV9P/gu6S+6u78mNRdc4dSE7uD9LKItwGfVxqZcBG526gLK0q6ldSoOiCv+xRwsqR7cgw3Ah9rEkOnrs7vKyLiDklnAbd1xh4RdwJI+hpwg6QlpPcXTyt87qKc3C+XNLVJF5ENEx5N0mpN0piIWJRbrbcBO0TEE4McwzzSheShfqGEvcq4BW91d6XSaxtXAE4Y7ORuNpTcgjczqylfZDUzqykneDOzmnKCNzOrKSd4M7OacoI3M6up/wc7kJLqg0ul1QAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "locus_to_id(\"8may_medhighMut/DB/SE001_aa.fa\")\n",
    "\n",
    "get_locus('8may_medhighMut/DB')\n",
    "\n",
    "counter = 1\n",
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    assign_neighbors(node.genome)\n",
    "    print(\"neighbors assigned\")\n",
    "    get_synteny_blocks(node.genome)\n",
    "    print(\"next node\")\n",
    "    assign_block_to_genes(node.genome)\n",
    "    \n",
    "    #print(node)\n",
    "    # here we go through all nodes, and if it is an ancestral, note, we make some plots\n",
    "    # in my 3 species example, there are 2 ancestral nodes, so two plots\n",
    "    \n",
    "    if isinstance(node.genome, pyham.AncestralGenome):\n",
    "        # links together the primary reconstruction's synteny blocks, using windows of 20\n",
    "        joining_blocks(20, node.genome)\n",
    "                \n",
    "        ### Plot ###\n",
    "        \n",
    "        title = node.genome.name  \n",
    "        raw_values = []\n",
    "        \n",
    "        for block in node.genome.new_synteny:\n",
    "            count = 0\n",
    "            \n",
    "            for element in block:\n",
    "                # goes through the elements of that block, the HOGs\n",
    " #               print(block, element)\n",
    "         \n",
    "                # if that element (HOG) is a pyham.HOG, increase count, it's the number of HOGs we've alignd the synteny of in this block\n",
    "                if isinstance(element, pyham.HOG):\n",
    "                    count += 1\n",
    "                    \n",
    " #               print(count)\n",
    "                    \n",
    "            raw_values.append(count)\n",
    "            \n",
    " #       print(raw_values)\n",
    "        \n",
    "        label = list(set(raw_values))\n",
    "        values = [0 for i in range(len(label))]\n",
    "        \n",
    "        for i in range(len(label)):\n",
    "            \n",
    "            for element in raw_values:\n",
    "                \n",
    "                if label[i] == element:\n",
    "                    values[i] += 1\n",
    "\n",
    "        plt.bar(label, values)\n",
    "        plt.title(title)\n",
    "        plt.xlabel('Abstractgenes per block')\n",
    "        plt.ylabel('Number of blocks')\n",
    "        plt.savefig('results/secondary_reconstruction_plot{}.png'.format(counter))\n",
    "        counter += 1\n",
    "        plt.show()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Gene(1)] 1 G1_SE001, sequence type: type1, locus: 1\n",
      "[Gene(2)] 2 G2_SE001, sequence type: type1, locus: 2\n",
      "[Gene(3)] 3 G3_SE001, sequence type: type1, locus: 3\n",
      "[Gene(4)] 4 G4_SE001, sequence type: type1, locus: 4\n",
      "[Gene(5)] 5 G5_SE001, sequence type: type1, locus: 5\n",
      "[Gene(6)] 6 G6_SE001, sequence type: type1, locus: 6\n",
      "[Gene(7)] 7 G7_SE001, sequence type: type1, locus: 7\n",
      "[Gene(8)] 8 G8_SE001, sequence type: type1, locus: 8\n",
      "[Gene(9)] 9 G9_SE001, sequence type: type1, locus: 9\n",
      "[Gene(10)] 10 G10_SE001, sequence type: type1, locus: 10\n",
      "[Gene(11)] 11 G11_SE001, sequence type: type1, locus: 11\n",
      "[Gene(12)] 12 G12_SE001, sequence type: type1, locus: 12\n",
      "[Gene(13)] 13 G13_SE001, sequence type: type1, locus: 13\n",
      "[Gene(14)] 14 G14_SE001, sequence type: type1, locus: 14\n",
      "[Gene(15)] 15 G15_SE001, sequence type: type1, locus: 15\n",
      "[Gene(16)] 16 G16_SE001, sequence type: type1, locus: 16\n",
      "[Gene(17)] 17 G17_SE001, sequence type: type1, locus: 17\n",
      "[Gene(18)] 18 G18_SE001, sequence type: type1, locus: 18\n",
      "[Gene(19)] 19 G19_SE001, sequence type: type1, locus: 19\n",
      "[Gene(20)] 20 G20_SE001, sequence type: type1, locus: 20\n",
      "[Gene(21)] 21 G21_SE001, sequence type: type1, locus: 21\n",
      "[Gene(22)] 22 G22_SE001, sequence type: type1, locus: 22\n",
      "[Gene(23)] 23 G23_SE001, sequence type: type1, locus: 23\n",
      "[Gene(24)] 24 G24_SE001, sequence type: type1, locus: 24\n",
      "[Gene(25)] 25 G25_SE001, sequence type: type1, locus: 25\n",
      "[Gene(26)] 1 G1_SE002, sequence type: type1, locus: 1\n",
      "[Gene(27)] 2 G2_SE002, sequence type: type1, locus: 2\n",
      "[Gene(28)] 3 G3_SE002, sequence type: type1, locus: 3\n",
      "[Gene(29)] 4 G4_SE002, sequence type: type1, locus: 4\n",
      "[Gene(30)] 5 G5_SE002, sequence type: type1, locus: 5\n",
      "[Gene(31)] 6 G6_SE002, sequence type: type1, locus: 6\n",
      "[Gene(32)] 7 G7_SE002, sequence type: type1, locus: 7\n",
      "[Gene(33)] 8 G8_SE002, sequence type: type1, locus: 8\n",
      "[Gene(34)] 9 G9_SE002, sequence type: type1, locus: 9\n",
      "[Gene(35)] 10 G10_SE002, sequence type: type1, locus: 10\n",
      "[Gene(36)] 11 G11_SE002, sequence type: type1, locus: 11\n",
      "[Gene(37)] 12 G12_SE002, sequence type: type1, locus: 12\n",
      "[Gene(38)] 13 G13_SE002, sequence type: type1, locus: 13\n",
      "[Gene(39)] 14 G14_SE002, sequence type: type1, locus: 14\n",
      "[Gene(40)] 15 G15_SE002, sequence type: type1, locus: 15\n",
      "[Gene(41)] 16 G16_SE002, sequence type: type1, locus: 16\n",
      "[Gene(42)] 17 G17_SE002, sequence type: type1, locus: 17\n",
      "[Gene(43)] 18 G18_SE002, sequence type: type1, locus: 18\n",
      "[Gene(44)] 19 G19_SE002, sequence type: type1, locus: 19\n",
      "[Gene(45)] 20 G20_SE002, sequence type: type1, locus: 20\n",
      "[Gene(46)] 21 G21_SE002, sequence type: type1, locus: 21\n",
      "[Gene(47)] 22 G22_SE002, sequence type: type1, locus: 22\n",
      "[Gene(48)] 23 G23_SE002, sequence type: type1, locus: 23\n",
      "[Gene(49)] 24 G24_SE002, sequence type: type1, locus: 24\n",
      "[Gene(50)] 25 G25_SE002, sequence type: type1, locus: 25\n",
      "[Gene(51)] 1 G1_SE003, sequence type: type1, locus: 1\n",
      "[Gene(52)] 2 G2_SE003, sequence type: type1, locus: 2\n",
      "[Gene(53)] 3 G3_SE003, sequence type: type1, locus: 3\n",
      "[Gene(54)] 4 G4_SE003, sequence type: type1, locus: 4\n",
      "[Gene(55)] 5 G5_SE003, sequence type: type1, locus: 5\n",
      "[Gene(56)] 6 G6_SE003, sequence type: type1, locus: 6\n",
      "[Gene(57)] 7 G7_SE003, sequence type: type1, locus: 7\n",
      "[Gene(58)] 8 G8_SE003, sequence type: type1, locus: 8\n",
      "[Gene(59)] 9 G9_SE003, sequence type: type1, locus: 9\n",
      "[Gene(60)] 10 G10_SE003, sequence type: type1, locus: 10\n",
      "[Gene(61)] 11 G11_SE003, sequence type: type1, locus: 11\n",
      "[Gene(62)] 12 G12_SE003, sequence type: type1, locus: 12\n",
      "[Gene(63)] 13 G13_SE003, sequence type: type1, locus: 13\n",
      "[Gene(64)] 14 G14_SE003, sequence type: type1, locus: 14\n",
      "[Gene(65)] 15 G15_SE003, sequence type: type1, locus: 15\n",
      "[Gene(66)] 16 G16_SE003, sequence type: type1, locus: 16\n",
      "[Gene(67)] 17 G17_SE003, sequence type: type1, locus: 17\n",
      "[Gene(68)] 18 G18_SE003, sequence type: type1, locus: 18\n",
      "[Gene(69)] 19 G19_SE003, sequence type: type1, locus: 19\n",
      "[Gene(70)] 20 G20_SE003, sequence type: type1, locus: 20\n",
      "[Gene(71)] 21 G21_SE003, sequence type: type1, locus: 21\n",
      "[Gene(72)] 22 G22_SE003, sequence type: type1, locus: 22\n",
      "[Gene(73)] 23 G23_SE003, sequence type: type1, locus: 23\n",
      "[Gene(74)] 24 G24_SE003, sequence type: type1, locus: 24\n",
      "[Gene(75)] 25 G25_SE003, sequence type: type1, locus: 25\n"
     ]
    }
   ],
   "source": [
    "# simple inversion example\n",
    "\n",
    "# Path to nwk file\n",
    "nwk_path = \"working_inversion_3spp/RealTree.nwk\"\n",
    "\n",
    "# The newick tree is extracted as a string\n",
    "tree_str = pyham.utils.get_newick_string(nwk_path, type=\"nwk\")\n",
    "\n",
    "# Path to OrthoXML file\n",
    "orthoxml_path = \"working_inversion_3spp/Output/HierarchicalGroups.orthoxml\"\n",
    "\n",
    "get_locus('working_inversion_3spp/DB')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "neighbors assigned\n",
      "[[Gene(1), Gene(2), Gene(3), Gene(4), Gene(5), Gene(6), Gene(7), Gene(8), Gene(9), Gene(10), Gene(11), Gene(12), Gene(13), Gene(14), Gene(15), Gene(16), Gene(17), Gene(18), Gene(19), Gene(20), Gene(21), Gene(22), Gene(23), Gene(24), Gene(25)]]\n",
      "next node\n",
      "neighbors assigned\n",
      "[[Gene(26), Gene(27), Gene(28), Gene(29), Gene(30), Gene(31), Gene(32), Gene(33), Gene(34), Gene(35), Gene(36), Gene(37), Gene(38), Gene(39), Gene(40), Gene(41), Gene(42), Gene(43), Gene(44), Gene(45), Gene(46), Gene(47), Gene(48), Gene(49), Gene(50)]]\n",
      "next node\n",
      "neighbors assigned\n",
      "[[Gene(51), Gene(52), Gene(53), Gene(54), Gene(55), Gene(56), Gene(57), Gene(58), Gene(59), Gene(60), Gene(61), Gene(62), Gene(63), Gene(64), Gene(65), Gene(66), Gene(67), Gene(68), Gene(69), Gene(70), Gene(71), Gene(72), Gene(73), Gene(74), Gene(75)]]\n",
      "next node\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'HOG' object has no attribute 'name'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-55-921ff4939eb2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mcounter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mnode\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mham_analysis_no_name\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtaxonomy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtraverse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'postorder'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0massign_neighbors\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"neighbors assigned\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mget_synteny_blocks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-54-c221d574be1e>\u001b[0m in \u001b[0;36massign_neighbors\u001b[0;34m(genome)\u001b[0m\n\u001b[1;32m     37\u001b[0m                     \u001b[0mgenome_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mchild\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mchild\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m# otherwise add the child\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mancestral_gene\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m \u001b[0;31m#            print(genome, children)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m \u001b[0;31m#            print(genome_dict)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'HOG' object has no attribute 'name'"
     ]
    }
   ],
   "source": [
    "counter = 1\n",
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    assign_neighbors(node.genome)\n",
    "    print(\"neighbors assigned\")\n",
    "    get_synteny_blocks(node.genome)\n",
    "    print(\"next node\")\n",
    "    assign_block_to_genes(node.genome)\n",
    "    \n",
    "    # here we go through all nodes, and if it is an ancestral, note, we make some plots\n",
    "    # in my 3 species example, there are 2 ancestral nodes, so two plots\n",
    "    \n",
    "    if isinstance(node.genome, pyham.AncestralGenome):\n",
    "        # links together the primary reconstruction's synteny blocks, using windows of 20\n",
    "        joining_blocks(20, node.genome)\n",
    "                \n",
    "        ### Plot ###        \n",
    "        title = node.genome.name  \n",
    "        raw_values = []\n",
    "        for block in node.genome.new_synteny:\n",
    "            count = 0\n",
    "            \n",
    "            for element in block:\n",
    "                # goes through the elements of that block, the HOGs         \n",
    "                # if that element (HOG) is a pyham.HOG, increase count, it's the number of HOGs we've alignd the synteny of in this block\n",
    "                if isinstance(element, pyham.HOG):\n",
    "                    count += 1  \n",
    "                    \n",
    "            raw_values.append(count)\n",
    "            \n",
    "        label = list(set(raw_values))\n",
    "        values = [0 for i in range(len(label))]\n",
    "        \n",
    "        for i in range(len(label)):\n",
    "            for element in raw_values:                \n",
    "                if label[i] == element:\n",
    "                    values[i] += 1\n",
    "\n",
    "        plt.bar(label, values)\n",
    "        plt.title(title)\n",
    "        plt.xlabel('Abstractgenes per block')\n",
    "        plt.ylabel('Number of blocks')\n",
    "        plt.savefig('results/secondary_reconstruction_plot{}.png'.format(counter))\n",
    "        counter += 1\n",
    "        plt.show()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py3]",
   "language": "python",
   "name": "conda-env-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
