{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.1.7'"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pyham\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Path to nwk file\n",
    "##nwk_path = \"test_data/EstimatedSpeciesTree.nwk\"\n",
    "#nwk_path = \"my_3spp_simulated_test_data/RealTree.nwk\"\n",
    "nwk_path = \"lower_mut_3spp/RealTree.nwk\"\n",
    "\n",
    "# The newick tree is extracted as a string\n",
    "tree_str = pyham.utils.get_newick_string(nwk_path, type=\"nwk\")\n",
    "\n",
    "# Path to OrthoXML file\n",
    "##orthoxml_path = \"test_data/HierarchicalGroups.orthoxml\"\n",
    "#orthoxml_path = \"my_3spp_simulated_test_data/Output/HierarchicalGroups.orthoxml\"\n",
    "orthoxml_path = \"lower_mut_3spp/Output/HierarchicalGroups.orthoxml\"\n",
    "\n",
    "#ham_analysis = pyham.Ham(tree_str, orthoxml_path, use_internal_name=True)\n",
    "ham_analysis_no_name = pyham.Ham(tree_str, orthoxml_path, use_internal_name=False)\n",
    "\n",
    "pyham.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "def locus_to_id(file):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function extracts loci from a fasta file and store the information as keys in a dictionary \n",
    "    and the related protein ID as value.\n",
    "\n",
    "    Args:\n",
    "        file (:obj:`str`): path to the file of interest.\n",
    "\n",
    "    Returns:\n",
    "        a dictionary mapping the locus (:obj:`str`) with the protein ID (:obj:`str`).\n",
    "    \"\"\"\n",
    "    \n",
    "    locus_to_id = {}\n",
    "    for seq_record in SeqIO.parse(file, 'fasta'):\n",
    "        locus_to_id[seq_record.description.split('locus: ')[-1]] = seq_record.description\n",
    "        \n",
    "    return locus_to_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1': 'G1_SE001, sequence type: type1, locus: 1',\n",
       " '2': 'G2_SE001, sequence type: type1, locus: 2',\n",
       " '3': 'G3_SE001, sequence type: type1, locus: 3',\n",
       " '4': 'G4_SE001, sequence type: type1, locus: 4',\n",
       " '5': 'G5_SE001, sequence type: type1, locus: 5',\n",
       " '6': 'G6_SE001, sequence type: type1, locus: 6',\n",
       " '7': 'G7_SE001, sequence type: type1, locus: 7',\n",
       " '8': 'G8_SE001, sequence type: type1, locus: 8',\n",
       " '9': 'G9_SE001, sequence type: type1, locus: 9',\n",
       " '10': 'G10_SE001, sequence type: type1, locus: 10',\n",
       " '11': 'G11_SE001, sequence type: type1, locus: 11',\n",
       " '12': 'G12_SE001, sequence type: type1, locus: 12',\n",
       " '13': 'G13_SE001, sequence type: type1, locus: 13',\n",
       " '14': 'G14_SE001, sequence type: type1, locus: 14',\n",
       " '15': 'G15_SE001, sequence type: type1, locus: 15',\n",
       " '16': 'G16_SE001, sequence type: type1, locus: 16',\n",
       " '17': 'G17_SE001, sequence type: type1, locus: 17',\n",
       " '18': 'G18_SE001, sequence type: type1, locus: 18',\n",
       " '19': 'G19_SE001, sequence type: type1, locus: 19',\n",
       " '20': 'G20_SE001, sequence type: type1, locus: 20',\n",
       " '21': 'G21_SE001, sequence type: type1, locus: 21',\n",
       " '22': 'G22_SE001, sequence type: type1, locus: 22',\n",
       " '23': 'G23_SE001, sequence type: type1, locus: 23',\n",
       " '24': 'G24_SE001, sequence type: type1, locus: 24',\n",
       " '25': 'G25_SE001, sequence type: type1, locus: 25'}"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# locus_to_id(\"test_data/DB/SE001_aa.fa\")\n",
    "##locus_to_id(\"my_3spp_simulated_test_data/DB/SE001_aa.fa\")\n",
    "locus_to_id(\"lower_mut_3spp/DB/SE001_aa.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_locus(directory):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function goes through each file of a folder and build a dictionary containing loci \n",
    "    as keys and the related protein ID as values. For each entry in the dictionary, the function \n",
    "    updates the corresponding :obj:`pyham.abstractgene.Gene` by adding its locus.\n",
    "\n",
    "    Args:\n",
    "        directory (:obj:`str`): directory of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "\n",
    "    path = directory\n",
    "    for file in os.listdir(path):\n",
    "        ##print('File:', file)\n",
    "        path_to_file = path + '/' + file\n",
    "        locus_dictionary = locus_to_id(path_to_file)\n",
    " \n",
    "        for (loci, gene_id) in locus_dictionary.items():\n",
    "            gene = ham_analysis_no_name.get_genes_by_external_id(gene_id)\n",
    "            \n",
    "            # each gene should only map to 1 other gene, so should always be length 1...\n",
    "            assert len(gene) == 1, 'gene_id maps to multiple entries'\n",
    "            gene[0].locus = loci\n",
    "            #for element in gene:\n",
    "            #    element.locus = loci\n",
    "#                print(element.locus)\n",
    "                \n",
    "        ham_analysis_no_name.get_extant_genome_by_name('.'.join(file.split('.')[:-1])).locus_assignment = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get_locus('test_data/DB')\n",
    "get_locus('my_3spp_simulated_test_data/DB')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_by_locus(genome, locus):\n",
    "    \n",
    "    \"\"\"\n",
    "    Get a :obj:`pyham.abstractgene.Gene` that match the query locus. \n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant genome of interest.\n",
    "        locus (:obj:`int`): locus.\n",
    "\n",
    "    Returns:\n",
    "        :obj:`pyham.abstractgene.Gene`.\n",
    "    \"\"\"\n",
    "    \n",
    "    for gene in genome.genes:\n",
    "        if int(gene.locus) == locus:\n",
    "            return gene \n",
    "    \n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_neighbors(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function assigns neighbors to each :obj:`pyham.abstractgene.Gene` or :obj:`pyham.abstractgene.HOG`.\n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    if isinstance(genome, pyham.ExtantGenome):\n",
    "        \n",
    "        for extant_gene in genome.genes:\n",
    "            extant_gene.neighbors = (get_gene_by_locus(genome, int(extant_gene.locus)-1), \n",
    "                                     get_gene_by_locus(genome, int(extant_gene.locus)+1))            \n",
    "                        \n",
    "    if isinstance(genome, pyham.AncestralGenome):\n",
    "    \n",
    "        for ancestral_gene in genome.genes:  \n",
    "            \n",
    "            # test here, see if does the same thing, children should only be direct descendants?\n",
    "            #children = [child for child in ancestral_gene.children]\n",
    "            children = ancestral_gene.children\n",
    "            \n",
    "            ######################################################################################\n",
    "            # Dictionary containing genome names as keys (i.e. SE004) and a list of children as value #\n",
    "            ######################################################################################\n",
    "            \n",
    "            genome_dict = {}  \n",
    "            \n",
    "            for child in children:\n",
    "                if child.genome.name in genome_dict:\n",
    "                    genome_dict[child.genome.name].append(child) # if already in dictionary, add it (duplicate)\n",
    "                else:\n",
    "                    genome_dict[child.genome.name] = [child] # otherwise add the child\n",
    "\n",
    "#            print(ancestral_gene)\n",
    "#            print(genome, children)\n",
    "#            print(genome_dict)\n",
    "            \n",
    "            ############################################\n",
    "            # Children are replaced by their neighbors #      \n",
    "            ############################################\n",
    "            \n",
    "            for genome, gene_collection in genome_dict.items(): #stoca \n",
    "                for i in range(len(gene_collection)):\n",
    "                    gene_collection[i] = list(gene_collection[i].neighbors)\n",
    "            \n",
    "#            print(genome_dict)\n",
    "\n",
    "            ################################################################################\n",
    "            # Dictionary containing genomes as keys (i.e. SE004) and a list of parent HOGs #\n",
    "            ################################################################################\n",
    "            \n",
    "            parent_dict = {} # gets a list of all potential neighboring hogs to the ancestral_gene\n",
    "            for genome, neighbors_collection in genome_dict.items(): #stoca \n",
    "                print(parent_dict)\n",
    "                for neighbors_pair in neighbors_collection:\n",
    "                    for neighbor in neighbors_pair: # don't really need the for loop bc only ever 2 neighbors\n",
    "          ##              print(neighbor, neighbor.parent)#, genome, ancestral_gene, parent_dict)\n",
    "                        if genome in parent_dict:   ## KJG - I don't understand this if/else - if we already put the focal genome in the parent dictionary, we do the same thing as if we didn't?\n",
    "          ##                  print(\"genome in parent dict\")\n",
    "                            if neighbor == None:\n",
    "                                parent_dict[genome].add(None)\n",
    "                            else:\n",
    "                                if neighbor.parent != ancestral_gene: #if theres' a duplication this might be true\n",
    "                                    parent_dict[genome].add(neighbor.parent) #add the neighbors parents - the neighboring hogs\n",
    "                                    print(neighbor, neighbor.parent, \"in parent dict\")\n",
    "                        else:\n",
    "                            if neighbor == None:\n",
    "                                parent_dict[genome] = {None}    ## ah, this makes focal genome 'none' while above does 'add(None)' in addition to current genome?\n",
    "                            else:\n",
    "                                if neighbor.parent != ancestral_gene: # fill the parent dict by looking to see \n",
    "                                    parent_dict[genome] = {neighbor.parent}  ## again, no 'add', just make\n",
    "                                    print(neighbor, neighbor.parent, \"not in parent dict\")\n",
    "          ##              print(neighbor, neighbor.parent, genome, ancestral_gene, parent_dict)\n",
    "#parent_dict = defaultdict(list)\n",
    "#if neighbor == None:\n",
    "#     parent_dict[genome].add(None)\n",
    "#else:\n",
    "# if neighbor.parent != ancestral_gene:\n",
    "# parent_dict[genome].add(neighbor.parent)\n",
    "            print(genome_dict)\n",
    "            print(parent_dict)\n",
    "\n",
    "            l = []  # List of sets\n",
    "            for genome, parent in parent_dict.items():  \n",
    "                print(genome, parent, l)\n",
    "                l.append(parent)  # parent is a set\n",
    "                ancestral_gene.list_parents = l            # List of potential neighbors in case of conflict\n",
    "                \n",
    "            final_neighbors = list(l[0].intersection(*l))  # Intersection of the sets\n",
    "            \n",
    "            if len(final_neighbors) == 2:\n",
    "                ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "                print(\"fin\", ancestral_gene.neighbors)\n",
    "            \n",
    "            elif len(final_neighbors) == 1:\n",
    "                none_counter = 0\n",
    "                potential_neighbor = None\n",
    "                \n",
    "                for collection in l:\n",
    "                    if len(collection) == 2:\n",
    "                        for element in collection:\n",
    "                            if element == final_neighbors[0]:\n",
    "                                continue\n",
    "                            elif element == None:\n",
    "                                none_counter += 1\n",
    "                            else:\n",
    "                                potential_neighbor = element\n",
    "                \n",
    "                    else:\n",
    "                        break\n",
    "                \n",
    "                if none_counter == 1:  # To deal with missing orthology\n",
    "                    final_neighbors.append(potential_neighbor)\n",
    "                \n",
    "                if len(final_neighbors) == 1:\n",
    "                    final_neighbors.append(None)\n",
    "                    ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "                    print(\"len1\", ancestral_gene.neighbors)\n",
    "                else:  \n",
    "                    ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "                    print(\"else\", ancestral_gene.neighbors)\n",
    "            \n",
    "            elif len(final_neighbors) == 0 or len(final_neighbors) > 2:\n",
    "                final_neighbors = [None, None]\n",
    "                ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "\n",
    "        ##        print(\"anc\", ancestral_gene.neighbors)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_synteny_blocks(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function aims to reconstruct synteny blocks using neighbors of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        list of lists containing :obj:`pyham.abstractgene.Gene` or :obj:`pyham.abstractgene.HOG` \n",
    "        in a specific order.\n",
    "    \"\"\"\n",
    "\n",
    "    syn_blocks = []\n",
    "    seen = []\n",
    "    gene_collection = genome.genes\n",
    "    tmp = []\n",
    "        \n",
    "    for abstract_gene in gene_collection: # go through all the genes (extant or ancestral) for the genome\n",
    "        \n",
    "        if abstract_gene in seen: # if the gene is already there, skip out of the loop to the next gene\n",
    "            continue\n",
    "                \n",
    "        block1 = []\n",
    "        block2 = []\n",
    "        \n",
    "        current1 = abstract_gene\n",
    "        current2 = abstract_gene  # Used for the block2\n",
    "        \n",
    "        block1.append(current1)  # [X] where X is the first abstract gene # add the current gene to block 1\n",
    "        seen.append(current1) # append the current gene to seen\n",
    "        \n",
    "        ######################\n",
    "        # First end (block1) #\n",
    "        ######################\n",
    "        \n",
    "        next_gene = current1.neighbors[0] # find the neighbors of the current gene\n",
    "        \n",
    "        while next_gene != None: # if there is a next gene\n",
    "            \n",
    "            if next_gene in seen:\n",
    "                next_gene = None  \n",
    "                continue\n",
    "            \n",
    "            previous = current1 # then put the current gene as previous\n",
    "            current1 = next_gene # and put the next gene as current\n",
    "            \n",
    "            if previous in current1.neighbors:   # if the previous gene is a neighbor of the new current gene\n",
    "                block1.append(current1) # append the new current gene to block 1\n",
    "                seen.append(current1)   # append the new current gene to seen\n",
    "                # this step should always be true and the else statement is a check that breaks if not true\n",
    "            \n",
    "            else:                                  \n",
    "                break\n",
    "                print(\"previous gene is not a neighbor of the current gene\")\n",
    "                \n",
    "            if previous == current1.neighbors[0]: # if the previous gene is the first neighbor of the current gene\n",
    "                next_gene = current1.neighbors[1] # then the next gene equals the 2nd neighbor of the current gene\n",
    "            \n",
    "            else:\n",
    "                next_gene = current1.neighbors[0] # otherwise the next gene is the first neighbor of the current gene\n",
    "            # then iterate the while loop\n",
    "\n",
    "        #######################\n",
    "        # Second end (block2) #\n",
    "        #######################\n",
    "        \n",
    "        next_gene = current2.neighbors[1] # current2 is the same original abstract gene from above\n",
    "                                          # take its 2nd neighbor and make it the next gene\n",
    "            \n",
    "        while next_gene != None:    # if there is a next gene\n",
    "            \n",
    "            if next_gene in seen:   # if it's already seen, next = none and iterate the while loop\n",
    "                next_gene = None\n",
    "                continue\n",
    "            \n",
    "            previous = current2    # make the current gene become the previous gene\n",
    "            current2 = next_gene   # make the next gene become the current gene\n",
    "            \n",
    "            if previous in current2.neighbors:   # if the previous gene is a neighbor of the new next current gene\n",
    "                block2.append(current2)  # append the new current gene to block 2\n",
    "                seen.append(current2)    # append the new current gene to seen\n",
    "            \n",
    "            else:                                  \n",
    "                break\n",
    "                print(\"previous gene is not a neighbor of the current2 gene\")\n",
    "                \n",
    "            if previous == current2.neighbors[0]: # if the previous gene is the first neighbor of the current gene\n",
    "                next_gene = current2.neighbors[1] # then the next gene equals the 2nd neighbor of the current gene\n",
    "            \n",
    "            else:\n",
    "                next_gene = current2.neighbors[0] # otherwise the next gene is the first neighbor of the current gene\n",
    "            # then iterate the while loop\n",
    "            \n",
    "#        print(abstract_gene)\n",
    "#        print(current1.neighbors)\n",
    "#        print(block1)\n",
    "#        print(block2)\n",
    "        syn_blocks.append(block1[::-1] + block2)\n",
    "        \n",
    "    print(syn_blocks)\n",
    "    genome.synteny = syn_blocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_block_to_genes(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    The function links each :obj:`pyham.abstractgene.AbstractGene` to the synteny block it belongs to. \n",
    "    It creates a dictionary containing the genome of :obj:`pyham.abstractgene` as key and the synteny \n",
    "    block as value.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    for block in genome.synteny:\n",
    "    \n",
    "        for abstract_gene in block:\n",
    "            genomes_to_block = {}  # Dictionary\n",
    "            genomes_to_block[genome.name] = block  \n",
    "            abstract_gene.blocks = genomes_to_block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_children(genome, children):  \n",
    "    \n",
    "    \"\"\"\n",
    "    Due to tandem duplication, some :obj:`pyham.abstractgene.HOG` can have more than two children.\n",
    "    The function refines the list of children by removing duplicates.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        children (:obj:`list` of :obj:`pyham.abstractgene.AbstractGene`): a list of direct descendants AbstractGene.\n",
    "    \n",
    "    Returns:\n",
    "        list of :obj:`pyham.abstractgene.AbstractGene` without duplicates.\n",
    "    \"\"\"\n",
    "    \n",
    "    unique_genome = []\n",
    "    unique_children = []\n",
    "    \n",
    "    if len(children) == 2:  # Two children comming from two different genomes\n",
    "        \n",
    "        if children[0].genome != children[1].genome:\n",
    "            return children\n",
    "        \n",
    "        else:  # Two children comming from the same genome (i.e. duplication)\n",
    "            unique_children.append(None)\n",
    "            return unique_children\n",
    "    \n",
    "    else:  # In case of duplications, only one child per specie is taken into account\n",
    "        \n",
    "        for child in children:  \n",
    "            \n",
    "            if child.genome in unique_genome:\n",
    "                continue\n",
    "\n",
    "            else:\n",
    "                \n",
    "                for neighbor in child.neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        continue\n",
    "\n",
    "                    elif neighbor.parent:\n",
    "                        \n",
    "                        if neighbor.parent in child.parent.blocks[genome.name]:  # To make sure that, for both unique children, we go in the same direction. \n",
    "                            continue\n",
    "                        \n",
    "                        unique_genome.append(child.genome)\n",
    "                        unique_children.append(child)\n",
    "\n",
    "                    else:  # If the child has no parent HOG\n",
    "                        unique_genome.append(child.genome)\n",
    "                        unique_children.append(child)\n",
    "        \n",
    "        if len(unique_children) == 0:\n",
    "            unique_children.append(None)\n",
    "            return unique_children  \n",
    "        \n",
    "        else:\n",
    "            return unique_children  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_unique_children_from_unit_block(genome, children, genes_seen):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function aims to be used after refining the :obj:`list` genes_seen. It uses the :obj:`list` \n",
    "    genes_seen to check whether the :obj:`list` current_children returned by the function get_unique_children \n",
    "    are correctly selected. Indeed, the function get_unique_children encounters a limitation with single HOG.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        children (:obj:`list` of :obj:`pyham.abstractgene.AbstractGene`): a list of direct descendants AbstractGene.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        \n",
    "    Returns:\n",
    "        list of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "    \"\"\"\n",
    "    \n",
    "    unique_genome = []\n",
    "    unique_children = []\n",
    "    \n",
    "    if len(children) == 2:  # Two children comming from two different genomes\n",
    "        \n",
    "        if children[0].genome != children[1].genome:\n",
    "            return children\n",
    "        \n",
    "        else:  # Two children comming from the same genome (i.e. duplication)\n",
    "            unique_children.append(None)\n",
    "            return unique_children\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        for child in children:  \n",
    "            \n",
    "            if child.genome in unique_genome:\n",
    "                continue\n",
    "            \n",
    "            for neighbor in child.neighbors:\n",
    "                \n",
    "                if neighbor == None:\n",
    "                    continue\n",
    "\n",
    "                if neighbor.parent in child.parent.blocks[genome.name] or neighbor in genes_seen:  # To make sure that for both unique children we go in the same direction \n",
    "                    continue\n",
    "                \n",
    "                unique_genome.append(child.genome)\n",
    "                unique_children.append(child)\n",
    "                \n",
    "        if len(unique_children) == 0:\n",
    "            unique_children.append(None)\n",
    "            return unique_children  # Return a list without duplicates\n",
    "        \n",
    "        else:\n",
    "            return unique_children  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Link:\n",
    "    \n",
    "    \"\"\"\n",
    "    Link is a class representing linking regions. Linking regions are unresolved regions with a weak synteny \n",
    "    signal that links two synteny blocks together. A Link is defined by its collection of poorly resolved genes.\n",
    "    \n",
    "    Attributes:\n",
    "        | linking_region: dictionary of :obj:`pyham.abstractgene.AbstractGene`\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.linking_region = {}\n",
    "        \n",
    "    def get_linking_region(self, genome, block, common_block, starting_children):\n",
    "        \n",
    "        \"\"\"\n",
    "        Create a linking region filled with :obj:`pyham.abstractgene.AbstractGene`.\n",
    "        \n",
    "        Args:\n",
    "            genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "            block (:obj:`list`): the synteny block from which the assessement starts.\n",
    "            common_block (:obj:`list`): the synteny block to be linked with the initial block.\n",
    "            starting_children (:obj:`list`): list of children coming from the last :obj:`pyham.abstractgene.AbstractGene` of the initial block.\n",
    "        \n",
    "        Returns:\n",
    "            return linking_region: dictionary of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "        \"\"\"\n",
    "        \n",
    "        initial_children = starting_children\n",
    "        seen = []\n",
    "\n",
    "        for child in initial_children:  # Creation of the dictionary's entries ({species1: [], species2: []})\n",
    "\n",
    "            if child.genome.name in self.linking_region:\n",
    "                continue\n",
    "\n",
    "            else:\n",
    "                self.linking_region[child.genome.name] = [] \n",
    "\n",
    "        next_children = []\n",
    "\n",
    "        for child in initial_children:\n",
    "            current_child = child\n",
    "            growing_link = True\n",
    "\n",
    "            while growing_link:\n",
    "                \n",
    "                for neighbor in current_child.neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor != current_child.next_neighbor:\n",
    "                        continue\n",
    "\n",
    "                    if neighbor.parent:\n",
    "\n",
    "                        if neighbor.parent.blocks[genome.name] != common_block:\n",
    "                            \n",
    "                            if neighbor.parent not in self.linking_region[neighbor.genome.name]:\n",
    "                                self.linking_region[neighbor.genome.name].append(neighbor.parent)\n",
    "                                current_child = neighbor\n",
    "                                seen.append(neighbor)\n",
    "\n",
    "                            else:\n",
    "                                current_child = neighbor\n",
    "                                seen.append(neighbor)\n",
    "\n",
    "                        else:\n",
    "                            growing_link = False\n",
    "\n",
    "                    else:\n",
    "                        self.linking_region[neighbor.genome.name].append(neighbor)  # The child is added to the linking region anyway.\n",
    "                        current_child = neighbor\n",
    "                        seen.append(neighbor)\n",
    "\n",
    "    def __repr__(self):\n",
    "        return '<{}>'.format(self.__class__.__name__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "def refine_genes_seen(genome, genes_seen, current_children):\n",
    "    \n",
    "    \"\"\"\n",
    "    Regarding the secondary reconstruction, when the growing process continue by starting from a single \n",
    "    :obj:`pyham.abstractgene.HOG`, both neighbors of one of its children can be within the \n",
    "    :obj:`list` genes_seen. The function will remove all children that come after the children of interest\n",
    "    within the :obj:`list` genes_seen.\n",
    "    \n",
    "    Ars:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        current_children (:obj:`list`): list of children to begin the investigation with at the next iteration.\n",
    "        \n",
    "    Returns:\n",
    "        return the refined :obj:`list` genes_seen.\n",
    "    \"\"\"\n",
    "    \n",
    "    if len(current_children) == 2:\n",
    "        tmp = genes_seen[:]  # Make an independent copy of the initial list\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == current_children[0]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == current_children[0].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:]\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == current_children[1]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == current_children[1].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:] \n",
    "\n",
    "    return genes_seen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "def adjust_genes_seen(genome, genes_seen, starting_children):\n",
    "    \n",
    "    \"\"\"\n",
    "    Same principle as for the function refine_genes_seen but used when the growing process continue by\n",
    "    sarting from an element at the extremity of a synteny block. Neighbors from one of its children can\n",
    "    be present within the :obj:`list` genes_seen. The function will remove all children that come after \n",
    "    the children of interest within the :obj:`list` genes_seen.\n",
    "    \n",
    "    Ars:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        starting_children (:obj:`list`): list of children to begin the investigation with at the next iteration.\n",
    "        \n",
    "    Returns:\n",
    "        return the refined :obj:`list` genes_seen.\n",
    "    \"\"\"\n",
    "    \n",
    "    if len(starting_children) == 2:\n",
    "        tmp = genes_seen[:]  # Make an independent copy of the initial list\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == starting_children[0]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == starting_children[0].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:]\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == starting_children[1]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == starting_children[1].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:] \n",
    "\n",
    "    return genes_seen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "def joining_blocks(window, genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    The function performs the secondary reconstruction. Synteny blocks from the primary reconstruction\n",
    "    are linked together using linking regions.\n",
    "    \n",
    "    Ars:\n",
    "        window (:obj:`int`): a window of assessment\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        \n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    for hog in genome.genes:  #### To be used with the function \"add_remaining_blocks\"\n",
    "        hog.processed = False\n",
    "    \n",
    "    new_blocks = []\n",
    "    block_record = []  # To keep track of the block already assessed. Useful to link several blocks in a row.\n",
    "    \n",
    "    for block in genome.synteny:\n",
    "            \n",
    "        if block in block_record or len(block) == 1:  \n",
    "            continue\n",
    "        \n",
    "        block2 = block  # Used to reconstruct the other side\n",
    "        \n",
    "        counter = 0\n",
    "        children = [child for child in block[-1].children]\n",
    "        current_children = get_unique_children(genome, children)  # Remove duplicates from children\n",
    "        starting_children = current_children  # To keep track of the initial children considered (used to build a linking region)\n",
    "        \n",
    "        if len(current_children) == 1 or len(current_children) > 2:  # If missing information (==1) or other special case (>2), the block is no longer taken into account\n",
    "            continue\n",
    "        \n",
    "        block_record.append(block)   \n",
    "        \n",
    "        blocks_seen = [set() for child in current_children]  # To save already assessed blocks and perform the intersection\n",
    "        genes_seen = []  # To keep track of already assessed genes\n",
    "        intersect = None\n",
    "        extended_block = []\n",
    "        next_children = [None, None]\n",
    "        genes_seen.append(current_children[0])  \n",
    "        genes_seen.append(current_children[1])  \n",
    "        \n",
    "        while counter < window:\n",
    "            none_type_as_neighbor = 0\n",
    "            block_found = False\n",
    "            reaching_already_assessed = False  \n",
    "            \n",
    "            if None in current_children:\n",
    "                break\n",
    "            \n",
    "            # Without the following block, when a common block is reached from one extremity, the next iteration \n",
    "            # starts from the opposite one. However, children originating from that extreme element are not\n",
    "            # added within the list of genes seen.\n",
    "            if current_children[0] and current_children[1] not in genes_seen:\n",
    "                genes_seen.append(current_children[0])\n",
    "                genes_seen.append(current_children[1])\n",
    "            \n",
    "            for i in range(len(current_children)):\n",
    "                \n",
    "                for neighbor in current_children[i].neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        none_type_as_neighbor += 1\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor.parent:  # Check whether the neighbor is a singleton (i.e. does it have a parent or not?)\n",
    "                        \n",
    "                        if neighbor.parent in block or neighbor in genes_seen:  # Avoid to take into account elements from the initial block or those already assessed\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            blocks_seen[i].add(tuple(neighbor.parent.blocks[genome.name]))\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region \n",
    "                            break\n",
    "                    \n",
    "                    else:\n",
    "                        \n",
    "                        if neighbor in genes_seen:\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region\n",
    "                \n",
    "            if none_type_as_neighbor > 0:  # If children have only one neighbor\n",
    "                break\n",
    "\n",
    "            counter += 1\n",
    "            current_children = next_children[:]\n",
    "            intersect = list(blocks_seen[0].intersection(*blocks_seen))  # Intersection\n",
    "            \n",
    "            if len(intersect) == 1:  # ex. intersect = [('a', 'c')]\n",
    "\n",
    "                    if list(intersect[0]) in block_record:\n",
    "                        reaching_already_assessed = True\n",
    "                        break\n",
    "                        \n",
    "                    block_found = True  \n",
    "                    counter = 0  # Reset the counter to pursue the assessement with further blocks\n",
    "                    common_block = list(intersect[0])  # Extraction of the last common block. \n",
    "                    block_record.append(common_block)\n",
    "\n",
    "                    link = Link()  # Class for linking regions\n",
    "                    link.get_linking_region(genome, block, common_block, starting_children)  # Creation of the linking region\n",
    "                    \n",
    "                    if not extended_block:  # If the list if empty (i.e. for a first block of a serie)\n",
    "                        extended_block.append(block)  # The initial block\n",
    "                        extended_block.append(link)  # The linking region\n",
    "                        extended_block.append(common_block)  # The common block\n",
    "\n",
    "                    else:  # For all the block but the first one, \"block\" corresponds to the previous common block\n",
    "                        extended_block.append(link)\n",
    "                        extended_block.append(common_block)\n",
    "\n",
    "                    block = common_block  # The new block becomes the last common block found within the while loop\n",
    "\n",
    "                    if len(block) > 1:\n",
    "                        \n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its first element (i.e. element 0), the next iteration should start from its last element (i.e. element -1) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block[0].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block[0].children)[0] and get_unique_children(genome, block[0].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block[-1].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)  \n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block[0].children)\n",
    "                            break\n",
    "\n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its last element (i.e. element -1), the next iteration should start from its first element (i.e. element 0) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block[-1].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block[-1].children)[0] and get_unique_children(genome, block[-1].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block[0].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block[-1].children)\n",
    "                            break\n",
    "                    \n",
    "                    ################################################################################################################\n",
    "                    # When we reach a common unit block, the next iteration should start from the same unique HOG (i.e. element 0) #\n",
    "                    ################################################################################################################\n",
    "\n",
    "                    elif len(block) == 1:  # If the block has only one HOG...\n",
    "                        current_children = get_unique_children(genome, block[0].children)                         \n",
    "                        \n",
    "                        if len(current_children) == 2:\n",
    "                            genes_seen = refine_genes_seen(genome, genes_seen, current_children)\n",
    "                            current_children = check_unique_children_from_unit_block(genome, block[0].children, genes_seen)\n",
    "\n",
    "                        starting_children = current_children\n",
    "                        \n",
    "                    blocks_seen = [set() for child in current_children]  # The variable is restarted for the next iteration\n",
    "\n",
    "            if reaching_already_assessed:  \n",
    "                break\n",
    "                \n",
    "            if len(current_children) == 2:\n",
    "                if current_children[0].genome != current_children[1].genome:\n",
    "                    continue\n",
    "                \n",
    "                break  # To exit the while loop\n",
    "                \n",
    "            else:  \n",
    "                break  # To exit the while loop\n",
    "        \n",
    "        tmp = [[], []]\n",
    "        if extended_block:\n",
    "            for item in extended_block:\n",
    "                if isinstance(item, Link):\n",
    "                    tmp[1].append(item)\n",
    "\n",
    "                else:\n",
    "                    for element in item:\n",
    "                        tmp[1].append(element)\n",
    "        \n",
    "        else:\n",
    "            for element in block:\n",
    "                tmp[1].append(element)  # If a new initial block reach directly an already assessed block (i.e. common block), the initial block is stored\n",
    "                \n",
    "###############################################################################################################\n",
    "\n",
    "        counter = 0\n",
    "        children = [child for child in block2[0].children]\n",
    "        current_children = get_unique_children(genome, children)  # Remove duplicates from children\n",
    "        starting_children = current_children\n",
    "        \n",
    "        if len(current_children) == 1 or len(current_children) > 2:  # If missing information (==1) or other special case (>2), the block is no longer taken into account\n",
    "            new_blocks.append(tmp[1])\n",
    "            continue\n",
    "        \n",
    "        blocks_seen = [set() for child in current_children]  # To save already assessed blocks and perform the intersection\n",
    "        genes_seen = []  # To keep track of already assessed genes (used for gene without parent to make to we go in one direction)\n",
    "        intersect = None\n",
    "        extended_block = []\n",
    "        next_children = [None, None]\n",
    "        genes_seen.append(current_children[0])  \n",
    "        genes_seen.append(current_children[1])  \n",
    "        \n",
    "        while counter < window:\n",
    "            none_type_as_neighbor = 0\n",
    "            block_found = False\n",
    "            reaching_already_assessed = False  \n",
    "            \n",
    "            if None in current_children:\n",
    "                break\n",
    "            \n",
    "            # Without the following block, when a common block is reached from one extremity, the next iteration \n",
    "            # starts from the opposite one. However, children originating from the that extreme element are not\n",
    "            # added within the list of genes seen.\n",
    "            if current_children[0] and current_children[1] not in genes_seen:\n",
    "                genes_seen.append(current_children[0])\n",
    "                genes_seen.append(current_children[1])\n",
    "            \n",
    "            for i in range(len(current_children)):\n",
    "                \n",
    "                for neighbor in current_children[i].neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        none_type_as_neighbor += 1\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor.parent:  # Check whether the neighbor is a singleton (i.e. does it have a parent or not?)\n",
    "                        \n",
    "                        if neighbor.parent in block2 or neighbor in genes_seen:  # Avoid to take into account elements from the initial block or those already assessed\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            blocks_seen[i].add(tuple(neighbor.parent.blocks[genome.name]))\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region \n",
    "                            break  \n",
    "\n",
    "                    else:\n",
    "                        \n",
    "                        if neighbor in genes_seen:\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region\n",
    "                \n",
    "            if none_type_as_neighbor > 0:  # If children have only one neighbor\n",
    "                break\n",
    "\n",
    "            counter += 1\n",
    "            current_children = next_children[:]\n",
    "            intersect = list(blocks_seen[0].intersection(*blocks_seen))  # Intersection\n",
    "            \n",
    "            if len(intersect) == 1:  # ex. intersect = [('a', 'c')]\n",
    "\n",
    "                    if list(intersect[0]) in block_record:\n",
    "                        reaching_already_assessed = True\n",
    "                        break\n",
    "                        \n",
    "                    block_found = True  \n",
    "                    counter = 0  # Reset the counter to pursue the assessement with further blocks\n",
    "                    common_block = list(intersect[0])  # Extraction of the last common block. \n",
    "                    block_record.append(common_block)\n",
    "\n",
    "                    link = Link()  # Class for linking regions\n",
    "                    link.get_linking_region(genome, block2, common_block, starting_children)  # Creation of the linking region\n",
    "                    \n",
    "                    extended_block.append(link)  # The linking region\n",
    "                    extended_block.append(common_block)  # The common block\n",
    "\n",
    "                    block2 = common_block  # The new block becomes the last common block found within the while loop\n",
    "\n",
    "                    if len(block2) > 1:\n",
    "                        \n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its first element (i.e. element 0), the next iteration should start from its last element (i.e. element -1) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block2[0].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block2[0].children)[0] and get_unique_children(genome, block2[0].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block2[-1].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block2[0].children)\n",
    "                            break\n",
    "\n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its last element (i.e. element -1), the next iteration should start from its first element (i.e. element 0) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block2[-1].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block2[-1].children)[0] and get_unique_children(genome, block2[-1].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block2[0].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  # 05.01.18\n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)  # 05.01.18\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block2[-1].children)\n",
    "                            break\n",
    "                    \n",
    "                    ################################################################################################################\n",
    "                    # When we reach a common unit block, the next iteration should start from the same unique HOG (i.e. element 0) #\n",
    "                    ################################################################################################################\n",
    "\n",
    "                    elif len(block2) == 1:  # If the block has only one HOG...\n",
    "                        current_children = get_unique_children(genome, block2[0].children) \n",
    "                        \n",
    "                        if len(current_children) == 2:\n",
    "                            genes_seen = refine_genes_seen(genome, genes_seen, current_children)\n",
    "                            current_children = check_unique_children_from_unit_block(genome, block2[0].children, genes_seen)\n",
    "\n",
    "                        starting_children = current_children\n",
    "                        \n",
    "                    blocks_seen = [set() for child in current_children]  # The variable is restarted for the next iteration\n",
    "\n",
    "            if reaching_already_assessed:  \n",
    "                break\n",
    "                \n",
    "            if len(current_children) == 2:\n",
    "                if current_children[0].genome != current_children[1].genome:\n",
    "                    continue\n",
    "                \n",
    "                break  # To exit the while loop\n",
    "                \n",
    "            else:  \n",
    "                break  # To exit the while loop\n",
    "\n",
    "        if extended_block:\n",
    "            for item in extended_block:\n",
    "                if isinstance(item, Link):\n",
    "                    tmp[0].append(item)\n",
    "\n",
    "                else:\n",
    "                    for element in item:\n",
    "                        tmp[0].append(element)\n",
    "        \n",
    "        # In tmp[1] there is always something, namely either block or extended block.\n",
    "        # The list tmp[0] can be empty\n",
    "        if tmp[0] and tmp[1]:\n",
    "            tmp[0] = tmp[0][::-1]\n",
    "            new_blocks.append(tmp[0] + tmp[1])\n",
    "            \n",
    "        elif tmp[1] and not tmp[0]:\n",
    "            new_blocks.append(tmp[1])\n",
    "            \n",
    "    genome.new_synteny = new_blocks\n",
    "\n",
    "    add_remaining_blocks(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_remaining_blocks(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    Once the secondary reconstruction is done. Some blocks or single HOGs from the primary reconstruction may\n",
    "    remained unassessed. The function add such items within the collection of extended synteny blocks.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        \n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    c = 0\n",
    "    \n",
    "    for block in genome.new_synteny:\n",
    "        \n",
    "        for element in block:\n",
    "            \n",
    "            if not isinstance(element, pyham.HOG):\n",
    "                \n",
    "                for key, value in element.linking_region.items():\n",
    "                    \n",
    "                    for gene in value:\n",
    "                        \n",
    "                        if isinstance(gene, pyham.HOG):\n",
    "                            gene.processed = True\n",
    "                            c += 1\n",
    "\n",
    "            else:\n",
    "                element.processed = True\n",
    "                c += 1\n",
    "    \n",
    "    not_processed = 0\n",
    "    \n",
    "    for gene in genome.genes:\n",
    "        \n",
    "        if gene.processed == False and gene.blocks[genome.name] not in genome.new_synteny:\n",
    "            genome.new_synteny.append(gene.blocks[genome.name])\n",
    "            not_processed += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--SE001_aa\n",
      "SE001_aa\n",
      "[Gene(1), Gene(2), Gene(3), Gene(4), Gene(5), Gene(6), Gene(7), Gene(8), Gene(9), Gene(10), Gene(11), Gene(12), Gene(13), Gene(14), Gene(15), Gene(16), Gene(17), Gene(18), Gene(19), Gene(20), Gene(21), Gene(22), Gene(23), Gene(24), Gene(25)]\n",
      "\n",
      "--SE002_aa\n",
      "SE002_aa\n",
      "[Gene(26), Gene(27), Gene(28), Gene(29), Gene(30), Gene(31), Gene(32), Gene(33), Gene(34), Gene(35), Gene(36), Gene(37), Gene(38), Gene(39), Gene(40), Gene(41), Gene(42), Gene(43), Gene(44), Gene(45), Gene(46), Gene(47), Gene(48), Gene(49), Gene(50)]\n",
      "\n",
      "--SE003_aa\n",
      "SE003_aa\n",
      "[Gene(51), Gene(52), Gene(53), Gene(54), Gene(55), Gene(56), Gene(57), Gene(58), Gene(59), Gene(60), Gene(61), Gene(62), Gene(63), Gene(64), Gene(65), Gene(66), Gene(67), Gene(68), Gene(69), Gene(70), Gene(71), Gene(72), Gene(73), Gene(74), Gene(75)]\n",
      "\n",
      "   /-SE002_aa\n",
      "--|\n",
      "   \\-SE003_aa\n",
      "SE002_aa/SE003_aa\n",
      "[<HOG(1)>, <HOG(2)>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>, <HOG()>]\n",
      "\n",
      "   /-SE001_aa\n",
      "--|\n",
      "  |   /-SE002_aa\n",
      "   \\-|\n",
      "      \\-SE003_aa\n",
      "SE001_aa/SE002_aa/SE003_aa\n",
      "[<HOG(3)>, <HOG(4)>, <HOG(5)>, <HOG(6)>, <HOG(7)>, <HOG(8)>, <HOG(9)>, <HOG(10)>, <HOG(11)>, <HOG(12)>, <HOG(13)>]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    print(node)\n",
    "    print(node.genome)\n",
    "    print(node.genome.genes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    \n",
    "    genome=node.genome   \n",
    "    if isinstance(genome, pyham.ExtantGenome):\n",
    "        \n",
    "        for extant_gene in genome.genes:\n",
    "            extant_gene.neighbors = (get_gene_by_locus(genome, int(extant_gene.locus)-1), \n",
    "                                     get_gene_by_locus(genome, int(extant_gene.locus)+1))\n",
    "            #print(extant_gene.neighbors)\n",
    "                        \n",
    "    if isinstance(genome, pyham.AncestralGenome):\n",
    "    \n",
    "##        print(genome.genes)\n",
    "        for ancestral_gene in genome.genes:  \n",
    "            children = ancestral_gene.children\n",
    "            \n",
    "##            print(ancestral_gene)\n",
    "##            print(genome, children)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "neighbors assigned\n",
      "[[Gene(1), Gene(2), Gene(3), Gene(4), Gene(5), Gene(6), Gene(7), Gene(8), Gene(9), Gene(10), Gene(11), Gene(12), Gene(13), Gene(14), Gene(15), Gene(16), Gene(17), Gene(18), Gene(19), Gene(20), Gene(21), Gene(22), Gene(23), Gene(24), Gene(25)]]\n",
      "next node\n",
      "neighbors assigned\n",
      "[[Gene(26), Gene(27), Gene(28), Gene(29), Gene(30), Gene(31), Gene(32), Gene(33), Gene(34), Gene(35), Gene(36), Gene(37), Gene(38), Gene(39), Gene(40), Gene(41), Gene(42), Gene(43), Gene(44), Gene(45), Gene(46), Gene(47), Gene(48), Gene(49), Gene(50)]]\n",
      "next node\n",
      "neighbors assigned\n",
      "[[Gene(51), Gene(52), Gene(53), Gene(54), Gene(55), Gene(56), Gene(57), Gene(58), Gene(59), Gene(60), Gene(61), Gene(62), Gene(63), Gene(64), Gene(65), Gene(66), Gene(67), Gene(68), Gene(69), Gene(70), Gene(71), Gene(72), Gene(73), Gene(74), Gene(75)]]\n",
      "next node\n",
      "{'SE002_aa': [Gene(27)], 'SE003_aa': [Gene(67)]}\n",
      "Gene(26) None not in parent dict\n",
      "Gene(28) None in parent dict\n",
      "Gene(66) <HOG()> not in parent dict\n",
      "Gene(68) <HOG(2)> in parent dict\n",
      "{'SE002_aa': [[Gene(26), Gene(28)]], 'SE003_aa': [[Gene(66), Gene(68)]]}\n",
      "{'SE002_aa': {None}, 'SE003_aa': {<HOG(2)>, <HOG()>}}\n",
      "{'SE002_aa': [Gene(41)], 'SE003_aa': [Gene(68)]}\n",
      "Gene(40) None not in parent dict\n",
      "Gene(42) None in parent dict\n",
      "Gene(67) <HOG(1)> not in parent dict\n",
      "Gene(69) None in parent dict\n",
      "{'SE002_aa': [[Gene(40), Gene(42)]], 'SE003_aa': [[Gene(67), Gene(69)]]}\n",
      "{'SE002_aa': {None}, 'SE003_aa': {<HOG(1)>, None}}\n",
      "len1 (None, None)\n",
      "{'SE002_aa': [Gene(50)], 'SE003_aa': [Gene(75)]}\n",
      "Gene(49) None not in parent dict\n",
      "Gene(74) None not in parent dict\n",
      "{'SE002_aa': [[Gene(49), None]], 'SE003_aa': [[Gene(74), None]]}\n",
      "{'SE002_aa': {None}, 'SE003_aa': {None}}\n",
      "len1 (None, None)\n",
      "{'SE003_aa': [Gene(71)]}\n",
      "Gene(70) <HOG()> not in parent dict\n",
      "Gene(72) None in parent dict\n",
      "{'SE003_aa': [[Gene(70), Gene(72)]]}\n",
      "{'SE003_aa': {<HOG()>, None}}\n",
      "fin (<HOG()>, None)\n",
      "{'SE002_aa': [Gene(48)]}\n",
      "Gene(47) <HOG()> not in parent dict\n",
      "Gene(49) None in parent dict\n",
      "{'SE002_aa': [[Gene(47), Gene(49)]]}\n",
      "{'SE002_aa': {<HOG()>, None}}\n",
      "fin (<HOG()>, None)\n",
      "{'SE002_aa': [Gene(44)]}\n",
      "Gene(43) None not in parent dict\n",
      "Gene(45) <HOG()> in parent dict\n",
      "{'SE002_aa': [[Gene(43), Gene(45)]]}\n",
      "{'SE002_aa': {<HOG()>, None}}\n",
      "fin (<HOG()>, None)\n",
      "{'SE002_aa': [Gene(29)]}\n",
      "Gene(28) None not in parent dict\n",
      "Gene(30) None in parent dict\n",
      "{'SE002_aa': [[Gene(28), Gene(30)]]}\n",
      "{'SE002_aa': {None}}\n",
      "len1 (None, None)\n",
      "{'SE002_aa': [Gene(45)], 'SE003_aa': [Gene(70)]}\n",
      "Gene(44) <HOG()> not in parent dict\n",
      "Gene(46) None in parent dict\n",
      "Gene(69) None not in parent dict\n",
      "Gene(71) <HOG()> in parent dict\n",
      "{'SE002_aa': [[Gene(44), Gene(46)]], 'SE003_aa': [[Gene(69), Gene(71)]]}\n",
      "{'SE002_aa': {<HOG()>, None}, 'SE003_aa': {<HOG()>, None}}\n",
      "len1 (None, None)\n",
      "{'SE003_aa': [Gene(65)]}\n",
      "Gene(64) None not in parent dict\n",
      "Gene(66) <HOG()> in parent dict\n",
      "{'SE003_aa': [[Gene(64), Gene(66)]]}\n",
      "{'SE003_aa': {<HOG()>, None}}\n",
      "fin (<HOG()>, None)\n",
      "{'SE002_aa': [Gene(36)], 'SE003_aa': [Gene(66)]}\n",
      "Gene(35) None not in parent dict\n",
      "Gene(37) None in parent dict\n",
      "Gene(65) <HOG()> not in parent dict\n",
      "Gene(67) <HOG(1)> in parent dict\n",
      "{'SE002_aa': [[Gene(35), Gene(37)]], 'SE003_aa': [[Gene(65), Gene(67)]]}\n",
      "{'SE002_aa': {None}, 'SE003_aa': {<HOG()>, <HOG(1)>}}\n",
      "{'SE002_aa': [Gene(31)]}\n",
      "Gene(30) None not in parent dict\n",
      "Gene(32) None in parent dict\n",
      "{'SE002_aa': [[Gene(30), Gene(32)]]}\n",
      "{'SE002_aa': {None}}\n",
      "len1 (None, None)\n",
      "{'SE002_aa': [Gene(47)]}\n",
      "Gene(46) None not in parent dict\n",
      "Gene(48) <HOG()> in parent dict\n",
      "{'SE002_aa': [[Gene(46), Gene(48)]]}\n",
      "{'SE002_aa': {<HOG()>, None}}\n",
      "fin (<HOG()>, None)\n",
      "{'SE003_aa': [Gene(73)]}\n",
      "Gene(72) None not in parent dict\n",
      "Gene(74) None in parent dict\n",
      "{'SE003_aa': [[Gene(72), Gene(74)]]}\n",
      "{'SE003_aa': {None}}\n",
      "len1 (None, None)\n",
      "{'SE003_aa': [Gene(58)]}\n",
      "Gene(57) None not in parent dict\n",
      "Gene(59) None in parent dict\n",
      "{'SE003_aa': [[Gene(57), Gene(59)]]}\n",
      "{'SE003_aa': {None}}\n",
      "len1 (None, None)\n",
      "{'SE003_aa': [Gene(56)]}\n",
      "Gene(55) None not in parent dict\n",
      "Gene(57) None in parent dict\n",
      "{'SE003_aa': [[Gene(55), Gene(57)]]}\n",
      "{'SE003_aa': {None}}\n",
      "len1 (None, None)\n",
      "neighbors assigned\n",
      "[[<HOG(1)>], [<HOG(2)>], [<HOG()>], [<HOG()>], [<HOG()>, <HOG()>], [<HOG()>], [<HOG()>], [<HOG()>], [<HOG()>], [<HOG()>], [<HOG()>], [<HOG()>], [<HOG()>], [<HOG()>]]\n",
      "next node\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAcsElEQVR4nO3deZgdVZ3/8fcnEFBIIAkJDFsIOAgDiAO2IgKiLIogi4gji7IMP4OOIo48KjoCjqiDoyjjChlZZVPWYQkaBgkMAoEOYUkIu1ERMI0sIWEP398f5zQpLrdvKt1d93anPq/nqedWnVO36luVm+89faruKUUEZmZWHyM6HYCZmbWXE7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048dugk7S9pJskPSPpSUm/l/ROSYdKWixpYcO0Tn7fOEmXSlok6Y+SDmzY7oG5fJGkyySNy+UrSzot1z0raZakD7XpWL8m6TuF+T/kY3pE0q8K602X9ELDcV9RqN9Z0r2SnpN0naQNCnUrSzpd0gJJj0v6YqFuM0ndkp7K0/9K2qwdx27DlxO/DSpJqwFXAj8GxgHrAv8OvJhXuTkiRjVMj+a6nwIvAWsBBwE/l7R53u7mwKnAJ3P9c8DP8vtWBP4M7AisDhwL/FrSpAoPtdfuwFRJh+TYdomIUUAXcG3Dup9rOO49ASSNBy7JcY8DuoFfFd73DWBjYAPg/cCXJe2W6x4F9svvGw9cDlww6Edpy5eI8ORp0CZSwnu6j7pDgRv7qFuVlPTfWij7JXBinv8OcF6h7i15/dF9bO8u4KNLiXUs6UuqB3gqz69XqD8MmAs8CzwMHNHk/fOBFYCfACe32Nd04P/1UTcZuKnhXDwPbJqX/wJ8oFB/AnBBk+2sCHwWeK7Ev9MewCxgAelL8xsN9RcCjwPPADcAm3f6s+Vp8Ca3+G2w3Q8slnSWpA9JGlvyfW8FFkfE/YWyO4HN8/zmeRmAiHiI/EXRuCFJa+XyOUvZ5wjgDFJLeiIp2f6kUD8f+DCwGulL4IeSti7UfxC4NiIWA7cAB0v6kqQuSSssZd9Fjce2CHgI2Dyfv3WK9bz+vAAg6WngBdJfWt8psc9FwMHAGNKXwGck7VOov5r0V8aawO3AuctwPDbEOfHboIqIBcD2QAD/DfRIujwnY4B3S3q6MD2Uy0eRWpdFzwCjS9YDIGkkKUmdFRH3LiXWv0XExRHxXEQ8C3yb1F3UW39VRDwUyfXANGCHwib2AKbmdc8BjiR9GVwPzJd0TMMuf9Rw7CeUOLZRheU+jzsixpC6uT5Hasm3FBHTI+LuiHg1Iu4Czm849tMj4tmIeJHU1fR2Sasvbbs2PDjx26CLiLkRcWhErAdsQWqxnpyrb4mIMYXpLbl8IallXbQaqZulTD2SRpC6h14iJcCWJK0i6dR8UXgBqUtjTG9rPf/Fcku+QP00qT9/fGFfuwK/KRz3uRGxC6kV/Wngm5I+WNjl5xuO/dgSx7awsNz0uAv7XwScApwtac2lHPs2+SJyj6Rncry9x7aCpBMlPZTPy7z8tvGttmnDhxO/VSq3us8kfQG0cj+woqSNC2VvZ0l3zZy8DICkjYCV8/uQJOA00oXfj0bEyyXCOxrYBNgmIlYD3rtk81oZuBj4PrBWblFPBZTXeScwLyJ6mhzzyxFxIek6w9KOu9mxrUq6hjEnIp4CHivW8/rz0mgEsArponor55EuBK8fEauTvjB6j+1AYG9gF9JfEZN6QytxLDYMOPHboJK0qaSjJa2Xl9cHDiD1gfcpt1YvIbWSV5W0HSn5/DKvci6wp6QdcmL8JnBJ7qIB+DnwD8CeEfF8yXBHk/r1n863hh5fqFuJ9MXSA7ySbw/9QKH+tW6efJyHStpD0mhJI/L6mwMzSsRxKbCFpI9KehNwHHBXoavqbODrksZK2hT4FOnLFEm7Stoqt9JXA35AulA9t8SxPxkRL0h6FynZF+teBP5G+hIpc83AhpNOX132tHxNpJbmr0l3oizKr6eSuicOBRaTui+K0zvze8cBl+X3/Qk4sGHbB+byRcD/AONy+QakawovNGz3oKXEug7pbpuFpL8cjsjbWTHXfxb4K/A06QvoAuBbua4b6Cpsa1/g96SkuwC4Gzi0UD+9SXwzC/W7APeSvoimA5MKdSsDp+ft/hX4YqHuY/l9C0lfUlOBLUv8O+0H/JHUZXQl6aL2ObluVD6/z+Z1Ds7n5e87/fnyNDiT8j+0mZWUL1TfAawT/g9kw5C7esyW3eqkVreTvg1LTvy2XMvDKDQOEbFQ0tX93WZE3B8R5w9mnFWQNKePYz+o07FZZ7mrx8ysZlbsdABljB8/PiZNmtTpMMzMhpWZM2c+ERETGsuHReKfNGkS3d3dnQ7DzGxYkfTHZuXu4zczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrmWHxy92BmHTMVZ0OwYaoeSfu0ekQzDrCLX4zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaqSzxSzpd0nxJswtl35N0r6S7JF0qaUxV+zczs+aqbPGfCezWUHYNsEVEbAncD3y1wv2bmVkTlSX+iLgBeLKhbFpEvJIXbwHWq2r/ZmbWXCf7+P8ZuLqD+zczq6WOJH5J/wa8ApzbYp3Jkroldff09LQvODOz5VzbE7+kQ4APAwdFRPS1XkRMiYiuiOiaMGFC+wI0M1vOtfUJXJJ2A74C7BgRz7Vz32ZmllR5O+f5wM3AJpIekXQ48BNgNHCNpDsknVLV/s3MrLnKWvwRcUCT4tOq2p+ZmZXjX+6amdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVTGWJX9LpkuZLml0oGyfpGkkP5NexVe3fzMyaq7LFfyawW0PZMcC1EbExcG1eNjOzNqos8UfEDcCTDcV7A2fl+bOAfarav5mZNdfuPv61IuIxgPy6Zpv3b2ZWe0P24q6kyZK6JXX39PR0Ohwzs+VGuxP/XyWtDZBf5/e1YkRMiYiuiOiaMGFC2wI0M1vetTvxXw4ckucPAf6nzfs3M6u9Km/nPB+4GdhE0iOSDgdOBHaV9ACwa142M7M2WrGqDUfEAX1U7VzVPs3MbOmG7MVdMzOrhhO/mVnNOPGbmdWME7+ZWc0sNfFL+k9Jq0kaKelaSU9I+kQ7gjMzs8FXpsX/gYhYAHwYeAR4K/ClSqMyM7PKlEn8I/Pr7sD5EdE48JqZmQ0jZe7jv0LSvcDzwL9ImgC8UG1YZmZWlaW2+CPiGGBboCsiXgaeA/aqOjAzM6tGmYu7p0fEUxGxuLcIOK/asMzMrCpl+vj/IunnAPlRidOAcyqNyszMKlOmq+dYYIGkU0hJ/6SIOKPyyMzMrBJ9XtyVtG9h8Vbg2PwakvaNiEuqDs7MzAZfq7t69mxYnkW6tXNPIAAnfjOzYajPxB8Rh7UzEDMza48yd/WcJWlMYXmspNOrDcvMzKpS5q6eLSPi6d6FiHgK2Kq6kMzMrEplEv+IfBsnAJLGUeGTu8zMrFplEvhJwE2SLsrLHwO+XV1IZmZWpaUm/og4W1I3sFMu2jci7qk2LDMzq0rZB7GMJA3V0DtvZmbDVJm7eo4CzgXGA2sC50g6surAzMysGmX6+A8HtomIRQCSvgvcDPy4ysDMzKwaZbp6BCwuLC9mSbdPv0j6V0lzJM2WdL6kNw1ke2ZmVl6ZFv8ZwAxJl+blfYDT+rtDSesCnwc2i4jnJf0a2B84s7/bNDOz8src1fMDSdOB7Ukt/cMiYtYg7PfNkl4GVgEeHeD2zMyspFajc44rLM7L02t1/X32bkT8RdL3gT+RHuc4LSKmNdn/ZGAywMSJE/uzKzMza6JVi38maRTOZv35AWzUnx3mXwHvDWwIPA1cKOkTEfG6h7tExBRgCkBXV1f0Z19mZvZGrUbn3LCife4C/CEiegAkXQK8Bz/Vy8ysLUqNuZMfyrI9qaX/fxFx2QD2+Sfg3ZJWIXX17Ax0D2B7Zma2DMr8gOtnwKeBu4HZwKcl/bS/O4yIGcBFwO15myPIXTpmZla9Mi3+HYEtIiIgjc9PStj9FhHHA8cPZBtmZtY/ZX7AdR9QvK1mfeCuasIxM7Oqtbqd8wpSn/7qwFxJt+blbYCb2hOemZkNtlZdPd9vWxRmZtY2rW7nvL6dgZiZWXuUHY/fzMyWE078ZmY102fil3Rtfv1u+8IxM7Oqtbq4u7akHYG9JF1Aw5g9EXF7pZGZmVklWiX+44BjgPWAHzTUBUsevm5mZsNIq7t6LgIuknRsRJzQxpjMzKxCZR7EcoKkvYD35qLpEXFltWGZmVlVygzS9h/AUcA9eToql5mZ2TBUZpC2PYB/jIhX4bVB2mYBX60yMDMzq0bZ+/jHFOZXryIQMzNrjzIt/v8AZkm6jnRL53txa9/MbNgqc3H3fEnTgXeSEv9XIuLxqgMzM7NqlHr0YkQ8BlxecSxmZtYGHqvHzKxmnPjNzGqmZeKXNELS7HYFY2Zm1WuZ+PO9+3dKmthqPTMzGz7KXNxdG5iTn7m7qLcwIvaqLCozM6tMmcT/75VHYWZmbbPUi7v52bvzgJF5/jZgQGPxSxoj6SJJ90qaK2nbgWzPzMzKKzNI26eAi4BTc9G6wGUD3O9/Ab+JiE2BtwNzB7g9MzMrqcztnJ8FtgMWAETEA8Ca/d2hpNVIwz6clrf3UkQ83d/tmZnZsimT+F+MiJd6FyStSHoCV39tBPQAZ0iaJekXklZtXEnSZEndkrp7enoGsDszMysqk/ivl/Q14M2SdgUuBK4YwD5XBLYGfh4RW5HuFDqmcaWImBIRXRHRNWHChAHszszMisok/mNILfS7gSOAqcDXB7DPR4BHImJGXr6I9EVgZmZtUGZ0zlfzw1dmkLp47ouIfnf1RMTjkv4saZOIuA/YmfRkLzMza4OlJn5JewCnAA+RhmXeUNIREXH1APZ7JHCupJWAh4HDBrAtMzNbBmV+wHUS8P6IeBBA0luAq4B+J/6IuAPo6u/7zcys/8r08c/vTfrZw8D8iuIxM7OK9dnil7Rvnp0jaSrwa1If/8dIv941M7NhqFVXz56F+b8CO+b5HmBsZRGZmVml+kz8EeELrmZmy6Eyd/VsSLoLZ1JxfQ/LbGY2PJW5q+cy0rg6VwCvVhuOmZlVrUzifyEiflR5JGZm1hZlEv9/SToemAa82FsYEQMak9/MzDqjTOJ/G/BJYCeWdPVEXjYzs2GmTOL/CLBRcWhmMzMbvsr8cvdOYEzVgZiZWXuUafGvBdwr6TZe38fv2znNzIahMon/+MqjMDOztikzHv/17QjEzMzao8wvd59lyTN2VwJGAosiYrUqAzMzs2qUafGPLi5L2gd4V2URmZlZpcrc1fM6EXEZvoffzGzYKtPVs29hcQTpyVn9fuaumZl1Vpm7eorj8r8CzAP2riQaMzOrXJk+fo/Lb2a2HGn16MXjWrwvIuKECuIxM7OKtWrxL2pStipwOLAG4MRvZjYMtXr04km985JGA0cBhwEXACf19T4zMxvaWt7OKWmcpG8Bd5G+JLaOiK9ExPyB7ljSCpJmSbpyoNsyM7PyWvXxfw/YF5gCvC0iFg7yvo8C5gL+BbCZWRu1avEfDawDfB14VNKCPD0racFAdippPWAP4BcD2Y6ZmS27Vn38y/yr3mVwMvBlYHRfK0iaDEwGmDhxYoWhmJnVS5XJvSlJHwbmR8TMVutFxJSI6IqIrgkTJrQpOjOz5V/bEz+wHbCXpHmkO4R2knROB+IwM6ultif+iPhqRKwXEZOA/YHfRcQn2h2HmVlddaLFb2ZmHVRmkLbKRMR0YHonYzAzqxu3+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGqm7Ylf0vqSrpM0V9IcSUe1OwYzszpbsQP7fAU4OiJulzQamCnpmoi4pwOxmJnVTttb/BHxWETcnuefBeYC67Y7DjOzuupoH7+kScBWwIwmdZMldUvq7unpaXdoZmbLrY4lfkmjgIuBL0TEgsb6iJgSEV0R0TVhwoT2B2hmtpzqSOKXNJKU9M+NiEs6EYOZWV114q4eAacBcyPiB+3ev5lZ3XWixb8d8ElgJ0l35Gn3DsRhZlZLbb+dMyJuBNTu/ZqZWeJf7pqZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNdOIJXGZWMOmYqzodgg1h807cY9C36Ra/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVTEcSv6TdJN0n6UFJx3QiBjOzump74pe0AvBT4EPAZsABkjZrdxxmZnXViRb/u4AHI+LhiHgJuADYuwNxmJnVUifG418X+HNh+RFgm8aVJE0GJufFhZLua0NsAzEeeKLTQZTgODN9d1A2M1zOJwyfWB1nwQA/pxs0K+xE4leTsnhDQcQUYEr14QwOSd0R0dXpOJbGcQ6u4RInDJ9YHWf1OtHV8wiwfmF5PeDRDsRhZlZLnUj8twEbS9pQ0krA/sDlHYjDzKyW2t7VExGvSPoc8FtgBeD0iJjT7jgqMFy6pRzn4BouccLwidVxVkwRb+heNzOz5Zh/uWtmVjNO/GZmNePEvxRLG15C0g8l3ZGn+yU9XahbXKir9AK2pNMlzZc0u496SfpRPo67JG1dqDtE0gN5OqTDcR6U47tL0k2S3l6omyfp7nw+uzsc5/skPVP49z2uUNfWIUlKxPqlQpyz8+dyXK5r5zldX9J1kuZKmiPpqCbrdPxzWjLOIfE57beI8NTHRLr4/BCwEbAScCewWYv1jyRdrO5dXtjGWN8LbA3M7qN+d+Bq0u8o3g3MyOXjgIfz69g8P7aDcb6nd/+kYT1mFOrmAeOHyPl8H3DlQD8z7Yi1Yd09gd916JyuDWyd50cD9zeem6HwOS0Z55D4nPZ3cou/tWUdXuIA4Py2RNYgIm4Anmyxyt7A2ZHcAoyRtDbwQeCaiHgyIp4CrgF261ScEXFTjgPgFtLvPNquxPnsS9uHJFnGWDv5GX0sIm7P888Cc0m/5C/q+Oe0TJxD5XPaX078rTUbXqLxgwqApA2ADYHfFYrfJKlb0i2S9qkuzFL6OpbSx9gBh5Naf70CmCZpZh7So9O2lXSnpKslbZ7Lhuz5lLQKKVleXCjuyDmVNAnYCpjRUDWkPqct4iwa6p/TN+jEkA3DSanhJbL9gYsiYnGhbGJEPCppI+B3ku6OiIcGPcpy+jqWZTnGtpH0ftJ/qO0Lxdvl87kmcI2ke3NrtxNuBzaIiIWSdgcuAzZmiJ7PbE/g9xFR/Oug7edU0ijSl88XImJBY3WTt3Tkc7qUOHvXGeqf06bc4m9tWYaX2J+GP6Ej4tH8+jAwndRy6JS+jmXIDaEhaUvgF8DeEfG33vLC+ZwPXErqVumIiFgQEQvz/FRgpKTxDMHzWdDqM9qWcyppJCmZnhsRlzRZZUh8TkvEOSw+p33q9EWGoTyR/iJ6mNSF03uhbvMm621CuqCjQtlYYOU8Px54gOov8k2i74uRe/D6i2a35vJxwB9yvGPz/LgOxjkReBB4T0P5qsDowvxNwG4djPPvev+9Sf+x/5TPbanPTDtjzfWrk64DrNqpc5rPz9nAyS3W6fjntGScQ+Zz2p/JXT0tRB/DS0j6JtAdEb23aB4AXBD5Xzv7B+BUSa+S/rI6MSLuqSpWSeeT7jQZL+kR4HhgZD6OU4CppDsmHgSeAw7LdU9KOoE0hhLAN+P1XQHtjvM4YA3gZ5IAXok0AuJawKW5bEXgvIj4TQfj3A/4jKRXgOeB/fO/f9uHJCkRK8BHgGkRsajw1raeU2A74JPA3ZLuyGVfIyXRofQ5LRPnkPic9peHbDAzqxn38ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME79VQtJHJIWkTQtl75N05TJs41BJ6wxSPGMk/ctgbGuoanV+84iR4/uxzTMl7Tfw6GwoceK3qhwA3Ej6tWh/HQo0TfySVljGbY0BlqvEL8m/w7F+ceK3QZfHONmONIZJY+JfTdKlku6RdIqkEZJWyC3L2Xkc83/Nrcwu4Nw8rvmbc6v1OEk3Ah+T9ClJt+WB0i7Og5Ahaa28jzvz9B7gROAteVvfy/v9WR5v/UpJU3tbtpLeIen6PMjWb/PokEiaLum7km5VevbCDrl8hbzN25TGZz8il68t6QYtGQd/hybnal5hm7dK+vtcPiEf02152i6Xf0PSFEnTSL8ubfSG89tkn1/M8cyW9IVC+cE5/jsl/bLJ+07I/07OG8Ndp3867Gn5m4BPAKfl+ZtYMrb5+4AXSGPVr0AaWnc/4B2kIXd73z8mv04Hugrl84AvF5bXKMx/Czgyz/+KNLAWeT+r0zCkQd7vVFLj5++Ap3LZyBzzhLzex8nPWMjxnJTndwf+N89PBr6e51cGuklDNhwN/FshjtFNztW8wjoHk8f4B84Dts/zE4G5ef4bwEzgzU221fT8FvYzPp/ru0nDCYwC5pDGkNocuI88jjx5OATgzHxe/hM4lcKwJJ6G7+Q/Fa0KBwAn5/kL8vLtefnWSIPW9Q41sD1wLbCRpB8DVwHTWmz7V4X5LSR9i9SNM4o0TALATqQkSqTRUp+RNLZhO9sDF0bEq8Djkq7L5ZsAW5BGVYSUQB8rvK93wK6ZpC8TgA8AWxb6wlcnjdR5G3C60oBfl0XEHTR3fuH1h3l+F2CzHAOklvzoPH95RDzfx7aand+LGo770shDN0i6BNiBNNLlRRHxBKQhEgrvOZb0oJEhOcSwLTsnfhtUktYgJd4tJAUpcYakL+dVGscIiYh4SunRdR8EPgv8E/DPfeyiONbMmcA+EXGnpENJLd7SobYonxMR2/ZR/2J+XcyS/z8i/bXx28aVJb2XNPDYLyV9LyKadc9Ek/kRwLaNCT5/ERTPQattNVtuddx9jd9yG/AOSeOiwnGcrH3cV2eDbT/SE5Q2iIhJEbE+aSTF3vHK3yVpw9xP/HHgxny3yYiIuJjUuux9zuqzpEff9WU08FhuUR9UKL8W+Ay81v++WpNt3Qh8NPf1r8WSL437gAmSts3vH6klD1npy29JA7aNzO95q6RVlR7OMz8i/hs4rXBcjT5eeL05z08DPte7gqR/XEoMvd5wfhvqbwD2kbSKpFVJg7f9H+mc/VP+4kb5mbzZb0jXSK4q/NVhw5hb/DbYDiAliaKLgQNJ3TQ35/q3kZLQpXn+jMJFw6/m1zOBUyQ9DzRrgR9LejLSH0n91r1J6ShgiqTDSS3zz0TEzZJ+r/RA8quBrwA7A7NJz1SdATwTES/lLpsfSVqd9H/kZFJfeF9+Qer2uV2pSd4D7EP6MvmSpJeBheTupyZWljSD1BA7IJd9HvippLtyDDcAn24RQ69m5/c1EXG7pDOBW3tjj4hZAJK+DVwvaTEwi3RXVe/7LsxJ/3JJu7foarJhwKNzWm1JGhXpCVprkBLhdhHxeJtjmEe6gP1EO/dr9eYWv9XZlZLGkB6YckK7k75Zp7jFb2ZWM764a2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjP/HwO6lbLFWyYuAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'SE001_aa': [Gene(5)], 'SE002_aa/SE003_aa': [<HOG()>, <HOG()>]}\n",
      "Gene(4) <HOG(11)> not in parent dict\n",
      "Gene(6) None in parent dict\n",
      "<HOG()> <HOG(7)> in parent dict\n",
      "{'SE001_aa': [[Gene(4), Gene(6)]], 'SE002_aa/SE003_aa': [[None, None], [<HOG()>, None]]}\n",
      "{'SE001_aa': {<HOG(11)>, None}, 'SE002_aa/SE003_aa': {<HOG(7)>, None}}\n",
      "len1 (None, None)\n",
      "{'SE001_aa': [Gene(7)], 'SE002_aa/SE003_aa': [<HOG()>]}\n",
      "Gene(6) None not in parent dict\n",
      "Gene(8) None in parent dict\n",
      "<HOG()> <HOG(10)> not in parent dict\n",
      "{'SE001_aa': [[Gene(6), Gene(8)]], 'SE002_aa/SE003_aa': [[<HOG()>, None]]}\n",
      "{'SE001_aa': {None}, 'SE002_aa/SE003_aa': {<HOG(10)>, None}}\n",
      "len1 (None, None)\n",
      "{'SE001_aa': [Gene(9)], 'SE002_aa/SE003_aa': [<HOG()>]}\n",
      "Gene(8) None not in parent dict\n",
      "Gene(10) <HOG(7)> in parent dict\n",
      "<HOG()> <HOG(7)> not in parent dict\n",
      "{'SE001_aa': [[Gene(8), Gene(10)]], 'SE002_aa/SE003_aa': [[<HOG()>, None]]}\n",
      "{'SE001_aa': {<HOG(7)>, None}, 'SE002_aa/SE003_aa': {<HOG(7)>, None}}\n",
      "fin (<HOG(7)>, None)\n",
      "{'SE001_aa': [Gene(12)], 'SE002_aa/SE003_aa': [<HOG()>]}\n",
      "Gene(11) None not in parent dict\n",
      "Gene(13) <HOG(7)> in parent dict\n",
      "{'SE001_aa': [[Gene(11), Gene(13)]], 'SE002_aa/SE003_aa': [[None, None]]}\n",
      "{'SE001_aa': {<HOG(7)>, None}, 'SE002_aa/SE003_aa': {None}}\n",
      "len1 (None, None)\n",
      "{'SE001_aa': [Gene(10), Gene(13)], 'SE002_aa/SE003_aa': [<HOG()>, <HOG()>]}\n",
      "Gene(9) <HOG(5)> not in parent dict\n",
      "Gene(11) None in parent dict\n",
      "Gene(12) <HOG(6)> in parent dict\n",
      "Gene(14) None in parent dict\n",
      "<HOG()> <HOG(8)> not in parent dict\n",
      "{'SE001_aa': [[Gene(9), Gene(11)], [Gene(12), Gene(14)]], 'SE002_aa/SE003_aa': [[<HOG()>, None], [None, None]]}\n",
      "{'SE001_aa': {<HOG(5)>, <HOG(6)>, None}, 'SE002_aa/SE003_aa': {<HOG(8)>, None}}\n",
      "len1 (None, None)\n",
      "{'SE001_aa': [Gene(15)], 'SE002_aa/SE003_aa': [<HOG()>]}\n",
      "Gene(14) None not in parent dict\n",
      "Gene(16) None in parent dict\n",
      "{'SE001_aa': [[Gene(14), Gene(16)]], 'SE002_aa/SE003_aa': [[None, None]]}\n",
      "{'SE001_aa': {None}, 'SE002_aa/SE003_aa': {None}}\n",
      "len1 (None, None)\n",
      "{'SE001_aa': [Gene(18)], 'SE002_aa/SE003_aa': [<HOG()>]}\n",
      "Gene(17) <HOG(12)> not in parent dict\n",
      "Gene(19) None in parent dict\n",
      "{'SE001_aa': [[Gene(17), Gene(19)]], 'SE002_aa/SE003_aa': [[None, None]]}\n",
      "{'SE001_aa': {<HOG(12)>, None}, 'SE002_aa/SE003_aa': {None}}\n",
      "len1 (None, None)\n",
      "{'SE001_aa': [Gene(22)], 'SE002_aa/SE003_aa': [<HOG()>]}\n",
      "Gene(21) None not in parent dict\n",
      "Gene(23) None in parent dict\n",
      "<HOG()> <HOG(4)> not in parent dict\n",
      "{'SE001_aa': [[Gene(21), Gene(23)]], 'SE002_aa/SE003_aa': [[<HOG()>, None]]}\n",
      "{'SE001_aa': {None}, 'SE002_aa/SE003_aa': {<HOG(4)>, None}}\n",
      "len1 (None, None)\n",
      "{'SE001_aa': [Gene(4)], 'SE002_aa/SE003_aa': [<HOG()>]}\n",
      "Gene(3) None not in parent dict\n",
      "Gene(5) <HOG(3)> in parent dict\n",
      "{'SE001_aa': [[Gene(3), Gene(5)]], 'SE002_aa/SE003_aa': [[None, None]]}\n",
      "{'SE001_aa': {<HOG(3)>, None}, 'SE002_aa/SE003_aa': {None}}\n",
      "len1 (None, None)\n",
      "{'SE001_aa': [Gene(17)], 'SE002_aa/SE003_aa': [<HOG()>]}\n",
      "Gene(16) None not in parent dict\n",
      "Gene(18) <HOG(9)> in parent dict\n",
      "{'SE001_aa': [[Gene(16), Gene(18)]], 'SE002_aa/SE003_aa': [[None, None]]}\n",
      "{'SE001_aa': {<HOG(9)>, None}, 'SE002_aa/SE003_aa': {None}}\n",
      "len1 (None, None)\n",
      "{'SE001_aa': [Gene(25)], 'SE002_aa/SE003_aa': [<HOG()>]}\n",
      "Gene(24) None not in parent dict\n",
      "{'SE001_aa': [[Gene(24), None]], 'SE002_aa/SE003_aa': [[None, None]]}\n",
      "{'SE001_aa': {None}, 'SE002_aa/SE003_aa': {None}}\n",
      "len1 (None, None)\n",
      "neighbors assigned\n",
      "[[<HOG(3)>], [<HOG(4)>], [<HOG(5)>], [<HOG(6)>], [<HOG(7)>], [<HOG(8)>], [<HOG(9)>], [<HOG(10)>], [<HOG(11)>], [<HOG(12)>], [<HOG(13)>]]\n",
      "next node\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAebElEQVR4nO3deZgcVbnH8e8vCYuQhCQkcNkDyCIgm4OorAIqm4KAC6gYLtcgKqCiiF5ZruhVVFxQEKLsq5dVRFQQWUTWhDUxgIIRETCDQEIioIT3/nHOQKUzS81S3TNTv8/z1NPVdarrvFWdvH3mVNUpRQRmZlYfI1odgJmZNZcTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxD1OStpV0q6R5kp6R9HtJW0maImmRpAUN06r5cxMkXSFpoaS/SDqgYbsH5OULJV0paUKh7FOSpkt6SdLZTdzXL0n638L8n/M+PS7pp4X1bpT0YsN+/7xQvrOkByX9U9INktYqlC0j6UxJ8yU9JemzhbK3SLouH+d2SZdIWqVJ+/6wpPUlrS7pMklP5+/8AUlT8jqTJUUn3/kHcrkknSjpH3n6piQV6thc0ox8XGZI2rxQ9mlJj+bj8oSk70oa1Yx9t75z4h+GJI0FrgZ+AEwAVgP+B3gpr3JbRIxumJ7IZacA/wJWBj4E/EjSxnm7GwOnAx/J5f8ETi1U/QTwVeDMKvevE7sD10j6aI5tl4gYDbQB1zes+6mG/X43gKSJwOXAMaRjNh34aeFzxwPrAWsBbweOkrRrLhsPTAMm5/LngbMGeicbSVoXGBERDwPnAX/N9a8IHAj8veEj4xr2vWP/pgJ7A5sBmwJ7AofkOpYGfgacT9rPc4Cf5eUAPwe2jIixwCZ5G4dXsb82gCLC0zCbSAnvuS7KpgC3dFG2PCnpr19Ydh7wjTz/v8CFhbJ18/pjGrbzVeDskrGOJ/1ItQPP5vnVC+UHAbNJyfRR4JBOPj8XGAn8EPheN3XdCPxXF2VTgVsbjsULwIb5/d+AdxbKTwAu7mJbWwLPl9j3PYB7gPmkpH18Q/klwFPAPOBmYOOG8sOBk/P8AmDzLuqZDAQwqovyW4GphfcHA7fn+XfmfVeh/DFg1062syLwG+DUEvv+/bzP84EZwHaFsjcDtwHPAU/m73XpVv6fGm6TW/zD08PAIknnSNpN0viSn1sfWBSpBdnhPmDjPL9xfg9ARDxC/qHoR6wjSK3jtYA1Scn2h4XyuaQW6FjSj8B3JW1ZKH8XcH1ELAJuBw6U9HlJbZJG9iKOxn1bCDwCbJyP36rFchY/Lo22B2aVqHMhqWU+jvQjcKikvQvlvyT9lbEScDdwQcPndwd+kedvB06R9EFJa5aou2ixfWfJ7/z+yBk5u79Q3tH9Nx94mtTiP71EnXcBm5P+uroQuETSsrlsEfAZYCLwVmBn4BO93CfrhhP/MBQR84FtSa28HwPtkq6StHJe5S2SnitMj+Tlo0mty6J5wJiS5X2J9R8RcVlE/DMinge+BuxQKP9FRDwSyU3AtcB2hU3sAVyT1z0fOIz0Y3ATMFfS0Q1Vntyw7yeU2LfRhfeNZYuRtClwLPD5Evt+Y0Q8EBGvRMT9wEUN+35mRDwfES+Rupo2k7RCrmc5YKu8nwDvA35H6qr6s6R7JW3VUOXTDfv+hi72fR4wOvfz9/idR8SFkbp61gdOY8kups72/fz83b8cEScBywAb5LIZEXF7LptD+iHZoZvNWS858Q9TETE7IqZExOqkvtdVge/l4tsjYlxhWjcvX0BqWReNJXWzlCnvNUnLSTo9nzCeT+rSGNfRWs9/sdyeT5w+R2rlTsxlI4B3AL8q7PcFEbELqRX9ceArkt5VqPLwhn0/psS+LSi873K/Jb2e1Eo/IiJ+V2Lft84nkdslzcvxduzbSEnfkPRIPi5z8scm5tedSV1TL+b9fjYijo6IjUnnX+4FriyepAUmNuz77C72fSywILfyS3/nEfFH0l86pzaWdbLvR0qanU9EPwesUNj39SVdnU+izyd1MU7sbnvWO078NRARDwJnk34AuvMwMErSeoVlm/Fat8Ws/B4ASeuQWmrFrqHeOpLU0ts6txq3f23zWga4DPg2sHJEjCO17juS2VbAnIhob9xoRPw7Ii4hdUv0tN+w5L4tTzqHMSsiniX1NW9WWL94XFC6Aug3wAkRcV6J+iB1cVwFrBERK5Bayx37dgCwF7ALKSlO7qgqvxa7eRYTEU+TjtmqpK6Uniy27yz5nW/a8AOyKV13ZY0iHbcuSdoO+ALwfmB8/l7n8dq+/Qh4EFgv/5v4UqHMBoAT/zAkacPcolo9v18D2J/UD9yl3K99OamVvLykbUjJpyORXQC8W9J2OTF+Bbg8d9EgaVTupx0JjJS0bIlL+8aQ+vWfU7o09LhC2dKkH5Z24GVJu5FONnZ4tZsn1z9F0h6SxkgakdffGLijhxgArgA2kbRv3odjSX3bD+byc4EvSxovaUPgY6QfUyStBvwWOCUiTitRV3Hfn4mIFyW9mZTsi2UvAf8AliO1eot2a9j3EyVtkr+DMcChwJ8i4h8l4jgX+Kyk1ZQu6z2yY99IJ8QXAYcrXdL6qbz8t7ne/5K0Up7fCPgiS15J1dl+v0z6XkdJOpbF/6oYQzrpuyAf60NL7IP1RqvPLnsa+Il0+eb/ka7GWJhfTyf955pC+o+8oGHaKn92AnBl/txjwAEN2z4gL19IusxvQqHseNJ5heJ0fA+xrkpKLgtIfzkcQuEKFOCTpD7j50g/QBcDX81l04G2wrb2AX5PujpoPvAAMKVQfiPwYsN+zyiU70Jqab6Q151cKFuGdJnq/BzPZwtlx+WYFzumJb6n/YC/kLpNriad1D4/l43Ox/f5vM6BuY7Xk/6CmdmwrR8Af8x1t+ftvSGXTe4svo59ILWmvwk8k6dvsvhVPFuQrrx5gXSSeYtC2Vn5eCwkdUd9C1i2h/0eCZyRj+WTwFH5s7vk8u3z97CAdN7iK3RxJZqnvk3KB9psSMknqu8FVo2a/SOWdBSpv/6oVsdiQ5PvsLOhagVSi7VWST+bQ7pxyqxP3OK3ykn6EukEXaPfRcRuzY6nmSTNIt2j0OiQiGi8Ln/YyCdwf9lZWaS7qq2FnPjNzGpmSHT1TJw4MSZPntzqMMzMhpQZM2Y8HRGTGpcPicQ/efJkpk+f3uowzMyGFEl/6Wy5r+M3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczq5khceduf0w+utOHFJmZDXpzvrFHJdt1i9/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MaqayxC/pTElzJc0sLJsg6TpJf8yv46uq38zMOldli/9sYNeGZUcD10fEesD1+b2ZmTVRZYk/Im4GnmlYvBdwTp4/B9i7qvrNzKxzze7jXzkingTIrys1uX4zs9obtCd3JU2VNF3S9Pb29laHY2Y2bDQ78f9d0ioA+XVuVytGxLSIaIuItkmTJjUtQDOz4a7Zif8q4KN5/qPAz5pcv5lZ7VV5OedFwG3ABpIel3Qw8A3gHZL+CLwjvzczsyYaVdWGI2L/Lop2rqpOMzPr2aA9uWtmZtVw4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaaUnil/QZSbMkzZR0kaRlWxGHmVkdNT3xS1oNOBxoi4hNgJHAB5sdh5lZXbWqq2cU8DpJo4DlgCdaFIeZWe00PfFHxN+AbwOPAU8C8yLi2sb1JE2VNF3S9Pb29maHaWY2bPWY+CV9U9JYSUtJul7S05I+3NcKJY0H9gLWBlYFlu9sexExLSLaIqJt0qRJfa3OzMwalGnxvzMi5gN7Ao8D6wOf70eduwB/joj2iPg3cDnwtn5sz8zMeqFM4l8qv+4OXBQRz/SzzseAt0haTpKAnYHZ/dymmZmVVCbx/1zSg0AbcL2kScCLfa0wIu4ALgXuBh7IMUzr6/bMzKx3RvW0QkQcLelEYH5ELJL0T+A9/ak0Io4DjuvPNszMrG/KnNw9MyKejYhFHYuAC6sNy8zMqlKmq+dvkn4Er16Rcy1wfqVRmZlZZXpM/BFxDDBf0mmkpH9SRJxVeWRmZlaJLvv4Je1TeHsncEx+DUn7RMTlVQdnZmYDr7uTu+9ueH8P6dLOdwNBuv7ezMyGmC4Tf0Qc1MxAzMysOcpc1XOOpHGF9+MlnVltWGZmVpUyV/VsGhHPdbyJiGeBLaoLyczMqlQm8Y/Il3ECIGkCJW78MjOzwalMAj8JuFXSpfn9+4CvVReSmZlVqcyQDedKmg7slBftExF/qDYsMzOrStkHsSxFGqqhY97MzIaoMlf1HAFcAEwEVgLOl3RY1YGZmVk1yvTxHwxsHRELAfJInbcBP6gyMDMzq0aZrh4BiwrvF/Fat4+ZmQ0xZVr8ZwF3SLoiv98bOKO6kMzMrEplrur5jqQbgW1JLf2DIuKeqgMzM7NqdDc654TC2zl5erVsAJ69a2ZmLdBdi38GaRTOzvrzA1inkojMzKxS3Y3OuXYzAzEzs+YoNeZOfijLtqSW/u8i4spKozIzs8qUuYHrVODjwAPATODjkk6pOjAzM6tGmRb/DsAmERGQxucn/QiYmdkQVOYGroeANQvv1wDuryYcMzOrWneXc/6c1Ke/AjBb0p35/dbArc0Jz8zMBlp3XT3fbloUZmbWNN1dznlTMwMxM7PmKDsev5mZDRNO/GZmNdNl4pd0fX49caArlTRO0qWSHpQ0W9JbB7oOMzPrXHcnd1eRtAPwHkkX0zBmT0Tc3Y96vw/8KiL2k7Q0sFw/tmVmZr3QXeI/FjgaWB34TkNZ8NrD13tF0lhge2AKQET8C/hXX7ZlZma9191VPZcCl0o6JiJOGMA61wHagbMkbUYaBfSIjkc7dpA0FZgKsOaaay6xETMz65seT+5GxAmS3iPp23nas591jgK2BH4UEVsAC0l/WTTWOy0i2iKibdKkSf2s0szMOpQZpO3rwBHAH/J0RF7WV48Dj0fEHfn9paQfAjMza4Iyg7TtAWweEa/Aq4O03QN8sS8VRsRTkv4qaYOIeAjYmfSDYmZmTVBqPH5gHNDxqMUVBqDew4AL8hU9jwIHDcA2zcyshDKJ/+vAPZJuIF3SuT19bO13iIh7gbb+bMPMzPqmx8QfERdJuhHYipT4vxART1UdmJmZVaNUV09EPAlcVXEsZmbWBB6rx8ysZpz4zcxqptvEL2mEpJnNCsbMzKrXbeLP1+7fJ8ljJpiZDRNlTu6uAszKz9x9dTydiHhPZVGZmVllyiT+/6k8CjMza5oy1/HfJGktYL2I+I2k5YCR1YdmZmZVKDNI28dIA6mdnhetBlxZZVBmZladMpdzfhLYBpgPEBF/BFaqMigzM6tOmcT/Un5KFgCSRpGewGVmZkNQmcR/k6QvAa+T9A7gEuDn1YZlZmZVKZP4jyY9KvEB4BDgGuDLVQZlZmbVKXNVzyv54St3kLp4HooId/WYmQ1RPSZ+SXsApwGPkIZlXlvSIRHxy6qDMzOzgVfmBq6TgLdHxJ8AJK0L/AJw4jczG4LK9PHP7Uj62aPA3IriMTOzinXZ4pe0T56dJeka4P9IffzvA+5qQmxmZlaB7rp63l2Y/zuwQ55vB8ZXFpGZmVWqy8QfEQc1MxAzM2uOMlf1rA0cBkwuru9hmc3MhqYyV/VcCZxBulv3lWrDMTOzqpVJ/C9GxMmVR2JmZk1RJvF/X9JxwLXASx0LI+LuyqIyM7PKlEn8bwQ+AuzEa109kd+bmdkQUybxvxdYpzg0s5mZDV1l7ty9DxhXdSBmZtYcZVr8KwMPSrqLxfv4+3U5p6SRwHTgbxGxZ3+2ZWZm5ZVJ/MdVVPcRwGxgbEXbNzOzTpQZj/+mga5U0urAHsDXgM8O9PbNzKxrPfbxS3pe0vw8vShpkaT5/az3e8BRdHNDmKSpkqZLmt7e3t7P6szMrEOPiT8ixkTE2DwtC+wL/LCvFUrakzTU84we6p0WEW0R0TZp0qS+VmdmZg3KXNWzmIi4kv5dw78N8B5Jc4CLgZ0knd+P7ZmZWS+UGaRtn8LbEUAb6QauPomILwJfzNveEfhcRHy4r9szM7PeKXNVT3Fc/peBOcBelURjZmaVK3NVT2Xj8kfEjcCNVW3fzMyW1N2jF4/t5nMRESdUEI+ZmVWsuxb/wk6WLQ8cDKwIOPGbmQ1B3T168aSOeUljSHfaHkS6Euekrj5nZmaDW7d9/JImkO6s/RBwDrBlRDzbjMDMzKwa3fXxfwvYB5gGvDEiFjQtKjMzq0x3N3AdCawKfBl4ojBsw/MDMGSDmZm1SHd9/L2+q9fMzAY/J3czs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrmaYnfklrSLpB0mxJsyQd0ewYzMzqbFQL6nwZODIi7pY0Bpgh6bqI+EMLYjEzq52mt/gj4smIuDvPPw/MBlZrdhxmZnXV0j5+SZOBLYA7OimbKmm6pOnt7e3NDs3MbNhqWeKXNBq4DPh0RMxvLI+IaRHRFhFtkyZNan6AZmbDVEsSv6SlSEn/goi4vBUxmJnVVSuu6hFwBjA7Ir7T7PrNzOquFS3+bYCPADtJujdPu7cgDjOzWmr65ZwRcQugZtdrZmaJ79w1M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrmZYkfkm7SnpI0p8kHd2KGMzM6qrpiV/SSOAUYDdgI2B/SRs1Ow4zs7pqRYv/zcCfIuLRiPgXcDGwVwviMDOrpVEtqHM14K+F948DWzeuJGkqMDW/XSDpoT7WNxF4uo+frZLj6h3H1TuOq3cGZVw6sd9xrdXZwlYkfnWyLJZYEDENmNbvyqTpEdHW3+0MNMfVO46rdxxX79QtrlZ09TwOrFF4vzrwRAviMDOrpVYk/ruA9SStLWlp4IPAVS2Iw8yslpre1RMRL0v6FPBrYCRwZkTMqrDKfncXVcRx9Y7j6h3H1Tu1iksRS3Svm5nZMOY7d83MasaJ38ysZoZN4i8zDISk90v6g6RZki4cDHFJ+q6ke/P0sKTnBklca0q6QdI9ku6XtPsgiWstSdfnmG6UtHoTYjpT0lxJM7sol6STc8z3S9qy6phKxrWhpNskvSTpc82IqWRcH8rH6X5Jt0rabBDFtleO615J0yVtOxjiKqy3laRFkvbrV4URMeQn0kniR4B1gKWB+4CNGtZZD7gHGJ/frzQY4mpY/zDSye6Wx0U6qXRont8ImDNI4roE+Gie3wk4rwlxbQ9sCczsonx34Jeke1TeAtxRdUwl41oJ2Ar4GvC5ZsRUMq63Ff4f7tas41UyttG8du5zU+DBwRBXXmck8FvgGmC//tQ3XFr8ZYaB+BhwSkQ8CxARcwdJXEX7AxcNkrgCGJvnV6A591qUiWsj4Po8f0Mn5QMuIm4Gnulmlb2AcyO5HRgnaZVWxxURcyPiLuDfVcfSUG9Pcd3a8f8QuJ10L09TlIhtQeQsCyxPJzeXtiKu7DDgMqDfuWu4JP7OhoFYrWGd9YH1Jf1e0u2Sdh0kcQGpCwNYm/SLPhjiOh74sKTHSS2MwwZJXPcB++b59wJjJK3YhNi6U/p7tiUcTPpradCQ9F5JDwK/AP6z1fEASFqN9O/9tIHY3nBJ/GWGgRhF6u7ZkdSy/omkcYMgrg4fBC6NiEUVxtOhTFz7A2dHxOqkrozzJFX976VMXJ8DdpB0D7AD8Dfg5Yrj6klvvmfLJL2dlPi/0OpYiiLiiojYENgbOKHV8WTfA74wUPmhFWP1VKHMMBCPA7dHxL+BP+dB39Yj3Uncyrg6fBD4ZIWxFJWJ62BgV4CIuE3SsqSBrKrsIusxroh4AtgHQNJoYN+ImFdhTGV4GJJekrQp8BNgt4j4R6vj6UxE3CxpXUkTI6LVA7i1ARdLgvT/cHdJL0fElX3Z2HBp8ZcZBuJK4O0AkiaSun4eHQRxIWkDYDxwW8Xx9Caux4Cdc3xvAJYF2lsdl6SJhb88vgicWXFMZVwFHJiv7nkLMC8inmx1UIOVpDWBy4GPRMTDrY6nSNLrlbNrvjpraaDlP0wRsXZETI6IycClwCf6mvRhmLT4o4thICR9BZgeEVflsndK+gOwCPh81S2NknFB6la5uHBSqVIl4zoS+LGkz5C6LaZUHV/JuHYEvi4pgJtpwl9Jki7K9U7M5zyOA5bKMZ9GOgeyO/An4J/AQVXHVCYuSf8BTCedpH9F0qdJV0nNb2VcwLHAisCpOce+HE0aGbNEbPuSfsT/DbwAfKAZ/y9LxDWw9TUp15iZ2SAxXLp6zMysJCd+M7OaceI3M6sZJ34zs5px4jczqxknfqtEvu09JG1YWLajpKt7sY0pklYdoHjGSfrEQGxrsOru+Eqak+9f6e02z+73SJA26DjxW1X2B24h3YTVV1OAThO/pJG93NY4YFglfknD4j4caz4nfhtweSiFbUjDPjQm/rGSrlB6LsJpkkZIGplbljMlPSDpM7mV2QZckMdGf11utR4r6RbgfZI+JukuSfdJukzScrn+lXMd9+XpbcA3gHXztr6V6z1V6dkMV0u6pqNlK+lNkm6SNEPSr5VH2lQa//9ESXcqPTthu7x8ZN7mXUpjuR+Sl68i6eZc58yO9RuO1ZzCNu+U9Pq8fFLep7vytE1efrykaZKuBc7t5PAvcXw7qfOzOZ6Z+aaujuUH5vjvk3ReJ587IX9PzhtDXTPGmvZUrwn4MHBGnr8V2DLP7wi8SBpvfyRwHbAf8CbgusLnx+XXG4G2wvI5wFGF9ysW5r8KHJbnfwp8Os+PJA0rPZnCWOe53mtIjZ//AJ7Ny5bKMU/K632A/IyEHM9JeX534Dd5firw5Ty/DOlu2bVJdz//dyGOMZ0cqzmFdQ4Ers7zFwLb5vk1gdl5/nhgBvC6TrbV6fEt1DMxH+sHSEMOjwZmAVsAGwMPARPz+hPy69n5uHwTOJ1806enoT35T0Wrwv6k0QQhjam/P3B3fn9nRDwKr96mvi1pfP11JP2ANBTutd1s+6eF+U0kfZXUjTOaNNQDpAe0HAgQaTTDeZLGN2xnW+CSiHgFeErSDXn5BsAmwHV5OIGRQHHcncvz6wzSjwnAO4FNC33hK/DaAIBnSloKuDIi7u1iny4qvH43z+8CbJRjgNSSH5Pnr4qIF7rYVmfH99KG/b4iIhbmdS4HtiMNy3Fp5MHIIqI4NvwxpIelTO2iThtinPhtQCmNjb8TKSkHKXGGpKPyKo1jhEREPKv0+L13kcbeeT9dj4O+sDB/NrB3RNwnaQqpxVs61G6Wz4qIt3ZR/lJ+XcRr/39E+mvj140rS9oe2IM0rPW3IqKz7pnoZH4E8NbGBJ9/CIrHoLttdfa+u/3uavyWu4A3SZrQ8INgQ5T76myg7Ud6GtVakUYTXAP4M6mlCfBmpdE3R5C6UW7JV5uMiIjLSK3LjmfWPg+MoWtjgCdzi/pDheXXA4fCq/3vYzvZ1i3Avrmvf2Ve+9F4CJgk6a3580tJ2riHff41cGiOA0nrS1pe6eE6cyPix8AZhf1q9IHCa8cIrdcCn+pYQdLmPcTQYYnj21B+M7C3pOUkLU96uMfvSMfs/fmHG0kTCp/5FekcyS8Kf3XYEOYWvw20/UlJougy4ABSN81tufyNpCR0RZ4/S4sPtwypRX+apBeAzlrgxwB3AH8h9Vt3JKUjgGmSDia1zA+N9EyB3ys9zPqXpId/7AzMBB7O25kXEf/KXTYnS1qB9H/ke6S+8K78hNTtc7dSk7yd9BCPHYHPK430uIDc/dSJZSTdQWqI7Z+XHQ6cIun+HMPNwMe7iaFDZ8f3VRFxt6SzgTs7Yo+IewAkfQ24SdIi0vOppxQ+d0lO+ldJ2r2briYbAjw6p9WWpNERsSC3cu8EtomIp5ocwxzSCexWP+jDasQtfquzq5Uev7k0cEKzk75Zq7jFb2ZWMz65a2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjP/D15esoUMYLcGAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "counter = 1\n",
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    assign_neighbors(node.genome)\n",
    "    print(\"neighbors assigned\")\n",
    "    get_synteny_blocks(node.genome)\n",
    "    print(\"next node\")\n",
    "    assign_block_to_genes(node.genome)\n",
    "    \n",
    "    #print(node)\n",
    "    # here we go through all nodes, and if it is an ancestral, note, we make some plots\n",
    "    # in my 3 species example, there are 2 ancestral nodes, so two plots\n",
    "    \n",
    "    if isinstance(node.genome, pyham.AncestralGenome):\n",
    "        # links together the primary reconstruction's synteny blocks, using windows of 20\n",
    "        joining_blocks(20, node.genome)\n",
    "                \n",
    "        ### Plot ###\n",
    "        \n",
    "        title = node.genome.name  \n",
    "        raw_values = []\n",
    "        \n",
    "        for block in node.genome.new_synteny:\n",
    "            count = 0\n",
    "            \n",
    "            for element in block:\n",
    "                # goes through the elements of that block, the HOGs\n",
    " #               print(block, element)\n",
    "         \n",
    "                # if that element (HOG) is a pyham.HOG, increase count, it's the number of HOGs we've alignd the synteny of in this block\n",
    "                if isinstance(element, pyham.HOG):\n",
    "                    count += 1\n",
    "                    \n",
    " #               print(count)\n",
    "                    \n",
    "            raw_values.append(count)\n",
    "            \n",
    " #       print(raw_values)\n",
    "        \n",
    "        label = list(set(raw_values))\n",
    "        values = [0 for i in range(len(label))]\n",
    "        \n",
    "        for i in range(len(label)):\n",
    "            \n",
    "            for element in raw_values:\n",
    "                \n",
    "                if label[i] == element:\n",
    "                    values[i] += 1\n",
    "\n",
    "        plt.bar(label, values)\n",
    "        plt.title(title)\n",
    "        plt.xlabel('Abstractgenes per block')\n",
    "        plt.ylabel('Number of blocks')\n",
    "        plt.savefig('results/secondary_reconstruction_plot{}.png'.format(counter))\n",
    "        counter += 1\n",
    "        plt.show()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAcsElEQVR4nO3deZgdVZ3/8fcnEFBIIAkJDFsIOAgDiAO2IgKiLIogi4gji7IMP4OOIo48KjoCjqiDoyjjChlZZVPWYQkaBgkMAoEOYUkIu1ERMI0sIWEP398f5zQpLrdvKt1d93anPq/nqedWnVO36luVm+89faruKUUEZmZWHyM6HYCZmbWXE7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048dugk7S9pJskPSPpSUm/l/ROSYdKWixpYcO0Tn7fOEmXSlok6Y+SDmzY7oG5fJGkyySNy+UrSzot1z0raZakD7XpWL8m6TuF+T/kY3pE0q8K602X9ELDcV9RqN9Z0r2SnpN0naQNCnUrSzpd0gJJj0v6YqFuM0ndkp7K0/9K2qwdx27DlxO/DSpJqwFXAj8GxgHrAv8OvJhXuTkiRjVMj+a6nwIvAWsBBwE/l7R53u7mwKnAJ3P9c8DP8vtWBP4M7AisDhwL/FrSpAoPtdfuwFRJh+TYdomIUUAXcG3Dup9rOO49ASSNBy7JcY8DuoFfFd73DWBjYAPg/cCXJe2W6x4F9svvGw9cDlww6Edpy5eI8ORp0CZSwnu6j7pDgRv7qFuVlPTfWij7JXBinv8OcF6h7i15/dF9bO8u4KNLiXUs6UuqB3gqz69XqD8MmAs8CzwMHNHk/fOBFYCfACe32Nd04P/1UTcZuKnhXDwPbJqX/wJ8oFB/AnBBk+2sCHwWeK7Ev9MewCxgAelL8xsN9RcCjwPPADcAm3f6s+Vp8Ca3+G2w3Q8slnSWpA9JGlvyfW8FFkfE/YWyO4HN8/zmeRmAiHiI/EXRuCFJa+XyOUvZ5wjgDFJLeiIp2f6kUD8f+DCwGulL4IeSti7UfxC4NiIWA7cAB0v6kqQuSSssZd9Fjce2CHgI2Dyfv3WK9bz+vAAg6WngBdJfWt8psc9FwMHAGNKXwGck7VOov5r0V8aawO3AuctwPDbEOfHboIqIBcD2QAD/DfRIujwnY4B3S3q6MD2Uy0eRWpdFzwCjS9YDIGkkKUmdFRH3LiXWv0XExRHxXEQ8C3yb1F3UW39VRDwUyfXANGCHwib2AKbmdc8BjiR9GVwPzJd0TMMuf9Rw7CeUOLZRheU+jzsixpC6uT5Hasm3FBHTI+LuiHg1Iu4Czm849tMj4tmIeJHU1fR2Sasvbbs2PDjx26CLiLkRcWhErAdsQWqxnpyrb4mIMYXpLbl8IallXbQaqZulTD2SRpC6h14iJcCWJK0i6dR8UXgBqUtjTG9rPf/Fcku+QP00qT9/fGFfuwK/KRz3uRGxC6kV/Wngm5I+WNjl5xuO/dgSx7awsNz0uAv7XwScApwtac2lHPs2+SJyj6Rncry9x7aCpBMlPZTPy7z8tvGttmnDhxO/VSq3us8kfQG0cj+woqSNC2VvZ0l3zZy8DICkjYCV8/uQJOA00oXfj0bEyyXCOxrYBNgmIlYD3rtk81oZuBj4PrBWblFPBZTXeScwLyJ6mhzzyxFxIek6w9KOu9mxrUq6hjEnIp4CHivW8/rz0mgEsArponor55EuBK8fEauTvjB6j+1AYG9gF9JfEZN6QytxLDYMOPHboJK0qaSjJa2Xl9cHDiD1gfcpt1YvIbWSV5W0HSn5/DKvci6wp6QdcmL8JnBJ7qIB+DnwD8CeEfF8yXBHk/r1n863hh5fqFuJ9MXSA7ySbw/9QKH+tW6efJyHStpD0mhJI/L6mwMzSsRxKbCFpI9KehNwHHBXoavqbODrksZK2hT4FOnLFEm7Stoqt9JXA35AulA9t8SxPxkRL0h6FynZF+teBP5G+hIpc83AhpNOX132tHxNpJbmr0l3oizKr6eSuicOBRaTui+K0zvze8cBl+X3/Qk4sGHbB+byRcD/AONy+QakawovNGz3oKXEug7pbpuFpL8cjsjbWTHXfxb4K/A06QvoAuBbua4b6Cpsa1/g96SkuwC4Gzi0UD+9SXwzC/W7APeSvoimA5MKdSsDp+ft/hX4YqHuY/l9C0lfUlOBLUv8O+0H/JHUZXQl6aL2ObluVD6/z+Z1Ds7n5e87/fnyNDiT8j+0mZWUL1TfAawT/g9kw5C7esyW3eqkVreTvg1LTvy2XMvDKDQOEbFQ0tX93WZE3B8R5w9mnFWQNKePYz+o07FZZ7mrx8ysZlbsdABljB8/PiZNmtTpMMzMhpWZM2c+ERETGsuHReKfNGkS3d3dnQ7DzGxYkfTHZuXu4zczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrmWHxy92BmHTMVZ0OwYaoeSfu0ekQzDrCLX4zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaqSzxSzpd0nxJswtl35N0r6S7JF0qaUxV+zczs+aqbPGfCezWUHYNsEVEbAncD3y1wv2bmVkTlSX+iLgBeLKhbFpEvJIXbwHWq2r/ZmbWXCf7+P8ZuLqD+zczq6WOJH5J/wa8ApzbYp3Jkroldff09LQvODOz5VzbE7+kQ4APAwdFRPS1XkRMiYiuiOiaMGFC+wI0M1vOtfUJXJJ2A74C7BgRz7Vz32ZmllR5O+f5wM3AJpIekXQ48BNgNHCNpDsknVLV/s3MrLnKWvwRcUCT4tOq2p+ZmZXjX+6amdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVTGWJX9LpkuZLml0oGyfpGkkP5NexVe3fzMyaq7LFfyawW0PZMcC1EbExcG1eNjOzNqos8UfEDcCTDcV7A2fl+bOAfarav5mZNdfuPv61IuIxgPy6Zpv3b2ZWe0P24q6kyZK6JXX39PR0Ohwzs+VGuxP/XyWtDZBf5/e1YkRMiYiuiOiaMGFC2wI0M1vetTvxXw4ckucPAf6nzfs3M6u9Km/nPB+4GdhE0iOSDgdOBHaV9ACwa142M7M2WrGqDUfEAX1U7VzVPs3MbOmG7MVdMzOrhhO/mVnNOPGbmdWME7+ZWc0sNfFL+k9Jq0kaKelaSU9I+kQ7gjMzs8FXpsX/gYhYAHwYeAR4K/ClSqMyM7PKlEn8I/Pr7sD5EdE48JqZmQ0jZe7jv0LSvcDzwL9ImgC8UG1YZmZWlaW2+CPiGGBboCsiXgaeA/aqOjAzM6tGmYu7p0fEUxGxuLcIOK/asMzMrCpl+vj/IunnAPlRidOAcyqNyszMKlOmq+dYYIGkU0hJ/6SIOKPyyMzMrBJ9XtyVtG9h8Vbg2PwakvaNiEuqDs7MzAZfq7t69mxYnkW6tXNPIAAnfjOzYajPxB8Rh7UzEDMza48yd/WcJWlMYXmspNOrDcvMzKpS5q6eLSPi6d6FiHgK2Kq6kMzMrEplEv+IfBsnAJLGUeGTu8zMrFplEvhJwE2SLsrLHwO+XV1IZmZWpaUm/og4W1I3sFMu2jci7qk2LDMzq0rZB7GMJA3V0DtvZmbDVJm7eo4CzgXGA2sC50g6surAzMysGmX6+A8HtomIRQCSvgvcDPy4ysDMzKwaZbp6BCwuLC9mSbdPv0j6V0lzJM2WdL6kNw1ke2ZmVl6ZFv8ZwAxJl+blfYDT+rtDSesCnwc2i4jnJf0a2B84s7/bNDOz8src1fMDSdOB7Ukt/cMiYtYg7PfNkl4GVgEeHeD2zMyspFajc44rLM7L02t1/X32bkT8RdL3gT+RHuc4LSKmNdn/ZGAywMSJE/uzKzMza6JVi38maRTOZv35AWzUnx3mXwHvDWwIPA1cKOkTEfG6h7tExBRgCkBXV1f0Z19mZvZGrUbn3LCife4C/CEiegAkXQK8Bz/Vy8ysLUqNuZMfyrI9qaX/fxFx2QD2+Sfg3ZJWIXX17Ax0D2B7Zma2DMr8gOtnwKeBu4HZwKcl/bS/O4yIGcBFwO15myPIXTpmZla9Mi3+HYEtIiIgjc9PStj9FhHHA8cPZBtmZtY/ZX7AdR9QvK1mfeCuasIxM7Oqtbqd8wpSn/7qwFxJt+blbYCb2hOemZkNtlZdPd9vWxRmZtY2rW7nvL6dgZiZWXuUHY/fzMyWE078ZmY102fil3Rtfv1u+8IxM7Oqtbq4u7akHYG9JF1Aw5g9EXF7pZGZmVklWiX+44BjgPWAHzTUBUsevm5mZsNIq7t6LgIuknRsRJzQxpjMzKxCZR7EcoKkvYD35qLpEXFltWGZmVlVygzS9h/AUcA9eToql5mZ2TBUZpC2PYB/jIhX4bVB2mYBX60yMDMzq0bZ+/jHFOZXryIQMzNrjzIt/v8AZkm6jnRL53txa9/MbNgqc3H3fEnTgXeSEv9XIuLxqgMzM7NqlHr0YkQ8BlxecSxmZtYGHqvHzKxmnPjNzGqmZeKXNELS7HYFY2Zm1WuZ+PO9+3dKmthqPTMzGz7KXNxdG5iTn7m7qLcwIvaqLCozM6tMmcT/75VHYWZmbbPUi7v52bvzgJF5/jZgQGPxSxoj6SJJ90qaK2nbgWzPzMzKKzNI26eAi4BTc9G6wGUD3O9/Ab+JiE2BtwNzB7g9MzMrqcztnJ8FtgMWAETEA8Ca/d2hpNVIwz6clrf3UkQ83d/tmZnZsimT+F+MiJd6FyStSHoCV39tBPQAZ0iaJekXklZtXEnSZEndkrp7enoGsDszMysqk/ivl/Q14M2SdgUuBK4YwD5XBLYGfh4RW5HuFDqmcaWImBIRXRHRNWHChAHszszMisok/mNILfS7gSOAqcDXB7DPR4BHImJGXr6I9EVgZmZtUGZ0zlfzw1dmkLp47ouIfnf1RMTjkv4saZOIuA/YmfRkLzMza4OlJn5JewCnAA+RhmXeUNIREXH1APZ7JHCupJWAh4HDBrAtMzNbBmV+wHUS8P6IeBBA0luAq4B+J/6IuAPo6u/7zcys/8r08c/vTfrZw8D8iuIxM7OK9dnil7Rvnp0jaSrwa1If/8dIv941M7NhqFVXz56F+b8CO+b5HmBsZRGZmVml+kz8EeELrmZmy6Eyd/VsSLoLZ1JxfQ/LbGY2PJW5q+cy0rg6VwCvVhuOmZlVrUzifyEiflR5JGZm1hZlEv9/SToemAa82FsYEQMak9/MzDqjTOJ/G/BJYCeWdPVEXjYzs2GmTOL/CLBRcWhmMzMbvsr8cvdOYEzVgZiZWXuUafGvBdwr6TZe38fv2znNzIahMon/+MqjMDOztikzHv/17QjEzMzao8wvd59lyTN2VwJGAosiYrUqAzMzs2qUafGPLi5L2gd4V2URmZlZpcrc1fM6EXEZvoffzGzYKtPVs29hcQTpyVn9fuaumZl1Vpm7eorj8r8CzAP2riQaMzOrXJk+fo/Lb2a2HGn16MXjWrwvIuKECuIxM7OKtWrxL2pStipwOLAG4MRvZjYMtXr04km985JGA0cBhwEXACf19T4zMxvaWt7OKWmcpG8Bd5G+JLaOiK9ExPyB7ljSCpJmSbpyoNsyM7PyWvXxfw/YF5gCvC0iFg7yvo8C5gL+BbCZWRu1avEfDawDfB14VNKCPD0racFAdippPWAP4BcD2Y6ZmS27Vn38y/yr3mVwMvBlYHRfK0iaDEwGmDhxYoWhmJnVS5XJvSlJHwbmR8TMVutFxJSI6IqIrgkTJrQpOjOz5V/bEz+wHbCXpHmkO4R2knROB+IwM6ultif+iPhqRKwXEZOA/YHfRcQn2h2HmVlddaLFb2ZmHVRmkLbKRMR0YHonYzAzqxu3+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGqm7Ylf0vqSrpM0V9IcSUe1OwYzszpbsQP7fAU4OiJulzQamCnpmoi4pwOxmJnVTttb/BHxWETcnuefBeYC67Y7DjOzuupoH7+kScBWwIwmdZMldUvq7unpaXdoZmbLrY4lfkmjgIuBL0TEgsb6iJgSEV0R0TVhwoT2B2hmtpzqSOKXNJKU9M+NiEs6EYOZWV114q4eAacBcyPiB+3ev5lZ3XWixb8d8ElgJ0l35Gn3DsRhZlZLbb+dMyJuBNTu/ZqZWeJf7pqZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNdOIJXGZWMOmYqzodgg1h807cY9C36Ra/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxm5nVTEcSv6TdJN0n6UFJx3QiBjOzump74pe0AvBT4EPAZsABkjZrdxxmZnXViRb/u4AHI+LhiHgJuADYuwNxmJnVUifG418X+HNh+RFgm8aVJE0GJufFhZLua0NsAzEeeKLTQZTgODN9d1A2M1zOJwyfWB1nwQA/pxs0K+xE4leTsnhDQcQUYEr14QwOSd0R0dXpOJbGcQ6u4RInDJ9YHWf1OtHV8wiwfmF5PeDRDsRhZlZLnUj8twEbS9pQ0krA/sDlHYjDzKyW2t7VExGvSPoc8FtgBeD0iJjT7jgqMFy6pRzn4BouccLwidVxVkwRb+heNzOz5Zh/uWtmVjNO/GZmNePEvxRLG15C0g8l3ZGn+yU9XahbXKir9AK2pNMlzZc0u496SfpRPo67JG1dqDtE0gN5OqTDcR6U47tL0k2S3l6omyfp7nw+uzsc5/skPVP49z2uUNfWIUlKxPqlQpyz8+dyXK5r5zldX9J1kuZKmiPpqCbrdPxzWjLOIfE57beI8NTHRLr4/BCwEbAScCewWYv1jyRdrO5dXtjGWN8LbA3M7qN+d+Bq0u8o3g3MyOXjgIfz69g8P7aDcb6nd/+kYT1mFOrmAeOHyPl8H3DlQD8z7Yi1Yd09gd916JyuDWyd50cD9zeem6HwOS0Z55D4nPZ3cou/tWUdXuIA4Py2RNYgIm4Anmyxyt7A2ZHcAoyRtDbwQeCaiHgyIp4CrgF261ScEXFTjgPgFtLvPNquxPnsS9uHJFnGWDv5GX0sIm7P888Cc0m/5C/q+Oe0TJxD5XPaX078rTUbXqLxgwqApA2ADYHfFYrfJKlb0i2S9qkuzFL6OpbSx9gBh5Naf70CmCZpZh7So9O2lXSnpKslbZ7Lhuz5lLQKKVleXCjuyDmVNAnYCpjRUDWkPqct4iwa6p/TN+jEkA3DSanhJbL9gYsiYnGhbGJEPCppI+B3ku6OiIcGPcpy+jqWZTnGtpH0ftJ/qO0Lxdvl87kmcI2ke3NrtxNuBzaIiIWSdgcuAzZmiJ7PbE/g9xFR/Oug7edU0ijSl88XImJBY3WTt3Tkc7qUOHvXGeqf06bc4m9tWYaX2J+GP6Ej4tH8+jAwndRy6JS+jmXIDaEhaUvgF8DeEfG33vLC+ZwPXErqVumIiFgQEQvz/FRgpKTxDMHzWdDqM9qWcyppJCmZnhsRlzRZZUh8TkvEOSw+p33q9EWGoTyR/iJ6mNSF03uhbvMm621CuqCjQtlYYOU8Px54gOov8k2i74uRe/D6i2a35vJxwB9yvGPz/LgOxjkReBB4T0P5qsDowvxNwG4djPPvev+9Sf+x/5TPbanPTDtjzfWrk64DrNqpc5rPz9nAyS3W6fjntGScQ+Zz2p/JXT0tRB/DS0j6JtAdEb23aB4AXBD5Xzv7B+BUSa+S/rI6MSLuqSpWSeeT7jQZL+kR4HhgZD6OU4CppDsmHgSeAw7LdU9KOoE0hhLAN+P1XQHtjvM4YA3gZ5IAXok0AuJawKW5bEXgvIj4TQfj3A/4jKRXgOeB/fO/f9uHJCkRK8BHgGkRsajw1raeU2A74JPA3ZLuyGVfIyXRofQ5LRPnkPic9peHbDAzqxn38ZuZ1YwTv5lZzTjxm5nVjBO/mVnNOPGbmdWME79VQtJHJIWkTQtl75N05TJs41BJ6wxSPGMk/ctgbGuoanV+84iR4/uxzTMl7Tfw6GwoceK3qhwA3Ej6tWh/HQo0TfySVljGbY0BlqvEL8m/w7F+ceK3QZfHONmONIZJY+JfTdKlku6RdIqkEZJWyC3L2Xkc83/Nrcwu4Nw8rvmbc6v1OEk3Ah+T9ClJt+WB0i7Og5Ahaa28jzvz9B7gROAteVvfy/v9WR5v/UpJU3tbtpLeIen6PMjWb/PokEiaLum7km5VevbCDrl8hbzN25TGZz8il68t6QYtGQd/hybnal5hm7dK+vtcPiEf02152i6Xf0PSFEnTSL8ubfSG89tkn1/M8cyW9IVC+cE5/jsl/bLJ+07I/07OG8Ndp3867Gn5m4BPAKfl+ZtYMrb5+4AXSGPVr0AaWnc/4B2kIXd73z8mv04Hugrl84AvF5bXKMx/Czgyz/+KNLAWeT+r0zCkQd7vVFLj5++Ap3LZyBzzhLzex8nPWMjxnJTndwf+N89PBr6e51cGuklDNhwN/FshjtFNztW8wjoHk8f4B84Dts/zE4G5ef4bwEzgzU221fT8FvYzPp/ru0nDCYwC5pDGkNocuI88jjx5OATgzHxe/hM4lcKwJJ6G7+Q/Fa0KBwAn5/kL8vLtefnWSIPW9Q41sD1wLbCRpB8DVwHTWmz7V4X5LSR9i9SNM4o0TALATqQkSqTRUp+RNLZhO9sDF0bEq8Djkq7L5ZsAW5BGVYSUQB8rvK93wK6ZpC8TgA8AWxb6wlcnjdR5G3C60oBfl0XEHTR3fuH1h3l+F2CzHAOklvzoPH95RDzfx7aand+LGo770shDN0i6BNiBNNLlRRHxBKQhEgrvOZb0oJEhOcSwLTsnfhtUktYgJd4tJAUpcYakL+dVGscIiYh4SunRdR8EPgv8E/DPfeyiONbMmcA+EXGnpENJLd7SobYonxMR2/ZR/2J+XcyS/z8i/bXx28aVJb2XNPDYLyV9LyKadc9Ek/kRwLaNCT5/ERTPQattNVtuddx9jd9yG/AOSeOiwnGcrH3cV2eDbT/SE5Q2iIhJEbE+aSTF3vHK3yVpw9xP/HHgxny3yYiIuJjUuux9zuqzpEff9WU08FhuUR9UKL8W+Ay81v++WpNt3Qh8NPf1r8WSL437gAmSts3vH6klD1npy29JA7aNzO95q6RVlR7OMz8i/hs4rXBcjT5eeL05z08DPte7gqR/XEoMvd5wfhvqbwD2kbSKpFVJg7f9H+mc/VP+4kb5mbzZb0jXSK4q/NVhw5hb/DbYDiAliaKLgQNJ3TQ35/q3kZLQpXn+jMJFw6/m1zOBUyQ9DzRrgR9LejLSH0n91r1J6ShgiqTDSS3zz0TEzZJ+r/RA8quBrwA7A7NJz1SdATwTES/lLpsfSVqd9H/kZFJfeF9+Qer2uV2pSd4D7EP6MvmSpJeBheTupyZWljSD1BA7IJd9HvippLtyDDcAn24RQ69m5/c1EXG7pDOBW3tjj4hZAJK+DVwvaTEwi3RXVe/7LsxJ/3JJu7foarJhwKNzWm1JGhXpCVprkBLhdhHxeJtjmEe6gP1EO/dr9eYWv9XZlZLGkB6YckK7k75Zp7jFb2ZWM764a2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjP/HwO6lbLFWyYuAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAebElEQVR4nO3deZgcVbnH8e8vCYuQhCQkcNkDyCIgm4OorAIqm4KAC6gYLtcgKqCiiF5ZruhVVFxQEKLsq5dVRFQQWUTWhDUxgIIRETCDQEIioIT3/nHOQKUzS81S3TNTv8/z1NPVdarrvFWdvH3mVNUpRQRmZlYfI1odgJmZNZcTv5lZzTjxm5nVjBO/mVnNOPGbmdWME7+ZWc048ZuZ1YwTv5lZzTjxD1OStpV0q6R5kp6R9HtJW0maImmRpAUN06r5cxMkXSFpoaS/SDqgYbsH5OULJV0paUKh7FOSpkt6SdLZTdzXL0n638L8n/M+PS7pp4X1bpT0YsN+/7xQvrOkByX9U9INktYqlC0j6UxJ8yU9JemzhbK3SLouH+d2SZdIWqVJ+/6wpPUlrS7pMklP5+/8AUlT8jqTJUUn3/kHcrkknSjpH3n6piQV6thc0ox8XGZI2rxQ9mlJj+bj8oSk70oa1Yx9t75z4h+GJI0FrgZ+AEwAVgP+B3gpr3JbRIxumJ7IZacA/wJWBj4E/EjSxnm7GwOnAx/J5f8ETi1U/QTwVeDMKvevE7sD10j6aI5tl4gYDbQB1zes+6mG/X43gKSJwOXAMaRjNh34aeFzxwPrAWsBbweOkrRrLhsPTAMm5/LngbMGeicbSVoXGBERDwPnAX/N9a8IHAj8veEj4xr2vWP/pgJ7A5sBmwJ7AofkOpYGfgacT9rPc4Cf5eUAPwe2jIixwCZ5G4dXsb82gCLC0zCbSAnvuS7KpgC3dFG2PCnpr19Ydh7wjTz/v8CFhbJ18/pjGrbzVeDskrGOJ/1ItQPP5vnVC+UHAbNJyfRR4JBOPj8XGAn8EPheN3XdCPxXF2VTgVsbjsULwIb5/d+AdxbKTwAu7mJbWwLPl9j3PYB7gPmkpH18Q/klwFPAPOBmYOOG8sOBk/P8AmDzLuqZDAQwqovyW4GphfcHA7fn+XfmfVeh/DFg1062syLwG+DUEvv+/bzP84EZwHaFsjcDtwHPAU/m73XpVv6fGm6TW/zD08PAIknnSNpN0viSn1sfWBSpBdnhPmDjPL9xfg9ARDxC/qHoR6wjSK3jtYA1Scn2h4XyuaQW6FjSj8B3JW1ZKH8XcH1ELAJuBw6U9HlJbZJG9iKOxn1bCDwCbJyP36rFchY/Lo22B2aVqHMhqWU+jvQjcKikvQvlvyT9lbEScDdwQcPndwd+kedvB06R9EFJa5aou2ixfWfJ7/z+yBk5u79Q3tH9Nx94mtTiP71EnXcBm5P+uroQuETSsrlsEfAZYCLwVmBn4BO93CfrhhP/MBQR84FtSa28HwPtkq6StHJe5S2SnitMj+Tlo0mty6J5wJiS5X2J9R8RcVlE/DMinge+BuxQKP9FRDwSyU3AtcB2hU3sAVyT1z0fOIz0Y3ATMFfS0Q1Vntyw7yeU2LfRhfeNZYuRtClwLPD5Evt+Y0Q8EBGvRMT9wEUN+35mRDwfES+Rupo2k7RCrmc5YKu8nwDvA35H6qr6s6R7JW3VUOXTDfv+hi72fR4wOvfz9/idR8SFkbp61gdOY8kups72/fz83b8cEScBywAb5LIZEXF7LptD+iHZoZvNWS858Q9TETE7IqZExOqkvtdVge/l4tsjYlxhWjcvX0BqWReNJXWzlCnvNUnLSTo9nzCeT+rSGNfRWs9/sdyeT5w+R2rlTsxlI4B3AL8q7PcFEbELqRX9ceArkt5VqPLwhn0/psS+LSi873K/Jb2e1Eo/IiJ+V2Lft84nkdslzcvxduzbSEnfkPRIPi5z8scm5tedSV1TL+b9fjYijo6IjUnnX+4FriyepAUmNuz77C72fSywILfyS3/nEfFH0l86pzaWdbLvR0qanU9EPwesUNj39SVdnU+izyd1MU7sbnvWO078NRARDwJnk34AuvMwMErSeoVlm/Fat8Ws/B4ASeuQWmrFrqHeOpLU0ts6txq3f23zWga4DPg2sHJEjCO17juS2VbAnIhob9xoRPw7Ii4hdUv0tN+w5L4tTzqHMSsiniX1NW9WWL94XFC6Aug3wAkRcV6J+iB1cVwFrBERK5Bayx37dgCwF7ALKSlO7qgqvxa7eRYTEU+TjtmqpK6Uniy27yz5nW/a8AOyKV13ZY0iHbcuSdoO+ALwfmB8/l7n8dq+/Qh4EFgv/5v4UqHMBoAT/zAkacPcolo9v18D2J/UD9yl3K99OamVvLykbUjJpyORXQC8W9J2OTF+Bbg8d9EgaVTupx0JjJS0bIlL+8aQ+vWfU7o09LhC2dKkH5Z24GVJu5FONnZ4tZsn1z9F0h6SxkgakdffGLijhxgArgA2kbRv3odjSX3bD+byc4EvSxovaUPgY6QfUyStBvwWOCUiTitRV3Hfn4mIFyW9mZTsi2UvAf8AliO1eot2a9j3EyVtkr+DMcChwJ8i4h8l4jgX+Kyk1ZQu6z2yY99IJ8QXAYcrXdL6qbz8t7ne/5K0Up7fCPgiS15J1dl+v0z6XkdJOpbF/6oYQzrpuyAf60NL7IP1RqvPLnsa+Il0+eb/ka7GWJhfTyf955pC+o+8oGHaKn92AnBl/txjwAEN2z4gL19IusxvQqHseNJ5heJ0fA+xrkpKLgtIfzkcQuEKFOCTpD7j50g/QBcDX81l04G2wrb2AX5PujpoPvAAMKVQfiPwYsN+zyiU70Jqab6Q151cKFuGdJnq/BzPZwtlx+WYFzumJb6n/YC/kLpNriad1D4/l43Ox/f5vM6BuY7Xk/6CmdmwrR8Af8x1t+ftvSGXTe4svo59ILWmvwk8k6dvsvhVPFuQrrx5gXSSeYtC2Vn5eCwkdUd9C1i2h/0eCZyRj+WTwFH5s7vk8u3z97CAdN7iK3RxJZqnvk3KB9psSMknqu8FVo2a/SOWdBSpv/6oVsdiQ5PvsLOhagVSi7VWST+bQ7pxyqxP3OK3ykn6EukEXaPfRcRuzY6nmSTNIt2j0OiQiGi8Ln/YyCdwf9lZWaS7qq2FnPjNzGpmSHT1TJw4MSZPntzqMMzMhpQZM2Y8HRGTGpcPicQ/efJkpk+f3uowzMyGFEl/6Wy5r+M3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczq5khceduf0w+utOHFJmZDXpzvrFHJdt1i9/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MasaJ38ysZpz4zcxqxonfzKxmnPjNzGrGid/MrGac+M3MaqayxC/pTElzJc0sLJsg6TpJf8yv46uq38zMOldli/9sYNeGZUcD10fEesD1+b2ZmTVRZYk/Im4GnmlYvBdwTp4/B9i7qvrNzKxzze7jXzkingTIrys1uX4zs9obtCd3JU2VNF3S9Pb29laHY2Y2bDQ78f9d0ioA+XVuVytGxLSIaIuItkmTJjUtQDOz4a7Zif8q4KN5/qPAz5pcv5lZ7VV5OedFwG3ABpIel3Qw8A3gHZL+CLwjvzczsyYaVdWGI2L/Lop2rqpOMzPr2aA9uWtmZtVw4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaaUnil/QZSbMkzZR0kaRlWxGHmVkdNT3xS1oNOBxoi4hNgJHAB5sdh5lZXbWqq2cU8DpJo4DlgCdaFIeZWe00PfFHxN+AbwOPAU8C8yLi2sb1JE2VNF3S9Pb29maHaWY2bPWY+CV9U9JYSUtJul7S05I+3NcKJY0H9gLWBlYFlu9sexExLSLaIqJt0qRJfa3OzMwalGnxvzMi5gN7Ao8D6wOf70eduwB/joj2iPg3cDnwtn5sz8zMeqFM4l8qv+4OXBQRz/SzzseAt0haTpKAnYHZ/dymmZmVVCbx/1zSg0AbcL2kScCLfa0wIu4ALgXuBh7IMUzr6/bMzKx3RvW0QkQcLelEYH5ELJL0T+A9/ak0Io4DjuvPNszMrG/KnNw9MyKejYhFHYuAC6sNy8zMqlKmq+dvkn4Er16Rcy1wfqVRmZlZZXpM/BFxDDBf0mmkpH9SRJxVeWRmZlaJLvv4Je1TeHsncEx+DUn7RMTlVQdnZmYDr7uTu+9ueH8P6dLOdwNBuv7ezMyGmC4Tf0Qc1MxAzMysOcpc1XOOpHGF9+MlnVltWGZmVpUyV/VsGhHPdbyJiGeBLaoLyczMqlQm8Y/Il3ECIGkCJW78MjOzwalMAj8JuFXSpfn9+4CvVReSmZlVqcyQDedKmg7slBftExF/qDYsMzOrStkHsSxFGqqhY97MzIaoMlf1HAFcAEwEVgLOl3RY1YGZmVk1yvTxHwxsHRELAfJInbcBP6gyMDMzq0aZrh4BiwrvF/Fat4+ZmQ0xZVr8ZwF3SLoiv98bOKO6kMzMrEplrur5jqQbgW1JLf2DIuKeqgMzM7NqdDc654TC2zl5erVsAJ69a2ZmLdBdi38GaRTOzvrzA1inkojMzKxS3Y3OuXYzAzEzs+YoNeZOfijLtqSW/u8i4spKozIzs8qUuYHrVODjwAPATODjkk6pOjAzM6tGmRb/DsAmERGQxucn/QiYmdkQVOYGroeANQvv1wDuryYcMzOrWneXc/6c1Ke/AjBb0p35/dbArc0Jz8zMBlp3XT3fbloUZmbWNN1dznlTMwMxM7PmKDsev5mZDRNO/GZmNdNl4pd0fX49caArlTRO0qWSHpQ0W9JbB7oOMzPrXHcnd1eRtAPwHkkX0zBmT0Tc3Y96vw/8KiL2k7Q0sFw/tmVmZr3QXeI/FjgaWB34TkNZ8NrD13tF0lhge2AKQET8C/hXX7ZlZma9191VPZcCl0o6JiJOGMA61wHagbMkbUYaBfSIjkc7dpA0FZgKsOaaay6xETMz65seT+5GxAmS3iPp23nas591jgK2BH4UEVsAC0l/WTTWOy0i2iKibdKkSf2s0szMOpQZpO3rwBHAH/J0RF7WV48Dj0fEHfn9paQfAjMza4Iyg7TtAWweEa/Aq4O03QN8sS8VRsRTkv4qaYOIeAjYmfSDYmZmTVBqPH5gHNDxqMUVBqDew4AL8hU9jwIHDcA2zcyshDKJ/+vAPZJuIF3SuT19bO13iIh7gbb+bMPMzPqmx8QfERdJuhHYipT4vxART1UdmJmZVaNUV09EPAlcVXEsZmbWBB6rx8ysZpz4zcxqptvEL2mEpJnNCsbMzKrXbeLP1+7fJ8ljJpiZDRNlTu6uAszKz9x9dTydiHhPZVGZmVllyiT+/6k8CjMza5oy1/HfJGktYL2I+I2k5YCR1YdmZmZVKDNI28dIA6mdnhetBlxZZVBmZladMpdzfhLYBpgPEBF/BFaqMigzM6tOmcT/Un5KFgCSRpGewGVmZkNQmcR/k6QvAa+T9A7gEuDn1YZlZmZVKZP4jyY9KvEB4BDgGuDLVQZlZmbVKXNVzyv54St3kLp4HooId/WYmQ1RPSZ+SXsApwGPkIZlXlvSIRHxy6qDMzOzgVfmBq6TgLdHxJ8AJK0L/AJw4jczG4LK9PHP7Uj62aPA3IriMTOzinXZ4pe0T56dJeka4P9IffzvA+5qQmxmZlaB7rp63l2Y/zuwQ55vB8ZXFpGZmVWqy8QfEQc1MxAzM2uOMlf1rA0cBkwuru9hmc3MhqYyV/VcCZxBulv3lWrDMTOzqpVJ/C9GxMmVR2JmZk1RJvF/X9JxwLXASx0LI+LuyqIyM7PKlEn8bwQ+AuzEa109kd+bmdkQUybxvxdYpzg0s5mZDV1l7ty9DxhXdSBmZtYcZVr8KwMPSrqLxfv4+3U5p6SRwHTgbxGxZ3+2ZWZm5ZVJ/MdVVPcRwGxgbEXbNzOzTpQZj/+mga5U0urAHsDXgM8O9PbNzKxrPfbxS3pe0vw8vShpkaT5/az3e8BRdHNDmKSpkqZLmt7e3t7P6szMrEOPiT8ixkTE2DwtC+wL/LCvFUrakzTU84we6p0WEW0R0TZp0qS+VmdmZg3KXNWzmIi4kv5dw78N8B5Jc4CLgZ0knd+P7ZmZWS+UGaRtn8LbEUAb6QauPomILwJfzNveEfhcRHy4r9szM7PeKXNVT3Fc/peBOcBelURjZmaVK3NVT2Xj8kfEjcCNVW3fzMyW1N2jF4/t5nMRESdUEI+ZmVWsuxb/wk6WLQ8cDKwIOPGbmQ1B3T168aSOeUljSHfaHkS6Euekrj5nZmaDW7d9/JImkO6s/RBwDrBlRDzbjMDMzKwa3fXxfwvYB5gGvDEiFjQtKjMzq0x3N3AdCawKfBl4ojBsw/MDMGSDmZm1SHd9/L2+q9fMzAY/J3czs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrmaYnfklrSLpB0mxJsyQd0ewYzMzqbFQL6nwZODIi7pY0Bpgh6bqI+EMLYjEzq52mt/gj4smIuDvPPw/MBlZrdhxmZnXV0j5+SZOBLYA7OimbKmm6pOnt7e3NDs3MbNhqWeKXNBq4DPh0RMxvLI+IaRHRFhFtkyZNan6AZmbDVEsSv6SlSEn/goi4vBUxmJnVVSuu6hFwBjA7Ir7T7PrNzOquFS3+bYCPADtJujdPu7cgDjOzWmr65ZwRcQugZtdrZmaJ79w1M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrGSd+M7OaceI3M6sZJ34zs5px4jczqxknfjOzmnHiNzOrmZYkfkm7SnpI0p8kHd2KGMzM6qrpiV/SSOAUYDdgI2B/SRs1Ow4zs7pqRYv/zcCfIuLRiPgXcDGwVwviMDOrpVEtqHM14K+F948DWzeuJGkqMDW/XSDpoT7WNxF4uo+frZLj6h3H1TuOq3cGZVw6sd9xrdXZwlYkfnWyLJZYEDENmNbvyqTpEdHW3+0MNMfVO46rdxxX79QtrlZ09TwOrFF4vzrwRAviMDOrpVYk/ruA9SStLWlp4IPAVS2Iw8yslpre1RMRL0v6FPBrYCRwZkTMqrDKfncXVcRx9Y7j6h3H1Tu1iksRS3Svm5nZMOY7d83MasaJ38ysZoZN4i8zDISk90v6g6RZki4cDHFJ+q6ke/P0sKTnBklca0q6QdI9ku6XtPsgiWstSdfnmG6UtHoTYjpT0lxJM7sol6STc8z3S9qy6phKxrWhpNskvSTpc82IqWRcH8rH6X5Jt0rabBDFtleO615J0yVtOxjiKqy3laRFkvbrV4URMeQn0kniR4B1gKWB+4CNGtZZD7gHGJ/frzQY4mpY/zDSye6Wx0U6qXRont8ImDNI4roE+Gie3wk4rwlxbQ9sCczsonx34Jeke1TeAtxRdUwl41oJ2Ar4GvC5ZsRUMq63Ff4f7tas41UyttG8du5zU+DBwRBXXmck8FvgGmC//tQ3XFr8ZYaB+BhwSkQ8CxARcwdJXEX7AxcNkrgCGJvnV6A591qUiWsj4Po8f0Mn5QMuIm4Gnulmlb2AcyO5HRgnaZVWxxURcyPiLuDfVcfSUG9Pcd3a8f8QuJ10L09TlIhtQeQsCyxPJzeXtiKu7DDgMqDfuWu4JP7OhoFYrWGd9YH1Jf1e0u2Sdh0kcQGpCwNYm/SLPhjiOh74sKTHSS2MwwZJXPcB++b59wJjJK3YhNi6U/p7tiUcTPpradCQ9F5JDwK/AP6z1fEASFqN9O/9tIHY3nBJ/GWGgRhF6u7ZkdSy/omkcYMgrg4fBC6NiEUVxtOhTFz7A2dHxOqkrozzJFX976VMXJ8DdpB0D7AD8Dfg5Yrj6klvvmfLJL2dlPi/0OpYiiLiiojYENgbOKHV8WTfA74wUPmhFWP1VKHMMBCPA7dHxL+BP+dB39Yj3Uncyrg6fBD4ZIWxFJWJ62BgV4CIuE3SsqSBrKrsIusxroh4AtgHQNJoYN+ImFdhTGV4GJJekrQp8BNgt4j4R6vj6UxE3CxpXUkTI6LVA7i1ARdLgvT/cHdJL0fElX3Z2HBp8ZcZBuJK4O0AkiaSun4eHQRxIWkDYDxwW8Xx9Caux4Cdc3xvAJYF2lsdl6SJhb88vgicWXFMZVwFHJiv7nkLMC8inmx1UIOVpDWBy4GPRMTDrY6nSNLrlbNrvjpraaDlP0wRsXZETI6IycClwCf6mvRhmLT4o4thICR9BZgeEVflsndK+gOwCPh81S2NknFB6la5uHBSqVIl4zoS+LGkz5C6LaZUHV/JuHYEvi4pgJtpwl9Jki7K9U7M5zyOA5bKMZ9GOgeyO/An4J/AQVXHVCYuSf8BTCedpH9F0qdJV0nNb2VcwLHAisCpOce+HE0aGbNEbPuSfsT/DbwAfKAZ/y9LxDWw9TUp15iZ2SAxXLp6zMysJCd+M7OaceI3M6sZJ34zs5px4jczqxknfqtEvu09JG1YWLajpKt7sY0pklYdoHjGSfrEQGxrsOru+Eqak+9f6e02z+73SJA26DjxW1X2B24h3YTVV1OAThO/pJG93NY4YFglfknD4j4caz4nfhtweSiFbUjDPjQm/rGSrlB6LsJpkkZIGplbljMlPSDpM7mV2QZckMdGf11utR4r6RbgfZI+JukuSfdJukzScrn+lXMd9+XpbcA3gHXztr6V6z1V6dkMV0u6pqNlK+lNkm6SNEPSr5VH2lQa//9ESXcqPTthu7x8ZN7mXUpjuR+Sl68i6eZc58yO9RuO1ZzCNu+U9Pq8fFLep7vytE1efrykaZKuBc7t5PAvcXw7qfOzOZ6Z+aaujuUH5vjvk3ReJ587IX9PzhtDXTPGmvZUrwn4MHBGnr8V2DLP7wi8SBpvfyRwHbAf8CbgusLnx+XXG4G2wvI5wFGF9ysW5r8KHJbnfwp8Os+PJA0rPZnCWOe53mtIjZ//AJ7Ny5bKMU/K632A/IyEHM9JeX534Dd5firw5Ty/DOlu2bVJdz//dyGOMZ0cqzmFdQ4Ers7zFwLb5vk1gdl5/nhgBvC6TrbV6fEt1DMxH+sHSEMOjwZmAVsAGwMPARPz+hPy69n5uHwTOJ1806enoT35T0Wrwv6k0QQhjam/P3B3fn9nRDwKr96mvi1pfP11JP2ANBTutd1s+6eF+U0kfZXUjTOaNNQDpAe0HAgQaTTDeZLGN2xnW+CSiHgFeErSDXn5BsAmwHV5OIGRQHHcncvz6wzSjwnAO4FNC33hK/DaAIBnSloKuDIi7u1iny4qvH43z+8CbJRjgNSSH5Pnr4qIF7rYVmfH99KG/b4iIhbmdS4HtiMNy3Fp5MHIIqI4NvwxpIelTO2iThtinPhtQCmNjb8TKSkHKXGGpKPyKo1jhEREPKv0+L13kcbeeT9dj4O+sDB/NrB3RNwnaQqpxVs61G6Wz4qIt3ZR/lJ+XcRr/39E+mvj140rS9oe2IM0rPW3IqKz7pnoZH4E8NbGBJ9/CIrHoLttdfa+u/3uavyWu4A3SZrQ8INgQ5T76myg7Ud6GtVakUYTXAP4M6mlCfBmpdE3R5C6UW7JV5uMiIjLSK3LjmfWPg+MoWtjgCdzi/pDheXXA4fCq/3vYzvZ1i3Avrmvf2Ve+9F4CJgk6a3580tJ2riHff41cGiOA0nrS1pe6eE6cyPix8AZhf1q9IHCa8cIrdcCn+pYQdLmPcTQYYnj21B+M7C3pOUkLU96uMfvSMfs/fmHG0kTCp/5FekcyS8Kf3XYEOYWvw20/UlJougy4ABSN81tufyNpCR0RZ4/S4sPtwypRX+apBeAzlrgxwB3AH8h9Vt3JKUjgGmSDia1zA+N9EyB3ys9zPqXpId/7AzMBB7O25kXEf/KXTYnS1qB9H/ke6S+8K78hNTtc7dSk7yd9BCPHYHPK430uIDc/dSJZSTdQWqI7Z+XHQ6cIun+HMPNwMe7iaFDZ8f3VRFxt6SzgTs7Yo+IewAkfQ24SdIi0vOppxQ+d0lO+ldJ2r2briYbAjw6p9WWpNERsSC3cu8EtomIp5ocwxzSCexWP+jDasQtfquzq5Uev7k0cEKzk75Zq7jFb2ZWMz65a2ZWM078ZmY148RvZlYzTvxmZjXjxG9mVjP/D15esoUMYLcGAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plots for the primary reconstruction\n",
    "\n",
    "counter = 1\n",
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    \n",
    "    if isinstance(node.genome, pyham.AncestralGenome):\n",
    "        title = node.genome.name  \n",
    "        raw_values = [len(block) for block in node.genome.synteny]  \n",
    "        label = list(set(raw_values))\n",
    "        values = [0 for i in range(len(label))]\n",
    "        \n",
    "        for i in range(len(label)):\n",
    "            \n",
    "            for element in raw_values:\n",
    "                \n",
    "                if label[i] == element:\n",
    "                    values[i] += 1\n",
    "\n",
    "        plt.bar(list(label), values)\n",
    "        plt.title(title)\n",
    "        plt.xlabel('Abstractgenes per block')\n",
    "        plt.ylabel('Number of blocks')\n",
    "        plt.savefig('results/primary_reconstruction_plot{}.png'.format(counter))\n",
    "        counter += 1\n",
    "        plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py3]",
   "language": "python",
   "name": "conda-env-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
