{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyham\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Path to nwk file\n",
    "#nwk_path = \"./data/events/RealTree.nwk\"\n",
    "nwk_path2 = \"./data/no_events/Output/EstimatedSpeciesTree.nwk\"\n",
    "\n",
    "# The newick tree is extracted as a string\n",
    "tree_str = pyham.utils.get_newick_string(nwk_path, type=\"nwk\")\n",
    "\n",
    "# Path to OrthoXML file\n",
    "orthoxml_path = \"./data/events/hogs.orthoxml\"\n",
    "\n",
    "#ham_analysis = pyham.Ham(tree_str, orthoxml_path, use_internal_name=True)\n",
    "ham_analysis_no_name = pyham.Ham(tree_str, orthoxml_path, use_internal_name=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pyham.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def locus_to_id(file):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function extracts loci from a fasta file and store the information as keys in a dictionary \n",
    "    and the related protein ID as value.\n",
    "\n",
    "    Args:\n",
    "        file (:obj:`str`): path to the file of interest.\n",
    "\n",
    "    Returns:\n",
    "        a dictionary mapping the locus (:obj:`str`) with the protein ID (:obj:`str`).\n",
    "    \"\"\"\n",
    "    \n",
    "    locus_to_id = {}\n",
    "    for seq_record in SeqIO.parse(file, 'fasta'):\n",
    "        locus_to_id[int(seq_record.description.split('locus: ')[-1])] = seq_record.id[:-1]\n",
    "        \n",
    "    return locus_to_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_locus(directory):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function goes through each file of a folder and build a dictionary containing loci \n",
    "    as keys and the related protein ID as values. For each entry in the dictionary, the function \n",
    "    updates the corresponding :obj:`pyham.abstractgene.Gene` by adding its locus.\n",
    "\n",
    "    Args:\n",
    "        directory (:obj:`str`): directory of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "\n",
    "    path = directory\n",
    "    for file in os.listdir(path):\n",
    "        path_to_file = path + '/' + file\n",
    "        locus_dictionary = locus_to_id(path_to_file)\n",
    " \n",
    "        for entry in locus_dictionary:\n",
    "            gene = ham_analysis_no_name2.get_genes_by_external_id(locus_dictionary[entry]) \n",
    "            for element in gene:\n",
    "                element.locus = entry\n",
    "                \n",
    "        ham_analysis_no_name.get_extant_genome_by_name(file[:-6]).locus_assignment = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_locus('/Users/karimsaied/Documents/biology/master/mls_2018/master_project/data/events/DB')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_by_locus(genome, locus):\n",
    "    \n",
    "    \"\"\"\n",
    "    Get a :obj:`pyham.abstractgene.Gene` that match the query locus. \n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant genome of interest.\n",
    "        locus (:obj:`int`): locus.\n",
    "\n",
    "    Returns:\n",
    "        :obj:`pyham.abstractgene.Gene`.\n",
    "    \"\"\"\n",
    "    \n",
    "    for gene in genome.genes:\n",
    "        if gene.locus == locus:\n",
    "            return gene \n",
    "    \n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_neighbors(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function assigns neighbors to each :obj:`pyham.abstractgene.Gene` or :obj:`pyham.abstractgene.HOG`.\n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    if isinstance(genome, pyham.ExtantGenome):\n",
    "        \n",
    "        for extant_gene in genome.genes:\n",
    "            extant_gene.neighbors = (get_gene_by_locus(genome, extant_gene.locus-1), \n",
    "                                     get_gene_by_locus(genome, extant_gene.locus+1))            \n",
    "                        \n",
    "    if isinstance(genome, pyham.AncestralGenome):\n",
    "    \n",
    "        for ancestral_gene in genome.genes:  \n",
    "            \n",
    "            children = [child for child in ancestral_gene.children]\n",
    "            \n",
    "            ######################################################################################\n",
    "            # Dictionary containing genome names as keys (i.e. SE004) and a list of children as value #\n",
    "            ######################################################################################\n",
    "            \n",
    "            genome_dict = {}  \n",
    "            \n",
    "            for child in children:\n",
    "                if child.genome.name in genome_dict:\n",
    "                    genome_dict[child.genome.name].append(child)\n",
    "                else:\n",
    "                    genome_dict[child.genome.name] = [child]\n",
    "            \n",
    "            ############################################\n",
    "            # Children are replaced by their neighbors #      \n",
    "            ############################################\n",
    "            \n",
    "            for genome, gene_collection in genome_dict.items(): #stoca \n",
    "                for i in range(len(gene_collection)):\n",
    "                    gene_collection[i] = list(gene_collection[i].neighbors)\n",
    "            \n",
    "            ################################################################################\n",
    "            # Dictionary containing genomes as keys (i.e. SE004) and a list of parent HOGs #\n",
    "            ################################################################################\n",
    "            \n",
    "            parent_dict = {}\n",
    "            for genome, neighbors_collection in genome_dict.items(): #stoca \n",
    "                for neighbors_pair in neighbors_collection:\n",
    "                    for neighbor in neighbors_pair:\n",
    "                        if genome in parent_dict:\n",
    "                            if neighbor == None:\n",
    "                                parent_dict[genome].add(None)\n",
    "                            else:\n",
    "                                if neighbor.parent != ancestral_gene:\n",
    "                                    parent_dict[genome].add(neighbor.parent)\n",
    "                        else:\n",
    "                            if neighbor == None:\n",
    "                                parent_dict[genome] = {None}\n",
    "                            else:\n",
    "                                if neighbor.parent != ancestral_gene:\n",
    "                                    parent_dict[genome] = {neighbor.parent}\n",
    "            \n",
    "            l = []  # List of sets\n",
    "            for genome, parent in parent_dict.items():  \n",
    "                l.append(parent)  # parent is a set\n",
    "                ancestral_gene.list_parents = l            # List of potential neighbors in case of conflict\n",
    "                \n",
    "            final_neighbors = list(l[0].intersection(*l))  # Intersection of the sets\n",
    "            \n",
    "            if len(final_neighbors) == 2:\n",
    "                ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "            \n",
    "            elif len(final_neighbors) == 1:\n",
    "                none_counter = 0\n",
    "                potential_neighbor = None\n",
    "                \n",
    "                for collection in l:\n",
    "                    if len(collection) == 2:\n",
    "                        for element in collection:\n",
    "                            if element == final_neighbors[0]:\n",
    "                                continue\n",
    "                            elif element == None:\n",
    "                                none_counter += 1\n",
    "                            else:\n",
    "                                potential_neighbor = element\n",
    "                \n",
    "                    else:\n",
    "                        break\n",
    "                \n",
    "                if none_counter == 1:  # To deal with missing orthology\n",
    "                    final_neighbors.append(potential_neighbor)\n",
    "                \n",
    "                if len(final_neighbors) == 1:\n",
    "                    final_neighbors.append(None)\n",
    "                    ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "                else:  \n",
    "                    ancestral_gene.neighbors = tuple(final_neighbors)\n",
    "            \n",
    "            elif len(final_neighbors) == 0 or len(final_neighbors) > 2:\n",
    "                final_neighbors = [None, None]\n",
    "                ancestral_gene.neighbors = tuple(final_neighbors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_synteny_blocks(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function aims to reconstruct synteny blocks using neighbors of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "\n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        list of lists containing :obj:`pyham.abstractgene.Gene` or :obj:`pyham.abstractgene.HOG` \n",
    "        in a specific order.\n",
    "    \"\"\"\n",
    "\n",
    "    syn_blocks = []\n",
    "    seen = []\n",
    "    gene_collection = genome.genes\n",
    "    tmp = []\n",
    "        \n",
    "    for abstract_gene in gene_collection:\n",
    "        \n",
    "        if abstract_gene in seen:\n",
    "            continue\n",
    "                \n",
    "        block1 = []\n",
    "        block2 = []\n",
    "        \n",
    "        current1 = abstract_gene\n",
    "        current2 = abstract_gene  # Used for the block2\n",
    "        \n",
    "        block1.append(current1)  # [X] where X is the first abstract gene\n",
    "        seen.append(current1)\n",
    "        \n",
    "        ######################\n",
    "        # First end (block1) #\n",
    "        ######################\n",
    "        \n",
    "        next_gene = current1.neighbors[0]\n",
    "        \n",
    "        while next_gene != None:\n",
    "            \n",
    "            if next_gene in seen:\n",
    "                next_gene = None  \n",
    "                continue\n",
    "            \n",
    "            previous = current1\n",
    "            current1 = next_gene\n",
    "            \n",
    "            if previous in current1.neighbors:  \n",
    "                block1.append(current1)\n",
    "                seen.append(current1)\n",
    "            \n",
    "            else:                                  \n",
    "                break\n",
    "                \n",
    "            if previous == current1.neighbors[0]:\n",
    "                next_gene = current1.neighbors[1]\n",
    "            \n",
    "            else:\n",
    "                next_gene = current1.neighbors[0]\n",
    "\n",
    "        #######################\n",
    "        # Second end (block2) #\n",
    "        #######################\n",
    "        \n",
    "        next_gene = current2.neighbors[1]\n",
    "            \n",
    "        while next_gene != None:\n",
    "            \n",
    "            if next_gene in seen:\n",
    "                next_gene = None\n",
    "                continue\n",
    "            \n",
    "            previous = current2\n",
    "            current2 = next_gene\n",
    "            \n",
    "            if previous in current2.neighbors:\n",
    "                block2.append(current2)\n",
    "                seen.append(current2)\n",
    "            \n",
    "            else:                                  \n",
    "                break\n",
    "                \n",
    "            if previous == current2.neighbors[0]:\n",
    "                next_gene = current2.neighbors[1]\n",
    "            \n",
    "            else:\n",
    "                next_gene = current2.neighbors[0]\n",
    "        \n",
    "        syn_blocks.append(block1[::-1] + block2)\n",
    "        \n",
    "    genome.synteny = syn_blocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_block_to_genes(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    The function links each :obj:`pyham.abstractgene.AbstractGene` to the synteny block it belongs to. \n",
    "    It creates a dictionary containing the genome of :obj:`pyham.abstractgene` as key and the synteny \n",
    "    block as value.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "\n",
    "    Returns:\n",
    "        xxx\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    for block in genome.synteny:\n",
    "    \n",
    "        for abstract_gene in block:\n",
    "            genomes_to_block = {}  # Dictionary\n",
    "            genomes_to_block[genome.name] = block  \n",
    "            abstract_gene.blocks = genomes_to_block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_children(genome, children):  \n",
    "    \n",
    "    \"\"\"\n",
    "    Due to tandem duplication, some :obj:`pyham.abstractgene.HOG` can have more than two children.\n",
    "    The function refines the list of children by removing duplicates.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        children (:obj:`list` of :obj:`pyham.abstractgene.AbstractGene`): a list of direct descendants AbstractGene.\n",
    "    \n",
    "    Returns:\n",
    "        list of :obj:`pyham.abstractgene.AbstractGene` without duplicates.\n",
    "    \"\"\"\n",
    "    \n",
    "    unique_genome = []\n",
    "    unique_children = []\n",
    "    \n",
    "    if len(children) == 2:  # Two children comming from two different genomes\n",
    "        \n",
    "        if children[0].genome != children[1].genome:\n",
    "            return children\n",
    "        \n",
    "        else:  # Two children comming from the same genome (i.e. duplication)\n",
    "            unique_children.append(None)\n",
    "            return unique_children\n",
    "    \n",
    "    else:  # In case of duplications, only one child per specie is taken into account\n",
    "        \n",
    "        for child in children:  \n",
    "            \n",
    "            if child.genome in unique_genome:\n",
    "                continue\n",
    "\n",
    "            else:\n",
    "                \n",
    "                for neighbor in child.neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        continue\n",
    "\n",
    "                    elif neighbor.parent:\n",
    "                        \n",
    "                        if neighbor.parent in child.parent.blocks[genome.name]:  # To make sure that, for both unique children, we go in the same direction. \n",
    "                            continue\n",
    "                        \n",
    "                        unique_genome.append(child.genome)\n",
    "                        unique_children.append(child)\n",
    "\n",
    "                    else:  # If the child has no parent HOG\n",
    "                        unique_genome.append(child.genome)\n",
    "                        unique_children.append(child)\n",
    "        \n",
    "        if len(unique_children) == 0:\n",
    "            unique_children.append(None)\n",
    "            return unique_children  \n",
    "        \n",
    "        else:\n",
    "            return unique_children  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_unique_children_from_unit_block(genome, children, genes_seen):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function aims to be used after refining the :obj:`list` genes_seen. It uses the :obj:`list` \n",
    "    genes_seen to check whether the :obj:`list` current_children returned by the function get_unique_children \n",
    "    are correctly selected. Indeed, the function get_unique_children encounters a limitation with single HOG.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        children (:obj:`list` of :obj:`pyham.abstractgene.AbstractGene`): a list of direct descendants AbstractGene.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        \n",
    "    Returns:\n",
    "        list of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "    \"\"\"\n",
    "    \n",
    "    unique_genome = []\n",
    "    unique_children = []\n",
    "    \n",
    "    if len(children) == 2:  # Two children comming from two different genomes\n",
    "        \n",
    "        if children[0].genome != children[1].genome:\n",
    "            return children\n",
    "        \n",
    "        else:  # Two children comming from the same genome (i.e. duplication)\n",
    "            unique_children.append(None)\n",
    "            return unique_children\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        for child in children:  \n",
    "            \n",
    "            if child.genome in unique_genome:\n",
    "                continue\n",
    "            \n",
    "            for neighbor in child.neighbors:\n",
    "                \n",
    "                if neighbor == None:\n",
    "                    continue\n",
    "\n",
    "                if neighbor.parent in child.parent.blocks[genome.name] or neighbor in genes_seen:  # To make sure that for both unique children we go in the same direction \n",
    "                    continue\n",
    "                \n",
    "                unique_genome.append(child.genome)\n",
    "                unique_children.append(child)\n",
    "                \n",
    "        if len(unique_children) == 0:\n",
    "            unique_children.append(None)\n",
    "            return unique_children  # Return a list without duplicates\n",
    "        \n",
    "        else:\n",
    "            return unique_children  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Link:\n",
    "    \n",
    "    \"\"\"\n",
    "    Link is a class representing linking regions. Linking regions are unresolved regions with a weak synteny \n",
    "    signal that links two synteny blocks together. A Link is defined by its collection of poorly resolved genes.\n",
    "    \n",
    "    Attributes:\n",
    "        | linking_region: dictionary of :obj:`pyham.abstractgene.AbstractGene`\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.linking_region = {}\n",
    "        \n",
    "    def get_linking_region(self, genome, block, common_block, starting_children):\n",
    "        \n",
    "        \"\"\"\n",
    "        Create a linking region filled with :obj:`pyham.abstractgene.AbstractGene`.\n",
    "        \n",
    "        Args:\n",
    "            genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "            block (:obj:`list`): the synteny block from which the assessement starts.\n",
    "            common_block (:obj:`list`): the synteny block to be linked with the initial block.\n",
    "            starting_children (:obj:`list`): list of children coming from the last :obj:`pyham.abstractgene.AbstractGene` of the initial block.\n",
    "        \n",
    "        Returns:\n",
    "            return linking_region: dictionary of :obj:`pyham.abstractgene.AbstractGene`.\n",
    "        \"\"\"\n",
    "        \n",
    "        initial_children = starting_children\n",
    "        seen = []\n",
    "\n",
    "        for child in initial_children:  # Creation of the dictionary's entries ({species1: [], species2: []})\n",
    "\n",
    "            if child.genome.name in self.linking_region:\n",
    "                continue\n",
    "\n",
    "            else:\n",
    "                self.linking_region[child.genome.name] = [] \n",
    "\n",
    "        next_children = []\n",
    "\n",
    "        for child in initial_children:\n",
    "            current_child = child\n",
    "            growing_link = True\n",
    "\n",
    "            while growing_link:\n",
    "                \n",
    "                for neighbor in current_child.neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor != current_child.next_neighbor:\n",
    "                        continue\n",
    "\n",
    "                    if neighbor.parent:\n",
    "\n",
    "                        if neighbor.parent.blocks[genome.name] != common_block:\n",
    "                            \n",
    "                            if neighbor.parent not in self.linking_region[neighbor.genome.name]:\n",
    "                                self.linking_region[neighbor.genome.name].append(neighbor.parent)\n",
    "                                current_child = neighbor\n",
    "                                seen.append(neighbor)\n",
    "\n",
    "                            else:\n",
    "                                current_child = neighbor\n",
    "                                seen.append(neighbor)\n",
    "\n",
    "                        else:\n",
    "                            growing_link = False\n",
    "\n",
    "                    else:\n",
    "                        self.linking_region[neighbor.genome.name].append(neighbor)  # The child is added to the linking region anyway.\n",
    "                        current_child = neighbor\n",
    "                        seen.append(neighbor)\n",
    "\n",
    "    def __repr__(self):\n",
    "        return '<{}>'.format(self.__class__.__name__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def refine_genes_seen(genome, genes_seen, current_children):\n",
    "    \n",
    "    \"\"\"\n",
    "    Regarding the secondary reconstruction, when the growing process continue by starting from a single \n",
    "    :obj:`pyham.abstractgene.HOG`, both neighbors of one of its children can be within the \n",
    "    :obj:`list` genes_seen. The function will remove all children that come after the children of interest\n",
    "    within the :obj:`list` genes_seen.\n",
    "    \n",
    "    Ars:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        current_children (:obj:`list`): list of children to begin the investigation with at the next iteration.\n",
    "        \n",
    "    Returns:\n",
    "        return the refined :obj:`list` genes_seen.\n",
    "    \"\"\"\n",
    "    \n",
    "    if len(current_children) == 2:\n",
    "        tmp = genes_seen[:]  # Make an independent copy of the initial list\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == current_children[0]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == current_children[0].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:]\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == current_children[1]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == current_children[1].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:] \n",
    "\n",
    "    return genes_seen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def adjust_genes_seen(genome, genes_seen, starting_children):\n",
    "    \n",
    "    \"\"\"\n",
    "    Same principle as for the function refine_genes_seen but used when the growing process continue by\n",
    "    sarting from an element at the extremity of a synteny block. Neighbors from one of its children can\n",
    "    be present within the :obj:`list` genes_seen. The function will remove all children that come after \n",
    "    the children of interest within the :obj:`list` genes_seen.\n",
    "    \n",
    "    Ars:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        genes_seen (:obj:`list`): a list of genes already assessed.\n",
    "        starting_children (:obj:`list`): list of children to begin the investigation with at the next iteration.\n",
    "        \n",
    "    Returns:\n",
    "        return the refined :obj:`list` genes_seen.\n",
    "    \"\"\"\n",
    "    \n",
    "    if len(starting_children) == 2:\n",
    "        tmp = genes_seen[:]  # Make an independent copy of the initial list\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == starting_children[0]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == starting_children[0].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:]\n",
    "\n",
    "        for i in range(len(genes_seen)):\n",
    "            \n",
    "            if genes_seen[i] == starting_children[1]:\n",
    "                \n",
    "                for gene in genes_seen[i+1:]:\n",
    "                    \n",
    "                    if gene.genome == starting_children[1].genome:\n",
    "                        \n",
    "                        for g in tmp:\n",
    "                            \n",
    "                            if g == gene:\n",
    "                                tmp.remove(g)\n",
    "\n",
    "        genes_seen = tmp[:] \n",
    "\n",
    "    return genes_seen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def joining_blocks(window, genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    The function performs the secondary reconstruction. Synteny blocks from the primary reconstruction\n",
    "    are linked together using linking regions.\n",
    "    \n",
    "    Ars:\n",
    "        window (:obj:`int`): a window of assessment\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        \n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    for hog in genome.genes:  #### To be used with the function \"add_remaining_blocks\"\n",
    "        hog.processed = False\n",
    "    \n",
    "    new_blocks = []\n",
    "    block_record = []  # To keep track of the block already assessed. Useful to link several blocks in a row.\n",
    "    \n",
    "    for block in genome.synteny:\n",
    "            \n",
    "        if block in block_record or len(block) == 1:  \n",
    "            continue\n",
    "        \n",
    "        block2 = block  # Used to reconstruct the other side\n",
    "        \n",
    "        counter = 0\n",
    "        children = [child for child in block[-1].children]\n",
    "        current_children = get_unique_children(genome, children)  # Remove duplicates from children\n",
    "        starting_children = current_children  # To keep track of the initial children considered (used to build a linking region)\n",
    "        \n",
    "        if len(current_children) == 1 or len(current_children) > 2:  # If missing information (==1) or other special case (>2), the block is no longer taken into account\n",
    "            continue\n",
    "        \n",
    "        block_record.append(block)   \n",
    "        \n",
    "        blocks_seen = [set() for child in current_children]  # To save already assessed blocks and perform the intersection\n",
    "        genes_seen = []  # To keep track of already assessed genes\n",
    "        intersect = None\n",
    "        extended_block = []\n",
    "        next_children = [None, None]\n",
    "        genes_seen.append(current_children[0])  \n",
    "        genes_seen.append(current_children[1])  \n",
    "        \n",
    "        while counter < window:\n",
    "            none_type_as_neighbor = 0\n",
    "            block_found = False\n",
    "            reaching_already_assessed = False  \n",
    "            \n",
    "            if None in current_children:\n",
    "                break\n",
    "            \n",
    "            # Without the following block, when a common block is reached from one extremity, the next iteration \n",
    "            # starts from the opposite one. However, children originating from that extreme element are not\n",
    "            # added within the list of genes seen.\n",
    "            if current_children[0] and current_children[1] not in genes_seen:\n",
    "                genes_seen.append(current_children[0])\n",
    "                genes_seen.append(current_children[1])\n",
    "            \n",
    "            for i in range(len(current_children)):\n",
    "                \n",
    "                for neighbor in current_children[i].neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        none_type_as_neighbor += 1\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor.parent:  # Check whether the neighbor is a singleton (i.e. does it have a parent or not?)\n",
    "                        \n",
    "                        if neighbor.parent in block or neighbor in genes_seen:  # Avoid to take into account elements from the initial block or those already assessed\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            blocks_seen[i].add(tuple(neighbor.parent.blocks[genome.name]))\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region \n",
    "                            break\n",
    "                    \n",
    "                    else:\n",
    "                        \n",
    "                        if neighbor in genes_seen:\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region\n",
    "                \n",
    "            if none_type_as_neighbor > 0:  # If children have only one neighbor\n",
    "                break\n",
    "\n",
    "            counter += 1\n",
    "            current_children = next_children[:]\n",
    "            intersect = list(blocks_seen[0].intersection(*blocks_seen))  # Intersection\n",
    "            \n",
    "            if len(intersect) == 1:  # ex. intersect = [('a', 'c')]\n",
    "\n",
    "                    if list(intersect[0]) in block_record:\n",
    "                        reaching_already_assessed = True\n",
    "                        break\n",
    "                        \n",
    "                    block_found = True  \n",
    "                    counter = 0  # Reset the counter to pursue the assessement with further blocks\n",
    "                    common_block = list(intersect[0])  # Extraction of the last common block. \n",
    "                    block_record.append(common_block)\n",
    "\n",
    "                    link = Link()  # Class for linking regions\n",
    "                    link.get_linking_region(genome, block, common_block, starting_children)  # Creation of the linking region\n",
    "                    \n",
    "                    if not extended_block:  # If the list if empty (i.e. for a first block of a serie)\n",
    "                        extended_block.append(block)  # The initial block\n",
    "                        extended_block.append(link)  # The linking region\n",
    "                        extended_block.append(common_block)  # The common block\n",
    "\n",
    "                    else:  # For all the block but the first one, \"block\" corresponds to the previous common block\n",
    "                        extended_block.append(link)\n",
    "                        extended_block.append(common_block)\n",
    "\n",
    "                    block = common_block  # The new block becomes the last common block found within the while loop\n",
    "\n",
    "                    if len(block) > 1:\n",
    "                        \n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its first element (i.e. element 0), the next iteration should start from its last element (i.e. element -1) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block[0].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block[0].children)[0] and get_unique_children(genome, block[0].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block[-1].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)  \n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block[0].children)\n",
    "                            break\n",
    "\n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its last element (i.e. element -1), the next iteration should start from its first element (i.e. element 0) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block[-1].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block[-1].children)[0] and get_unique_children(genome, block[-1].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block[0].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block[-1].children)\n",
    "                            break\n",
    "                    \n",
    "                    ################################################################################################################\n",
    "                    # When we reach a common unit block, the next iteration should start from the same unique HOG (i.e. element 0) #\n",
    "                    ################################################################################################################\n",
    "\n",
    "                    elif len(block) == 1:  # If the block has only one HOG...\n",
    "                        current_children = get_unique_children(genome, block[0].children)                         \n",
    "                        \n",
    "                        if len(current_children) == 2:\n",
    "                            genes_seen = refine_genes_seen(genome, genes_seen, current_children)\n",
    "                            current_children = check_unique_children_from_unit_block(genome, block[0].children, genes_seen)\n",
    "\n",
    "                        starting_children = current_children\n",
    "                        \n",
    "                    blocks_seen = [set() for child in current_children]  # The variable is restarted for the next iteration\n",
    "\n",
    "            if reaching_already_assessed:  \n",
    "                break\n",
    "                \n",
    "            if len(current_children) == 2:\n",
    "                if current_children[0].genome != current_children[1].genome:\n",
    "                    continue\n",
    "                \n",
    "                break  # To exit the while loop\n",
    "                \n",
    "            else:  \n",
    "                break  # To exit the while loop\n",
    "        \n",
    "        tmp = [[], []]\n",
    "        if extended_block:\n",
    "            for item in extended_block:\n",
    "                if isinstance(item, Link):\n",
    "                    tmp[1].append(item)\n",
    "\n",
    "                else:\n",
    "                    for element in item:\n",
    "                        tmp[1].append(element)\n",
    "        \n",
    "        else:\n",
    "            for element in block:\n",
    "                tmp[1].append(element)  # If a new initial block reach directly an already assessed block (i.e. common block), the initial block is stored\n",
    "                \n",
    "###############################################################################################################\n",
    "\n",
    "        counter = 0\n",
    "        children = [child for child in block2[0].children]\n",
    "        current_children = get_unique_children(genome, children)  # Remove duplicates from children\n",
    "        starting_children = current_children\n",
    "        \n",
    "        if len(current_children) == 1 or len(current_children) > 2:  # If missing information (==1) or other special case (>2), the block is no longer taken into account\n",
    "            new_blocks.append(tmp[1])\n",
    "            continue\n",
    "        \n",
    "        blocks_seen = [set() for child in current_children]  # To save already assessed blocks and perform the intersection\n",
    "        genes_seen = []  # To keep track of already assessed genes (used for gene without parent to make to we go in one direction)\n",
    "        intersect = None\n",
    "        extended_block = []\n",
    "        next_children = [None, None]\n",
    "        genes_seen.append(current_children[0])  \n",
    "        genes_seen.append(current_children[1])  \n",
    "        \n",
    "        while counter < window:\n",
    "            none_type_as_neighbor = 0\n",
    "            block_found = False\n",
    "            reaching_already_assessed = False  \n",
    "            \n",
    "            if None in current_children:\n",
    "                break\n",
    "            \n",
    "            # Without the following block, when a common block is reached from one extremity, the next iteration \n",
    "            # starts from the opposite one. However, children originating from the that extreme element are not\n",
    "            # added within the list of genes seen.\n",
    "            if current_children[0] and current_children[1] not in genes_seen:\n",
    "                genes_seen.append(current_children[0])\n",
    "                genes_seen.append(current_children[1])\n",
    "            \n",
    "            for i in range(len(current_children)):\n",
    "                \n",
    "                for neighbor in current_children[i].neighbors:\n",
    "                    \n",
    "                    if neighbor == None:\n",
    "                        none_type_as_neighbor += 1\n",
    "                        continue\n",
    "                        \n",
    "                    if neighbor.parent:  # Check whether the neighbor is a singleton (i.e. does it have a parent or not?)\n",
    "                        \n",
    "                        if neighbor.parent in block2 or neighbor in genes_seen:  # Avoid to take into account elements from the initial block or those already assessed\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            blocks_seen[i].add(tuple(neighbor.parent.blocks[genome.name]))\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region \n",
    "                            break  \n",
    "\n",
    "                    else:\n",
    "                        \n",
    "                        if neighbor in genes_seen:\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            next_children[i] = neighbor\n",
    "                            genes_seen.append(neighbor)\n",
    "                            current_children[i].next_neighbor = neighbor  # Used in the function for the linking region\n",
    "                \n",
    "            if none_type_as_neighbor > 0:  # If children have only one neighbor\n",
    "                break\n",
    "\n",
    "            counter += 1\n",
    "            current_children = next_children[:]\n",
    "            intersect = list(blocks_seen[0].intersection(*blocks_seen))  # Intersection\n",
    "            \n",
    "            if len(intersect) == 1:  # ex. intersect = [('a', 'c')]\n",
    "\n",
    "                    if list(intersect[0]) in block_record:\n",
    "                        reaching_already_assessed = True\n",
    "                        break\n",
    "                        \n",
    "                    block_found = True  \n",
    "                    counter = 0  # Reset the counter to pursue the assessement with further blocks\n",
    "                    common_block = list(intersect[0])  # Extraction of the last common block. \n",
    "                    block_record.append(common_block)\n",
    "\n",
    "                    link = Link()  # Class for linking regions\n",
    "                    link.get_linking_region(genome, block2, common_block, starting_children)  # Creation of the linking region\n",
    "                    \n",
    "                    extended_block.append(link)  # The linking region\n",
    "                    extended_block.append(common_block)  # The common block\n",
    "\n",
    "                    block2 = common_block  # The new block becomes the last common block found within the while loop\n",
    "\n",
    "                    if len(block2) > 1:\n",
    "                        \n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its first element (i.e. element 0), the next iteration should start from its last element (i.e. element -1) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block2[0].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block2[0].children)[0] and get_unique_children(genome, block2[0].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block2[-1].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  \n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block2[0].children)\n",
    "                            break\n",
    "\n",
    "                        #################################################################################################################################################\n",
    "                        # If we reach a common block toward its last element (i.e. element -1), the next iteration should start from its first element (i.e. element 0) #\n",
    "                        #################################################################################################################################################\n",
    "\n",
    "                        if len(get_unique_children(genome, block2[-1].children)) == 2:\n",
    "                            \n",
    "                            if get_unique_children(genome, block2[-1].children)[0] and get_unique_children(genome, block2[-1].children)[1] in genes_seen:\n",
    "                                current_children = get_unique_children(genome, block2[0].children)  \n",
    "                                starting_children = current_children\n",
    "                                \n",
    "                                if len(starting_children) == 2:  # 05.01.18\n",
    "                                    genes_seen = adjust_genes_seen(genome, genes_seen, starting_children)  # 05.01.18\n",
    "\n",
    "                        else:\n",
    "                            current_children = get_unique_children(genome, block2[-1].children)\n",
    "                            break\n",
    "                    \n",
    "                    ################################################################################################################\n",
    "                    # When we reach a common unit block, the next iteration should start from the same unique HOG (i.e. element 0) #\n",
    "                    ################################################################################################################\n",
    "\n",
    "                    elif len(block2) == 1:  # If the block has only one HOG...\n",
    "                        current_children = get_unique_children(genome, block2[0].children) \n",
    "                        \n",
    "                        if len(current_children) == 2:\n",
    "                            genes_seen = refine_genes_seen(genome, genes_seen, current_children)\n",
    "                            current_children = check_unique_children_from_unit_block(genome, block2[0].children, genes_seen)\n",
    "\n",
    "                        starting_children = current_children\n",
    "                        \n",
    "                    blocks_seen = [set() for child in current_children]  # The variable is restarted for the next iteration\n",
    "\n",
    "            if reaching_already_assessed:  \n",
    "                break\n",
    "                \n",
    "            if len(current_children) == 2:\n",
    "                if current_children[0].genome != current_children[1].genome:\n",
    "                    continue\n",
    "                \n",
    "                break  # To exit the while loop\n",
    "                \n",
    "            else:  \n",
    "                break  # To exit the while loop\n",
    "\n",
    "        if extended_block:\n",
    "            for item in extended_block:\n",
    "                if isinstance(item, Link):\n",
    "                    tmp[0].append(item)\n",
    "\n",
    "                else:\n",
    "                    for element in item:\n",
    "                        tmp[0].append(element)\n",
    "        \n",
    "        # In tmp[1] there is always something, namely either block or extended block.\n",
    "        # The list tmp[0] can be empty\n",
    "        if tmp[0] and tmp[1]:\n",
    "            tmp[0] = tmp[0][::-1]\n",
    "            new_blocks.append(tmp[0] + tmp[1])\n",
    "            \n",
    "        elif tmp[1] and not tmp[0]:\n",
    "            new_blocks.append(tmp[1])\n",
    "            \n",
    "    genome.new_synteny = new_blocks\n",
    "\n",
    "    add_remaining_blocks(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_remaining_blocks(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    Once the secondary reconstruction is done. Some blocks or single HOGs from the primary reconstruction may\n",
    "    remained unassessed. The function add such items within the collection of extended synteny blocks.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        \n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    c = 0\n",
    "    \n",
    "    for block in genome.new_synteny:\n",
    "        \n",
    "        for element in block:\n",
    "            \n",
    "            if not isinstance(element, pyham.HOG):\n",
    "                \n",
    "                for key, value in element.linking_region.items():\n",
    "                    \n",
    "                    for gene in value:\n",
    "                        \n",
    "                        if isinstance(gene, pyham.HOG):\n",
    "                            gene.processed = True\n",
    "                            c += 1\n",
    "\n",
    "            else:\n",
    "                element.processed = True\n",
    "                c += 1\n",
    "    \n",
    "    not_processed = 0\n",
    "    \n",
    "    for gene in genome.genes:\n",
    "        \n",
    "        if gene.processed == False and gene.blocks[genome.name] not in genome.new_synteny:\n",
    "            genome.new_synteny.append(gene.blocks[genome.name])\n",
    "            not_processed += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reconstruction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter = 1\n",
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    assign_neighbors(node.genome)\n",
    "    get_synteny_blocks(node.genome)\n",
    "    assign_block_to_genes(node.genome)\n",
    "    \n",
    "    if isinstance(node.genome, pyham.AncestralGenome):\n",
    "        joining_blocks(20, node.genome)\n",
    "        \n",
    "        ### Plot ###\n",
    "        \n",
    "        title = node.genome.name  \n",
    "        raw_values = []\n",
    "        \n",
    "        for block in node.genome.new_synteny:\n",
    "            count = 0\n",
    "            \n",
    "            for element in block:\n",
    "                \n",
    "                if isinstance(element, pyham.HOG):\n",
    "                    count += 1\n",
    "                    \n",
    "            raw_values.append(count)\n",
    "        \n",
    "        label = list(set(raw_values))\n",
    "        values = [0 for i in range(len(label))]\n",
    "        \n",
    "        for i in range(len(label)):\n",
    "            \n",
    "            for element in raw_values:\n",
    "                \n",
    "                if label[i] == element:\n",
    "                    values[i] += 1\n",
    "\n",
    "        plt.bar(label, values)\n",
    "        plt.title(title)\n",
    "        plt.xlabel('Abstractgenes per block')\n",
    "        plt.ylabel('Number of blocks')\n",
    "        plt.savefig('results/plots/secondary_reconstruction/plot{}.png'.format(counter))\n",
    "        counter += 1\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plots for the primary reconstruction\n",
    "\n",
    "counter = 1\n",
    "for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "    \n",
    "    if isinstance(node.genome, pyham.AncestralGenome):\n",
    "        title = node.genome.name  \n",
    "        raw_values = [len(block) for block in node.genome.synteny]  \n",
    "        label = list(set(raw_values))\n",
    "        values = [0 for i in range(len(label))]\n",
    "        \n",
    "        for i in range(len(label)):\n",
    "            \n",
    "            for element in raw_values:\n",
    "                \n",
    "                if label[i] == element:\n",
    "                    values[i] += 1\n",
    "\n",
    "        plt.bar(list(label), values)\n",
    "        plt.title(title)\n",
    "        plt.xlabel('Abstractgenes per block')\n",
    "        plt.ylabel('Number of blocks')\n",
    "        plt.savefig('results/plots/primary_reconstruction/plot{}.png'.format(counter))\n",
    "        counter += 1\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fraction_in_blocks():\n",
    "    \n",
    "    \"\"\"\n",
    "    Compute the fraction of :obj:`pyham.abstractgene.HOG` within synteny blocks and extended synteny blocks.\n",
    "    \n",
    "    Args:\n",
    "        xxx\n",
    "        \n",
    "    Returns:\n",
    "        return a text file.\n",
    "    \"\"\"\n",
    "    \n",
    "    with open('results/fraction_of_hog_in_blocks.txt', 'w+') as stat:\n",
    "        stat.write('{:120s} {}{}{}\\n'.format(' ', 'r1', '\\t', 'r2'))\n",
    "        \n",
    "        for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "            HOGs_in_blocks_r1 = 0\n",
    "            HOGs_in_blocks_r2 = 0\n",
    "            \n",
    "            HOGs_resolved_r1 = None\n",
    "            HOGs_resolved_r2 = None\n",
    "            \n",
    "            if isinstance(node.genome, pyham.AncestralGenome):\n",
    "                \n",
    "                for block in node.genome.synteny:\n",
    "                    \n",
    "                    if len(block) == 1:\n",
    "                        continue\n",
    "                        \n",
    "                    for element in block:\n",
    "                            HOGs_in_blocks_r1 += 1\n",
    "                            \n",
    "                for block in node.genome.new_synteny:\n",
    "                    \n",
    "                    if len(block) == 1: #and len(block[0].children) > 1:\n",
    "                        continue\n",
    "                        \n",
    "                    for element in block:\n",
    "                        \n",
    "                        if isinstance(element, pyham.HOG):\n",
    "                            HOGs_in_blocks_r2 += 1\n",
    "                \n",
    "                HOGs_resolved_r1 = HOGs_in_blocks_r1/len(node.genome.genes)\n",
    "                HOGs_resolved_r2 = HOGs_in_blocks_r2/len(node.genome.genes)\n",
    "                stat.write('{:120s} {}{}{}\\n'.format(node.genome.name, round(HOGs_resolved_r1, 3), '\\t', round(HOGs_resolved_r2, 3)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def number_and_length(species):\n",
    "    \n",
    "    \"\"\"\n",
    "    Compute the number of synteny blocks as well as the average length for a given genome.\n",
    "    \n",
    "    Args:\n",
    "        species (:obj:`string`): the name of the genome\n",
    "        \n",
    "    Returns:\n",
    "        return a text file.\n",
    "    \"\"\"\n",
    "    \n",
    "    with open('results/number_and_length.txt', 'w+') as stat:\n",
    "        stat.write(species+'\\n')\n",
    "        stat.write('{}{}1st reconstruction{}2nd reconstruction\\n'.format('\\t','\\t','\\t'))\n",
    "    \n",
    "        for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "            \n",
    "            if node.genome.name == species:\n",
    "                block_counter_1 = 0\n",
    "                block_counter_2 = 0\n",
    "                \n",
    "                total_length_1 = 0\n",
    "                total_length_2 = 0\n",
    "                \n",
    "                mean_length_1 = None\n",
    "                mean_length_2 = None\n",
    "\n",
    "                for block in node.genome.synteny:\n",
    "                    \n",
    "                    if len(block) > 1:  # Perhaps modify to take into account single HOGs with only one child?\n",
    "                        block_counter_1 += 1\n",
    "                        total_length_1 += len(block)\n",
    "                \n",
    "                for block in node.genome.new_synteny:\n",
    "                    \n",
    "                    if len(block) > 1:\n",
    "                        block_counter_2 += 1\n",
    "\n",
    "                        for element in block:\n",
    "                            \n",
    "                            if isinstance(element, pyham.HOG):\n",
    "                                total_length_2 += 1\n",
    "                    \n",
    "                mean_length_1 = total_length_1/block_counter_1\n",
    "                mean_length_2 = total_length_2/block_counter_2\n",
    "                \n",
    "                stat.write('Block number{}{}{}{}{}{}\\n'.format('\\t', block_counter_1, '\\t', '\\t', '\\t', block_counter_2))\n",
    "                stat.write('Mean length{}{}{}{}{}{}'.format('\\t', round(mean_length_1, 2), '\\t', '\\t', '\\t', round(mean_length_2, 2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def decrease():\n",
    "    \n",
    "    \"\"\"\n",
    "    Compute the decrease of single HOGs in term of pourcentage.\n",
    "    \n",
    "    Args:\n",
    "        xxx\n",
    "        \n",
    "    Returns:\n",
    "        return a text file.\n",
    "    \"\"\"\n",
    "    \n",
    "    decreases = []\n",
    "\n",
    "    with open('results/percentage_decrease.txt', 'w+') as stat:\n",
    "        stat.write('{:120s} {}\\n'.format(' ', 'Decrease %'))\n",
    "\n",
    "        for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "            single_r1 = 0\n",
    "            single_r2 = 0\n",
    "            if isinstance(node.genome, pyham.AncestralGenome):\n",
    "                for block in node.genome.synteny:\n",
    "                    if len(block) == 1:\n",
    "                        single_r1 += 1\n",
    "\n",
    "                for block in node.genome.new_synteny:\n",
    "                    if len(block) == 1:\n",
    "                        single_r2 += 1\n",
    "\n",
    "                perc = ((single_r1 - single_r2)/single_r1)*100\n",
    "                decreases.append(round(perc, 1))\n",
    "                stat.write('{:120s} {}\\n'.format(node.genome.name, round(perc, 1)))\n",
    "    print(decreases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def one_child_hogs_in_link():\n",
    "    \n",
    "    \"\"\"\n",
    "    Compute the amount of HOGs containing a single descendant in one species and that are present\n",
    "    within linking regions.\n",
    "    \n",
    "    Args:\n",
    "        xxx\n",
    "        \n",
    "    Returns:\n",
    "        return a text file.\n",
    "    \"\"\"\n",
    "    \n",
    "    with open('results/one_child_hogs_in_link.txt', 'w+') as stat:\n",
    "        stat.write('{:120s} {}\\n'.format(' ', 'One-child HOGs %'))\n",
    "        \n",
    "        for node in ham_analysis_no_name.taxonomy.tree.traverse('postorder'):\n",
    "\n",
    "            if isinstance(node.genome, pyham.AncestralGenome):\n",
    "\n",
    "                unit_HOGs = 0\n",
    "                one_child_HOGs = 0\n",
    "\n",
    "                for block in node.genome.new_synteny:\n",
    "\n",
    "                    for element in block:\n",
    "\n",
    "                        if isinstance(element, Link):\n",
    "\n",
    "                            for key, value in element.linking_region.items():\n",
    "\n",
    "                                for item in value:\n",
    "\n",
    "                                    if item.genome == node.genome:\n",
    "\n",
    "                                        if len(item.blocks[node.genome.name]) == 1:\n",
    "\n",
    "                                            unit_HOGs += 1\n",
    "\n",
    "                                            if len(item.children) == 1:\n",
    "\n",
    "                                                one_child_HOGs += 1\n",
    "\n",
    "                if unit_HOGs > 0:\n",
    "                    perc = (one_child_HOGs/unit_HOGs)*100\n",
    "                    stat.write('{:120s} {}\\n'.format(node.genome.name, round(perc, 1)))\n",
    "\n",
    "                else:\n",
    "                    stat.write('{:120s} {}\\n'.format(node.genome.name, 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_N50(genome):\n",
    "    \n",
    "    \"\"\"\n",
    "    Compute the N50 and L50 values for a given genome. These values are used to assess the quality \n",
    "    of the reconstruction.\n",
    "    \n",
    "    Args:\n",
    "        genome (:obj:`pyham.genome.Genome`): extant or ancestral genome of interest.\n",
    "        \n",
    "    Returns:\n",
    "        xxx\n",
    "    \"\"\"\n",
    "    \n",
    "    counter = 0\n",
    "    blocks_count = 0\n",
    "    \n",
    "    tmp_block = None\n",
    "    L50 = None\n",
    "    \n",
    "    synteny_blocks = []\n",
    "    genome_length = len(genome.genes)\n",
    "    \n",
    "    for block in genome.synteny:\n",
    "        if len(block) > 1:\n",
    "            synteny_blocks.append(block)\n",
    "            \n",
    "    synteny_blocks.sort(key = len, reverse = True)\n",
    "    \n",
    "    for block in synteny_blocks:\n",
    "        if counter > (genome_length/2):\n",
    "            return len(tmp_block), blocks_count  # N50, L50\n",
    "            break\n",
    "        \n",
    "        tmp_block = block\n",
    "        counter += len(block)\n",
    "        blocks_count += 1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py3]",
   "language": "python",
   "name": "conda-env-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
